# 10 Sex Linked Markers {#sec-s10}

```{r setup, include=FALSE}
library(knitr)
library(formatR)
library(tidyverse)
knitr::opts_chunk$set(tidy = TRUE, tidy.opts = list(width.cutoff = 72), 
                      cache = TRUE, 
                      echo = TRUE)
```

 *Session leads*
 
![](images/Presenters10.png)


## Required packages

```{r, warning=FALSE, message=FALSE}
library(dartR.base)
library(dartR.sexlinked)
```

## Dataset 1 - ZW//ZZ - The Yellow Tufted Honeyeater

![The Yellow Tufted Honeyeater](images/Yellow_Tufted_Honeyeater.jpg)

### Load data

::: panel-tabset
##### code

```{r, output = FALSE}
data("YTH")

YTH                   # Explore the dataset
YTH@n.loc             # Number of SNPs
length(YTH@ind.names) # Number of individuals

```

##### data

```{r, echo = FALSE}
data("YTH")
YTH                   # Explore the dataset
```

##### SNPs

```{r, echo = FALSE}
YTH@n.loc             # Number of SNPs
```

##### Individuals

```{r, echo = FALSE}
length(YTH@ind.names) # Number of individuals

```
:::

### Run `filter.sex.linked`

This function identifies sex-linked and autosomal loci present in a SNP dataset (i.e., genlight object) using individuals with known sex. It identifies five types of loci: w-linked or y-linked, sex-biased, z-linked or x-linked, gametologous and autosomal.

*The genlight object must contain in gl\@other\$ind.metrics a column named "id", and a column named "sex" in which individuals with known-sex are assigned 'M' for male, or 'F' for female. The function ignores individuals that are assigned anything else or nothing at all (unknown-sex).*

![ZW/ZZ sex chromosomes](images/ZW.png)

::: callout-tip
**NOTE**

Set `ncores` to more than 1 (default) if you have more than 50,000 SNPs, since it could actually slow down the analysis with smaller datasets.
:::

::: panel-tabset
##### code

```{r, output = FALSE}
head(YTH@other$ind.metrics) # Check that ind.metrics has the necessary columns
```

##### table

```{r, echo=FALSE}
knitr::kable(head(YTH@other$ind.metrics)) # Check that ind.metrics has the necessary columns
```
:::

::: panel-tabset
##### code

```{r, output= FALSE}
res <- dartR.sexlinked::filter.sex.linked(gl = YTH, system = "zw")
```

##### output

```{r, echo=FALSE}
res <- dartR.sexlinked::filter.sex.linked(gl = YTH, system = "zw")
```
:::

::: callout-note
### Exercise

![](images/task.png){#id .class width="48" height="48"}

How many males and females does the dataset contain?

How many sex-linked loci were found?
:::

Now check the output:

::: panel-tabset
##### code

```{r, output = FALSE}
res$w.linked       # Notice that it says "w-linked"
res$z.linked       # Notice that it says "z-linked"
res$gametolog
res$sex.biased
res$autosomal
knitr::kable(head(res$results.table))  # The output table
```

##### w.linked

```{r, echo = FALSE}
res$w.linked 
```

##### z.linked

##### gametolog

##### sex.biased

##### autosomal

##### table
:::

The output consists of a genlight object for each type of loci, plus a results table.

### Run `infer.sex`

This function uses the complete output of function filter.sex.linked (list of 6 objects) to infer the sex of all individuals in the dataset. Specifically, the function uses 3 types of sex-linked loci (W-/Y-linked, Z-/X-linked, and gametologs), assigns a preliminary genetic sex for each type of sex-linked loci available, and outputs an `agreed sex`.

```{r}
sexID <- dartR.sexlinked::infer.sex(gl_sex_filtered = res, system = "zw", seed = 124)

knitr::kable(head(sexID))
```

::: callout-warning
**IMPORTANT** We created this function with the explicit intent that a human checks the evidence for the `agreed sex` that do NOT agree for all types of sex-linked loci (denoted as '\*M' or '\*F'). This human can then use their criterion to validate these assignments.
:::

::: callout-note
### Exercise

![](images/task.png){.class width="48" height="48"}

Can you find individuals for which the `agreed sex` is uncertain (i.e., has an asterisk "\*")?
:::

## EXERCISE

::: callout-note
### Exercise

![](images/task.png){.class width="48" height="48"}

Imagine you are analyzing the genetic data of a population and you want to look at its genetic structure and genetic diversity. You get to work by filtering your beautiful set of SNPs. Because you are a rigorous, you want to test whether using function `filter.sex.linked` to properly filter out sex-linked loci changes the results of the populations analyses. So you: (1) process your SNP dataset by applying standard filters and run analyses, and (2) process the dataset by filtering out sex-linked loci first, applying standard filters second, and then run analyses.

Choose one of the following datasets (or your own data) and report:

1.  How many sex-linked markers are present?

2.  How many individuals had a wrong sexID?

3.  Do you see any changes in your PCA/structure analyses when you filtered out sex-linked markers versus when you did not?

4.  Do you see any differences in genetic diversity and fixation indices when you filtered out sex-linked markers versus when you did not?
:::

![](images/beer.png){.class width="200" height="200"} ![](images/Eastern_Yellow_Robin.jpg){alt="The Eastern Yellow Robin" width="349"}

![](images/Bull_Shark.jpg){.class width="350" height="200"} ![](images/Blue_shark.jpg){.class width="350" height="200"}

### Exercise data 1 - Your own data

::: {.infobox data-latex="{task}"}
**HINT**

You can have a look at the exercise data below for inspiration.
:::

#### 1. Number of sex-linked markers?

#### 2. Individuals with wrong sexID?

#### 3. Changes in PCA before and after removing the SLM?

#### 4. Differences in genetic diversity and fixation indices between autosomal and SLM?

### Exercise data 2 - The Eastern Yellow Robin

Data from @robledo2023

![The Eastern Yellow Robin](images/Eastern_Yellow_Robin.jpg)

## Further Readings

Devloo-Delva, F., Burridge, C. P., Kyne, P. M., Brunnschweiler, J. M., Chapman, D. D., Charvet, P., ... & Feutry, P. (2023). From rivers to ocean basins: The role of ocean barriers and philopatry in the genetic structuring of a cosmopolitan coastal predator. Ecology and Evolution, 13(2), e9837. https://doi.org/10.1002/ece3.9837

Nikolic, N., Devloo-Delva, F., Bailleul, D., Noskova, E., Rougeux, C., Delord, C., ... & Arnaud‐Haond, S. (2023). Stepping up to genome scan allows stock differentiation in the worldwide distributed blue shark *Prionace glauca*. Molecular Ecology, 32(5), 1000-1019. https://doi.org/10.1111/mec.16822

Robledo-Ruiz, D. A., Austin, L., Amos, J. N., Castrejón-Figueroa, J., Harley, D. K. P., Magrath, M. J. L., Sunnucks, P., & Pavlova, A. (2023). Easy-to-use R functions to separate reduced-representation genomic datasets into sex-linked and autosomal loci, and conduct sex assignment. Molecular Ecology Resources, 00, 1--21. https://doi.org/10.1111/1755-0998.13844
