<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Kioloa PopGen - 10&nbsp; Sex Linked Markers</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./session11.html" rel="next">
<link href="./session09.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="css.css">
<link rel="stylesheet" href="include/webex.css">
</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./session10.html"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Sex Linked Markers</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Kioloa PopGen</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome!</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./schedule.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Program</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./speakers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Speaker Bios</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./install.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Install dartRverse</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session01.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Intro to dartR</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session02.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Pop Gen In Conservation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session03.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Sequencing Technologies</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session04.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Data Management, Reproducibility &amp; Integrity</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session05.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Effective Population Size</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session06.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Management of Small Populations</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session07_CB_Yuma.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Natural Selection</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session08.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Landscape Genetics</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session09.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Lineage Divergence</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session10.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Sex Linked Markers</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session11.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">From Genes to Kin: Dissecting Relatedness &amp; Kinship</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session12.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Genetic Structure</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session13_JD.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Combining Genomic Resources</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./break.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">__________</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./useful.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">More info</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">dartR Tutorials</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#required-packages" id="toc-required-packages" class="nav-link active" data-scroll-target="#required-packages"><em>Required packages</em></a></li>
  <li><a href="#dataset-1---zwzz---the-yellow-tufted-honeyeater" id="toc-dataset-1---zwzz---the-yellow-tufted-honeyeater" class="nav-link" data-scroll-target="#dataset-1---zwzz---the-yellow-tufted-honeyeater">Dataset 1 - ZW//ZZ - The Yellow Tufted Honeyeater</a>
  <ul class="collapse">
  <li><a href="#load-data" id="toc-load-data" class="nav-link" data-scroll-target="#load-data">Load data</a></li>
  <li><a href="#run-gl.filter.sexlinked" id="toc-run-gl.filter.sexlinked" class="nav-link" data-scroll-target="#run-gl.filter.sexlinked">Run <code>gl.filter.sexlinked</code></a></li>
  <li><a href="#run-gl.infer.sex" id="toc-run-gl.infer.sex" class="nav-link" data-scroll-target="#run-gl.infer.sex">Run <code>gl.infer.sex</code></a></li>
  </ul></li>
  <li><a href="#dataset-2---xxxy---the-leadbeaters-possum" id="toc-dataset-2---xxxy---the-leadbeaters-possum" class="nav-link" data-scroll-target="#dataset-2---xxxy---the-leadbeaters-possum">Dataset 2 - XX/XY - The Leadbeater’s possum</a>
  <ul class="collapse">
  <li><a href="#load-data-1" id="toc-load-data-1" class="nav-link" data-scroll-target="#load-data-1">Load data</a></li>
  <li><a href="#run-gl.filter.sexlinked-1" id="toc-run-gl.filter.sexlinked-1" class="nav-link" data-scroll-target="#run-gl.filter.sexlinked-1">Run <code>gl.filter.sexlinked</code></a></li>
  <li><a href="#run-gl.infer.sex-1" id="toc-run-gl.infer.sex-1" class="nav-link" data-scroll-target="#run-gl.infer.sex-1">Run <code>gl.infer.sex</code></a></li>
  </ul></li>
  <li><a href="#exercise-4" id="toc-exercise-4" class="nav-link" data-scroll-target="#exercise-4">EXERCISE</a></li>
  <li><a href="#further-study" id="toc-further-study" class="nav-link" data-scroll-target="#further-study"><em>Further study</em></a>
  <ul class="collapse">
  <li><a href="#exercise-data-1---your-own-data" id="toc-exercise-data-1---your-own-data" class="nav-link" data-scroll-target="#exercise-data-1---your-own-data">Exercise data 1 - Your own data</a></li>
  <li><a href="#exercise-data-2---the-eastern-yellow-robin" id="toc-exercise-data-2---the-eastern-yellow-robin" class="nav-link" data-scroll-target="#exercise-data-2---the-eastern-yellow-robin">Exercise data 2 - The Eastern Yellow Robin</a></li>
  <li><a href="#exercise-data-3---bull-shark" id="toc-exercise-data-3---bull-shark" class="nav-link" data-scroll-target="#exercise-data-3---bull-shark">Exercise data 3 - Bull shark</a></li>
  <li><a href="#exercise-data-4---blue-shark" id="toc-exercise-data-4---blue-shark" class="nav-link" data-scroll-target="#exercise-data-4---blue-shark">Exercise data 4 - Blue shark</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Sex Linked Markers</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p><em>Session presenters</em></p>
<p><img src="images/Presenters10.png" class="img-fluid"></p>
<section id="required-packages" class="level2">
<h2 class="anchored" data-anchor-id="required-packages"><em>Required packages</em></h2>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-1_4541195b6efc622db5f45e297fcff002">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dartR.base)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dartR.sexlinked)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="dataset-1---zwzz---the-yellow-tufted-honeyeater" class="level2">
<h2 class="anchored" data-anchor-id="dataset-1---zwzz---the-yellow-tufted-honeyeater">Dataset 1 - ZW//ZZ - The Yellow Tufted Honeyeater</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Yellow_Tufted_Honeyeater.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">The Yellow Tufted Honeyeater</figcaption>
</figure>
</div>
<section id="load-data" class="level3">
<h3 class="anchored" data-anchor-id="load-data">Load data</h3>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-2_5ed4bd858f683d74f8289fb291a9a7bf">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">#data("YTH")</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">'./data/YTH.rda'</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>YTH                   <span class="co"># Explore the dataset</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 609 genotypes,  994 SNPs , size: 49.9 Mb

    missing data: 139174 (=22.99 %) scored as NA

 ** Genetic data
   @gen: list of 609 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  609 individual labels
   @loc.names:  994 locus labels
   @loc.all:  994 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 12-516)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, pop, sex, sex_original, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Lichenostomus_HeHo_v1, ChromPos_Lichenostomus_HeHo_v1, AlnCnt_Lichenostomus_HeHo_v1, AlnEvalue_Lichenostomus_HeHo_v1, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, maf 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>YTH<span class="sc">@</span>n.loc             <span class="co"># Number of SNPs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 994</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(YTH<span class="sc">@</span>ind.names) <span class="co"># Number of individuals</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 609</code></pre>
</div>
</div>
</section>
<section id="run-gl.filter.sexlinked" class="level3">
<h3 class="anchored" data-anchor-id="run-gl.filter.sexlinked">Run <code>gl.filter.sexlinked</code></h3>
<p>This function identifies sex-linked and autosomal loci present in a SNP dataset (i.e., genlight object) using individuals with known sex. It identifies five types of loci: w-linked or y-linked, sex-biased, z-linked or x-linked, gametologous and autosomal.</p>
<p><em>The genlight object must contain in gl@other$ind.metrics a column named “id”, and a column named “sex” in which individuals with known-sex are assigned ‘M’ for male, or ‘F’ for female. The function ignores individuals that are assigned anything else or nothing at all (unknown-sex).</em></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/ZW.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">ZW/ZZ sex chromosomes</figcaption>
</figure>
</div>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Caution
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>NOTE</strong></p>
<p>Set <code>ncores</code> to more than 1 (default) if you have more than 50,000 SNPs, since it could actually slow down the analysis with smaller datasets.</p>
</div>
</div>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-3_68732323b42aab4c952a4157f192c745">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(YTH<span class="sc">@</span>other<span class="sc">$</span>ind.metrics)) <span class="co"># Check that ind.metrics has the necessary columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 13%">
<col style="width: 13%">
<col style="width: 12%">
<col style="width: 5%">
<col style="width: 17%">
<col style="width: 17%">
<col style="width: 20%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">id</th>
<th style="text-align: left;">pop</th>
<th style="text-align: left;">sex</th>
<th style="text-align: left;">sex_original</th>
<th style="text-align: left;">service</th>
<th style="text-align: left;">plate_location</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ANWC46839</td>
<td style="text-align: left;">ANWC46839</td>
<td style="text-align: left;">Melanops</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">DLich17-2918</td>
<td style="text-align: left;">1-A1</td>
</tr>
<tr class="even">
<td style="text-align: left;">W49</td>
<td style="text-align: left;">W49</td>
<td style="text-align: left;">Cassidix</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">DLich17-2918</td>
<td style="text-align: left;">1-A10</td>
</tr>
<tr class="odd">
<td style="text-align: left;">W90</td>
<td style="text-align: left;">W90</td>
<td style="text-align: left;">Cassidix</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">DLich17-2918</td>
<td style="text-align: left;">1-A12</td>
</tr>
<tr class="even">
<td style="text-align: left;">C25</td>
<td style="text-align: left;">C25</td>
<td style="text-align: left;">Cassidix</td>
<td style="text-align: left;">M</td>
<td style="text-align: left;">M</td>
<td style="text-align: left;">DLich17-2918</td>
<td style="text-align: left;">1-A2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">C8</td>
<td style="text-align: left;">C8</td>
<td style="text-align: left;">Cassidix</td>
<td style="text-align: left;">M</td>
<td style="text-align: left;">M</td>
<td style="text-align: left;">DLich17-2918</td>
<td style="text-align: left;">1-A3</td>
</tr>
<tr class="even">
<td style="text-align: left;">W70</td>
<td style="text-align: left;">W70</td>
<td style="text-align: left;">Cassidix</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">DLich17-2918</td>
<td style="text-align: left;">1-A4</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">gl.filter.sexlinked</span>(<span class="at">gl =</span> YTH, <span class="at">system =</span> <span class="st">"zw"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Detected 276 females and 333 males.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Starting phase 1. May take a while...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building call rate plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done. Starting phase 2.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building heterozygosity plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-3-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-3-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done building heterozygosity plots.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**FINISHED** Total of analyzed loci: 994.
Found 506 sex-linked loci:
   52 W-linked loci
   273 sex-biased loci
   165 Z-linked loci
   16 ZW gametologs.
And 488 autosomal loci.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-3-4.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/task.png" id="id" class="class" width="48" height="48"></p>
<p>How many males and females does the dataset contain?</p>
<p>How many sex-linked loci were found?</p>
</div>
</div>
<p>Now check the output:</p>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-4_14241064a290a6ad55bed496a3a7f6ac">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>w.linked       <span class="co"># Notice that it says "w-linked"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 609 genotypes,  52 SNPs , size: 48.9 Mb

    missing data: 17304 (=54.64 %) scored as NA

 ** Genetic data
   @gen: list of 609 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  609 individual labels
   @loc.names:  52 locus labels
   @loc.all:  52 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 12-516)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, pop, sex, sex_original, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Lichenostomus_HeHo_v1, ChromPos_Lichenostomus_HeHo_v1, AlnCnt_Lichenostomus_HeHo_v1, AlnEvalue_Lichenostomus_HeHo_v1, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, maf 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>z.linked       <span class="co"># Notice that it says "z-linked"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 609 genotypes,  165 SNPs , size: 48.9 Mb

    missing data: 2990 (=2.98 %) scored as NA

 ** Genetic data
   @gen: list of 609 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  609 individual labels
   @loc.names:  165 locus labels
   @loc.all:  165 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 12-516)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, pop, sex, sex_original, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Lichenostomus_HeHo_v1, ChromPos_Lichenostomus_HeHo_v1, AlnCnt_Lichenostomus_HeHo_v1, AlnEvalue_Lichenostomus_HeHo_v1, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, maf 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>gametolog</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 609 genotypes,  16 SNPs , size: 48.8 Mb

    missing data: 580 (=5.95 %) scored as NA

 ** Genetic data
   @gen: list of 609 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  609 individual labels
   @loc.names:  16 locus labels
   @loc.all:  16 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 12-516)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, pop, sex, sex_original, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Lichenostomus_HeHo_v1, ChromPos_Lichenostomus_HeHo_v1, AlnCnt_Lichenostomus_HeHo_v1, AlnEvalue_Lichenostomus_HeHo_v1, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, maf 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>sex.biased</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 609 genotypes,  273 SNPs , size: 49.2 Mb

    missing data: 46048 (=27.7 %) scored as NA

 ** Genetic data
   @gen: list of 609 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  609 individual labels
   @loc.names:  273 locus labels
   @loc.all:  273 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 12-516)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, pop, sex, sex_original, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Lichenostomus_HeHo_v1, ChromPos_Lichenostomus_HeHo_v1, AlnCnt_Lichenostomus_HeHo_v1, AlnEvalue_Lichenostomus_HeHo_v1, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, maf 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>autosomal</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 609 genotypes,  488 SNPs , size: 49.4 Mb

    missing data: 72252 (=24.31 %) scored as NA

 ** Genetic data
   @gen: list of 609 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  609 individual labels
   @loc.names:  488 locus labels
   @loc.all:  488 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 12-516)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, pop, sex, sex_original, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Lichenostomus_HeHo_v1, ChromPos_Lichenostomus_HeHo_v1, AlnCnt_Lichenostomus_HeHo_v1, AlnEvalue_Lichenostomus_HeHo_v1, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, maf 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(res<span class="sc">$</span>results.table))  <span class="co"># The output table</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 5%">
<col style="width: 2%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 3%">
<col style="width: 4%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 3%">
<col style="width: 4%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">index</th>
<th style="text-align: right;">count.F.miss</th>
<th style="text-align: right;">count.M.miss</th>
<th style="text-align: right;">count.F.scored</th>
<th style="text-align: right;">count.M.scored</th>
<th style="text-align: right;">ratio</th>
<th style="text-align: right;">p.value</th>
<th style="text-align: right;">p.adjusted</th>
<th style="text-align: right;">scoringRate.F</th>
<th style="text-align: right;">scoringRate.M</th>
<th style="text-align: left;">w.linked</th>
<th style="text-align: left;">sex.biased</th>
<th style="text-align: right;">count.F.het</th>
<th style="text-align: right;">count.M.het</th>
<th style="text-align: right;">count.F.hom</th>
<th style="text-align: right;">count.M.hom</th>
<th style="text-align: right;">stat</th>
<th style="text-align: right;">stat.p.value</th>
<th style="text-align: right;">stat.p.adjusted</th>
<th style="text-align: right;">heterozygosity.F</th>
<th style="text-align: right;">heterozygosity.M</th>
<th style="text-align: left;">z.linked</th>
<th style="text-align: left;">zw.gametolog</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">27382025-26-T/C</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">61</td>
<td style="text-align: right;">25</td>
<td style="text-align: right;">215</td>
<td style="text-align: right;">308</td>
<td style="text-align: right;">3.4882302</td>
<td style="text-align: right;">0.0000003</td>
<td style="text-align: right;">0.0000016</td>
<td style="text-align: right;">0.7789855</td>
<td style="text-align: right;">0.9249249</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">73</td>
<td style="text-align: right;">215</td>
<td style="text-align: right;">235</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">NA</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.2370130</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
</tr>
<tr class="even">
<td style="text-align: left;">27338005-34-A/G</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">12</td>
<td style="text-align: right;">13</td>
<td style="text-align: right;">264</td>
<td style="text-align: right;">320</td>
<td style="text-align: right;">1.1186728</td>
<td style="text-align: right;">0.8390268</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.9565217</td>
<td style="text-align: right;">0.9609610</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">144</td>
<td style="text-align: right;">264</td>
<td style="text-align: right;">176</td>
<td style="text-align: right;">0.0046581</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.4500000</td>
<td style="text-align: left;">TRUE</td>
<td style="text-align: left;">FALSE</td>
</tr>
<tr class="odd">
<td style="text-align: left;">27331627-16-T/G</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">108</td>
<td style="text-align: right;">159</td>
<td style="text-align: right;">168</td>
<td style="text-align: right;">174</td>
<td style="text-align: right;">0.7039235</td>
<td style="text-align: right;">0.0334301</td>
<td style="text-align: right;">0.0860868</td>
<td style="text-align: right;">0.6086957</td>
<td style="text-align: right;">0.5225225</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">168</td>
<td style="text-align: right;">172</td>
<td style="text-align: right;">0.5128690</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0114943</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
</tr>
<tr class="even">
<td style="text-align: left;">53948461-35-G/A</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">46</td>
<td style="text-align: right;">64</td>
<td style="text-align: right;">230</td>
<td style="text-align: right;">269</td>
<td style="text-align: right;">0.8408645</td>
<td style="text-align: right;">0.4593502</td>
<td style="text-align: right;">0.7791707</td>
<td style="text-align: right;">0.8333333</td>
<td style="text-align: right;">0.8078078</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">29</td>
<td style="text-align: right;">27</td>
<td style="text-align: right;">201</td>
<td style="text-align: right;">242</td>
<td style="text-align: right;">1.2924846</td>
<td style="text-align: right;">0.3950852</td>
<td style="text-align: right;">0.9051780</td>
<td style="text-align: right;">0.1260870</td>
<td style="text-align: right;">0.1003717</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
</tr>
<tr class="odd">
<td style="text-align: left;">27360874-8-A/G</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">41</td>
<td style="text-align: right;">63</td>
<td style="text-align: right;">235</td>
<td style="text-align: right;">270</td>
<td style="text-align: right;">0.7480768</td>
<td style="text-align: right;">0.1956613</td>
<td style="text-align: right;">0.4120495</td>
<td style="text-align: right;">0.8514493</td>
<td style="text-align: right;">0.8108108</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">25</td>
<td style="text-align: right;">38</td>
<td style="text-align: right;">210</td>
<td style="text-align: right;">232</td>
<td style="text-align: right;">0.7272744</td>
<td style="text-align: right;">0.2807852</td>
<td style="text-align: right;">0.7009152</td>
<td style="text-align: right;">0.1063830</td>
<td style="text-align: right;">0.1407407</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
</tr>
<tr class="even">
<td style="text-align: left;">27377678-32-C/A</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">30</td>
<td style="text-align: right;">33</td>
<td style="text-align: right;">246</td>
<td style="text-align: right;">300</td>
<td style="text-align: right;">1.1084581</td>
<td style="text-align: right;">0.7894511</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.8913043</td>
<td style="text-align: right;">0.9009009</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">243</td>
<td style="text-align: right;">294</td>
<td style="text-align: right;">0.6054740</td>
<td style="text-align: right;">0.5234966</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.0121951</td>
<td style="text-align: right;">0.0200000</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The output consists of a genlight object for each type of loci, plus a results table.</p>
</section>
<section id="run-gl.infer.sex" class="level3">
<h3 class="anchored" data-anchor-id="run-gl.infer.sex">Run <code>gl.infer.sex</code></h3>
<p>This function uses the complete output of function <code>gl.filter.sexlinked</code> (list of 6 objects) to infer the sex of all individuals in the dataset. Specifically, the function uses 3 types of sex-linked loci (W-/Y-linked, Z-/X-linked, and gametologs), assigns a preliminary genetic sex for each type of sex-linked loci available, and outputs an <code>agreed sex</code>.</p>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-5_adb779aa6b1b95b714cf1da29dc92a76">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>sexID <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">gl.infer.sex</span>(<span class="at">gl_sex_filtered =</span> res, <span class="at">system =</span> <span class="st">"zw"</span>, <span class="at">seed =</span> <span class="dv">124</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>***FINISHED***</code></pre>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(sexID))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 8%">
<col style="width: 8%">
<col style="width: 11%">
<col style="width: 7%">
<col style="width: 6%">
<col style="width: 11%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 12%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 9%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">id</th>
<th style="text-align: left;">w.linked.sex</th>
<th style="text-align: right;">#missing</th>
<th style="text-align: right;">#called</th>
<th style="text-align: left;">z.linked.sex</th>
<th style="text-align: right;">#Hom.z</th>
<th style="text-align: right;">#Het.z</th>
<th style="text-align: left;">gametolog.sex</th>
<th style="text-align: right;">#Hom.g</th>
<th style="text-align: right;">#Het.g</th>
<th style="text-align: left;">agreed.sex</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ANWC46839</td>
<td style="text-align: left;">ANWC46839</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">51</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">141</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="even">
<td style="text-align: left;">W49</td>
<td style="text-align: left;">W49</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">52</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">156</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="odd">
<td style="text-align: left;">W90</td>
<td style="text-align: left;">W90</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">48</td>
<td style="text-align: right;">4</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">162</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="even">
<td style="text-align: left;">C25</td>
<td style="text-align: left;">C25</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">52</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">52</td>
<td style="text-align: right;">113</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">C8</td>
<td style="text-align: left;">C8</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">52</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">48</td>
<td style="text-align: right;">116</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">W70</td>
<td style="text-align: left;">W70</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">49</td>
<td style="text-align: right;">3</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">152</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">F</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>IMPORTANT</strong> We created this function with the explicit intent that a human checks the evidence for the <code>agreed sex</code> that do NOT agree for all types of sex-linked loci (denoted as ‘*M’ or ‘*F’). This human can then use their criterion to validate these assignments.</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/task.png" class="class" width="48" height="48"></p>
<p>Can you find individuals for which the <code>agreed sex</code> is uncertain (i.e., has an asterisk “*”)?</p>
</div>
</div>
</section>
</section>
<section id="dataset-2---xxxy---the-leadbeaters-possum" class="level2">
<h2 class="anchored" data-anchor-id="dataset-2---xxxy---the-leadbeaters-possum">Dataset 2 - XX/XY - The Leadbeater’s possum</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Leadbeaters_possum.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">The Leadbeater’s possum</figcaption>
</figure>
</div>
<section id="load-data-1" class="level3">
<h3 class="anchored" data-anchor-id="load-data-1">Load data</h3>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-6_2240e64b1dfb006b0232010368ee4032">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co">#data("LBP")</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">'./data/LBP.rda'</span>)</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>LBP                   <span class="co"># Explore the dataset</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 376 genotypes,  1,000 SNPs , size: 5.2 Mb

    missing data: 20670 (=5.5 %) scored as NA

 ** Genetic data
   @gen: list of 376 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  376 individual labels
   @loc.names:  1000 locus labels
   @loc.all:  1000 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 95-281)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, sex, pop, Year.collected, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Possum_v2, ChromPos_Possum_v2, AlnCnt_Possum_v2, AlnEvalue_Possum_v2, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, monomorphs, maf, OneRatio, PIC 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>LBP<span class="sc">@</span>n.loc             <span class="co"># Number of SNPs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(LBP<span class="sc">@</span>ind.names) <span class="co"># Number of individuals</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 376</code></pre>
</div>
</div>
</section>
<section id="run-gl.filter.sexlinked-1" class="level3">
<h3 class="anchored" data-anchor-id="run-gl.filter.sexlinked-1">Run <code>gl.filter.sexlinked</code></h3>
<p>This function identifies sex-linked and autosomal loci present in a SNP dataset (genlight object) using individuals with known sex. It identifies five types of loci: w-linked or y-linked, sex-biased, z-linked or x-linked, gametologous and autosomal.</p>
<p><em>The genlight object must contain in gl@other$ind.metrics a column named “id”, and a column named “sex” in which individuals with known-sex are assigned ‘M’ for male, or ‘F’ for female. The function ignores individuals that are assigned anything else or nothing at all (unknown-sex).</em></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/XY.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">XX/XY sex chromosomes</figcaption>
</figure>
</div>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-7_538839a6b9b5ded6103ecf70487c9c12">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(LBP<span class="sc">@</span>other<span class="sc">$</span>ind.metrics)) <span class="co"># Check that ind.metrics has the necessary columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 15%">
<col style="width: 23%">
<col style="width: 20%">
<col style="width: 23%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">id</th>
<th style="text-align: left;">sex</th>
<th style="text-align: left;">pop</th>
<th style="text-align: right;">Year.collected</th>
<th style="text-align: left;">service</th>
<th style="text-align: left;">plate_location</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Y2</td>
<td style="text-align: left;">Y2</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">Yellingbo</td>
<td style="text-align: right;">1997</td>
<td style="text-align: left;">DLpos17-2786</td>
<td style="text-align: left;">1-A1</td>
</tr>
<tr class="even">
<td style="text-align: left;">Y16</td>
<td style="text-align: left;">Y16</td>
<td style="text-align: left;">M</td>
<td style="text-align: left;">Yellingbo</td>
<td style="text-align: right;">2001</td>
<td style="text-align: left;">DLpos17-2786</td>
<td style="text-align: left;">1-A10</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Y17</td>
<td style="text-align: left;">Y17</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">Yellingbo</td>
<td style="text-align: right;">1997</td>
<td style="text-align: left;">DLpos17-2786</td>
<td style="text-align: left;">1-A11</td>
</tr>
<tr class="even">
<td style="text-align: left;">Y18</td>
<td style="text-align: left;">Y18</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">Yellingbo</td>
<td style="text-align: right;">1999</td>
<td style="text-align: left;">DLpos17-2786</td>
<td style="text-align: left;">1-A12</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Y3</td>
<td style="text-align: left;">Y3</td>
<td style="text-align: left;">F</td>
<td style="text-align: left;">Yellingbo</td>
<td style="text-align: right;">1997</td>
<td style="text-align: left;">DLpos17-2786</td>
<td style="text-align: left;">1-A2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Y4</td>
<td style="text-align: left;">Y4</td>
<td style="text-align: left;">M</td>
<td style="text-align: left;">Yellingbo</td>
<td style="text-align: right;">1997</td>
<td style="text-align: left;">DLpos17-2786</td>
<td style="text-align: left;">1-A3</td>
</tr>
</tbody>
</table>
</div>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">gl.filter.sexlinked</span>(<span class="at">gl =</span> LBP, <span class="at">system =</span> <span class="st">"xy"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Detected 162 females and 211 males.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Starting phase 1. May take a while...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building call rate plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done. Starting phase 2.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building heterozygosity plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-7-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-7-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done building heterozygosity plots.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**FINISHED** Total of analyzed loci: 1000.
Found 77 sex-linked loci:
   1 Y-linked loci
   9 sex-biased loci
   66 X-linked loci
   1 XY gametologs.
And 923 autosomal loci.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-7-4.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/task.png" class="class" width="48" height="48"></p>
<p>How many males and females does the dataset contain?</p>
<p>How many sex-linked loci were found?</p>
</div>
</div>
<p>Now check the output:</p>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-8_05c50798ed978c1a3bb1b7fa754a9cc5">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>y.linked      <span class="co"># Notice that it says "y-linked"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 376 genotypes,  1 SNPs , size: 4.7 Mb

    missing data: 164 (=43.62 %) scored as NA

 ** Genetic data
   @gen: list of 376 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  376 individual labels
   @loc.names:  1 locus labels
   @loc.all:  1 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 95-281)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, sex, pop, Year.collected, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Possum_v2, ChromPos_Possum_v2, AlnCnt_Possum_v2, AlnEvalue_Possum_v2, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, monomorphs, maf, OneRatio, PIC 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>x.linked      <span class="co"># Notice that it says "x-linked"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 376 genotypes,  66 SNPs , size: 4.7 Mb

    missing data: 827 (=3.33 %) scored as NA

 ** Genetic data
   @gen: list of 376 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  376 individual labels
   @loc.names:  66 locus labels
   @loc.all:  66 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 95-281)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, sex, pop, Year.collected, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Possum_v2, ChromPos_Possum_v2, AlnCnt_Possum_v2, AlnEvalue_Possum_v2, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, monomorphs, maf, OneRatio, PIC 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>gametolog</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 376 genotypes,  1 SNPs , size: 4.7 Mb

    missing data: 0 (=0 %) scored as NA

 ** Genetic data
   @gen: list of 376 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  376 individual labels
   @loc.names:  1 locus labels
   @loc.all:  1 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 95-281)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, sex, pop, Year.collected, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Possum_v2, ChromPos_Possum_v2, AlnCnt_Possum_v2, AlnEvalue_Possum_v2, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, monomorphs, maf, OneRatio, PIC 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>sex.biased</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 376 genotypes,  9 SNPs , size: 4.7 Mb

    missing data: 853 (=25.21 %) scored as NA

 ** Genetic data
   @gen: list of 376 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  376 individual labels
   @loc.names:  9 locus labels
   @loc.all:  9 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 95-281)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, sex, pop, Year.collected, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Possum_v2, ChromPos_Possum_v2, AlnCnt_Possum_v2, AlnEvalue_Possum_v2, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, monomorphs, maf, OneRatio, PIC 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>autosomal</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> ********************
 *** DARTR OBJECT ***
 ********************

 ** 376 genotypes,  923 SNPs , size: 5.2 Mb

    missing data: 18826 (=5.42 %) scored as NA

 ** Genetic data
   @gen: list of 376 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 ** Additional data
   @ind.names:  376 individual labels
   @loc.names:  923 locus labels
   @loc.all:  923 allele labels
   @position: integer storing positions of the SNPs [within 69 base sequence]
   @pop: population of each individual (group size range: 95-281)
   @other: a list containing: loc.metrics, ind.metrics, loc.metrics.flags, verbose, history 
    @other$ind.metrics: id, sex, pop, Year.collected, service, plate_location 
    @other$loc.metrics: AlleleID, CloneID, AlleleSequence, TrimmedSequence, Chrom_Possum_v2, ChromPos_Possum_v2, AlnCnt_Possum_v2, AlnEvalue_Possum_v2, SNP, SnpPosition, CallRate, OneRatioRef, OneRatioSnp, FreqHomRef, FreqHomSnp, FreqHets, PICRef, PICSnp, AvgPIC, AvgCountRef, AvgCountSnp, RepAvg, clone, uid, rdepth, monomorphs, maf, OneRatio, PIC 
   @other$latlon[g]: no coordinates attached</code></pre>
</div>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(res<span class="sc">$</span>results.table))  <span class="co"># The output table</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 5%">
<col style="width: 2%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 3%">
<col style="width: 4%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 3%">
<col style="width: 4%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">index</th>
<th style="text-align: right;">count.F.miss</th>
<th style="text-align: right;">count.M.miss</th>
<th style="text-align: right;">count.F.scored</th>
<th style="text-align: right;">count.M.scored</th>
<th style="text-align: right;">ratio</th>
<th style="text-align: right;">p.value</th>
<th style="text-align: right;">p.adjusted</th>
<th style="text-align: right;">scoringRate.F</th>
<th style="text-align: right;">scoringRate.M</th>
<th style="text-align: left;">y.linked</th>
<th style="text-align: left;">sex.biased</th>
<th style="text-align: right;">count.F.het</th>
<th style="text-align: right;">count.M.het</th>
<th style="text-align: right;">count.F.hom</th>
<th style="text-align: right;">count.M.hom</th>
<th style="text-align: right;">stat</th>
<th style="text-align: right;">stat.p.value</th>
<th style="text-align: right;">stat.p.adjusted</th>
<th style="text-align: right;">heterozygosity.F</th>
<th style="text-align: right;">heterozygosity.M</th>
<th style="text-align: left;">x.linked</th>
<th style="text-align: left;">xy.gametolog</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">28681424-34-G/T</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">162</td>
<td style="text-align: right;">210</td>
<td style="text-align: right;">1.2953770</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.9952607</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">161</td>
<td style="text-align: right;">210</td>
<td style="text-align: right;">1.303397</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.0061728</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
</tr>
<tr class="even">
<td style="text-align: left;">28678947-56-C/T</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">12</td>
<td style="text-align: right;">8</td>
<td style="text-align: right;">150</td>
<td style="text-align: right;">203</td>
<td style="text-align: right;">2.0261283</td>
<td style="text-align: right;">0.1638428</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0.9259259</td>
<td style="text-align: right;">0.9620853</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">141</td>
<td style="text-align: right;">196</td>
<td style="text-align: right;">1.784196</td>
<td style="text-align: right;">0.3044519</td>
<td style="text-align: right;">0.9598961</td>
<td style="text-align: right;">0.0600000</td>
<td style="text-align: right;">0.0344828</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
</tr>
<tr class="odd">
<td style="text-align: left;">28680567-32-T/G</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">12</td>
<td style="text-align: right;">12</td>
<td style="text-align: right;">150</td>
<td style="text-align: right;">199</td>
<td style="text-align: right;">1.3256351</td>
<td style="text-align: right;">0.5289429</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0.9259259</td>
<td style="text-align: right;">0.9431280</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">141</td>
<td style="text-align: right;">188</td>
<td style="text-align: right;">1.090635</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.0600000</td>
<td style="text-align: right;">0.0552764</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
</tr>
<tr class="even">
<td style="text-align: left;">28688313-7-C/G</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">162</td>
<td style="text-align: right;">211</td>
<td style="text-align: right;">1.3015303</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">156</td>
<td style="text-align: right;">211</td>
<td style="text-align: right;">8.076192</td>
<td style="text-align: right;">0.0459068</td>
<td style="text-align: right;">0.3917911</td>
<td style="text-align: right;">0.0370370</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
</tr>
<tr class="odd">
<td style="text-align: left;">28681679-51-C/T</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">22</td>
<td style="text-align: right;">30</td>
<td style="text-align: right;">140</td>
<td style="text-align: right;">181</td>
<td style="text-align: right;">0.9482168</td>
<td style="text-align: right;">0.8814171</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0.8641975</td>
<td style="text-align: right;">0.8578199</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">139</td>
<td style="text-align: right;">180</td>
<td style="text-align: right;">1.293900</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.0071429</td>
<td style="text-align: right;">0.0055249</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
</tr>
<tr class="even">
<td style="text-align: left;">28681994-14-G/A</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">162</td>
<td style="text-align: right;">210</td>
<td style="text-align: right;">1.2953770</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.9952607</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: right;">18</td>
<td style="text-align: right;">19</td>
<td style="text-align: right;">144</td>
<td style="text-align: right;">191</td>
<td style="text-align: right;">1.255791</td>
<td style="text-align: right;">0.6007790</td>
<td style="text-align: right;">1.0000000</td>
<td style="text-align: right;">0.1111111</td>
<td style="text-align: right;">0.0904762</td>
<td style="text-align: left;">FALSE</td>
<td style="text-align: left;">FALSE</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The output consists of a genlight object for each type of loci, plus a results table.</p>
</section>
<section id="run-gl.infer.sex-1" class="level3">
<h3 class="anchored" data-anchor-id="run-gl.infer.sex-1">Run <code>gl.infer.sex</code></h3>
<p>This function uses the output of function <code>gl.filter.sexlinked</code> (list of 6 objects) to infer the sex of all individuals in the dataset. It uses 3 types of sex-linked loci (W-/Y-linked, Z-/X-linked, and gametologs), assigns a preliminary genetic sex for each type of sex-linked loci available, and outputs an <code>agreed sex</code>.</p>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-9_c77a5bb8d45452f45164e3535b632197">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>sexID <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">gl.infer.sex</span>(<span class="at">gl_sex_filtered =</span> res, <span class="at">system =</span> <span class="st">"xy"</span>, <span class="at">seed =</span> <span class="dv">124</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Not enough gametologs (need at least 5). Assigning NA...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>***FINISHED***</code></pre>
</div>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(sexID))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 12%">
<col style="width: 7%">
<col style="width: 8%">
<col style="width: 12%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 13%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 10%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">id</th>
<th style="text-align: left;">y.linked.sex</th>
<th style="text-align: right;">#called</th>
<th style="text-align: right;">#missing</th>
<th style="text-align: left;">x.linked.sex</th>
<th style="text-align: right;">#Het.x</th>
<th style="text-align: right;">#Hom.x</th>
<th style="text-align: left;">gametolog.sex</th>
<th style="text-align: left;">#Het.g</th>
<th style="text-align: left;">#Hom.g</th>
<th style="text-align: left;">agreed.sex</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Y2</td>
<td style="text-align: left;">Y2</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">19</td>
<td style="text-align: right;">47</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="even">
<td style="text-align: left;">Y16</td>
<td style="text-align: left;">Y16</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">56</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Y17</td>
<td style="text-align: left;">Y17</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">27</td>
<td style="text-align: right;">37</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="even">
<td style="text-align: left;">Y18</td>
<td style="text-align: left;">Y18</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">62</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">*F</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Y3</td>
<td style="text-align: left;">Y3</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">63</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">Y4</td>
<td style="text-align: left;">Y4</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">63</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>What did the function mention about gametologs? How did that impact the results table? How many types of sex-linked loci were used to infer an <code>agreed sex</code>?</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/task.png" class="class" width="48" height="48"></p>
<p>Can you find individuals for which the <code>agreed sex</code> is uncertain (i.e., has an asterisk “*”)?</p>
</div>
</div>
</section>
</section>
<section id="exercise-4" class="level2">
<h2 class="anchored" data-anchor-id="exercise-4">EXERCISE</h2>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/task.png" class="class" width="48" height="48"></p>
<p>Imagine you are analyzing the genetic data of a population and you want to look at its genetic structure and genetic diversity. You get to work by filtering your beautiful set of SNPs. Because you are a rigorous, you want to test whether using function <code>gl.filter.sexlinked</code> to properly filter out sex-linked loci changes the results of the populations analyses. So you: (1) process your SNP dataset by applying standard filters and run analyses, and (2) process the dataset by filtering out sex-linked loci first, applying standard filters second, and then run analyses.</p>
<p>Choose one of the following datasets (or your own data) and report:</p>
<ol type="1">
<li><p>How many sex-linked markers are present?</p></li>
<li><p>How many individuals had a wrong sexID?</p></li>
<li><p>Do you see any changes in your PCA/structure analyses when you filtered out sex-linked markers versus when you did not?</p></li>
<li><p>Do you see any differences in genetic diversity and fixation indices when you filtered out sex-linked markers versus when you did not?</p></li>
</ol>
</div>
</div>
<p><img src="images/beer.png" class="class" width="350" height="200"> <img src="images/Eastern_Yellow_Robin.jpg" class="img-fluid" alt="The Eastern Yellow Robin" width="349"></p>
<p><img src="images/Bull_Shark.jpg" class="class" width="350" height="200"> <img src="images/Blue_shark.jpg" class="class" width="350" height="200"></p>
</section>
<section id="further-study" class="level2">
<h2 class="anchored" data-anchor-id="further-study"><em>Further study</em></h2>
<section id="exercise-data-1---your-own-data" class="level3">
<h3 class="anchored" data-anchor-id="exercise-data-1---your-own-data">Exercise data 1 - Your own data</h3>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<strong>HINT</strong>
</div>
</div>
<div class="callout-body-container callout-body">
<p>You can have a look at the exercise data below for inspiration.</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise - Your own data
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li><p>Number of sex-linked markers?</p></li>
<li><p>Individuals with wrong sexID?</p></li>
<li><p>Changes in PCA before and after removing the SLM?</p></li>
<li><p>Differences in genetic diversity and fixation indices between autosomal and SLM?</p></li>
</ol>
</div>
</div>
</section>
<section id="exercise-data-2---the-eastern-yellow-robin" class="level3">
<h3 class="anchored" data-anchor-id="exercise-data-2---the-eastern-yellow-robin">Exercise data 2 - The Eastern Yellow Robin</h3>
<p>Data from <span class="citation" data-cites="robledo2023">Robledo-Ruiz et al. (<a href="references.html#ref-robledo2023" role="doc-biblioref">2023</a>)</span></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Eastern_Yellow_Robin.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">The Eastern Yellow Robin</figcaption>
</figure>
</div>
<section id="load-data-2" class="level4">
<h4 class="anchored" data-anchor-id="load-data-2">Load data</h4>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-10_bc22c575cba2eaa973b59e755a68eb31">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="co">#data("EYR")</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">'./data/EYR.rda'</span>)</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>EYR<span class="sc">@</span>n.loc</span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(EYR<span class="sc">@</span>pop)</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(EYR<span class="sc">@</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>pop)</span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(EYR<span class="sc">@</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>sex, <span class="at">useNA =</span> <span class="st">"ifany"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1000

    Crusoe Muckleford      Timor     Wombat 
       238        421         52         71 

    Crusoe Muckleford      Timor     Wombat 
       238        421         52         71 

      F   M 
  1 352 429 </code></pre>
</div>
</div>
</section>
<section id="number-of-sex-linked-markers" class="level4">
<h4 class="anchored" data-anchor-id="number-of-sex-linked-markers">1. Number of sex-linked markers?</h4>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-11_d5a4e815c87a1461a4cb6a1c592f1536">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">gl.filter.sexlinked</span>(<span class="at">gl =</span> EYR, <span class="at">system =</span> <span class="st">"zw"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Detected 352 females and 429 males.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Starting phase 1. May take a while...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building call rate plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done. Starting phase 2.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building heterozygosity plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-11-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-11-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done building heterozygosity plots.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**FINISHED** Total of analyzed loci: 1000.
Found 150 sex-linked loci:
   16 W-linked loci
   82 sex-biased loci
   32 Z-linked loci
   20 ZW gametologs.
And 850 autosomal loci.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-11-4.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="individuals-with-wrong-sexid" class="level4">
<h4 class="anchored" data-anchor-id="individuals-with-wrong-sexid">2. Individuals with wrong sexID?</h4>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-12_b4361d5fb8c14bda204eeb15174f0a44">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>sexID <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">gl.infer.sex</span>(<span class="at">gl_sex_filtered =</span> res, <span class="at">system =</span> <span class="st">"zw"</span>, <span class="at">seed =</span> <span class="dv">124</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>***FINISHED***</code></pre>
</div>
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(sexID))</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(EYR<span class="sc">$</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>sex <span class="sc">!=</span> sexID<span class="sc">$</span>agreed.sex, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 11%">
<col style="width: 7%">
<col style="width: 6%">
<col style="width: 11%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 11%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 9%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">id</th>
<th style="text-align: left;">w.linked.sex</th>
<th style="text-align: right;">#missing</th>
<th style="text-align: right;">#called</th>
<th style="text-align: left;">z.linked.sex</th>
<th style="text-align: right;">#Hom.z</th>
<th style="text-align: right;">#Het.z</th>
<th style="text-align: left;">gametolog.sex</th>
<th style="text-align: right;">#Hom.g</th>
<th style="text-align: right;">#Het.g</th>
<th style="text-align: left;">agreed.sex</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">024-96401</td>
<td style="text-align: left;">024-96401</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">16</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">25</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">024-96401b</td>
<td style="text-align: left;">024-96401b</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">16</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">21</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">024-96402</td>
<td style="text-align: left;">024-96402</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">15</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">32</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="even">
<td style="text-align: left;">024-96403</td>
<td style="text-align: left;">024-96403</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">15</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">21</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">024-96404</td>
<td style="text-align: left;">024-96404</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">16</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">12</td>
<td style="text-align: right;">20</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">024-96405</td>
<td style="text-align: left;">024-96405</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">16</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">21</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5</td>
<td style="text-align: left;">M</td>
</tr>
</tbody>
</table>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 55</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/task.png" class="class" width="48" height="48"></p>
<p>Can you tell which misidentified sexes are due to uncertain genetic sex (indicated with *)?</p>
<p><strong>HINT</strong> Try using <code>grep(pattern = "\\*", x = sexID$agreed.sex)</code></p>
</div>
</div>
</section>
<section id="processing-snps-with-two-filtering-regimes" class="level4">
<h4 class="anchored" data-anchor-id="processing-snps-with-two-filtering-regimes">Processing SNPs with two filtering regimes</h4>
<section id="filtering-snps-only-with-standard-filters-sloppy" class="level5">
<h5 class="anchored" data-anchor-id="filtering-snps-only-with-standard-filters-sloppy">Filtering SNPs only with standard filters (sloppy)</h5>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-13_8f6a9ac31a7d65163d433f83bcbf4fb1">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter for read depth</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.rdepth</span>(EYR)  <span class="co"># This is the initial dataset</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.rdepth 
  Processing genlight object with SNP data
  Reporting Read Depth by Locus
  No. of loci = 1000 
  No. of individuals = 782 
    Minimum      :  2.6 
    1st quartile :  4.3 
    Median       :  5.6 
    Mean         :  5.9649 
    3r quartile  :  7.325 
    Maximum      :  13.2 
    Missing Rate Overall:  0.19 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>   Quantile Threshold Retained Percent Filtered Percent
1      100%      13.2        1     0.1      999    99.9
2       95%       9.9       51     5.1      949    94.9
3       90%       9.0      105    10.5      895    89.5
4       85%       8.3      151    15.1      849    84.9
5       80%       7.8      208    20.8      792    79.2
6       75%       7.3      258    25.8      742    74.2
7       70%       6.9      304    30.4      696    69.6
8       65%       6.5      354    35.4      646    64.6
9       60%       6.2      404    40.4      596    59.6
10      55%       5.9      451    45.1      549    54.9
11      50%       5.6      504    50.4      496    49.6
12      45%       5.3      563    56.3      437    43.7
13      40%       5.1      602    60.2      398    39.8
14      35%       4.8      659    65.9      341    34.1
15      30%       4.6      702    70.2      298    29.8
16      25%       4.3      752    75.2      248    24.8
17      20%       4.0      823    82.3      177    17.7
18      15%       3.9      852    85.2      148    14.8
19      10%       3.6      906    90.6       94     9.4
20       5%       3.3      956    95.6       44     4.4
21       0%       2.6     1000   100.0        0     0.0
Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.rdepth </code></pre>
</div>
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>EYR.sloppy <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.rdepth</span>(EYR, <span class="at">lower =</span> <span class="dv">3</span>, <span class="at">upper =</span> <span class="dv">11</span>, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-13-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter for loci call rate</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.callrate</span>(EYR.sloppy, <span class="at">method =</span> <span class="st">"loc"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.callrate 
  Processing genlight object with SNP data
  Reporting Call Rate by Locus
  No. of loci = 958 
  No. of individuals = 782 
    Minimum      :  0.20844 
    1st quartile :  0.7202688 
    Median       :  0.895141 
    Mean         :  0.8131871 
    3r quartile  :  0.950128 
    Maximum      :  0.988491 
    Missing Rate Overall:  0.1868 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-13-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.callrate </code></pre>
</div>
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>EYR.sloppy <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.callrate</span>(EYR.sloppy, <span class="at">method =</span> <span class="st">"loc"</span>,  <span class="at">threshold =</span> <span class="fl">0.75</span>, <span class="at">verbose =</span> <span class="dv">0</span>, <span class="at">recalc =</span> <span class="cn">TRUE</span>)</span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter for individual call rate</span></span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.callrate</span>(EYR.sloppy, <span class="at">method =</span> <span class="st">"ind"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.callrate 
  Processing genlight object with SNP data

  Reporting Call Rate by Individual
  No. of loci = 703 
  No. of individuals = 782 
    Minimum      :  0.03556188 
    1st quartile :  0.9174964 
    Median       :  0.9416785 
    Mean         :  0.9108097 
    3r quartile  :  0.9573257 
    Maximum      :  0.9829303 
    Missing Rate Overall:  0.0892 

Listing 4 populations and their average CallRates
  Monitor again after filtering
  Population CallRate   N
1     Crusoe   0.9027 238
2 Muckleford   0.9073 421
3      Timor   0.9402  52
4     Wombat   0.9371  71

Listing 20 individuals with the lowest CallRates
  Use this list to see which individuals will be lost on filtering by individual
  Set ind.to.list parameter to see more individuals
   Individual   CallRate
1    M18.29.1 0.03556188
2    M18.18.1 0.03982930
3    M18.47.2 0.06970128
4    C18.16.1 0.07112376
5   027-34168 0.07681366
6    C18.15.2 0.08534851
7    C18.21.2 0.08677098
8    M18.47.3 0.14224751
9    M18.35.2 0.17211949
10   M18.20.3 0.24039829
11   M20.70.2 0.39687055
12   C18.28.1 0.39971550
13   C18.17.2 0.46088193
14  027-34065 0.50640114
15   C18.14.1 0.50640114
16   M20.70.3 0.50782361
17  M20.110.1 0.52347084
18   M19.12.1 0.53342817
19    M19.8.1 0.54907539
20   M20.64.3 0.56045519

)</code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-13-4.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.callrate </code></pre>
</div>
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>EYR.sloppy <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.callrate</span>(EYR.sloppy, <span class="at">method =</span> <span class="st">"ind"</span>, <span class="at">threshold =</span> <span class="fl">0.65</span>, <span class="at">verbose =</span> <span class="dv">0</span>, <span class="at">recalc =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter for MAC (= 3)</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.maf</span>(EYR.sloppy)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.maf 
  Processing genlight object with SNP data
Starting :: 

 Starting dartR.base 

 Starting gl.report.maf 

  Reporting Minor Allele Frequency (MAF) by Locus for population Crusoe 
  No. of loci = 670 
  No. of individuals = 231 
    Minimum      :  0.0022 
    1st quantile :  0.064825 
    Median       :  0.1582 
    Mean         :  0.1793525 
    3r quantile  :  0.267475 
    Maximum      :  0.4975 
    Missing Rate Overall:  0.08 

  Reporting Minor Allele Frequency (MAF) by Locus for population Muckleford 
  No. of loci = 683 
  No. of individuals = 401 
    Minimum      :  0.0013 
    1st quantile :  0.05875 
    Median       :  0.1404 
    Mean         :  0.1729489 
    3r quantile  :  0.2617 
    Maximum      :  0.4985 
    Missing Rate Overall:  0.07 

  Reporting Minor Allele Frequency (MAF) by Locus for population Timor 
  No. of loci = 589 
  No. of individuals = 52 
    Minimum      :  0.0096 
    1st quantile :  0.0673 
    Median       :  0.1667 
    Mean         :  0.1914129 
    3r quantile  :  0.2872 
    Maximum      :  0.5 
    Missing Rate Overall:  0.06 

  Reporting Minor Allele Frequency (MAF) by Locus for population Wombat 
  No. of loci = 627 
  No. of individuals = 71 
    Minimum      :  0.007 
    1st quantile :  0.06385 
    Median       :  0.1449 
    Mean         :  0.1746703 
    3r quantile  :  0.2542 
    Maximum      :  0.5 
    Missing Rate Overall:  0.06 

  Reporting Minor Allele Frequency (MAF) by Locus OVERALL
  No. of loci = 703 
  No. of individuals = 755 
    Minimum      :  7e-04 
    1st quantile :  0.0618 
    Median       :  0.1443 
    Mean         :  0.1711869 
    3r quantile  :  0.254425 
    Maximum      :  0.5 
    Missing Rate Overall:  0.07 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-13-5.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>   Quantile Threshold Retained Percent Filtered Percent
1      100%    0.5000        1     0.1      699    99.9
2       95%    0.4431       36     5.1      664    94.9
3       90%    0.3812       71    10.1      629    89.9
4       85%    0.3346      105    15.0      595    85.0
5       80%    0.2873      141    20.1      559    79.9
6       75%    0.2537      176    25.1      524    74.9
7       70%    0.2279      210    30.0      490    70.0
8       65%    0.2009      246    35.1      454    64.9
9       60%    0.1779      280    40.0      420    60.0
10      55%    0.1595      315    45.0      385    55.0
11      50%    0.1442      351    50.1      349    49.9
12      45%    0.1210      386    55.1      314    44.9
13      40%    0.1024      422    60.3      278    39.7
14      35%    0.0872      455    65.0      245    35.0
15      30%    0.0755      490    70.0      210    30.0
16      25%    0.0618      526    75.1      174    24.9
17      20%    0.0487      561    80.1      139    19.9
18      15%    0.0346      595    85.0      105    15.0
19      10%    0.0227      631    90.1       69     9.9
20       5%    0.0053      666    95.1       34     4.9
21       0%    0.0007      700   100.0        0     0.0
Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.maf </code></pre>
</div>
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>EYR.sloppy <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.maf</span>(EYR.sloppy, <span class="at">threshold =</span> <span class="dv">3</span>, <span class="at">verbose =</span> <span class="dv">0</span>, <span class="at">recalc =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.select.colors 
  Warning: Number of required colors not specified, set to 9
  Library: RColorBrewer
  Palette: brewer.pal
  Showing and returning 2 of 9 colors for library RColorBrewer : palette Blues </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-13-6.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.select.colors </code></pre>
</div>
</div>
</section>
<section id="filtering-snps-with-gl.filter.sexlinked-and-standard-filters-correct" class="level5">
<h5 class="anchored" data-anchor-id="filtering-snps-with-gl.filter.sexlinked-and-standard-filters-correct">Filtering SNPs with gl.filter.sexlinked and standard filters (correct)</h5>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-14_a09146a0b639781275fc6daf35090f0a">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter for sex-linked loci</span></span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a>correct <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">gl.filter.sexlinked</span>(EYR, <span class="at">system =</span> <span class="st">"zw"</span>)  <span class="co"># This is the initial dataset</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-14-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-14-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-14-4.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We will use correct$autosomal for the next filters</span></span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter for read depth</span></span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.rdepth</span>(correct<span class="sc">$</span>autosomal)  <span class="co"># This is the filtered dataset</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.rdepth 
  Processing genlight object with SNP data
  Reporting Read Depth by Locus
  No. of loci = 850 
  No. of individuals = 782 
    Minimum      :  2.6 
    1st quartile :  4.3 
    Median       :  5.6 
    Mean         :  6.008941 
    3r quartile  :  7.4 
    Maximum      :  13.2 
    Missing Rate Overall:  0.18 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-14-5.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>   Quantile Threshold Retained Percent Filtered Percent
1      100%      13.2        1     0.1      849    99.9
2       95%       9.9       45     5.3      805    94.7
3       90%       9.1       88    10.4      762    89.6
4       85%       8.4      129    15.2      721    84.8
5       80%       7.9      173    20.4      677    79.6
6       75%       7.4      220    25.9      630    74.1
7       70%       6.9      264    31.1      586    68.9
8       65%       6.5      305    35.9      545    64.1
9       60%       6.2      350    41.2      500    58.8
10      55%       5.9      391    46.0      459    54.0
11      50%       5.6      435    51.2      415    48.8
12      45%       5.4      472    55.5      378    44.5
13      40%       5.1      519    61.1      331    38.9
14      35%       4.9      555    65.3      295    34.7
15      30%       4.6      603    70.9      247    29.1
16      25%       4.3      645    75.9      205    24.1
17      20%       4.0      705    82.9      145    17.1
18      15%       3.9      730    85.9      120    14.1
19      10%       3.6      774    91.1       76     8.9
20       5%       3.3      812    95.5       38     4.5
21       0%       2.6      850   100.0        0     0.0
Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.rdepth </code></pre>
</div>
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a>EYR.correct <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.rdepth</span>(correct<span class="sc">$</span>autosomal, <span class="at">lower =</span> <span class="dv">3</span>, <span class="at">upper =</span> <span class="dv">11</span>, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-14-6.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter for loci call rate</span></span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.callrate</span>(EYR.correct, <span class="at">method =</span> <span class="st">"loc"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.callrate 
  Processing genlight object with SNP data
  Reporting Call Rate by Locus
  No. of loci = 811 
  No. of individuals = 782 
    Minimum      :  0.20844 
    1st quartile :  0.7436065 
    Median       :  0.900256 
    Mean         :  0.8192658 
    3r quartile  :  0.951407 
    Maximum      :  0.988491 
    Missing Rate Overall:  0.1807 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-14-7.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.callrate </code></pre>
</div>
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>EYR.correct <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.callrate</span>(EYR.correct, <span class="at">method =</span> <span class="st">"loc"</span>,  <span class="at">threshold =</span> <span class="fl">0.75</span>, <span class="at">verbose =</span> <span class="dv">0</span>, <span class="at">recalc =</span> <span class="cn">TRUE</span>)</span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter for individual call rate</span></span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.callrate</span>(EYR.correct, <span class="at">method =</span> <span class="st">"ind"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.callrate 
  Processing genlight object with SNP data

  Reporting Call Rate by Individual
  No. of loci = 605 
  No. of individuals = 782 
    Minimum      :  0.03801653 
    1st quartile :  0.9173554 
    Median       :  0.9438017 
    Mean         :  0.9120479 
    3r quartile  :  0.9586777 
    Maximum      :  0.9818182 
    Missing Rate Overall:  0.088 

Listing 4 populations and their average CallRates
  Monitor again after filtering
  Population CallRate   N
1     Crusoe   0.9037 238
2 Muckleford   0.9090 421
3      Timor   0.9418  52
4     Wombat   0.9365  71

Listing 20 individuals with the lowest CallRates
  Use this list to see which individuals will be lost on filtering by individual
  Set ind.to.list parameter to see more individuals
   Individual   CallRate
1    M18.29.1 0.03801653
2    M18.18.1 0.04462810
3    M18.47.2 0.06776860
4    C18.16.1 0.07438017
5   027-34168 0.08760331
6    C18.15.2 0.08760331
7    C18.21.2 0.08760331
8    M18.47.3 0.13719008
9    M18.35.2 0.18016529
10   M18.20.3 0.24132231
11   C18.28.1 0.41487603
12   M20.70.2 0.42644628
13   C18.17.2 0.44628099
14  027-34065 0.49586777
15   C18.14.1 0.49752066
16  M20.110.1 0.53057851
17   M20.70.3 0.53553719
18   M19.12.1 0.54380165
19    M19.8.1 0.56694215
20   M19.33.2 0.57851240

)</code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-14-8.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.callrate </code></pre>
</div>
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>EYR.correct <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.callrate</span>(EYR.correct, <span class="at">method =</span> <span class="st">"ind"</span>, <span class="at">threshold =</span> <span class="fl">0.65</span>, <span class="at">verbose =</span> <span class="dv">0</span>, <span class="at">recalc =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter for MAC (= 3)</span></span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.maf</span>(EYR.correct)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.maf 
  Processing genlight object with SNP data
Starting :: 

 Starting dartR.base 

 Starting gl.report.maf 

  Reporting Minor Allele Frequency (MAF) by Locus for population Crusoe 
  No. of loci = 573 
  No. of individuals = 231 
    Minimum      :  0.0022 
    1st quantile :  0.06 
    Median       :  0.1488 
    Mean         :  0.1741178 
    3r quantile  :  0.2646 
    Maximum      :  0.4975 
    Missing Rate Overall:  0.08 

  Reporting Minor Allele Frequency (MAF) by Locus for population Muckleford 
  No. of loci = 585 
  No. of individuals = 401 
    Minimum      :  0.0013 
    1st quantile :  0.055 
    Median       :  0.129 
    Mean         :  0.1639928 
    3r quantile  :  0.2474 
    Maximum      :  0.4985 
    Missing Rate Overall:  0.07 

  Reporting Minor Allele Frequency (MAF) by Locus for population Timor 
  No. of loci = 504 
  No. of individuals = 52 
    Minimum      :  0.0096 
    1st quantile :  0.068275 
    Median       :  0.1635 
    Mean         :  0.1898613 
    3r quantile  :  0.286075 
    Maximum      :  0.5 
    Missing Rate Overall:  0.06 

  Reporting Minor Allele Frequency (MAF) by Locus for population Wombat 
  No. of loci = 536 
  No. of individuals = 71 
    Minimum      :  0.007 
    1st quantile :  0.062475 
    Median       :  0.13805 
    Mean         :  0.1706063 
    3r quantile  :  0.2509 
    Maximum      :  0.5 
    Missing Rate Overall:  0.06 

  Reporting Minor Allele Frequency (MAF) by Locus OVERALL
  No. of loci = 605 
  No. of individuals = 755 
    Minimum      :  7e-04 
    1st quantile :  0.05815 
    Median       :  0.13255 
    Mean         :  0.1628168 
    3r quantile  :  0.243425 
    Maximum      :  0.5 
    Missing Rate Overall:  0.07 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-14-9.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>   Quantile Threshold Retained Percent Filtered Percent
1      100%    0.5000        1     0.2      601    99.8
2       95%    0.4364       31     5.1      571    94.9
3       90%    0.3714       61    10.1      541    89.9
4       85%    0.3207       91    15.1      511    84.9
5       80%    0.2734      121    20.1      481    79.9
6       75%    0.2436      151    25.1      451    74.9
7       70%    0.2127      181    30.1      421    69.9
8       65%    0.1901      211    35.0      391    65.0
9       60%    0.1714      241    40.0      361    60.0
10      55%    0.1499      271    45.0      331    55.0
11      50%    0.1320      302    50.2      300    49.8
12      45%    0.1092      332    55.1      270    44.9
13      40%    0.0950      362    60.1      240    39.9
14      35%    0.0806      393    65.3      209    34.7
15      30%    0.0710      422    70.1      180    29.9
16      25%    0.0580      452    75.1      150    24.9
17      20%    0.0447      482    80.1      120    19.9
18      15%    0.0305      512    85.0       90    15.0
19      10%    0.0194      542    90.0       60    10.0
20       5%    0.0048      572    95.0       30     5.0
21       0%    0.0007      602   100.0        0     0.0
Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.maf </code></pre>
</div>
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>EYR.correct <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.maf</span>(EYR.correct, <span class="at">threshold =</span> <span class="dv">3</span>, <span class="at">verbose =</span> <span class="dv">0</span>, <span class="at">recalc =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.select.colors 
  Warning: Number of required colors not specified, set to 9
  Library: RColorBrewer
  Palette: brewer.pal
  Showing and returning 2 of 9 colors for library RColorBrewer : palette Blues </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-14-10.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.select.colors </code></pre>
</div>
</div>
</section>
</section>
<section id="changes-in-pca-before-and-after-removing-the-slm" class="level4">
<h4 class="anchored" data-anchor-id="changes-in-pca-before-and-after-removing-the-slm">3. Changes in PCA before and after removing the SLM?</h4>
<section id="pca-on-sloppy-dataset-only-standard-filters" class="level5">
<h5 class="anchored" data-anchor-id="pca-on-sloppy-dataset-only-standard-filters">PCA on sloppy dataset (only standard filters)</h5>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-15_0316e996a23f7468d90995b45fdd9829">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a>PCA.sloppy <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.pcoa</span>(EYR.sloppy, <span class="at">verbose =</span> <span class="dv">0</span>)</span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.pcoa.plot</span>(PCA.sloppy, EYR.sloppy, <span class="at">xaxis =</span> <span class="dv">1</span>, <span class="at">yaxis =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-15-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.pcoa.plot</span>(PCA.sloppy, EYR.sloppy, <span class="at">xaxis =</span> <span class="dv">2</span>, <span class="at">yaxis =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-15-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-15-4.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.colors 
Selected color type 2 
Completed: gl.colors 
Starting :: 
 Starting dartR.base 
 Starting gl.pcoa.plot 
  Processing an ordination file (glPca)
  Processing genlight object with SNP data
  Plotting populations in a space defined by the SNPs
  Preparing plot .... please wait
Completed: :: 
 Completed: dartR.base 
 Completed: gl.pcoa.plot 
Starting :: 
 Starting dartR.base 
 Starting gl.pcoa.plot 
  Processing an ordination file (glPca)
  Processing genlight object with SNP data
  Plotting populations in a space defined by the SNPs
  Preparing plot .... please wait
Completed: :: 
 Completed: dartR.base 
 Completed: gl.pcoa.plot </code></pre>
</div>
</div>
</section>
<section id="pca-on-correct-dataset-gl.filter.sexlinked-and-standard-filters" class="level5">
<h5 class="anchored" data-anchor-id="pca-on-correct-dataset-gl.filter.sexlinked-and-standard-filters">PCA on correct dataset (gl.filter.sexlinked and standard filters)</h5>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-16_29e68826b66ab37c239a00934725d55c">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a>PCA.correct <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.pcoa</span>(EYR.correct, <span class="at">verbose =</span> <span class="dv">0</span>)</span>
<span id="cb113-2"><a href="#cb113-2" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.pcoa.plot</span>(PCA.correct, EYR.correct, <span class="at">xaxis =</span> <span class="dv">1</span>, <span class="at">yaxis =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-16-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.pcoa.plot</span>(PCA.correct, EYR.correct, <span class="at">xaxis =</span> <span class="dv">2</span>, <span class="at">yaxis =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-16-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-16-4.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.colors 
Selected color type 2 
Completed: gl.colors 
Starting :: 
 Starting dartR.base 
 Starting gl.pcoa.plot 
  Processing an ordination file (glPca)
  Processing genlight object with SNP data
  Plotting populations in a space defined by the SNPs
  Preparing plot .... please wait
Completed: :: 
 Completed: dartR.base 
 Completed: gl.pcoa.plot 
Starting :: 
 Starting dartR.base 
 Starting gl.pcoa.plot 
  Processing an ordination file (glPca)
  Processing genlight object with SNP data
  Plotting populations in a space defined by the SNPs
  Preparing plot .... please wait
Completed: :: 
 Completed: dartR.base 
 Completed: gl.pcoa.plot </code></pre>
</div>
</div>
</section>
</section>
<section id="differences-in-genetic-diversity-and-fixation-indices-between-autosomal-and-slm" class="level4">
<h4 class="anchored" data-anchor-id="differences-in-genetic-diversity-and-fixation-indices-between-autosomal-and-slm">4. Differences in genetic diversity and fixation indices between autosomal and SLM?</h4>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-17_f917077bcb554085c3c7998a803aa666">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Basic stats</span></span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a>basic.sloppy  <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">utils.basic.stats</span>(EYR.sloppy)</span>
<span id="cb116-3"><a href="#cb116-3" aria-hidden="true" tabindex="-1"></a>basic.correct <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">utils.basic.stats</span>(EYR.correct)</span>
<span id="cb116-4"><a href="#cb116-4" aria-hidden="true" tabindex="-1"></a>basic.sloppy<span class="sc">$</span>overall</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Ho      Hs      Ht     Dst     Htp    Dstp     Fst    Fstp     Fis    Dest 
 0.1638  0.2442  0.2537  0.0095  0.2568  0.0126  0.0373  0.0491  0.3294  0.0167 
Gst_max   Gst_H 
 0.7061  0.0696 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a>basic.correct<span class="sc">$</span>overall</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Ho      Hs      Ht     Dst     Htp    Dstp     Fst    Fstp     Fis    Dest 
 0.1631  0.2379  0.2459  0.0080  0.2485  0.0107  0.0325  0.0429  0.3144  0.0140 
Gst_max   Gst_H 
 0.7135  0.0601 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Genetic diversity per pop</span></span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a>divers.sloppy <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.report.diversity</span>(EYR.sloppy, <span class="at">pbar =</span> <span class="cn">FALSE</span>, <span class="at">table =</span> <span class="cn">FALSE</span>, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Processing genlight object with SNP data</code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a>divers.correct <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.report.diversity</span>(EYR.correct, <span class="at">pbar =</span> <span class="cn">FALSE</span>, <span class="at">table =</span> <span class="cn">FALSE</span>, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Processing genlight object with SNP data</code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-17-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a>divers.sloppy<span class="sc">$</span>one_H_alpha</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    Crusoe Muckleford      Timor     Wombat 
 0.3995681  0.3947475  0.3621317  0.3686690 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a>divers.correct<span class="sc">$</span>one_H_alpha</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    Crusoe Muckleford      Timor     Wombat 
 0.3896088  0.3810663  0.3594440  0.3619161 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a>divers.sloppy<span class="sc">$</span>one_H_beta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                Crusoe Muckleford      Timor     Wombat
Crusoe              NA 0.02730206 0.09476805 0.06283335
Muckleford 0.006464867         NA 0.07400643 0.06687160
Timor      0.022772763 0.02518305         NA 0.08882423
Wombat     0.019504127 0.02191442 0.03822232         NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a>divers.correct<span class="sc">$</span>one_H_beta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                Crusoe Muckleford      Timor     Wombat
Crusoe              NA 0.02594333 0.08764100 0.06061531
Muckleford 0.005966702         NA 0.07063921 0.06381958
Timor      0.016777813 0.02104909         NA 0.08446058
Wombat     0.015541765 0.01981305 0.03062416         NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Fixation indices</span></span>
<span id="cb132-2"><a href="#cb132-2" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.fst.pop</span>(EYR.sloppy, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               Crusoe Muckleford      Timor Wombat
Crusoe             NA         NA         NA     NA
Muckleford 0.03234709         NA         NA     NA
Timor      0.04072932 0.05451318         NA     NA
Wombat     0.05484958 0.02376727 0.08401313     NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.fst.pop</span>(EYR.correct, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               Crusoe Muckleford      Timor Wombat
Crusoe             NA         NA         NA     NA
Muckleford 0.02795813         NA         NA     NA
Timor      0.03987369 0.04810759         NA     NA
Wombat     0.04446378 0.02270137 0.07114017     NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.fstat</span>(EYR.sloppy, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.colors 
Selected color type div 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-17-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>$Stat_matrices
$Stat_matrices$Fst
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0164 0.0214 0.0289
Muckleford 0.0164         NA 0.0288 0.0124
Timor      0.0214     0.0288     NA 0.0439
Wombat     0.0289     0.0124 0.0439     NA

$Stat_matrices$Fstp
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0323 0.0420 0.0561
Muckleford 0.0323         NA 0.0559 0.0244
Timor      0.0420     0.0559     NA 0.0840
Wombat     0.0561     0.0244 0.0840     NA

$Stat_matrices$Dest
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0113 0.0143 0.0193
Muckleford 0.0113         NA 0.0190 0.0080
Timor      0.0143     0.0190     NA 0.0283
Wombat     0.0193     0.0080 0.0283     NA

$Stat_matrices$Gst_H
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0521 0.0665 0.0892
Muckleford 0.0521         NA 0.0883 0.0385
Timor      0.0665     0.0883     NA 0.1308
Wombat     0.0892     0.0385 0.1308     NA


[[2]]
      Stat_tables.Crusoe_vs_Muckleford Stat_tables.Crusoe_vs_Timor
Fst                             0.0164                      0.0214
Fstp                            0.0323                      0.0420
Dest                            0.0113                      0.0143
Gst_H                           0.0521                      0.0665
      Stat_tables.Crusoe_vs_Wombat Stat_tables.Muckleford_vs_Timor
Fst                         0.0289                          0.0288
Fstp                        0.0561                          0.0559
Dest                        0.0193                          0.0190
Gst_H                       0.0892                          0.0883
      Stat_tables.Muckleford_vs_Wombat Stat_tables.Timor_vs_Wombat
Fst                             0.0124                      0.0439
Fstp                            0.0244                      0.0840
Dest                            0.0080                      0.0283
Gst_H                           0.0385                      0.1308</code></pre>
</div>
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.fstat</span>(EYR.correct, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.colors 
Selected color type div 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-17-4.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>$Stat_matrices
$Stat_matrices$Fst
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0141 0.0208 0.0232
Muckleford 0.0141         NA 0.0249 0.0117
Timor      0.0208     0.0249     NA 0.0368
Wombat     0.0232     0.0117 0.0368     NA

$Stat_matrices$Fstp
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0278 0.0408 0.0454
Muckleford 0.0278         NA 0.0487 0.0231
Timor      0.0408     0.0487     NA 0.0711
Wombat     0.0454     0.0231 0.0711     NA

$Stat_matrices$Dest
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0092 0.0135 0.0149
Muckleford 0.0092         NA 0.0159 0.0073
Timor      0.0135     0.0159     NA 0.0231
Wombat     0.0149     0.0073 0.0231     NA

$Stat_matrices$Gst_H
           Crusoe Muckleford  Timor Wombat
Crusoe         NA     0.0441 0.0640 0.0713
Muckleford 0.0441         NA 0.0758 0.0359
Timor      0.0640     0.0758     NA 0.1097
Wombat     0.0713     0.0359 0.1097     NA


[[2]]
      Stat_tables.Crusoe_vs_Muckleford Stat_tables.Crusoe_vs_Timor
Fst                             0.0141                      0.0208
Fstp                            0.0278                      0.0408
Dest                            0.0092                      0.0135
Gst_H                           0.0441                      0.0640
      Stat_tables.Crusoe_vs_Wombat Stat_tables.Muckleford_vs_Timor
Fst                         0.0232                          0.0249
Fstp                        0.0454                          0.0487
Dest                        0.0149                          0.0159
Gst_H                       0.0713                          0.0758
      Stat_tables.Muckleford_vs_Wombat Stat_tables.Timor_vs_Wombat
Fst                             0.0117                      0.0368
Fstp                            0.0231                      0.0711
Dest                            0.0073                      0.0231
Gst_H                           0.0359                      0.1097</code></pre>
</div>
</div>
</section>
</section>
<section id="exercise-data-3---bull-shark" class="level3">
<h3 class="anchored" data-anchor-id="exercise-data-3---bull-shark">Exercise data 3 - Bull shark</h3>
<p>Data from <span class="citation" data-cites="devloodelva_rivers_2023">Devloo‐Delva et al. (<a href="references.html#ref-devloodelva_rivers_2023" role="doc-biblioref">2023</a>)</span>.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Bull_Shark.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">The Bull Shark</figcaption>
</figure>
</div>
<section id="load-data-3" class="level4">
<h4 class="anchored" data-anchor-id="load-data-3">Load data</h4>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-18_3d6b73db2791994533e0bf1a6c380b93">
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">load</span>(<span class="st">"data/Bull_shark_DArTseq_genlight_for_sex-linked_markers.Rdata"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data.gl"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb144"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb144-1"><a href="#cb144-1" aria-hidden="true" tabindex="-1"></a>data.gl<span class="sc">@</span>n.loc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 93202</code></pre>
</div>
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(data.gl<span class="sc">@</span>pop)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
E-ATL  E-IO  Fiji Japan  N-IO W-ATL  W-IO W-PAC 
    2    36     8    14    20    36    40    26 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(data.gl<span class="sc">@</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>pop)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
E-ATL  E-IO  Fiji Japan  N-IO W-ATL  W-IO W-PAC 
    2    36     8    14    20    36    40    26 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb150"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb150-1"><a href="#cb150-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(data.gl<span class="sc">@</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>sex, <span class="at">useNA =</span> <span class="st">"ifany"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   F    M &lt;NA&gt; 
  85   64   33 </code></pre>
</div>
</div>
</section>
<section id="number-of-sex-linked-markers-1" class="level4">
<h4 class="anchored" data-anchor-id="number-of-sex-linked-markers-1">1. Number of sex-linked markers?</h4>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-19_b199fe6c4beae6654be4cd9d0d0d1e2d">
<div class="sourceCode cell-code" id="cb152"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb152-1"><a href="#cb152-1" aria-hidden="true" tabindex="-1"></a>ncores <span class="ot">&lt;-</span> <span class="fu">min</span>(<span class="dv">4</span>, parallel<span class="sc">::</span><span class="fu">detectCores</span>())</span>
<span id="cb152-2"><a href="#cb152-2" aria-hidden="true" tabindex="-1"></a><span class="co">#still takes some minutes to run</span></span>
<span id="cb152-3"><a href="#cb152-3" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">gl.filter.sexlinked</span>(<span class="at">gl =</span> data.gl, <span class="at">system =</span> <span class="st">"xy"</span>, <span class="at">plots =</span> <span class="cn">TRUE</span>, <span class="at">ncores =</span> ncores)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Detected 85 females and 64 males.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Starting phase 1. Working in parallel...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building call rate plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done. Starting phase 2.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building heterozygosity plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-19-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-19-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done building heterozygosity plots.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**FINISHED** Total of analyzed loci: 93202.
Found 9 sex-linked loci:
   2 Y-linked loci
   2 sex-biased loci
   4 X-linked loci
   1 XY gametologs.
And 93193 autosomal loci.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="session10_files/figure-html/unnamed-chunk-19-4.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="individuals-with-wrong-sexid-1" class="level4">
<h4 class="anchored" data-anchor-id="individuals-with-wrong-sexid-1">2. Individuals with wrong sexID?</h4>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-20_0b8d3132e8813ed8cf0c1baa14dbb902">
<div class="sourceCode cell-code" id="cb160"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb160-1"><a href="#cb160-1" aria-hidden="true" tabindex="-1"></a><span class="co">#check if you can increase the number of cores</span></span>
<span id="cb160-2"><a href="#cb160-2" aria-hidden="true" tabindex="-1"></a>sexID <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">gl.infer.sex</span>(<span class="at">gl_sex_filtered =</span> res, <span class="at">system =</span> <span class="st">"xy"</span>, <span class="at">seed =</span> <span class="dv">124</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Not enough gametologs (need at least 5). Assigning NA...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>***FINISHED***</code></pre>
</div>
<div class="sourceCode cell-code" id="cb163"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb163-1"><a href="#cb163-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(sexID))</span>
<span id="cb163-2"><a href="#cb163-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb163-3"><a href="#cb163-3" aria-hidden="true" tabindex="-1"></a>agreed.sex <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">*"</span>, <span class="at">replacement =</span> <span class="st">""</span>, <span class="at">x =</span> sexID<span class="sc">$</span>agreed.sex) <span class="co"># remove asterisk</span></span>
<span id="cb163-4"><a href="#cb163-4" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(data.gl<span class="sc">$</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>sex <span class="sc">!=</span> agreed.sex, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 10%">
<col style="width: 10%">
<col style="width: 10%">
<col style="width: 6%">
<col style="width: 7%">
<col style="width: 10%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 11%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 9%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">id</th>
<th style="text-align: left;">y.linked.sex</th>
<th style="text-align: right;">#called</th>
<th style="text-align: right;">#missing</th>
<th style="text-align: left;">x.linked.sex</th>
<th style="text-align: right;">#Het.x</th>
<th style="text-align: right;">#Hom.x</th>
<th style="text-align: left;">gametolog.sex</th>
<th style="text-align: left;">#Het.g</th>
<th style="text-align: left;">#Hom.g</th>
<th style="text-align: left;">agreed.sex</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">CL-FIJ002-F</td>
<td style="text-align: left;">CL-FIJ002-F</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="even">
<td style="text-align: left;">CL-FIJ003-M</td>
<td style="text-align: left;">CL-FIJ003-M</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">4</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CL-FIJ008-F</td>
<td style="text-align: left;">CL-FIJ008-F</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="even">
<td style="text-align: left;">CL-FIJ010-F</td>
<td style="text-align: left;">CL-FIJ010-F</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CL-FIJ015-F</td>
<td style="text-align: left;">CL-FIJ015-F</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
<tr class="even">
<td style="text-align: left;">CL-FIJ018-F</td>
<td style="text-align: left;">CL-FIJ018-F</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">2</td>
<td style="text-align: left;">F</td>
<td style="text-align: right;">3</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">F</td>
</tr>
</tbody>
</table>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 8</code></pre>
</div>
</div>
</section>
</section>
<section id="exercise-data-4---blue-shark" class="level3">
<h3 class="anchored" data-anchor-id="exercise-data-4---blue-shark">Exercise data 4 - Blue shark</h3>
<p>Data from <span class="citation" data-cites="nikolic_stepping_2023">Nikolic et al. (<a href="references.html#ref-nikolic_stepping_2023" role="doc-biblioref">2023</a>)</span>.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Blue_shark.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">The Blue Shark</figcaption>
</figure>
</div>
<section id="load-data-4" class="level4">
<h4 class="anchored" data-anchor-id="load-data-4">Load data</h4>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-21_18ee8b3b729543ed7c21fdf9af9c9021">
<div class="sourceCode cell-code" id="cb165"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb165-1"><a href="#cb165-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">load</span>(<span class="st">"data/Blue_shark_DArTseq_genlight_for_sex-linked_markers.Rdata"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data.gl"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb167"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb167-1"><a href="#cb167-1" aria-hidden="true" tabindex="-1"></a>data.gl<span class="sc">@</span>n.loc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 172384</code></pre>
</div>
<div class="sourceCode cell-code" id="cb169"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb169-1"><a href="#cb169-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(data.gl<span class="sc">@</span>pop)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   EIO   MED1   MED2   NATL  NEATL    NIO   NPAC   SAF1   SAF2 SWPAC1 SWPAC2 
     8     34     20     49     26     27      4     21     89     30     16 
SWPAC3    WIO 
    11     29 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb171"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb171-1"><a href="#cb171-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(data.gl<span class="sc">@</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>pop)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   EIO   MED1   MED2   NATL  NEATL    NIO   NPAC   SAF1   SAF2 SWPAC1 SWPAC2 
     8     34     20     49     26     27      4     21     89     30     16 
SWPAC3    WIO 
    11     29 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb173"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb173-1"><a href="#cb173-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(data.gl<span class="sc">@</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>sex, <span class="at">useNA =</span> <span class="st">"ifany"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   F    M &lt;NA&gt; 
 104  111  149 </code></pre>
</div>
</div>
</section>
<section id="number-of-sex-linked-markers-2" class="level4">
<h4 class="anchored" data-anchor-id="number-of-sex-linked-markers-2">1. Number of sex-linked markers?</h4>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-22_8051261d9a22ba5e59295ee1bbe37f80">
<div class="sourceCode cell-code" id="cb175"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb175-1"><a href="#cb175-1" aria-hidden="true" tabindex="-1"></a><span class="co">#check if you can increase the number of cores</span></span>
<span id="cb175-2"><a href="#cb175-2" aria-hidden="true" tabindex="-1"></a><span class="co"># ncores &lt;- min(4,parallel::detectCores())</span></span>
<span id="cb175-3"><a href="#cb175-3" aria-hidden="true" tabindex="-1"></a><span class="co"># resbull &lt;- dartR.sexlinked::gl.filter.sexlinked(gl = data.gl, system = "xy", plots = TRUE, ncores = ncores)</span></span>
<span id="cb175-4"><a href="#cb175-4" aria-hidden="true" tabindex="-1"></a><span class="co">#load results from previous run</span></span>
<span id="cb175-5"><a href="#cb175-5" aria-hidden="true" tabindex="-1"></a>resbull <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"./data/resbull.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="individuals-with-wrong-sexid-2" class="level4">
<h4 class="anchored" data-anchor-id="individuals-with-wrong-sexid-2">2. Individuals with wrong sexID?</h4>
<div class="cell" data-hash="session10_cache/html/unnamed-chunk-23_29b2f44684bbfc68ee356b220bdcc563">
<div class="sourceCode cell-code" id="cb176"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb176-1"><a href="#cb176-1" aria-hidden="true" tabindex="-1"></a><span class="co">#check if you can increase the number of cores</span></span>
<span id="cb176-2"><a href="#cb176-2" aria-hidden="true" tabindex="-1"></a>sexID <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">gl.infer.sex</span>(<span class="at">gl_sex_filtered =</span> resbull, <span class="at">system =</span> <span class="st">"xy"</span>, <span class="at">seed =</span> <span class="dv">124</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Not enough gametologs (need at least 5). Assigning NA...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>***FINISHED***</code></pre>
</div>
<div class="sourceCode cell-code" id="cb179"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb179-1"><a href="#cb179-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(sexID))</span>
<span id="cb179-2"><a href="#cb179-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb179-3"><a href="#cb179-3" aria-hidden="true" tabindex="-1"></a>agreed.sex <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">*"</span>, <span class="at">replacement =</span> <span class="st">""</span>, <span class="at">x =</span> sexID<span class="sc">$</span>agreed.sex) <span class="co"># remove asterisk</span></span>
<span id="cb179-4"><a href="#cb179-4" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(data.gl<span class="sc">$</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>sex <span class="sc">!=</span> agreed.sex, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 12%">
<col style="width: 7%">
<col style="width: 8%">
<col style="width: 12%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 12%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 10%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">id</th>
<th style="text-align: left;">y.linked.sex</th>
<th style="text-align: right;">#called</th>
<th style="text-align: right;">#missing</th>
<th style="text-align: left;">x.linked.sex</th>
<th style="text-align: right;">#Het.x</th>
<th style="text-align: right;">#Hom.x</th>
<th style="text-align: left;">gametolog.sex</th>
<th style="text-align: left;">#Het.g</th>
<th style="text-align: left;">#Hom.g</th>
<th style="text-align: left;">agreed.sex</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">60088</td>
<td style="text-align: left;">60088</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">22</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">60160</td>
<td style="text-align: left;">60160</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">21</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">60168</td>
<td style="text-align: left;">60168</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">22</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">60176</td>
<td style="text-align: left;">60176</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">21</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">60096</td>
<td style="text-align: left;">60096</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">22</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">60104</td>
<td style="text-align: left;">60104</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">19</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
</tbody>
</table>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 166</code></pre>
</div>
</div>


<div id="refs" class="references csl-bib-body hanging-indent" role="list" style="display: none">
<div id="ref-devloodelva_rivers_2023" class="csl-entry" role="listitem">
Devloo‐Delva, Floriaan, Christopher P. Burridge, Peter M. Kyne, Juerg M. Brunnschweiler, Demian D. Chapman, Patricia Charvet, Xiao Chen, et al. 2023. <span>“From Rivers to Ocean Basins: <span>The</span> Role of Ocean Barriers and Philopatry in the Genetic Structuring of a Cosmopolitan Coastal Predator.”</span> <em>Ecology and Evolution</em> 13 (2): e9837. <a href="https://doi.org/10.1002/ece3.9837">https://doi.org/10.1002/ece3.9837</a>.
</div>
<div id="ref-nikolic_stepping_2023" class="csl-entry" role="listitem">
Nikolic, Natacha, Floriaan Devloo-Delva, Diane Bailleul, Ekaterina Noskova, Clément Rougeux, Chrystelle Delord, Philippe Borsa, et al. 2023. <span>“Stepping up to Genome Scan Allows Stock Differentiation in the Worldwide Distributed Blue Shark <span>Prionace</span> Glauca.”</span> <em>Molecular Ecology</em> 32 (5): 1000–1019. <a href="https://doi.org/10.1111/mec.16822">https://doi.org/10.1111/mec.16822</a>.
</div>
<div id="ref-robledo2023" class="csl-entry" role="listitem">
Robledo-Ruiz, Diana A., Lana Austin, J. Nevil Amos, Jesús Castrejón-Figueroa, Daniel K. P. Harley, Michael J. L. Magrath, Paul Sunnucks, and Alexandra Pavlova. 2023. <span>“Easy-to-Use <span>R</span> Functions to Separate Reduced-Representation Genomic Datasets into Sex-Linked and Autosomal Loci, and Conduct Sex Assignment.”</span> <em>Molecular Ecology Resources</em> 0 (n/a). <a href="https://doi.org/10.1111/1755-0998.13844">https://doi.org/10.1111/1755-0998.13844</a>.
</div>
</div>
</section>
</section>
</section>

</main> <!-- /main -->
<script>

/* update total correct if #webex-total_correct exists */
update_total_correct = function() {
  console.log("webex: update total_correct");

  var t = document.getElementsByClassName("webex-total_correct");
  for (var i = 0; i < t.length; i++) {
    p = t[i].parentElement;
    var correct = p.getElementsByClassName("webex-correct").length;
    var solvemes = p.getElementsByClassName("webex-solveme").length;
    var radiogroups = p.getElementsByClassName("webex-radiogroup").length;
    var selects = p.getElementsByClassName("webex-select").length;

    t[i].innerHTML = correct + " of " + (solvemes + radiogroups + selects) + " correct";
  }
}

/* webex-solution button toggling function */
b_func = function() {
  console.log("webex: toggle hide");

  var cl = this.parentElement.classList;
  if (cl.contains('open')) {
    cl.remove("open");
  } else {
    cl.add("open");
  }
}

/* check answers */
check_func = function() {
  console.log("webex: check answers");

  var cl = this.parentElement.classList;
  if (cl.contains('unchecked')) {
    cl.remove("unchecked");
    this.innerHTML = "Hide Answers";
  } else {
    cl.add("unchecked");
    this.innerHTML = "Show Answers";
  }
}

/* function for checking solveme answers */
solveme_func = function(e) {
  console.log("webex: check solveme");

  var real_answers = JSON.parse(this.dataset.answer);
  var my_answer = this.value;
  var cl = this.classList;
  if (cl.contains("ignorecase")) {
    my_answer = my_answer.toLowerCase();
  }
  if (cl.contains("nospaces")) {
    my_answer = my_answer.replace(/ /g, "")
  }

  if (my_answer == "") {
    cl.remove("webex-correct");
    cl.remove("webex-incorrect");
  } else if (real_answers.includes(my_answer)) {
    cl.add("webex-correct");
    cl.remove("webex-incorrect");
  } else {
    cl.add("webex-incorrect");
    cl.remove("webex-correct");
  }

  // match numeric answers within a specified tolerance
  if(this.dataset.tol > 0){
    var tol = JSON.parse(this.dataset.tol);
    var matches = real_answers.map(x => Math.abs(x - my_answer) < tol)
    if (matches.reduce((a, b) => a + b, 0) > 0) {
      cl.add("webex-correct");
    } else {
      cl.remove("webex-correct");
    }
  }

  // added regex bit
  if (cl.contains("regex")){
    answer_regex = RegExp(real_answers.join("|"))
    if (answer_regex.test(my_answer)) {
      cl.add("webex-correct");
    }
  }

  update_total_correct();
}

/* function for checking select answers */
select_func = function(e) {
  console.log("webex: check select");

  var cl = this.classList

  /* add style */
  cl.remove("webex-incorrect");
  cl.remove("webex-correct");
  if (this.value == "answer") {
    cl.add("webex-correct");
  } else if (this.value != "blank") {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

/* function for checking radiogroups answers */
radiogroups_func = function(e) {
  console.log("webex: check radiogroups");

  var checked_button = document.querySelector('input[name=' + this.id + ']:checked');
  var cl = checked_button.parentElement.classList;
  var labels = checked_button.parentElement.parentElement.children;

  /* get rid of styles */
  for (i = 0; i < labels.length; i++) {
    labels[i].classList.remove("webex-incorrect");
    labels[i].classList.remove("webex-correct");
  }

  /* add style */
  if (checked_button.value == "answer") {
    cl.add("webex-correct");
  } else {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

window.onload = function() {
  console.log("webex onload");
  /* set up solution buttons */
  var buttons = document.getElementsByTagName("button");

  for (var i = 0; i < buttons.length; i++) {
    if (buttons[i].parentElement.classList.contains('webex-solution')) {
      buttons[i].onclick = b_func;
    }
  }

  var check_sections = document.getElementsByClassName("webex-check");
  console.log("check:", check_sections.length);
  for (var i = 0; i < check_sections.length; i++) {
    check_sections[i].classList.add("unchecked");

    let btn = document.createElement("button");
    btn.innerHTML = "Show Answers";
    btn.classList.add("webex-check-button");
    btn.onclick = check_func;
    check_sections[i].appendChild(btn);

    let spn = document.createElement("span");
    spn.classList.add("webex-total_correct");
    check_sections[i].appendChild(spn);
  }

  /* set up webex-solveme inputs */
  var solveme = document.getElementsByClassName("webex-solveme");

  for (var i = 0; i < solveme.length; i++) {
    /* make sure input boxes don't auto-anything */
    solveme[i].setAttribute("autocomplete","off");
    solveme[i].setAttribute("autocorrect", "off");
    solveme[i].setAttribute("autocapitalize", "off");
    solveme[i].setAttribute("spellcheck", "false");
    solveme[i].value = "";

    /* adjust answer for ignorecase or nospaces */
    var cl = solveme[i].classList;
    var real_answer = solveme[i].dataset.answer;
    if (cl.contains("ignorecase")) {
      real_answer = real_answer.toLowerCase();
    }
    if (cl.contains("nospaces")) {
      real_answer = real_answer.replace(/ /g, "");
    }
    solveme[i].dataset.answer = real_answer;

    /* attach checking function */
    solveme[i].onkeyup = solveme_func;
    solveme[i].onchange = solveme_func;

    solveme[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  /* set up radiogroups */
  var radiogroups = document.getElementsByClassName("webex-radiogroup");
  for (var i = 0; i < radiogroups.length; i++) {
    radiogroups[i].onchange = radiogroups_func;
  }

  /* set up selects */
  var selects = document.getElementsByClassName("webex-select");
  for (var i = 0; i < selects.length; i++) {
    selects[i].onchange = select_func;
    selects[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  update_total_correct();
}

</script>
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
          // default icon
          link.classList.add("external");
      }
    }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./session09.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Lineage Divergence</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./session11.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">From Genes to Kin: Dissecting Relatedness &amp; Kinship</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>