<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Kioloa PopGen - 11&nbsp; From Genes to Kin: Dissecting Relatedness &amp; Kinship</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./session12.html" rel="next">
<link href="./session10.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="css.css">
<link rel="stylesheet" href="include/webex.css">
</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./session11.html"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">From Genes to Kin: Dissecting Relatedness &amp; Kinship</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Kioloa PopGen</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome!</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./schedule.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Program</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./speakers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Speaker Bios</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./install.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Install dartRverse</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session01.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Intro to dartR</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session02.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Pop Gen In Conservation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session03.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Sequencing Technologies</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session04.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Data Management, Reproducibility &amp; Integrity</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session05.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Effective Population Size</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session06.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Management of Small Populations</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session07.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Natural Selection</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session08.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Landscape Genetics</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session09.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Lineage Divergence</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session10.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Sex Linked Markers</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session11.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">From Genes to Kin: Dissecting Relatedness &amp; Kinship</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session12.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Genetic Structure</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session13.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Combining Genomic Resources</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./break.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">__________</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./useful.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">More info</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">dartR Tutorials</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#required-packages" id="toc-required-packages" class="nav-link active" data-scroll-target="#required-packages"><em>Required packages</em></a></li>
  <li><a href="#introduction" id="toc-introduction" class="nav-link" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#the-quest-to-quash-the-cane-toad-conundrum-in-mandalong" id="toc-the-quest-to-quash-the-cane-toad-conundrum-in-mandalong" class="nav-link" data-scroll-target="#the-quest-to-quash-the-cane-toad-conundrum-in-mandalong">The Quest to Quash the Cane Toad Conundrum in Mandalong</a>
  <ul class="collapse">
  <li><a href="#colony" id="toc-colony" class="nav-link" data-scroll-target="#colony">COLONY</a></li>
  <li><a href="#filtering" id="toc-filtering" class="nav-link" data-scroll-target="#filtering">Filtering</a></li>
  <li><a href="#coancestry" id="toc-coancestry" class="nav-link" data-scroll-target="#coancestry">COANCESTRY</a></li>
  <li><a href="#emibd9" id="toc-emibd9" class="nav-link" data-scroll-target="#emibd9">EMIBD9</a></li>
  <li><a href="#gcta" id="toc-gcta" class="nav-link" data-scroll-target="#gcta">GCTA</a></li>
  <li><a href="#genomic-relationship-matrix-grm" id="toc-genomic-relationship-matrix-grm" class="nav-link" data-scroll-target="#genomic-relationship-matrix-grm">Genomic relationship matrix (GRM)</a></li>
  <li><a href="#plotting" id="toc-plotting" class="nav-link" data-scroll-target="#plotting">Plotting</a></li>
  </ul></li>
  <li><a href="#feathers-of-laughter-unveiling-the-mysteries-of-the-amazonian-giggletuft-hilaroplumis-amazoniana" id="toc-feathers-of-laughter-unveiling-the-mysteries-of-the-amazonian-giggletuft-hilaroplumis-amazoniana" class="nav-link" data-scroll-target="#feathers-of-laughter-unveiling-the-mysteries-of-the-amazonian-giggletuft-hilaroplumis-amazoniana">Feathers of Laughter: Unveiling the Mysteries of the Amazonian Giggletuft (Hilaroplumis amazoniana)”</a>
  <ul class="collapse">
  <li><a href="#coancestry-1" id="toc-coancestry-1" class="nav-link" data-scroll-target="#coancestry-1">COANCESTRY</a></li>
  <li><a href="#plotting-1" id="toc-plotting-1" class="nav-link" data-scroll-target="#plotting-1">Plotting</a></li>
  </ul></li>
  <li><a href="#from-the-brink-and-back-the-arabian-oryxs-oman-odyssey" id="toc-from-the-brink-and-back-the-arabian-oryxs-oman-odyssey" class="nav-link" data-scroll-target="#from-the-brink-and-back-the-arabian-oryxs-oman-odyssey">From the Brink and Back: The Arabian Oryx’s Oman Odyssey</a>
  <ul class="collapse">
  <li><a href="#load-dataset-and-filtering" id="toc-load-dataset-and-filtering" class="nav-link" data-scroll-target="#load-dataset-and-filtering">Load dataset and filtering</a></li>
  <li><a href="#grm" id="toc-grm" class="nav-link" data-scroll-target="#grm">GRM</a></li>
  <li><a href="#sequoia" id="toc-sequoia" class="nav-link" data-scroll-target="#sequoia">SEQUOIA</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">From Genes to Kin: Dissecting Relatedness &amp; Kinship</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p><em>Session Presenters</em></p>
<p><img src="images/Presenters11.png" class="img-fluid"></p>
<section id="required-packages" class="level3">
<h3 class="anchored" data-anchor-id="required-packages"><em>Required packages</em></h3>
<div class="cell" data-hash="session11_cache/html/unnamed-chunk-1_24ccd93d115c5bca351bead1a5be7986">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dartRverse)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(related)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reshape2)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sequoia)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(kinship2)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dartR.base)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>make sure you have the packages installed, see</em> <a href="./install.html">Install dartRverse</a></p>
</section>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>Two alleles are identical by descent (IBD) if they are identical copies of the same ancestral allele in a base population. Relatedness is the proportion of alleles shared between individuals that are IBD. Kinship is the probability that two alleles, one taken at random from each individual, are IBD (<a href="https://doi.org/10.1111/2041-210X.13963">Wang 2022</a>). Kinship considers half the genetic information because it looks at the probability that one individual inherits a particular allele from a common ancestor shared with another individual. Therefore, kinship is equal to half of relatedness.</p>
<p>Reletedness/kinship are not absolute states but are estimated relative to a reference population for which there is generally little information, so that we can estimate the kinship of a pair of individuals only relative to some other quantity.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>Below is a table modified from Speed &amp; Balding (<a href="https://doi.org/10.1038/nrg382">2015</a>) showing kinship values, and their confidence intervals (CI) for different relationships.</p>
<table class="table">
<thead>
<tr class="header">
<th style="text-align: left;">Relationship</th>
<th style="text-align: center;">Kinship</th>
<th style="text-align: center;">CI(95%)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Identical twins/clones/same individual</td>
<td style="text-align: center;">0.5</td>
<td style="text-align: center;">-</td>
</tr>
<tr class="even">
<td style="text-align: left;">Sibling/Parent-Offspring</td>
<td style="text-align: center;">0.25</td>
<td style="text-align: center;">(0.204, 0.296)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Half-sibling</td>
<td style="text-align: center;">0.125</td>
<td style="text-align: center;">(0.092, 0.158)</td>
</tr>
<tr class="even">
<td style="text-align: left;">First cousin</td>
<td style="text-align: center;">0.062</td>
<td style="text-align: center;">(0.038, 0.089)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Half-cousin</td>
<td style="text-align: center;">0.031</td>
<td style="text-align: center;">(0.012, 0.055)</td>
</tr>
<tr class="even">
<td style="text-align: left;">Second cousin</td>
<td style="text-align: center;">0.016</td>
<td style="text-align: center;">(0.004, 0.031)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Half-second cousin</td>
<td style="text-align: center;">0.008</td>
<td style="text-align: center;">(0.001, 0.020)</td>
</tr>
<tr class="even">
<td style="text-align: left;">Third cousin</td>
<td style="text-align: center;">0.004</td>
<td style="text-align: center;">(0.000, 0.012)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Unrelated</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">-</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="the-quest-to-quash-the-cane-toad-conundrum-in-mandalong" class="level2">
<h2 class="anchored" data-anchor-id="the-quest-to-quash-the-cane-toad-conundrum-in-mandalong">The Quest to Quash the Cane Toad Conundrum in Mandalong</h2>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="./images/cane_toad.png" class="img-fluid figure-img" width="300"></p>
<figcaption class="figure-caption"><em>Cane toad (Rhinella marinus)</em></figcaption>
</figure>
</div>
<p>Introduced to Australia in the 1930s to combat cane beetles, cane toads have since become a significant invasive threat, adapting to diverse habitats and reproducing prolifically. Lacking natural predators and possessing toxic secretions, they’ve harmed native wildlife and disrupted ecosystems across eastern and northern Australia.</p>
<p><em>The key issue here is figuring out if the young toads at Mandalong are locals, born and bred, or if they’ve crashed the party as a group from elsewhere.</em></p>
<section id="loading-data" class="level4">
<h4 class="anchored" data-anchor-id="loading-data">Loading data</h4>
<div class="cell" data-hash="session11_cache/html/unnamed-chunk-2_06e8a5256ae37ed4d58be62a884a5975">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>t1 <span class="ot">&lt;-</span> <span class="fu">gl.load</span>(<span class="st">"./data/cane_toad.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.load 
  Processing genlight object with SNP data
  Loaded object of type SNP from ./data/cane_toad.rds 
Starting gl.compliance.check 
  Processing genlight object with SNP data
  Checking coding of SNPs
    SNP data scored NA, 0, 1 or 2 confirmed
  Checking locus metrics and flags
  Recalculating locus metrics
  Checking for monomorphic loci
    No monomorphic loci detected
  Checking for loci with all missing data
    No loci with all missing data detected
  Checking whether individual names are unique.
  Checking for individual metrics
    Individual metrics confirmed
  Checking for population assignments
    Population assignments confirmed
  Spelling of coordinates checked and changed if necessary to 
            lat/lon
Completed: gl.compliance.check 
Completed: gl.load </code></pre>
</div>
</div>
</section>
<section id="finding-duplicated-samples" class="level4">
<h4 class="anchored" data-anchor-id="finding-duplicated-samples">Finding duplicated samples</h4>
<p>It is critical to identify and remove duplicated samples. To identify duplicated samples, we used the software Colony (<a href="https://doi.org/10.1111/j.1755-0998.2009.02787.x">Jones &amp; Wang, 2010</a>) and compare it with dartR function <code>gl.report.replicates</code>.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>Ideally, in a large dataset with related and unrelated individuals and several replicated individuals, such as in a capture/mark/recapture study, the first histogram should have four “peaks”. The first peak should represent unrelated individuals, the second peak should correspond to second-degree relationships (such as cousins), the third peak should represent first-degree relationships (like parent/offspring and full siblings), and the fourth peak should represent replicated individuals.</p>
<p>In order to ensure that replicated individuals are properly identified, it’s important to have a clear separation between the third and fourth peaks in the second histogram. This means that there should be bins with zero counts between these two peaks</p>
</div>
</div>
<div class="cell" data-hash="session11_cache/html/unnamed-chunk-3_9ca8cdf4d1c9c26c1116db35c551faf8">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>t1_dup <span class="ot">&lt;-</span> t1</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>res_dup <span class="ot">&lt;-</span> <span class="fu">gl.report.replicates</span>(t1_dup,<span class="at">perc_geno =</span> <span class="fl">0.925</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.report.replicates 
  Processing genlight object with SNP data</code></pre>
</div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.report.replicates </code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>res_dup2 <span class="ot">&lt;-</span> res_dup<span class="sc">$</span>table.rep</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Ordering pairwise estimates </span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>res_dup2 <span class="ot">&lt;-</span> res_dup2[<span class="fu">order</span>(res_dup2<span class="sc">$</span>perc,<span class="at">decreasing =</span> <span class="cn">TRUE</span>),]</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>res_dup_2 <span class="ot">&lt;-</span> res_dup<span class="sc">$</span>ind.list.rep</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>res_dup_2 <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res_dup_2,<span class="cf">function</span>(x){</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste</span>(x,<span class="at">collapse =</span> <span class="st">", "</span>)</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>res_dup_2 <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="fu">names</span>(res_dup_2),res_dup_2,<span class="at">sep =</span> <span class="st">", "</span>)</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>res_dup_2 <span class="ot">&lt;-</span> res_dup_2[<span class="fu">order</span>(res_dup_2)]</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(res_dup_2,<span class="at">sep=</span><span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>JC_0040, JC0068
JC_0048, JC_0053
JC0062, JC0063
JC0072, JC0073
JC0076, JC0077
RJCT10, RJCT51</code></pre>
</div>
</div>
</section>
<section id="colony" class="level3">
<h3 class="anchored" data-anchor-id="colony">COLONY</h3>
<p>We will use <a href="https://doi.org/10.1111/1755-0998.13844">Diana Robledo’s</a> function to convert a genlight object into COLONY format.</p>
<div class="cell" data-hash="session11_cache/html/unnamed-chunk-4_93b4ef2b624fc7dd921ca01f204869d8">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># loading gl2colony function from Diana Robledo </span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"./functions/gl2colony.R"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'crayon'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:ggplot2':

    %+%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:adegenet':

    chr</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>t3 <span class="ot">&lt;-</span> t1_dup</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>t3<span class="sc">$</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>offspring <span class="ot">&lt;-</span> <span class="st">"yes"</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>t3<span class="sc">$</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>mother <span class="ot">&lt;-</span> <span class="st">"no"</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>t3<span class="sc">$</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>father <span class="ot">&lt;-</span> <span class="st">"no"</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>t3<span class="sc">$</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>id <span class="ot">&lt;-</span> <span class="fu">indNames</span>(t3)</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="fu">gl2colony</span>(<span class="at">gl =</span> t3,</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>          <span class="at">filename_out =</span> <span class="st">"colony"</span>,</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>          <span class="at">probability_father =</span> <span class="fl">0.5</span>,</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>          <span class="at">probability_mother =</span> <span class="fl">0.5</span>,</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>          <span class="at">seed =</span> <span class="cn">NULL</span>,                <span class="co"># Seed for random number generator</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>          <span class="at">update_allele_freq =</span> <span class="dv">0</span>,     <span class="co"># 0/1=Not updating/updating allele frequency</span></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>          <span class="at">di_mono_ecious =</span> <span class="dv">2</span>,         <span class="co"># 2/1=Dioecious/Monoecious species</span></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>          <span class="at">inbreed =</span> <span class="dv">0</span>,                <span class="co"># 0/1=no inbreeding/inbreeding</span></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>          <span class="at">haplodiploid =</span> <span class="dv">0</span>,           <span class="co"># 0/1=Diploid species/HaploDiploid species</span></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>          <span class="at">polygamy_male =</span> <span class="dv">0</span>,          <span class="co"># 0/1=Polygamy/Monogamy for males</span></span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>          <span class="at">polygamy_female =</span> <span class="dv">0</span>,        <span class="co"># 0/1=Polygamy/Monogamy for females</span></span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>          <span class="at">clone_inference =</span> <span class="dv">1</span>,        <span class="co"># 0/1=Clone inference =No/Yes</span></span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>          <span class="at">scale_shibship =</span> <span class="dv">1</span>,         <span class="co"># 0/1=Scale full sibship=No/Yes</span></span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>          <span class="at">sibship_prior =</span> <span class="dv">0</span>,          <span class="co"># 0/1/2/3/4=No/Weak/Medium/Strong/Optimal sibship prior</span></span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a>          <span class="at">known_allele_freq =</span> <span class="dv">0</span>,      <span class="co"># 0/1=Unknown/Known population allele frequency</span></span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a>          <span class="at">num_runs =</span> <span class="dv">1</span>,               <span class="co"># Number of runs</span></span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a>          <span class="at">length_run =</span> <span class="dv">2</span>,             <span class="co"># 1/2/3/4=short/medium/long/very long run</span></span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a>          <span class="at">monitor_method =</span> <span class="dv">0</span>,         <span class="co"># 0/1=Monitor method by Iterate#/Time in second</span></span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a>          <span class="at">monitor_interval =</span> <span class="dv">10000</span>,   <span class="co"># Monitor interval in Iterate# / in seconds</span></span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a>          <span class="at">windows_gui =</span> <span class="dv">0</span>,            <span class="co"># 0/1=No/Yes for run with Windows GUI</span></span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true" tabindex="-1"></a>          <span class="at">likelihood =</span> <span class="dv">0</span>,             <span class="co"># 0/1/2=PairLikelihood score/Fulllikelihood/FPLS</span></span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true" tabindex="-1"></a>          <span class="at">precision_fl =</span> <span class="dv">2</span>,           <span class="co"># 0/1/2/3=Low/Medium/High/Very high precision with Full-likelihood</span></span>
<span id="cb13-29"><a href="#cb13-29" aria-hidden="true" tabindex="-1"></a>          <span class="at">marker_id =</span> <span class="st">'mk@'</span>,          <span class="co"># Marker Ids</span></span>
<span id="cb13-30"><a href="#cb13-30" aria-hidden="true" tabindex="-1"></a>          <span class="at">marker_type =</span> <span class="st">'0@'</span>,         <span class="co"># Marker types, 0/1=Codominant/Dominant</span></span>
<span id="cb13-31"><a href="#cb13-31" aria-hidden="true" tabindex="-1"></a>          <span class="at">allelic_dropout =</span> <span class="st">'0.000@'</span>, <span class="co"># Allelic dropout rate at each locus</span></span>
<span id="cb13-32"><a href="#cb13-32" aria-hidden="true" tabindex="-1"></a>          <span class="at">other_typ_err =</span> <span class="st">'0.05@'</span>,    <span class="co"># Other typing error rate at each locus</span></span>
<span id="cb13-33"><a href="#cb13-33" aria-hidden="true" tabindex="-1"></a>          <span class="at">paternity_exclusion_threshold =</span> <span class="st">'0 0'</span>,</span>
<span id="cb13-34"><a href="#cb13-34" aria-hidden="true" tabindex="-1"></a>          <span class="at">maternity_exclusion_threshold =</span> <span class="st">'0 0'</span>,</span>
<span id="cb13-35"><a href="#cb13-35" aria-hidden="true" tabindex="-1"></a>          <span class="at">paternal_sibship =</span> <span class="dv">0</span>,</span>
<span id="cb13-36"><a href="#cb13-36" aria-hidden="true" tabindex="-1"></a>          <span class="at">maternal_sibship =</span> <span class="dv">0</span>,</span>
<span id="cb13-37"><a href="#cb13-37" aria-hidden="true" tabindex="-1"></a>          <span class="at">excluded_paternity =</span> <span class="dv">0</span>,</span>
<span id="cb13-38"><a href="#cb13-38" aria-hidden="true" tabindex="-1"></a>          <span class="at">excluded_maternity =</span> <span class="dv">0</span>,</span>
<span id="cb13-39"><a href="#cb13-39" aria-hidden="true" tabindex="-1"></a>          <span class="at">excluded_paternal_sibships =</span> <span class="dv">0</span>,</span>
<span id="cb13-40"><a href="#cb13-40" aria-hidden="true" tabindex="-1"></a>          <span class="at">excluded_maternity_sibships =</span> <span class="dv">0</span></span>
<span id="cb13-41"><a href="#cb13-41" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Random seed set to 42763</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>112 Offspring detected. 
0 Fathers detected. 
0 Mothers detected.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Missing paternal and maternal IDs: only sibship inference will be possible.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Exporting Genlight object to COLONY2 format...</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>(100%) COLONY2 file successfully exported!</code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Uncomment to run .</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="co"># system.time(system(paste(path.colony, "IFN:colony")))</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="co"># # user  system elapsed </span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="co"># # 181.102   1.893 183.162 </span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="co"># duplicates &lt;- read.csv("./my_project.PairwiseCloneDyad")</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="co"># read the output</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a><span class="co">#duplicates &lt;- read.csv("./data/my_project.PairwiseCloneDyad")</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a><span class="co">#duplicates</span></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(res_dup_2,<span class="at">sep=</span><span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>JC_0040, JC0068
JC_0048, JC_0053
JC0062, JC0063
JC0072, JC0073
JC0076, JC0077
RJCT10, RJCT51</code></pre>
</div>
</div>
</section>
<section id="filtering" class="level3">
<h3 class="anchored" data-anchor-id="filtering">Filtering</h3>
<p>Initially, it’s essential to ascertain the optimal approach for filtering, selecting between individual missingness and loci missingness. This decision is critical as individuals of lower quality can significantly impact loci that are otherwise deemed acceptable. Similarly, loci of subpar quality contribute to an increase in missing data among individuals who would otherwise meet our criteria. In our scenario, the priority is to retain individuals over loci. This prioritization is based on the premise that if the missing loci are uniformly distributed across all individuals, opting for individual-based filtering will not markedly influence the outcomes when compared to loci-based filtering.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Previous studies showed that around 5,000 SNPs are sufficient to accurately estimate relatedness and kinship (<a href="https://doi.org/10.1111/mec.14833">Goudet et al., 2018</a>).</p>
</div>
</div>
<div class="cell" data-hash="session11_cache/html/unnamed-chunk-5_2e5e8a479f200eba7ddafd4f80e902e5">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co">#removing duplicates </span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> <span class="fu">gl.drop.ind</span>(t1,<span class="at">ind.list =</span> res_dup<span class="sc">$</span>ind.list.drop)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="co"># filtering on callrate by individual</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="fu">gl.report.callrate</span>(t2,<span class="at">method =</span> <span class="st">"ind"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> <span class="fu">gl.filter.callrate</span>(t2,<span class="at">method =</span> <span class="st">"ind"</span>, <span class="at">threshold =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-5-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># filtering monomorphs</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> <span class="fu">gl.filter.monomorphs</span>(t2,<span class="at">verbose =</span> <span class="dv">5</span>)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="co"># filtering on callrate by locus</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> <span class="fu">gl.filter.callrate</span>(t2,<span class="at">method =</span> <span class="st">"loc"</span>, <span class="at">threshold =</span> <span class="fl">0.95</span>,<span class="at">verbose =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-5-3.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># filtering on reproducibility </span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> <span class="fu">gl.filter.reproducibility</span>(t2,<span class="at">threshold =</span> <span class="fl">0.99</span>,<span class="at">verbose =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-5-4.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># filtering on read depth </span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> <span class="fu">gl.filter.rdepth</span>(t2,<span class="at">lower =</span> <span class="dv">10</span>, <span class="at">upper =</span> <span class="fu">max</span>(t2<span class="sc">$</span>other<span class="sc">$</span>loc.metrics<span class="sc">$</span>rdepth),<span class="at">verbose =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-5-5.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># filtering on Minor Allele Count (MAC)</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> <span class="fu">gl.filter.maf</span>(t2,<span class="at">threshold =</span> <span class="dv">3</span>,<span class="at">verbose =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-5-6.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-5-7.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># filtering secondaries</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> <span class="fu">gl.filter.secondaries</span>(t2,<span class="at">verbose =</span> <span class="dv">5</span>)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="co"># filtering on Hardy Weinberg proportions</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> <span class="fu">gl.filter.hwe</span>(t2,<span class="at">verbose =</span> <span class="dv">5</span>)</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="co"># assigning chromosome information</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span>chromosome <span class="ot">&lt;-</span> t2<span class="sc">$</span>other<span class="sc">$</span>loc.metrics<span class="sc">$</span>Chrom_Rhinella_marina_v1</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a><span class="co"># assigning SNP position</span></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span>position <span class="ot">&lt;-</span> t2<span class="sc">$</span>other<span class="sc">$</span>loc.metrics<span class="sc">$</span>ChromPosSnp_Rhinella_marina_v1</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a><span class="co"># report of linkage disequilibrium </span></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>ld_rep <span class="ot">&lt;-</span> <span class="fu">gl.report.ld.map</span>(t2,<span class="at">verbose =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-5-8.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-5-9.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-5-10.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># filtering on LD</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="co"># a threshold (R.squared = 0.2) is commonly used to imply that two loci are unlinked (Delourme et al., 2013; Li et al., 2014).</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>t2 <span class="ot">&lt;-</span> <span class="fu">gl.filter.ld</span>(t2,<span class="at">ld.report =</span> ld_rep, <span class="at">threshold =</span> <span class="fl">0.2</span>, <span class="at">verbose =</span> <span class="dv">5</span>)</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Uncomment to run.</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a><span class="co"># filtering on Hamming distance</span></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a><span class="co"># system.time(</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a><span class="co"># t2 &lt;- gl.filter.hamming(t2)</span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a><span class="co"># )</span></span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a><span class="co"># user  system elapsed </span></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a><span class="co"># 67.403   2.520  71.174 </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="coancestry" class="level3">
<h3 class="anchored" data-anchor-id="coancestry">COANCESTRY</h3>
<p>We first run COANCESTRY (<a href="https://doi.org/10.1111/j.1755-0998.2010.02885.x">Wang, 2011</a>) using <a href="https://github.com/timothyfrasier/related">related</a> package and Wang’s unbiased estimator (<a href="https://doi.org/10.1038/hdy.2017.52">Wang, 2017</a>).</p>
<div class="cell" data-hash="session11_cache/html/unnamed-chunk-6_cd116ca9fdf65134c60e0d75503a30e3">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># converting to related format </span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>related <span class="ot">&lt;-</span> <span class="fu">gl2related</span>(t2, <span class="at">save=</span><span class="cn">FALSE</span>)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="co"># using Wang's unbiased estimator</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>res_coancestry <span class="ot">&lt;-</span> <span class="fu">coancestry</span>(related, <span class="at">wang=</span><span class="dv">1</span>)</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>res_coancestry_2 <span class="ot">&lt;-</span> res_coancestry<span class="sc">$</span>relatedness[,<span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">6</span>)]</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>res_coancestry_3 <span class="ot">&lt;-</span> <span class="fu">cbind</span>(res_coancestry_2<span class="sc">$</span>ind2.id,res_coancestry_2<span class="sc">$</span>ind1.id,res_coancestry_2<span class="sc">$</span>wang)</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(res_coancestry_3) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(res_coancestry_2)</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>res_coancestry_4 <span class="ot">&lt;-</span> <span class="fu">rbind</span>(res_coancestry_2,res_coancestry_3)</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>res_coancestry_4<span class="sc">$</span>wang <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(res_coancestry_4<span class="sc">$</span>wang)</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>mat_coan <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">acast</span>(res_coancestry_4, ind1.id<span class="sc">~</span>ind2.id, <span class="at">value.var=</span><span class="st">"wang"</span>))</span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>mat_coan <span class="ot">&lt;-</span> <span class="fu">apply</span>(mat_coan, <span class="dv">2</span>, as.numeric)</span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(mat_coan) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(mat_coan)</span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>coan_col <span class="ot">&lt;-</span> mat_coan</span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a>coan_col[<span class="fu">upper.tri</span>(coan_col)] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a>coan_col <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">as.table</span>(<span class="fu">as.matrix</span>(coan_col)))</span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a><span class="co"># dividing by two to get kinship </span></span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a>coan_col<span class="sc">$</span>Freq <span class="ot">&lt;-</span> coan_col<span class="sc">$</span>Freq<span class="sc">/</span><span class="dv">2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="emibd9" class="level3">
<h3 class="anchored" data-anchor-id="emibd9">EMIBD9</h3>
<p>Now let’s estimate kinship using new Wang’s method which uses a likelihood approach as implemented in the software EMIBD9 (<a href="https://doi.org/10.1111/2041-210X.13963">Wang 2022</a>).</p>
<div class="cell" data-hash="session11_cache/html/unnamed-chunk-7_46895efc736fb2fde75e7214369141cd">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Uncomment to run </span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a> <span class="co"># system.time(EMIBD9 &lt;- gl.run.EMIBD9(t2,emibd9.path = path.EMIBD9))</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="co"># user  system elapsed </span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="co"># 444.003   0.226 449.960 </span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a><span class="co"># saveRDS(EMIBD9,"EMIBD9.rds")</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a><span class="co"># read output</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>EMIBD9 <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"./data/EMIBD9.rds"</span>)</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>EMIBD9_col <span class="ot">&lt;-</span> EMIBD9<span class="sc">$</span>rel</span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>EMIBD9_col[<span class="fu">upper.tri</span>(EMIBD9_col)] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>EMIBD9_col <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">as.table</span>(<span class="fu">as.matrix</span>(EMIBD9_col)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="gcta" class="level3">
<h3 class="anchored" data-anchor-id="gcta">GCTA</h3>
<p>Relatedness estimation using a genetic relationship matrix (<a href="https://doi.org/10.1016/j.ajhg.2011.02.002">Lee et al., 2011</a>) as implemented in the program GCTA (<a href="10.1016/j.ajhg.2010.11.011">Yang et al., 2011</a>).</p>
<div class="cell" data-hash="session11_cache/html/unnamed-chunk-8_c1439372d5dbe10f7a5f1084e6952cea">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># function to read the GRM binary file</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>ReadGRMBin <span class="ot">&lt;-</span> <span class="cf">function</span>(prefix, <span class="at">AllN=</span><span class="cn">FALSE</span>, <span class="at">size=</span><span class="dv">4</span>){</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>  sum_i <span class="ot">&lt;-</span> <span class="cf">function</span>(i){</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="fu">sum</span>(<span class="dv">1</span><span class="sc">:</span>i))</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>  BinFileName <span class="ot">&lt;-</span> <span class="fu">paste</span>(prefix,<span class="st">".grm.bin"</span>,<span class="at">sep=</span><span class="st">""</span>)</span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>  NFileName <span class="ot">&lt;-</span> <span class="fu">paste</span>(prefix,<span class="st">".grm.N.bin"</span>,<span class="at">sep=</span><span class="st">""</span>)</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>  IDFileName <span class="ot">&lt;-</span> <span class="fu">paste</span>(prefix,<span class="st">".grm.id"</span>,<span class="at">sep=</span><span class="st">""</span>)</span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>  id <span class="ot">&lt;-</span>  <span class="fu">read.table</span>(IDFileName)</span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a>  n <span class="ot">&lt;-</span> <span class="fu">dim</span>(id)[<span class="dv">1</span>]</span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a>  BinFile <span class="ot">&lt;-</span> <span class="fu">file</span>(BinFileName, <span class="st">"rb"</span>);</span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a>  grm <span class="ot">&lt;-</span> <span class="fu">readBin</span>(BinFile, <span class="at">n =</span>n<span class="sc">*</span>(n<span class="sc">+</span><span class="dv">1</span>)<span class="sc">/</span><span class="dv">2</span>, <span class="at">what=</span><span class="fu">numeric</span>(<span class="dv">0</span>), <span class="at">size=</span>size)</span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a>  NFile<span class="ot">=</span><span class="fu">file</span>(NFileName, <span class="st">"rb"</span>);</span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(AllN<span class="sc">==</span><span class="cn">TRUE</span>){</span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a>    N <span class="ot">&lt;-</span> <span class="fu">readBin</span>(NFile, <span class="at">n =</span> n <span class="sc">*</span> (n<span class="sc">+</span><span class="dv">1</span>)<span class="sc">/</span><span class="dv">2</span>, <span class="at">what=</span><span class="fu">numeric</span>(<span class="dv">0</span>), <span class="at">size=</span>size)</span>
<span id="cb31-17"><a href="#cb31-17" aria-hidden="true" tabindex="-1"></a>  }<span class="cf">else</span>{</span>
<span id="cb31-18"><a href="#cb31-18" aria-hidden="true" tabindex="-1"></a>    N <span class="ot">&lt;-</span> <span class="fu">readBin</span>(NFile, <span class="at">n=</span><span class="dv">1</span>, <span class="at">what=</span><span class="fu">numeric</span>(<span class="dv">0</span>), <span class="at">size=</span>size)</span>
<span id="cb31-19"><a href="#cb31-19" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb31-20"><a href="#cb31-20" aria-hidden="true" tabindex="-1"></a>  i <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span>n, sum_i)</span>
<span id="cb31-21"><a href="#cb31-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(<span class="at">diag =</span> grm[i], <span class="at">off =</span> grm[<span class="sc">-</span>i], <span class="at">id =</span> id, <span class="at">N =</span> N))</span>
<span id="cb31-22"><a href="#cb31-22" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb31-23"><a href="#cb31-23" aria-hidden="true" tabindex="-1"></a><span class="co"># using dummy chromosome and position information</span></span>
<span id="cb31-24"><a href="#cb31-24" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span>chromosome <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(<span class="fu">as.character</span>(<span class="st">"1"</span>))</span>
<span id="cb31-25"><a href="#cb31-25" aria-hidden="true" tabindex="-1"></a>t2<span class="sc">$</span>position <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nLoc</span>(t2)</span>
<span id="cb31-26"><a href="#cb31-26" aria-hidden="true" tabindex="-1"></a><span class="co"># converting to BED format using PLINK</span></span>
<span id="cb31-27"><a href="#cb31-27" aria-hidden="true" tabindex="-1"></a><span class="fu">gl2plink</span>(t2,</span>
<span id="cb31-28"><a href="#cb31-28" aria-hidden="true" tabindex="-1"></a>         <span class="at">bed.files =</span> <span class="cn">TRUE</span>,</span>
<span id="cb31-29"><a href="#cb31-29" aria-hidden="true" tabindex="-1"></a>         <span class="at">outpath =</span> <span class="fu">getwd</span>(),</span>
<span id="cb31-30"><a href="#cb31-30" aria-hidden="true" tabindex="-1"></a>         <span class="at">plink.bin.path =</span> path.plink)</span>
<span id="cb31-31"><a href="#cb31-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-32"><a href="#cb31-32" aria-hidden="true" tabindex="-1"></a><span class="co"># running GCTA</span></span>
<span id="cb31-33"><a href="#cb31-33" aria-hidden="true" tabindex="-1"></a><span class="fu">system</span>(<span class="fu">paste</span>(path.gcta,<span class="st">"--bfile gl_plink --make-grm-bin --out gcta_grm"</span>))</span>
<span id="cb31-34"><a href="#cb31-34" aria-hidden="true" tabindex="-1"></a><span class="co"># reading output</span></span>
<span id="cb31-35"><a href="#cb31-35" aria-hidden="true" tabindex="-1"></a>grm_GCTA <span class="ot">&lt;-</span> <span class="fu">ReadGRMBin</span>(<span class="st">"gcta_grm"</span>)</span>
<span id="cb31-36"><a href="#cb31-36" aria-hidden="true" tabindex="-1"></a><span class="co"># formatting data</span></span>
<span id="cb31-37"><a href="#cb31-37" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">nrow =</span> <span class="fu">nrow</span>(grm_GCTA<span class="sc">$</span>id),<span class="at">ncol =</span> <span class="fu">nrow</span>(grm_GCTA<span class="sc">$</span>id) )</span>
<span id="cb31-38"><a href="#cb31-38" aria-hidden="true" tabindex="-1"></a>mat[<span class="fu">upper.tri</span>(mat, <span class="at">diag =</span> <span class="cn">FALSE</span>)] <span class="ot">&lt;-</span> grm_GCTA<span class="sc">$</span>off</span>
<span id="cb31-39"><a href="#cb31-39" aria-hidden="true" tabindex="-1"></a>mat[<span class="fu">lower.tri</span>(mat)] <span class="ot">&lt;-</span> <span class="fu">t</span>(mat)[<span class="fu">lower.tri</span>(mat)]</span>
<span id="cb31-40"><a href="#cb31-40" aria-hidden="true" tabindex="-1"></a><span class="fu">diag</span>(mat) <span class="ot">&lt;-</span> grm_GCTA<span class="sc">$</span>diag</span>
<span id="cb31-41"><a href="#cb31-41" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(mat) <span class="ot">&lt;-</span> grm_GCTA<span class="sc">$</span>id<span class="sc">$</span>V2</span>
<span id="cb31-42"><a href="#cb31-42" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(mat) <span class="ot">&lt;-</span> grm_GCTA<span class="sc">$</span>id<span class="sc">$</span>V2</span>
<span id="cb31-43"><a href="#cb31-43" aria-hidden="true" tabindex="-1"></a>order_mat <span class="ot">&lt;-</span> <span class="fu">colnames</span>(mat)[<span class="fu">order</span>(<span class="fu">colnames</span>(mat))]</span>
<span id="cb31-44"><a href="#cb31-44" aria-hidden="true" tabindex="-1"></a>mat <span class="ot">&lt;-</span> mat[order_mat, order_mat]</span>
<span id="cb31-45"><a href="#cb31-45" aria-hidden="true" tabindex="-1"></a>GCTA_col <span class="ot">&lt;-</span> mat</span>
<span id="cb31-46"><a href="#cb31-46" aria-hidden="true" tabindex="-1"></a>GCTA_col[<span class="fu">upper.tri</span>(GCTA_col)] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb31-47"><a href="#cb31-47" aria-hidden="true" tabindex="-1"></a>GCTA_col <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">as.table</span>(<span class="fu">as.matrix</span>(GCTA_col)))</span>
<span id="cb31-48"><a href="#cb31-48" aria-hidden="true" tabindex="-1"></a><span class="co"># dividing by two to get kinship </span></span>
<span id="cb31-49"><a href="#cb31-49" aria-hidden="true" tabindex="-1"></a>GCTA_col<span class="sc">$</span>Freq <span class="ot">&lt;-</span> GCTA_col<span class="sc">$</span>Freq<span class="sc">/</span><span class="dv">2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="genomic-relationship-matrix-grm" class="level3">
<h3 class="anchored" data-anchor-id="genomic-relationship-matrix-grm">Genomic relationship matrix (GRM)</h3>
<p>Relatedness estimation using an additive relationship matrix (<a href="https://doi.org/10.1534/g3.112.004259">Endelman &amp; Jannink, 2012</a>) as implemented in the R package rrBLUP (<a href="https://doi.org/10.3835/plantgenome2011.08.0024">Endelman, 2011</a>).</p>
<p>The GRM is an estimate of the proportion of alleles that two individuals have in common. It is generated by estimating the covariance of the genotypes between two individuals, i.e.&nbsp;how much genotypes in the two individuals correspond with each other. This covariance depends on the probability that alleles at a random locus are identical by state (IBS). Two alleles are IBS if they represent the same allele. Two alleles are identical by descent (IBD) if one is a physical copy of the other or if they are both physical copies of the same ancestral allele. Note that IBD is complicated to determine. IBD implies IBS, but not conversely. However, as the number of SNPs in a dataset increases, the mean probability of IBS approaches the mean probability of IBD.</p>
<div class="cell" data-hash="session11_cache/html/unnamed-chunk-9_67d08a3d5af5234bcb044f60f7111526">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>GRM <span class="ot">&lt;-</span> <span class="fu">gl.grm</span>(t2,<span class="at">palette_convergent =</span> <span class="fu">gl.colors</span>(<span class="at">type=</span><span class="st">"con"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># formatting data</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>order_grm <span class="ot">&lt;-</span> <span class="fu">colnames</span>(GRM)[<span class="fu">order</span>(<span class="fu">colnames</span>(GRM))]</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>GRM <span class="ot">&lt;-</span> GRM[order_grm, order_grm]</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>GRM_col <span class="ot">&lt;-</span> GRM</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>GRM_col[<span class="fu">upper.tri</span>(GRM_col)] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>GRM_col <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">as.table</span>(<span class="fu">as.matrix</span>(GRM_col)))</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a><span class="co"># dividing by two to get kinship </span></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>GRM_col<span class="sc">$</span>Freq <span class="ot">&lt;-</span> GRM_col<span class="sc">$</span>Freq<span class="sc">/</span><span class="dv">2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plotting" class="level3">
<h3 class="anchored" data-anchor-id="plotting">Plotting</h3>
<div class="cell" data-hash="session11_cache/html/unnamed-chunk-10_dae62da9efb8845880e1f5136f770938">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Sibling/Parent-Offspring</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"./functions/gl.grm.network_2.R"</span>)</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>mat_coan <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">'./data/session11coancestry_mat_coan.rds'</span>)</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>grm_sib <span class="ot">&lt;-</span> <span class="fu">gl.grm.network_2</span>(mat_coan, t2,</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>                            <span class="at">node.size =</span> <span class="dv">10</span>,</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>                            <span class="at">method_relatedness =</span> <span class="st">"fr"</span>,</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>                            <span class="at">relatedness_factor =</span> <span class="fl">0.5</span>,</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>                            <span class="at">palette_discrete =</span> <span class="fu">gl.colors</span>(<span class="at">type =</span> <span class="st">"dis"</span>),</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>                            <span class="at">method =</span> <span class="st">"kk"</span>,</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>                            <span class="at">title=</span><span class="st">"Coancestry relatedness &gt; 0.5 </span><span class="sc">\n</span><span class="st">(parent–offspring &amp; full sibs)"</span></span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Half-sibling</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>grm_half_sib <span class="ot">&lt;-</span> <span class="fu">gl.grm.network_2</span>(mat_coan, t2,</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">node.size =</span> <span class="dv">10</span>,</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">method_relatedness =</span> <span class="st">"fr"</span>,</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">relatedness_factor =</span> <span class="fl">0.25</span>,</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">palette_discrete =</span> <span class="fu">gl.colors</span>(<span class="at">type =</span> <span class="st">"dis"</span>),</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">method =</span> <span class="st">"kk"</span>,</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">title=</span><span class="st">"Coancestry relatedness &gt; 0.25 </span><span class="sc">\n</span><span class="st">(parent–offspring &amp; full sibs + half sibs)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-10-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># First cousin</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>grm_first_cos <span class="ot">&lt;-</span> <span class="fu">gl.grm.network_2</span>(mat_coan, t2,</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">node.size =</span> <span class="dv">10</span>,</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">method_relatedness =</span> <span class="st">"fr"</span>,</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">relatedness_factor =</span> <span class="fl">0.125</span>,</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">palette_discrete =</span> <span class="fu">gl.colors</span>(<span class="at">type =</span> <span class="st">"dis"</span>),</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">method =</span> <span class="st">"kk"</span>,</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">title=</span><span class="st">"Coancestry relatedness &gt; 0.125 </span><span class="sc">\n</span><span class="st">(parent–offspring &amp; full sibs + half sibs + first cousins) "</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-10-3.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell" data-hash="session11_cache/html/unnamed-chunk-12_c031b1c90420fc2f479445832715771c">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># TABLE</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>rel_cal <span class="ot">&lt;-</span> <span class="fu">cbind</span>(coan_col,EMIBD9_col<span class="sc">$</span>Freq,GCTA_col<span class="sc">$</span>Freq,GRM_col<span class="sc">$</span>Freq)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>rel_cal <span class="ot">&lt;-</span> rel_cal[<span class="fu">complete.cases</span>(rel_cal<span class="sc">$</span>Freq),]</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>rel_cal <span class="ot">&lt;-</span> rel_cal[<span class="fu">order</span>(rel_cal<span class="sc">$</span>Freq,<span class="at">decreasing =</span> <span class="cn">TRUE</span>),]</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(rel_cal) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ind1"</span>,<span class="st">"ind2"</span>,<span class="st">"Coancestry"</span>,<span class="st">"EMIBD9"</span>,<span class="st">"GCTA"</span>,<span class="st">"rrBLUP"</span>)</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>rel_cal <span class="ot">&lt;-</span> rel_cal[<span class="fu">which</span>(rel_cal<span class="sc">$</span>Coancestry <span class="sc">&gt;=</span> <span class="fl">0.05</span> <span class="sc">&amp;</span> </span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>                           rel_cal<span class="sc">$</span>EMIBD9 <span class="sc">&gt;=</span> <span class="fl">0.05</span> <span class="sc">&amp;</span></span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>                           rel_cal<span class="sc">$</span>GCTA <span class="sc">&gt;=</span> <span class="fl">0.05</span> <span class="sc">&amp;</span></span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a>                           rel_cal<span class="sc">$</span>rrBLUP <span class="sc">&gt;=</span><span class="fl">0.05</span>),]</span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a>rel_plot_2 <span class="ot">&lt;-</span> <span class="fu">melt</span>(rel_cal, <span class="at">id.vars =</span> <span class="fu">c</span>(<span class="st">"ind1"</span>,<span class="st">"ind2"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Box and violin plots showing the distribution of kinship values from four methods. Relatedness estimates from Coancestry, GCTA and rrBLUP methods were divided by two to obtain kinship estimates. Kinship values shown in the plot, were filtered such as the minimum kinship value in all the datasets was &gt;0.05 for any pairwise relationship.</p>
<div class="cell" data-hash="session11_cache/html/unnamed-chunk-13_c73086f8adc68e04b8f534c466f34821">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(rel_plot_2,<span class="fu">aes</span>(<span class="at">y=</span>value,<span class="at">x=</span>variable,<span class="at">color =</span> variable,<span class="at">fill =</span> variable)) <span class="sc">+</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">position=</span><span class="fu">position_jitterdodge</span>(<span class="at">dodge.width=</span><span class="dv">1</span>),<span class="at">show.legend =</span> F) <span class="sc">+</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">alpha=</span><span class="fl">0.1</span>,<span class="at">scale =</span> <span class="st">"count"</span>)<span class="sc">+</span></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">alpha=</span><span class="fl">0.35</span>)<span class="sc">+</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">16</span>) <span class="sc">+</span></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>( <span class="at">legend.position =</span> <span class="st">"bottom"</span>,</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">axis.ticks.x=</span><span class="fu">element_blank</span>() ,</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">axis.text.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">axis.title.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">legend.title=</span><span class="fu">element_blank</span>(),</span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">legend.text=</span><span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">18</span>),</span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a>         <span class="at">legend.key.height=</span><span class="fu">unit</span>(<span class="dv">1</span>, <span class="st">"cm"</span>))<span class="sc">+</span></span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Kiship"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="feathers-of-laughter-unveiling-the-mysteries-of-the-amazonian-giggletuft-hilaroplumis-amazoniana" class="level2">
<h2 class="anchored" data-anchor-id="feathers-of-laughter-unveiling-the-mysteries-of-the-amazonian-giggletuft-hilaroplumis-amazoniana">Feathers of Laughter: Unveiling the Mysteries of the Amazonian Giggletuft (Hilaroplumis amazoniana)”</h2>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="./images/New_bird.png" class="img-fluid figure-img" width="300"></p>
<figcaption class="figure-caption"><em>Amazonian Giggletuft (Hilaroplumis amazoniana)</em></figcaption>
</figure>
</div>
<p>A new bird species has recently been identified in the Amazon region, characterized by its distinctive and intriguing appearance. Scientists are keen to explore the possibility of multiple paternity occurrences within clutches for this species. To delve into this question, the research team employs SNP data to conduct relatedness analyses. Egg samples have been collected from three separate clutches.</p>
<p>Understanding of mating dynamics of this species in the wild may help shape future management strategies, particularly in captive breeding programs. By adjusting ratios of males to females based on the mating system, we may be able to mimic wild matings as closely as possible.</p>
<div class="cell" data-hash="session11_cache/html/unnamed-chunk-14_77372764e7f093e8e31b2f5b4ee8a8ac">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># loading filtered file</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>new_bird <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"./data/new_bird.rds"</span>) </span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA visualization. </span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>pcoa_new_bird <span class="ot">&lt;-</span> <span class="fu">gl.pcoa</span>(new_bird)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># you can zoom in or out and move in all directions. you can also </span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="co"># unselect/select populations if you click in the labels of the populations in </span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="co"># the legend of the plot.</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a><span class="fu">gl.pcoa.plot</span>(pcoa_new_bird, new_bird,<span class="at">zaxis =</span> <span class="dv">3</span>, <span class="at">pt.size =</span> <span class="dv">4</span>)</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a><span class="co"># a relatedness matrix. </span></span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>new_bird_grm <span class="ot">&lt;-</span> <span class="fu">gl.grm</span>(new_bird,<span class="at">palette_convergent =</span> <span class="fu">gl.colors</span>(<span class="at">type=</span><span class="st">"con"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-14-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<section id="coancestry-1" class="level3">
<h3 class="anchored" data-anchor-id="coancestry-1">COANCESTRY</h3>
<div class="cell" data-hash="session11_cache/html/unnamed-chunk-15_c9d43e528881ea73610f937ff542f4fb">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>new_bird_related <span class="ot">&lt;-</span> <span class="fu">gl2related</span>(new_bird)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="co"># The mean relatedness for each clutch is calculated, with 1,000 bootstrap interactions performed to determine 95% confidence intervals.</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>res_coancestry <span class="ot">&lt;-</span> <span class="fu">coancestry</span>(new_bird_related,<span class="at">wang=</span><span class="dv">2</span>,<span class="at">ci95.num.bootstrap =</span> <span class="dv">1000</span>)</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>res_coancestry_2 <span class="ot">&lt;-</span> res_coancestry<span class="sc">$</span>relatedness[,<span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">6</span>)]</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>ind_1 <span class="ot">&lt;-</span> <span class="fu">rbindlist</span>(<span class="fu">lapply</span>(<span class="fu">lapply</span>(<span class="fu">str_split</span>(res_coancestry_2<span class="sc">$</span>ind1.id,<span class="at">pattern =</span> <span class="st">"_"</span>),t),as.data.frame))</span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>ind_2 <span class="ot">&lt;-</span> <span class="fu">rbindlist</span>(<span class="fu">lapply</span>(<span class="fu">lapply</span>(<span class="fu">str_split</span>(res_coancestry_2<span class="sc">$</span>ind2.id,<span class="at">pattern =</span> <span class="st">"_"</span>),t),as.data.frame))</span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>analysis_coancestry <span class="ot">&lt;-</span> <span class="fu">cbind</span>(ind_1,ind_2,res_coancestry_2<span class="sc">$</span>wang)</span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(analysis_coancestry) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"clutch_1"</span>,<span class="st">"ind_1"</span>,<span class="st">"clutch_2"</span>,<span class="st">"ind_2"</span>,<span class="st">"relatedness"</span>)</span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>analysis_coancestry<span class="sc">$</span>same_clutch <span class="ot">&lt;-</span> analysis_coancestry<span class="sc">$</span>clutch_1 <span class="sc">==</span> analysis_coancestry<span class="sc">$</span>clutch_2</span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>analysis_coancestry_2 <span class="ot">&lt;-</span> analysis_coancestry[<span class="fu">which</span>(analysis_coancestry<span class="sc">$</span>same_clutch<span class="sc">==</span>T),]</span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a>analysis_coancestry_2<span class="sc">$</span>clutch_1 <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(analysis_coancestry_2<span class="sc">$</span>clutch_1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plotting-1" class="level3">
<h3 class="anchored" data-anchor-id="plotting-1">Plotting</h3>
<p>All the individuals had a coefficient of relatedness within clutches greater than 0.25, with means within each clutch close to 0.50. We can therefore conclude that all individuals within each clutch were full siblings who share the same mother and father.</p>
<div class="cell" data-hash="session11_cache/html/unnamed-chunk-17_1a67c31de9422215a5c85e2acea30e91">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>pal <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"lightgoldenrod"</span>,<span class="st">"darkturquoise"</span>,<span class="st">"deeppink"</span>)</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(analysis_coancestry_2, <span class="fu">aes</span>(<span class="at">x=</span>clutch_1 ,<span class="at">y =</span> relatedness,<span class="at">group=</span>clutch_1)) <span class="sc">+</span>  </span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_violin</span>(<span class="fu">aes</span>(<span class="at">fill=</span>clutch_1)) <span class="sc">+</span></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">fill=</span>clutch_1),<span class="at">alpha=</span><span class="fl">0.5</span>,<span class="at">position =</span> <span class="fu">position_dodge</span>(<span class="at">width=</span><span class="fl">0.8</span>),<span class="at">show.legend =</span> F,  <span class="at">notch =</span> <span class="cn">FALSE</span>)<span class="sc">+</span></span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">colour=</span>clutch_1),<span class="at">position=</span><span class="fu">position_jitterdodge</span>(<span class="at">dodge.width=</span><span class="dv">1</span>),<span class="at">show.legend =</span> F) <span class="sc">+</span></span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> pal)<span class="sc">+</span></span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">22</span>) <span class="sc">+</span></span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>( <span class="at">legend.position =</span> <span class="st">"bottom"</span>,</span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a>           <span class="at">axis.ticks.x=</span><span class="fu">element_blank</span>() ,</span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a>           <span class="at">axis.text.x =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a>           <span class="at">axis.title.x=</span><span class="fu">element_blank</span>(),</span>
<span id="cb42-14"><a href="#cb42-14" aria-hidden="true" tabindex="-1"></a>           <span class="at">legend.title=</span><span class="fu">element_blank</span>(),</span>
<span id="cb42-15"><a href="#cb42-15" aria-hidden="true" tabindex="-1"></a>           <span class="at">legend.text=</span><span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">18</span>),</span>
<span id="cb42-16"><a href="#cb42-16" aria-hidden="true" tabindex="-1"></a>           <span class="at">legend.key.height=</span><span class="fu">unit</span>(<span class="fl">1.5</span>, <span class="st">"cm"</span>))</span>
<span id="cb42-17"><a href="#cb42-17" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="from-the-brink-and-back-the-arabian-oryxs-oman-odyssey" class="level2">
<h2 class="anchored" data-anchor-id="from-the-brink-and-back-the-arabian-oryxs-oman-odyssey">From the Brink and Back: The Arabian Oryx’s Oman Odyssey</h2>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="./images/Arabian_Oryx.png" class="img-fluid figure-img" width="300"></p>
<figcaption class="figure-caption"><em>Arabian oryx (Oryx leucoryx)</em></figcaption>
</figure>
</div>
<p>The Arabian oryx, once extinct in the wild by 1972, found a lifeline through global zoo and private collection breeding efforts. Thanks to “The World Herd” in the Phoenix Zoo and Saudi Arabian private reserves, these oryxes were reintroduced into the wild. Now, we’ve got about 1,000 roaming free and another 6,000-7,000 in captivity, marking a historic comeback from extinction to vulnerable status. Quite the turnaround for these majestic creatures in Oman!</p>
<p><img src="./images/oryx_timeline.png" class="img-fluid" style="width:100.0%" data-fig-align="left" alt="Arabian oryx (Oryx leucoryx)"> Timeline of the Arabian oryx reintroduction program in Oman and establishment of the “World Herd”. Top row displays the years when the animals were captured, translocated or released and the country where the event took place. Middle boxes show the major event and activity at that time. Bottom boxes show the number and sex of animals which had been used.</p>
<section id="load-dataset-and-filtering" class="level3">
<h3 class="anchored" data-anchor-id="load-dataset-and-filtering">Load dataset and filtering</h3>
<div class="cell" data-hash="session11_cache/html/unnamed-chunk-18_19d4f46030fa88ef2b05ee68d26ff6c7">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>oryx_gl <span class="ot">&lt;-</span> <span class="fu">gl.load</span>(<span class="st">"./data/oryx.rds"</span>)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>t1_dup <span class="ot">&lt;-</span> oryx_gl</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>res_dup <span class="ot">&lt;-</span> <span class="fu">gl.report.replicates</span>(t1_dup,<span class="at">perc_geno =</span> <span class="fl">0.925</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>res_dup2 <span class="ot">&lt;-</span> res_dup<span class="sc">$</span>table.rep</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Ordering pairwise estimates </span></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>res_dup2 <span class="ot">&lt;-</span> res_dup2[<span class="fu">order</span>(res_dup2<span class="sc">$</span>perc,<span class="at">decreasing =</span> <span class="cn">TRUE</span>),]</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>res_dup_2 <span class="ot">&lt;-</span> res_dup<span class="sc">$</span>ind.list.rep</span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>res_dup_2 <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res_dup_2,<span class="cf">function</span>(x){</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste</span>(x,<span class="at">collapse =</span> <span class="st">", "</span>)</span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>res_dup_2 <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="fu">names</span>(res_dup_2),res_dup_2,<span class="at">sep =</span> <span class="st">", "</span>)</span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>res_dup_2 <span class="ot">&lt;-</span> res_dup_2[<span class="fu">order</span>(res_dup_2)]</span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(res_dup_2,<span class="at">sep=</span><span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-13"><a href="#cb44-13" aria-hidden="true" tabindex="-1"></a><span class="co">#removing duplicates </span></span>
<span id="cb44-14"><a href="#cb44-14" aria-hidden="true" tabindex="-1"></a>oryx_gl <span class="ot">&lt;-</span> <span class="fu">gl.drop.ind</span>(oryx_gl,<span class="at">ind.list =</span> res_dup<span class="sc">$</span>ind.list.drop)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="grm" class="level3">
<h3 class="anchored" data-anchor-id="grm">GRM</h3>
<p>Heatmap of the probabilities of identity by descent in the group formed by the herds of WWR-Oman and WWR-UAE. As described by Endelman &amp; Jannink (68), each diagonal element has a mean that equals 1+f, where f is the inbreeding coefficient (i.e.&nbsp;the probability that the two alleles at a randomly chosen locus are IBD from the base population). As this probability lies between 0 and 1, the diagonal elements range from 1 to 2. Because the inbreeding coefficients are expressed relative to the current population, the mean of the off-diagonal elements is -(1+f)/n, where n is the number of loci. Yellow and red colours indicate those individuals that are more related to each other. The identification number of each individual is shown in the margins of the figure, where the last letter denotes whether the individual is male (M) or female (F).</p>
<div class="cell" data-hash="session11_cache/html/unnamed-chunk-19_98519d95819507f86c6e2b12f8c55103">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>oryx_related_matrix_all <span class="ot">&lt;-</span> <span class="fu">gl.grm</span>(oryx_gl,<span class="at">palette_convergent =</span> <span class="fu">gl.colors</span>(<span class="at">type=</span><span class="st">"con"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">gl.grm.network</span>(oryx_related_matrix_all,</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>               oryx_gl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-19-2.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="sequoia" class="level3">
<h3 class="anchored" data-anchor-id="sequoia">SEQUOIA</h3>
<div class="cell" data-hash="session11_cache/html/unnamed-chunk-20_51317a993c72a954b9091011b397320b">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>LifeHistData_2 <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(oryx_gl<span class="sc">$</span>ind.names)</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>LifeHistData_2<span class="sc">$</span>Sex <span class="ot">&lt;-</span> <span class="fu">as.character</span>(oryx_gl<span class="sc">$</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>Gender)</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>LifeHistData_2[LifeHistData_2<span class="sc">==</span><span class="st">"Female"</span>] <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>LifeHistData_2[LifeHistData_2<span class="sc">==</span><span class="st">"Male"</span>] <span class="ot">&lt;-</span> <span class="dv">2</span></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>LifeHistData_2<span class="sc">$</span>Sex <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(LifeHistData_2<span class="sc">$</span>Sex)</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>LifeHistData_2<span class="sc">$</span>BirthYear <span class="ot">&lt;-</span> <span class="fu">as.character</span>(oryx_gl<span class="sc">$</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>Age)</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>LifeHistData_2[LifeHistData_2<span class="sc">==</span><span class="st">"Old"</span>] <span class="ot">&lt;-</span> <span class="dv">2000</span></span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>LifeHistData_2[LifeHistData_2<span class="sc">==</span><span class="st">"Adult"</span>] <span class="ot">&lt;-</span> <span class="dv">2005</span></span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>LifeHistData_2[LifeHistData_2<span class="sc">==</span><span class="st">"Juvenile"</span>] <span class="ot">&lt;-</span> <span class="dv">2010</span></span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>LifeHistData_2[LifeHistData_2<span class="sc">==</span><span class="st">"Calf"</span>] <span class="ot">&lt;-</span> <span class="dv">2015</span></span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a>LifeHistData_2<span class="sc">$</span>BirthYear <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(LifeHistData_2<span class="sc">$</span>BirthYear)</span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-14"><a href="#cb47-14" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(LifeHistData_2)<span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ID"</span>,<span class="st">"Sex"</span>,<span class="st">"BirthYear"</span>)</span>
<span id="cb47-15"><a href="#cb47-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-16"><a href="#cb47-16" aria-hidden="true" tabindex="-1"></a>oryx_sequoia <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">as.data.frame</span>(oryx_gl))</span>
<span id="cb47-17"><a href="#cb47-17" aria-hidden="true" tabindex="-1"></a>oryx_sequoia[<span class="fu">is.na</span>(oryx_sequoia)] <span class="ot">&lt;-</span> <span class="st">"-9"</span></span>
<span id="cb47-18"><a href="#cb47-18" aria-hidden="true" tabindex="-1"></a>oryx_sequoia_2 <span class="ot">&lt;-</span> <span class="fu">cbind</span>(oryx_gl<span class="sc">$</span>ind.names,oryx_sequoia)</span>
<span id="cb47-19"><a href="#cb47-19" aria-hidden="true" tabindex="-1"></a>oryx_sequoia_2 <span class="ot">&lt;-</span> <span class="fu">mapply</span>(oryx_sequoia_2, <span class="at">FUN=</span>as.numeric)</span>
<span id="cb47-20"><a href="#cb47-20" aria-hidden="true" tabindex="-1"></a>oryx_sequoia_2 <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">data=</span>oryx_sequoia_2, <span class="at">ncol=</span><span class="fu">ncol</span>(oryx_sequoia)<span class="sc">+</span><span class="dv">1</span>, <span class="at">nrow=</span><span class="fu">nrow</span>(oryx_sequoia))</span>
<span id="cb47-21"><a href="#cb47-21" aria-hidden="true" tabindex="-1"></a>oryx_sequoia_2 <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(oryx_sequoia_2)</span>
<span id="cb47-22"><a href="#cb47-22" aria-hidden="true" tabindex="-1"></a>GenoM <span class="ot">&lt;-</span> <span class="fu">GenoConvert</span>(<span class="at">InFile=</span>oryx_sequoia_2, <span class="at">InFormat=</span><span class="st">"seq"</span>,<span class="at">Missing=</span><span class="st">"-9"</span>,<span class="at">IDcol=</span><span class="dv">1</span>)</span>
<span id="cb47-23"><a href="#cb47-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-24"><a href="#cb47-24" aria-hidden="true" tabindex="-1"></a>GenoM_2 <span class="ot">&lt;-</span> GenoM</span>
<span id="cb47-25"><a href="#cb47-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-26"><a href="#cb47-26" aria-hidden="true" tabindex="-1"></a>sequoia_res <span class="ot">&lt;-</span> <span class="fu">sequoia</span>(<span class="at">GenoM =</span> GenoM_2,</span>
<span id="cb47-27"><a href="#cb47-27" aria-hidden="true" tabindex="-1"></a>                       <span class="at">Plot =</span> <span class="cn">FALSE</span>,</span>
<span id="cb47-28"><a href="#cb47-28" aria-hidden="true" tabindex="-1"></a>                       <span class="at">UseAge =</span> <span class="cn">FALSE</span>,</span>
<span id="cb47-29"><a href="#cb47-29" aria-hidden="true" tabindex="-1"></a>                       <span class="at">MaxSibshipSize=</span><span class="dv">20</span>)</span>
<span id="cb47-30"><a href="#cb47-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-31"><a href="#cb47-31" aria-hidden="true" tabindex="-1"></a>sequoia_res_2 <span class="ot">&lt;-</span> <span class="fu">GetMaybeRel</span>(<span class="at">GenoM =</span> GenoM_2, </span>
<span id="cb47-32"><a href="#cb47-32" aria-hidden="true" tabindex="-1"></a>                             <span class="at">LifeHistData =</span>LifeHistData_2)</span>
<span id="cb47-33"><a href="#cb47-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-34"><a href="#cb47-34" aria-hidden="true" tabindex="-1"></a>pedigree_res <span class="ot">&lt;-</span> sequoia_res<span class="sc">$</span>Pedigree</span>
<span id="cb47-35"><a href="#cb47-35" aria-hidden="true" tabindex="-1"></a>pedigree_res<span class="sc">$</span>sex <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">as.character</span>(oryx_gl<span class="sc">$</span>other<span class="sc">$</span>ind.metrics<span class="sc">$</span>Gender),<span class="fu">rep</span>(<span class="st">"Female"</span>,<span class="dv">6</span>),<span class="fu">rep</span>(<span class="st">"Male"</span>,<span class="dv">6</span>))</span>
<span id="cb47-36"><a href="#cb47-36" aria-hidden="true" tabindex="-1"></a>family <span class="ot">&lt;-</span> <span class="fu">pedigree</span>(<span class="at">id=</span>pedigree_res<span class="sc">$</span>id, <span class="at">dadid=</span>pedigree_res<span class="sc">$</span>sire, <span class="at">momid=</span>pedigree_res<span class="sc">$</span>dam,<span class="at">sex=</span>pedigree_res<span class="sc">$</span>sex)</span>
<span id="cb47-37"><a href="#cb47-37" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(family)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session11_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid" width="672"></p>
</div>
</div>


</section>
</section>

</main> <!-- /main -->
<script>

/* update total correct if #webex-total_correct exists */
update_total_correct = function() {
  console.log("webex: update total_correct");

  var t = document.getElementsByClassName("webex-total_correct");
  for (var i = 0; i < t.length; i++) {
    p = t[i].parentElement;
    var correct = p.getElementsByClassName("webex-correct").length;
    var solvemes = p.getElementsByClassName("webex-solveme").length;
    var radiogroups = p.getElementsByClassName("webex-radiogroup").length;
    var selects = p.getElementsByClassName("webex-select").length;

    t[i].innerHTML = correct + " of " + (solvemes + radiogroups + selects) + " correct";
  }
}

/* webex-solution button toggling function */
b_func = function() {
  console.log("webex: toggle hide");

  var cl = this.parentElement.classList;
  if (cl.contains('open')) {
    cl.remove("open");
  } else {
    cl.add("open");
  }
}

/* check answers */
check_func = function() {
  console.log("webex: check answers");

  var cl = this.parentElement.classList;
  if (cl.contains('unchecked')) {
    cl.remove("unchecked");
    this.innerHTML = "Hide Answers";
  } else {
    cl.add("unchecked");
    this.innerHTML = "Show Answers";
  }
}

/* function for checking solveme answers */
solveme_func = function(e) {
  console.log("webex: check solveme");

  var real_answers = JSON.parse(this.dataset.answer);
  var my_answer = this.value;
  var cl = this.classList;
  if (cl.contains("ignorecase")) {
    my_answer = my_answer.toLowerCase();
  }
  if (cl.contains("nospaces")) {
    my_answer = my_answer.replace(/ /g, "")
  }

  if (my_answer == "") {
    cl.remove("webex-correct");
    cl.remove("webex-incorrect");
  } else if (real_answers.includes(my_answer)) {
    cl.add("webex-correct");
    cl.remove("webex-incorrect");
  } else {
    cl.add("webex-incorrect");
    cl.remove("webex-correct");
  }

  // match numeric answers within a specified tolerance
  if(this.dataset.tol > 0){
    var tol = JSON.parse(this.dataset.tol);
    var matches = real_answers.map(x => Math.abs(x - my_answer) < tol)
    if (matches.reduce((a, b) => a + b, 0) > 0) {
      cl.add("webex-correct");
    } else {
      cl.remove("webex-correct");
    }
  }

  // added regex bit
  if (cl.contains("regex")){
    answer_regex = RegExp(real_answers.join("|"))
    if (answer_regex.test(my_answer)) {
      cl.add("webex-correct");
    }
  }

  update_total_correct();
}

/* function for checking select answers */
select_func = function(e) {
  console.log("webex: check select");

  var cl = this.classList

  /* add style */
  cl.remove("webex-incorrect");
  cl.remove("webex-correct");
  if (this.value == "answer") {
    cl.add("webex-correct");
  } else if (this.value != "blank") {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

/* function for checking radiogroups answers */
radiogroups_func = function(e) {
  console.log("webex: check radiogroups");

  var checked_button = document.querySelector('input[name=' + this.id + ']:checked');
  var cl = checked_button.parentElement.classList;
  var labels = checked_button.parentElement.parentElement.children;

  /* get rid of styles */
  for (i = 0; i < labels.length; i++) {
    labels[i].classList.remove("webex-incorrect");
    labels[i].classList.remove("webex-correct");
  }

  /* add style */
  if (checked_button.value == "answer") {
    cl.add("webex-correct");
  } else {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

window.onload = function() {
  console.log("webex onload");
  /* set up solution buttons */
  var buttons = document.getElementsByTagName("button");

  for (var i = 0; i < buttons.length; i++) {
    if (buttons[i].parentElement.classList.contains('webex-solution')) {
      buttons[i].onclick = b_func;
    }
  }

  var check_sections = document.getElementsByClassName("webex-check");
  console.log("check:", check_sections.length);
  for (var i = 0; i < check_sections.length; i++) {
    check_sections[i].classList.add("unchecked");

    let btn = document.createElement("button");
    btn.innerHTML = "Show Answers";
    btn.classList.add("webex-check-button");
    btn.onclick = check_func;
    check_sections[i].appendChild(btn);

    let spn = document.createElement("span");
    spn.classList.add("webex-total_correct");
    check_sections[i].appendChild(spn);
  }

  /* set up webex-solveme inputs */
  var solveme = document.getElementsByClassName("webex-solveme");

  for (var i = 0; i < solveme.length; i++) {
    /* make sure input boxes don't auto-anything */
    solveme[i].setAttribute("autocomplete","off");
    solveme[i].setAttribute("autocorrect", "off");
    solveme[i].setAttribute("autocapitalize", "off");
    solveme[i].setAttribute("spellcheck", "false");
    solveme[i].value = "";

    /* adjust answer for ignorecase or nospaces */
    var cl = solveme[i].classList;
    var real_answer = solveme[i].dataset.answer;
    if (cl.contains("ignorecase")) {
      real_answer = real_answer.toLowerCase();
    }
    if (cl.contains("nospaces")) {
      real_answer = real_answer.replace(/ /g, "");
    }
    solveme[i].dataset.answer = real_answer;

    /* attach checking function */
    solveme[i].onkeyup = solveme_func;
    solveme[i].onchange = solveme_func;

    solveme[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  /* set up radiogroups */
  var radiogroups = document.getElementsByClassName("webex-radiogroup");
  for (var i = 0; i < radiogroups.length; i++) {
    radiogroups[i].onchange = radiogroups_func;
  }

  /* set up selects */
  var selects = document.getElementsByClassName("webex-select");
  for (var i = 0; i < selects.length; i++) {
    selects[i].onchange = select_func;
    selects[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  update_total_correct();
}

</script>
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
          // default icon
          link.classList.add("external");
      }
    }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./session10.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Sex Linked Markers</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./session12.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Genetic Structure</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>