<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Kioloa PopGen - 7&nbsp; Natural Selection</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./session08.html" rel="next">
<link href="./session06.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="css.css">
<link rel="stylesheet" href="include/webex.css">
</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./session07.html"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Natural Selection</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Kioloa PopGen</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome!</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./schedule.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Program</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./install.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Install dartRverse</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session01.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Intro to dartR</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session02.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Pop Gen In Conservation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session03.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Sequencing Technologies</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session04.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Data Management, Reproducibility &amp; Integrity</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session05.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Effective Population Size</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session06.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Management of Small Populations</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session07.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Natural Selection</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session08.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Landscape Genetics</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session09.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Lineage Divergence</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session10.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Sex Linked Markers</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session11.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">From Genes to Kin: Dissecting Relatedness &amp; Kinship</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session12.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Genetic Structure</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session13.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Combining Genomic Resources</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./break.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">__________</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./useful.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">More info</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">dartR Tutorials</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#required-packages" id="toc-required-packages" class="nav-link active" data-scroll-target="#required-packages"><em>Required packages</em></a></li>
  <li><a href="#gea-analysis-of-snp-and-environmental-data-using-rda" id="toc-gea-analysis-of-snp-and-environmental-data-using-rda" class="nav-link" data-scroll-target="#gea-analysis-of-snp-and-environmental-data-using-rda"><strong>GEA analysis of SNP and environmental data using RDA</strong></a></li>
  <li><a href="#data-preparation" id="toc-data-preparation" class="nav-link" data-scroll-target="#data-preparation">1. Data preparation</a>
  <ul class="collapse">
  <li><a href="#individual-based" id="toc-individual-based" class="nav-link" data-scroll-target="#individual-based">Individual based</a></li>
  <li><a href="#population-based" id="toc-population-based" class="nav-link" data-scroll-target="#population-based">Population based</a></li>
  <li><a href="#get-environmental-data" id="toc-get-environmental-data" class="nav-link" data-scroll-target="#get-environmental-data">Get environmental data</a></li>
  <li><a href="#extract-site-environmental-data" id="toc-extract-site-environmental-data" class="nav-link" data-scroll-target="#extract-site-environmental-data">Extract site environmental data</a></li>
  <li><a href="#no-missing-data" id="toc-no-missing-data" class="nav-link" data-scroll-target="#no-missing-data">No missing data</a></li>
  <li><a href="#plot-pca" id="toc-plot-pca" class="nav-link" data-scroll-target="#plot-pca">Plot PCA</a></li>
  </ul></li>
  <li><a href="#run-analysis" id="toc-run-analysis" class="nav-link" data-scroll-target="#run-analysis">3. Run analysis</a>
  <ul class="collapse">
  <li><a href="#plot-rda" id="toc-plot-rda" class="nav-link" data-scroll-target="#plot-rda">Plot RDA</a></li>
  </ul></li>
  <li><a href="#identify-candidate-loci" id="toc-identify-candidate-loci" class="nav-link" data-scroll-target="#identify-candidate-loci">4. Identify candidate loci</a>
  <ul class="collapse">
  <li><a href="#individual-level" id="toc-individual-level" class="nav-link" data-scroll-target="#individual-level">Individual level</a></li>
  <li><a href="#plot-rda-1" id="toc-plot-rda-1" class="nav-link" data-scroll-target="#plot-rda-1">Plot RDA</a></li>
  </ul></li>
  <li><a href="#further-study" id="toc-further-study" class="nav-link" data-scroll-target="#further-study"><em>Further Study</em></a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Natural Selection</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p><em>Session Presenters</em></p>
<p><img src="images/Presenters7.png" class="img-fluid"></p>
<section id="required-packages" class="level2">
<h2 class="anchored" data-anchor-id="required-packages"><em>Required packages</em></h2>
<div class="cell" data-hash="session07_cache/html/unnamed-chunk-1_fb5d76809ea623f0834012d3d145e89f">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#devtools::install_github("pygmyperch/melfuR")</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co">#BiocManager::install('qvalue')</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(adegenet)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(LEA)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vegan)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(fmsb)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(psych)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dartRverse)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(melfuR)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sdmpredictors)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sf)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(raster)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(robust)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(qvalue)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>make sure you have the packages installed, see</em> <a href="./install.html">Install dartRverse</a></p>
</section>
<section id="gea-analysis-of-snp-and-environmental-data-using-rda" class="level2">
<h2 class="anchored" data-anchor-id="gea-analysis-of-snp-and-environmental-data-using-rda"><strong>GEA analysis of SNP and environmental data using RDA</strong></h2>
<p>First lets load in some functions we will be using.</p>
<div class="cell" data-hash="session07_cache/html/unnamed-chunk-2_c4811a36e9ad6f2bc418e3f91af2c971">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"utils.R"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="data-preparation" class="level2">
<h2 class="anchored" data-anchor-id="data-preparation">1. Data preparation</h2>
<div class="cell" data-hash="session07_cache/html/unnamed-chunk-3_5b4986dd2aa522727ae9ace2c9035e62">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load genlight object</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"data/Mf5000_gl.RData"</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>Mf5000_gl</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> /// GENLIGHT OBJECT /////////

 // 249 genotypes,  5,000 binary SNPs, size: 1.1 Mb
 12917 (1.04 %) missing data

 // Basic content
   @gen: list of 249 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 // Optional content
   @ind.names:  249 individual labels
   @loc.names:  5000 locus labels
   @pop: population of each individual (group size range: 9-30)
   @other: a list containing: X </code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>Mf5000_gl<span class="sc">@</span>pop</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR
 [19] MBR MBR MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC
 [37] MDC MDC MDC MDC MDC MDC GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL
 [55] GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL
 [73] WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK BEN BEN
 [91] BEN BEN BEN BEN BEN BEN BEN BEN BEN BEN BEN BEN BOG BOG BOG BOG BOG BOG
[109] BOG BOG BOG BOG BOG BOG BOG BOG BOG BOG PEL PEL PEL PEL PEL PEL PEL PEL
[127] PEL GWY GWY GWY GWY GWY GWY GWY GWY GWY GWY GWY GWY DUM DUM DUM DUM DUM
[145] DUM DUM DUM DUM DUM DUM DUM DUM DUM MIB MIB MIB MIB MIB MIB MIB MIB MIB
[163] MIB MIB MIB MIB MIB MIB MIB MIB MIB MIB MIB STG STG STG STG STG STG STG
[181] STG STG STG STG STG STG STG STG STG STG STG STG STG OAK OAK OAK OAK OAK
[199] OAK OAK OAK OAK OAK OAK OAK OAK OAK OAK OAK OAK OAK WAR WAR WAR WAR WAR
[217] WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR KIL KIL KIL
[235] KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL
Levels: BEN BOG DUM GGL GWY KIL MBR MDC MIB OAK PEL STG WAK WAR</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Factors in R can make you question your life choices like nothing else...</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co"># re-order the pop levels to match the order of individuals in your data</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>Mf5000_gl<span class="sc">@</span>pop <span class="ot">&lt;-</span> <span class="fu">factor</span>(Mf5000_gl<span class="sc">@</span>pop, <span class="at">levels =</span> <span class="fu">as.character</span>(<span class="fu">unique</span>(Mf5000_gl<span class="sc">@</span>pop)))</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>Mf5000_gl<span class="sc">@</span>pop</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR
 [19] MBR MBR MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC
 [37] MDC MDC MDC MDC MDC MDC GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL
 [55] GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL
 [73] WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK BEN BEN
 [91] BEN BEN BEN BEN BEN BEN BEN BEN BEN BEN BEN BEN BOG BOG BOG BOG BOG BOG
[109] BOG BOG BOG BOG BOG BOG BOG BOG BOG BOG PEL PEL PEL PEL PEL PEL PEL PEL
[127] PEL GWY GWY GWY GWY GWY GWY GWY GWY GWY GWY GWY GWY DUM DUM DUM DUM DUM
[145] DUM DUM DUM DUM DUM DUM DUM DUM DUM MIB MIB MIB MIB MIB MIB MIB MIB MIB
[163] MIB MIB MIB MIB MIB MIB MIB MIB MIB MIB MIB STG STG STG STG STG STG STG
[181] STG STG STG STG STG STG STG STG STG STG STG STG STG OAK OAK OAK OAK OAK
[199] OAK OAK OAK OAK OAK OAK OAK OAK OAK OAK OAK OAK OAK WAR WAR WAR WAR WAR
[217] WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR KIL KIL KIL
[235] KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL
Levels: MBR MDC GGL WAK BEN BOG PEL GWY DUM MIB STG OAK WAR KIL</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># convert to genind</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>Mf5000.genind <span class="ot">&lt;-</span> <span class="fu">gl2gi</span>(Mf5000_gl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl2gi 
  Processing genlight object with SNP data
Matrix converted.. Prepare genind object...
Completed: gl2gi </code></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>Mf5000.genind</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>/// GENIND OBJECT /////////

 // 249 individuals; 5,000 loci; 10,000 alleles; size: 12.2 Mb

 // Basic content
   @tab:  249 x 10000 matrix of allele counts
   @loc.n.all: number of alleles per locus (range: 2-2)
   @loc.fac: locus factor for the 10000 columns of @tab
   @all.names: list of allele names for each locus
   @ploidy: ploidy of each individual  (range: 2-2)
   @type:  codom
   @call: df2genind(X = xx[, ], sep = "/", ncode = 1, ind.names = x@ind.names, 
    pop = x@pop, NA.char = "-", ploidy = 2)

 // Optional content
   @pop: population of each individual (group size range: 9-30)
   @other: a list containing: X </code></pre>
</div>
</div>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Ordination analyses cannot handle missing data…
</div>
</div>
<div class="callout-body-container callout-body">
<p>Impute missing data</p>
<p><strong>Decision time</strong></p>
<ol type="1">
<li><p>remove loci with missing data?</p></li>
<li><p>remove individuals with missing data?</p></li>
<li><p>impute missing data, ok how?</p>
<p>3a. most common genotype across all data?</p>
<p>3b. most common genotype per site/population/other?</p>
<p>3c. based on population allele frequencies characterised using snmf (admixture)?<br>
</p></li>
</ol>
</div>
</div>
<div class="cell" data-hash="session07_cache/html/unnamed-chunk-4_3e0d27b1c56d8e1315ada091f93c0825">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ADMIXTURE results most likely 6 pops</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>imputed.gi <span class="ot">&lt;-</span> melfuR<span class="sc">::</span><span class="fu">impute.data</span>(Mf5000.genind, <span class="at">K =</span> <span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Total missing data =  1.04 %
[1] "*********************************************************"
[1] " Conversion from the lfmm format to the geno format"
[1] "*********************************************************"

    - number of detected individuals:   249
    - number of detected loci:      5000

The project is saved into :
 dat.snmfProject 

To load the project, use:
 project = load.snmfProject("dat.snmfProject")

To remove the project, use:
 remove.snmfProject("dat.snmfProject")

[1] 42
[1] "*************************************"
[1] "*          create.dataset            *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)                 249
        -L (number of loci)                        5000
        -s (seed random init)                      42
        -r (percentage of masked data)             0.05
        -x (genotype file in .geno format)         C:\repositories\dartR_workshop\kioloa\dat.geno
        -o (output file in .geno format)           C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno

 Write genotype file with masked data, C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno:        OK.

[1] "*************************************"
[1] "* sNMF K = 6  repetition 1      *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)             249
        -L (number of loci)                    5000
        -K (number of ancestral pops)          6
        -x (input file)                        C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno
        -q (individual admixture file)         C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run1/dat_r1.6.Q
        -g (ancestral frequencies file)        C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run1/dat_r1.6.G
        -i (number max of iterations)          200
        -a (regularization parameter)          10
        -s (seed random init)                  42
        -e (tolerance error)                   1E-05
        -p (number of processes)               1
        - diploid

Read genotype file C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno:        OK.


Main algorithm:
    [                                                                           ]
    [=====================================]
Number of iterations: 98

Least-square error: 388910.818678
Write individual ancestry coefficient file C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run1/dat_r1.6.Q:       OK.
Write ancestral allele frequency coefficient file C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run1/dat_r1.6.G:    OK.

[1] "*************************************"
[1] "*    cross-entropy estimation       *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)         249
        -L (number of loci)                5000
        -K (number of ancestral pops)      6
        -x (genotype file)                 C:\repositories\dartR_workshop\kioloa\dat.geno
        -q (individual admixture)          C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run1/dat_r1.6.Q
        -g (ancestral frequencies)         C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run1/dat_r1.6.G
        -i (with masked genotypes)         C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno
        - diploid

Cross-Entropy (all data):    0.515314
Cross-Entropy (masked data):     0.550054
The project is saved into :
 dat.snmfProject 

To load the project, use:
 project = load.snmfProject("dat.snmfProject")

To remove the project, use:
 remove.snmfProject("dat.snmfProject")

[1] 2036314816
[1] "*************************************"
[1] "*          create.dataset            *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)                 249
        -L (number of loci)                        5000
        -s (seed random init)                      2036314816
        -r (percentage of masked data)             0.05
        -x (genotype file in .geno format)         C:\repositories\dartR_workshop\kioloa\dat.geno
        -o (output file in .geno format)           C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno

 Write genotype file with masked data, C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno:        OK.

[1] "*************************************"
[1] "* sNMF K = 6  repetition 2      *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)             249
        -L (number of loci)                    5000
        -K (number of ancestral pops)          6
        -x (input file)                        C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno
        -q (individual admixture file)         C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run2/dat_r2.6.Q
        -g (ancestral frequencies file)        C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run2/dat_r2.6.G
        -i (number max of iterations)          200
        -a (regularization parameter)          10
        -s (seed random init)                  2036314816
        -e (tolerance error)                   1E-05
        -p (number of processes)               1
        - diploid

Read genotype file C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno:        OK.


Main algorithm:
    [                                                                           ]
    [======================================]
Number of iterations: 102

Least-square error: 390122.868041
Write individual ancestry coefficient file C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run2/dat_r2.6.Q:       OK.
Write ancestral allele frequency coefficient file C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run2/dat_r2.6.G:    OK.

[1] "*************************************"
[1] "*    cross-entropy estimation       *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)         249
        -L (number of loci)                5000
        -K (number of ancestral pops)      6
        -x (genotype file)                 C:\repositories\dartR_workshop\kioloa\dat.geno
        -q (individual admixture)          C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run2/dat_r2.6.Q
        -g (ancestral frequencies)         C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run2/dat_r2.6.G
        -i (with masked genotypes)         C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno
        - diploid

Cross-Entropy (all data):    0.51454
Cross-Entropy (masked data):     0.558075
The project is saved into :
 dat.snmfProject 

To load the project, use:
 project = load.snmfProject("dat.snmfProject")

To remove the project, use:
 remove.snmfProject("dat.snmfProject")

[1] 1372102712
[1] "*************************************"
[1] "*          create.dataset            *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)                 249
        -L (number of loci)                        5000
        -s (seed random init)                      1372102712
        -r (percentage of masked data)             0.05
        -x (genotype file in .geno format)         C:\repositories\dartR_workshop\kioloa\dat.geno
        -o (output file in .geno format)           C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno

 Write genotype file with masked data, C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno:        OK.

[1] "*************************************"
[1] "* sNMF K = 6  repetition 3      *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)             249
        -L (number of loci)                    5000
        -K (number of ancestral pops)          6
        -x (input file)                        C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno
        -q (individual admixture file)         C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run3/dat_r3.6.Q
        -g (ancestral frequencies file)        C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run3/dat_r3.6.G
        -i (number max of iterations)          200
        -a (regularization parameter)          10
        -s (seed random init)                  1372102712
        -e (tolerance error)                   1E-05
        -p (number of processes)               1
        - diploid

Read genotype file C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno:        OK.


Main algorithm:
    [                                                                           ]
    [=====================================]
Number of iterations: 98

Least-square error: 390123.414335
Write individual ancestry coefficient file C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run3/dat_r3.6.Q:       OK.
Write ancestral allele frequency coefficient file C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run3/dat_r3.6.G:    OK.

[1] "*************************************"
[1] "*    cross-entropy estimation       *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)         249
        -L (number of loci)                5000
        -K (number of ancestral pops)      6
        -x (genotype file)                 C:\repositories\dartR_workshop\kioloa\dat.geno
        -q (individual admixture)          C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run3/dat_r3.6.Q
        -g (ancestral frequencies)         C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run3/dat_r3.6.G
        -i (with masked genotypes)         C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno
        - diploid

Cross-Entropy (all data):    0.515192
Cross-Entropy (masked data):     0.546855
The project is saved into :
 dat.snmfProject 

To load the project, use:
 project = load.snmfProject("dat.snmfProject")

To remove the project, use:
 remove.snmfProject("dat.snmfProject")

[1] 501538892
[1] "*************************************"
[1] "*          create.dataset            *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)                 249
        -L (number of loci)                        5000
        -s (seed random init)                      501538892
        -r (percentage of masked data)             0.05
        -x (genotype file in .geno format)         C:\repositories\dartR_workshop\kioloa\dat.geno
        -o (output file in .geno format)           C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno

 Write genotype file with masked data, C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno:        OK.

[1] "*************************************"
[1] "* sNMF K = 6  repetition 4      *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)             249
        -L (number of loci)                    5000
        -K (number of ancestral pops)          6
        -x (input file)                        C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno
        -q (individual admixture file)         C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run4/dat_r4.6.Q
        -g (ancestral frequencies file)        C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run4/dat_r4.6.G
        -i (number max of iterations)          200
        -a (regularization parameter)          10
        -s (seed random init)                  501538892
        -e (tolerance error)                   1E-05
        -p (number of processes)               1
        - diploid

Read genotype file C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno:        OK.


Main algorithm:
    [                                                                           ]
    [==================================]
Number of iterations: 92

Least-square error: 389969.387848
Write individual ancestry coefficient file C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run4/dat_r4.6.Q:       OK.
Write ancestral allele frequency coefficient file C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run4/dat_r4.6.G:    OK.

[1] "*************************************"
[1] "*    cross-entropy estimation       *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)         249
        -L (number of loci)                5000
        -K (number of ancestral pops)      6
        -x (genotype file)                 C:\repositories\dartR_workshop\kioloa\dat.geno
        -q (individual admixture)          C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run4/dat_r4.6.Q
        -g (ancestral frequencies)         C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run4/dat_r4.6.G
        -i (with masked genotypes)         C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno
        - diploid

Cross-Entropy (all data):    0.514871
Cross-Entropy (masked data):     0.554227
The project is saved into :
 dat.snmfProject 

To load the project, use:
 project = load.snmfProject("dat.snmfProject")

To remove the project, use:
 remove.snmfProject("dat.snmfProject")

[1] 1497835296
[1] "*************************************"
[1] "*          create.dataset            *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)                 249
        -L (number of loci)                        5000
        -s (seed random init)                      1497835296
        -r (percentage of masked data)             0.05
        -x (genotype file in .geno format)         C:\repositories\dartR_workshop\kioloa\dat.geno
        -o (output file in .geno format)           C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno

 Write genotype file with masked data, C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno:        OK.

[1] "*************************************"
[1] "* sNMF K = 6  repetition 5      *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)             249
        -L (number of loci)                    5000
        -K (number of ancestral pops)          6
        -x (input file)                        C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno
        -q (individual admixture file)         C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run5/dat_r5.6.Q
        -g (ancestral frequencies file)        C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run5/dat_r5.6.G
        -i (number max of iterations)          200
        -a (regularization parameter)          10
        -s (seed random init)                  1497835296
        -e (tolerance error)                   1E-05
        -p (number of processes)               1
        - diploid

Read genotype file C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno:        OK.


Main algorithm:
    [                                                                           ]
    [=======================================]
Number of iterations: 105

Least-square error: 390133.766506
Write individual ancestry coefficient file C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run5/dat_r5.6.Q:       OK.
Write ancestral allele frequency coefficient file C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run5/dat_r5.6.G:    OK.

[1] "*************************************"
[1] "*    cross-entropy estimation       *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)         249
        -L (number of loci)                5000
        -K (number of ancestral pops)      6
        -x (genotype file)                 C:\repositories\dartR_workshop\kioloa\dat.geno
        -q (individual admixture)          C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run5/dat_r5.6.Q
        -g (ancestral frequencies)         C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run5/dat_r5.6.G
        -i (with masked genotypes)         C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno
        - diploid

Cross-Entropy (all data):    0.515216
Cross-Entropy (masked data):     0.547343
The project is saved into :
 dat.snmfProject 

To load the project, use:
 project = load.snmfProject("dat.snmfProject")

To remove the project, use:
 remove.snmfProject("dat.snmfProject")

[1] 2046585485
[1] "*************************************"
[1] "*          create.dataset            *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)                 249
        -L (number of loci)                        5000
        -s (seed random init)                      2046585485
        -r (percentage of masked data)             0.05
        -x (genotype file in .geno format)         C:\repositories\dartR_workshop\kioloa\dat.geno
        -o (output file in .geno format)           C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno

 Write genotype file with masked data, C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno:        OK.

[1] "*************************************"
[1] "* sNMF K = 6  repetition 6      *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)             249
        -L (number of loci)                    5000
        -K (number of ancestral pops)          6
        -x (input file)                        C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno
        -q (individual admixture file)         C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run6/dat_r6.6.Q
        -g (ancestral frequencies file)        C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run6/dat_r6.6.G
        -i (number max of iterations)          200
        -a (regularization parameter)          10
        -s (seed random init)                  2046585485
        -e (tolerance error)                   1E-05
        -p (number of processes)               1
        - diploid

Read genotype file C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno:        OK.


Main algorithm:
    [                                                                           ]
    [==================================================]
Number of iterations: 133

Least-square error: 388974.296211
Write individual ancestry coefficient file C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run6/dat_r6.6.Q:       OK.
Write ancestral allele frequency coefficient file C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run6/dat_r6.6.G:    OK.

[1] "*************************************"
[1] "*    cross-entropy estimation       *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)         249
        -L (number of loci)                5000
        -K (number of ancestral pops)      6
        -x (genotype file)                 C:\repositories\dartR_workshop\kioloa\dat.geno
        -q (individual admixture)          C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run6/dat_r6.6.Q
        -g (ancestral frequencies)         C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run6/dat_r6.6.G
        -i (with masked genotypes)         C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno
        - diploid

Cross-Entropy (all data):    0.515075
Cross-Entropy (masked data):     0.55581
The project is saved into :
 dat.snmfProject 

To load the project, use:
 project = load.snmfProject("dat.snmfProject")

To remove the project, use:
 remove.snmfProject("dat.snmfProject")

[1] 1629520877
[1] "*************************************"
[1] "*          create.dataset            *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)                 249
        -L (number of loci)                        5000
        -s (seed random init)                      1629520877
        -r (percentage of masked data)             0.05
        -x (genotype file in .geno format)         C:\repositories\dartR_workshop\kioloa\dat.geno
        -o (output file in .geno format)           C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno

 Write genotype file with masked data, C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno:        OK.

[1] "*************************************"
[1] "* sNMF K = 6  repetition 7      *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)             249
        -L (number of loci)                    5000
        -K (number of ancestral pops)          6
        -x (input file)                        C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno
        -q (individual admixture file)         C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run7/dat_r7.6.Q
        -g (ancestral frequencies file)        C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run7/dat_r7.6.G
        -i (number max of iterations)          200
        -a (regularization parameter)          10
        -s (seed random init)                  1629520877
        -e (tolerance error)                   1E-05
        -p (number of processes)               1
        - diploid

Read genotype file C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno:        OK.


Main algorithm:
    [                                                                           ]
    [=========]
Number of iterations: 23

Least-square error: 390786.553313
Write individual ancestry coefficient file C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run7/dat_r7.6.Q:       OK.
Write ancestral allele frequency coefficient file C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run7/dat_r7.6.G:    OK.

[1] "*************************************"
[1] "*    cross-entropy estimation       *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)         249
        -L (number of loci)                5000
        -K (number of ancestral pops)      6
        -x (genotype file)                 C:\repositories\dartR_workshop\kioloa\dat.geno
        -q (individual admixture)          C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run7/dat_r7.6.Q
        -g (ancestral frequencies)         C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run7/dat_r7.6.G
        -i (with masked genotypes)         C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno
        - diploid

Cross-Entropy (all data):    0.518724
Cross-Entropy (masked data):     0.559747
The project is saved into :
 dat.snmfProject 

To load the project, use:
 project = load.snmfProject("dat.snmfProject")

To remove the project, use:
 remove.snmfProject("dat.snmfProject")

[1] 1792160697
[1] "*************************************"
[1] "*          create.dataset            *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)                 249
        -L (number of loci)                        5000
        -s (seed random init)                      1792160697
        -r (percentage of masked data)             0.05
        -x (genotype file in .geno format)         C:\repositories\dartR_workshop\kioloa\dat.geno
        -o (output file in .geno format)           C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno

 Write genotype file with masked data, C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno:        OK.

[1] "*************************************"
[1] "* sNMF K = 6  repetition 8      *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)             249
        -L (number of loci)                    5000
        -K (number of ancestral pops)          6
        -x (input file)                        C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno
        -q (individual admixture file)         C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run8/dat_r8.6.Q
        -g (ancestral frequencies file)        C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run8/dat_r8.6.G
        -i (number max of iterations)          200
        -a (regularization parameter)          10
        -s (seed random init)                  1792160697
        -e (tolerance error)                   1E-05
        -p (number of processes)               1
        - diploid

Read genotype file C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno:        OK.


Main algorithm:
    [                                                                           ]
    [=====================================================]
Number of iterations: 141

Least-square error: 389482.108053
Write individual ancestry coefficient file C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run8/dat_r8.6.Q:       OK.
Write ancestral allele frequency coefficient file C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run8/dat_r8.6.G:    OK.

[1] "*************************************"
[1] "*    cross-entropy estimation       *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)         249
        -L (number of loci)                5000
        -K (number of ancestral pops)      6
        -x (genotype file)                 C:\repositories\dartR_workshop\kioloa\dat.geno
        -q (individual admixture)          C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run8/dat_r8.6.Q
        -g (ancestral frequencies)         C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run8/dat_r8.6.G
        -i (with masked genotypes)         C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno
        - diploid

Cross-Entropy (all data):    0.515373
Cross-Entropy (masked data):     0.552094
The project is saved into :
 dat.snmfProject 

To load the project, use:
 project = load.snmfProject("dat.snmfProject")

To remove the project, use:
 remove.snmfProject("dat.snmfProject")

[1] 1478357482
[1] "*************************************"
[1] "*          create.dataset            *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)                 249
        -L (number of loci)                        5000
        -s (seed random init)                      1478357482
        -r (percentage of masked data)             0.05
        -x (genotype file in .geno format)         C:\repositories\dartR_workshop\kioloa\dat.geno
        -o (output file in .geno format)           C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno

 Write genotype file with masked data, C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno:        OK.

[1] "*************************************"
[1] "* sNMF K = 6  repetition 9      *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)             249
        -L (number of loci)                    5000
        -K (number of ancestral pops)          6
        -x (input file)                        C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno
        -q (individual admixture file)         C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run9/dat_r9.6.Q
        -g (ancestral frequencies file)        C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run9/dat_r9.6.G
        -i (number max of iterations)          200
        -a (regularization parameter)          10
        -s (seed random init)                  1478357482
        -e (tolerance error)                   1E-05
        -p (number of processes)               1
        - diploid

Read genotype file C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno:        OK.


Main algorithm:
    [                                                                           ]
    [==================================================]
Number of iterations: 134

Least-square error: 389483.213088
Write individual ancestry coefficient file C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run9/dat_r9.6.Q:       OK.
Write ancestral allele frequency coefficient file C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run9/dat_r9.6.G:    OK.

[1] "*************************************"
[1] "*    cross-entropy estimation       *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)         249
        -L (number of loci)                5000
        -K (number of ancestral pops)      6
        -x (genotype file)                 C:\repositories\dartR_workshop\kioloa\dat.geno
        -q (individual admixture)          C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run9/dat_r9.6.Q
        -g (ancestral frequencies)         C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run9/dat_r9.6.G
        -i (with masked genotypes)         C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno
        - diploid

Cross-Entropy (all data):    0.51554
Cross-Entropy (masked data):     0.549863
The project is saved into :
 dat.snmfProject 

To load the project, use:
 project = load.snmfProject("dat.snmfProject")

To remove the project, use:
 remove.snmfProject("dat.snmfProject")

[1] 1849722938
[1] "*************************************"
[1] "*          create.dataset            *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)                 249
        -L (number of loci)                        5000
        -s (seed random init)                      1849722938
        -r (percentage of masked data)             0.05
        -x (genotype file in .geno format)         C:\repositories\dartR_workshop\kioloa\dat.geno
        -o (output file in .geno format)           C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno

 Write genotype file with masked data, C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno:        OK.

[1] "*************************************"
[1] "* sNMF K = 6  repetition 10      *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)             249
        -L (number of loci)                    5000
        -K (number of ancestral pops)          6
        -x (input file)                        C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno
        -q (individual admixture file)         C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run10/dat_r10.6.Q
        -g (ancestral frequencies file)        C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run10/dat_r10.6.G
        -i (number max of iterations)          200
        -a (regularization parameter)          10
        -s (seed random init)                  1849722938
        -e (tolerance error)                   1E-05
        -p (number of processes)               1
        - diploid

Read genotype file C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno:        OK.


Main algorithm:
    [                                                                           ]
    [=================================]
Number of iterations: 87

Least-square error: 389102.635227
Write individual ancestry coefficient file C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run10/dat_r10.6.Q:     OK.
Write ancestral allele frequency coefficient file C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run10/dat_r10.6.G:  OK.

[1] "*************************************"
[1] "*    cross-entropy estimation       *"
[1] "*************************************"
summary of the options:

        -n (number of individuals)         249
        -L (number of loci)                5000
        -K (number of ancestral pops)      6
        -x (genotype file)                 C:\repositories\dartR_workshop\kioloa\dat.geno
        -q (individual admixture)          C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run10/dat_r10.6.Q
        -g (ancestral frequencies)         C:/repositories/dartR_workshop/kioloa/dat.snmf/K6/run10/dat_r10.6.G
        -i (with masked genotypes)         C:/repositories/dartR_workshop/kioloa/dat.snmf/masked/dat_I.geno
        - diploid

Cross-Entropy (all data):    0.515332
Cross-Entropy (masked data):     0.550284
The project is saved into :
 dat.snmfProject 

To load the project, use:
 project = load.snmfProject("dat.snmfProject")

To remove the project, use:
 remove.snmfProject("dat.snmfProject")

Missing genotype imputation for K = 6 
Missing genotype imputation for run = 3 
Results are written in the file:  dat.lfmm_imputed.lfmm 
Total missing data was =  1.04 %
Total missing data now =  0 %</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># order major/minor alleles (generally no reason to do this... but might be useful for someone)</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>imputed.sorted.gi <span class="ot">&lt;-</span> <span class="fu">sort_alleles</span>(imputed.gi)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="co"># check results</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>imputed.gi<span class="sc">@</span>tab[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       SNP_1.C SNP_1.G SNP_2.A SNP_2.T SNP_3.T SNP_3.A
MBR_10       1       1       1       1       2       0
MBR_11       1       1       1       1       2       0
MBR_12       1       1       2       0       2       0
MBR_14       0       2       2       0       2       0
MBR_15       0       2       1       1       2       0
MBR_16       0       2       2       0       2       0
MBR_17       0       2       0       2       2       0
MBR_18       0       2       1       1       2       0
MBR_19       1       1       2       0       2       0
MBR_1        0       2       1       1       2       0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>imputed.sorted.gi<span class="sc">@</span>tab[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       SNP_1.G SNP_1.C SNP_2.A SNP_2.T SNP_3.T SNP_3.A
MBR_10       1       1       1       1       2       0
MBR_11       1       1       1       1       2       0
MBR_12       1       1       2       0       2       0
MBR_14       2       0       2       0       2       0
MBR_15       2       0       1       1       2       0
MBR_16       2       0       2       0       2       0
MBR_17       2       0       0       2       2       0
MBR_18       2       0       1       1       2       0
MBR_19       1       1       2       0       2       0
MBR_1        2       0       1       1       2       0</code></pre>
</div>
</div>
<p>Format SNP data for ordination analysis in vegan. You want a matrix of allele counts per locus, per individual can use population allele frequencies instead</p>
<section id="individual-based" class="level3">
<h3 class="anchored" data-anchor-id="individual-based">Individual based</h3>
<div class="cell" data-hash="session07_cache/html/unnamed-chunk-5_a57c8e0680bce811d1f01cb66a757132">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># for individual based analyses</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="co"># get allele counts</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>alleles <span class="ot">&lt;-</span> imputed.sorted.gi<span class="sc">@</span>tab</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="co"># get genotypes (counts of reference allele) and clean up locus names</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>snps <span class="ot">&lt;-</span> alleles[,<span class="fu">seq</span>(<span class="dv">1</span>,<span class="fu">ncol</span>(alleles),<span class="dv">2</span>)]</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(snps) <span class="ot">&lt;-</span> <span class="fu">locNames</span>(imputed.sorted.gi)</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>snps[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       SNP_1 SNP_2 SNP_3 SNP_4 SNP_5 SNP_6 SNP_7 SNP_8 SNP_9 SNP_10
MBR_10     1     1     2     2     2     2     1     1     2      1
MBR_11     1     1     2     2     2     2     2     2     2      0
MBR_12     1     2     2     2     2     1     2     2     2      1
MBR_14     2     2     2     2     2     2     1     1     2      2
MBR_15     2     1     2     2     2     2     1     1     2      2
MBR_16     2     2     2     2     2     2     2     2     2      2
MBR_17     2     0     2     2     2     2     0     2     2      2
MBR_18     2     1     2     2     2     1     1     2     2      0
MBR_19     1     2     2     2     2     1     2     2     2      2
MBR_1      2     1     2     2     1     2     2     2     2      2</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Alternative: impute missing data with most common genotype across all data</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="co"># alleles &lt;- Mf5000.genind@tab</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="co"># snps &lt;- alleles[,seq(1,ncol(alleles),2)]</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="co"># colnames(snps) &lt;- locNames(Mf5000.genind)</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="co"># snps &lt;- apply(snps, 2, function(x) replace(x, is.na(x), as.numeric(names(which.max(table(x))))))</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="co"># check total % missing data</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="co"># (sum(is.na(snps)))/(dim(snps)[1]*dim(snps)[2])*100</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="population-based" class="level3">
<h3 class="anchored" data-anchor-id="population-based">Population based</h3>
<div class="cell" data-hash="session07_cache/html/unnamed-chunk-6_d25d1a817c654c99e95cc3c3eb3f9bea">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># for population based analyses</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="co"># get pop allele frequencies</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>gp <span class="ot">&lt;-</span> <span class="fu">genind2genpop</span>(imputed.sorted.gi)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 Converting data from a genind to a genpop object... 

...done.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>AF <span class="ot">&lt;-</span> <span class="fu">makefreq</span>(gp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 Finding allelic frequencies from a genpop object... 

...done.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># drop one (redundant) allele per locus</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>AF <span class="ot">&lt;-</span> AF[,<span class="fu">seq</span>(<span class="dv">1</span>,<span class="fu">ncol</span>(AF),<span class="dv">2</span>)]</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(AF) <span class="ot">&lt;-</span> <span class="fu">locNames</span>(gp)</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(AF) <span class="ot">&lt;-</span> <span class="fu">levels</span>(imputed.sorted.gi<span class="sc">@</span>pop)</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>AF[<span class="dv">1</span><span class="sc">:</span><span class="dv">14</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         SNP_1     SNP_2     SNP_3     SNP_4     SNP_5     SNP_6
MBR 0.87500000 0.8250000 0.9750000 1.0000000 0.9750000 0.8250000
MDC 0.70454545 0.7045455 0.8409091 0.9545455 0.8863636 0.8181818
GGL 0.71666667 0.7833333 0.9000000 0.9166667 0.8500000 0.8000000
WAK 0.93750000 1.0000000 1.0000000 1.0000000 1.0000000 0.3125000
BEN 0.92857143 1.0000000 1.0000000 1.0000000 1.0000000 0.6071429
BOG 0.75000000 0.8125000 0.7812500 1.0000000 0.7812500 0.9062500
PEL 0.83333333 0.7222222 0.9444444 0.9444444 0.8888889 1.0000000
GWY 0.75000000 0.9166667 1.0000000 1.0000000 1.0000000 0.8333333
DUM 0.82142857 1.0000000 0.9285714 0.9285714 1.0000000 0.8571429
MIB 0.92500000 0.7750000 0.9250000 0.8250000 0.9750000 0.8500000
STG 0.97500000 0.8750000 0.6500000 0.8000000 1.0000000 0.8000000
OAK 0.02777778 0.6388889 0.6388889 0.6666667 0.3333333 1.0000000
WAR 0.20000000 0.9000000 0.7000000 0.5750000 0.2750000 1.0000000
KIL 0.25000000 0.7777778 0.5277778 0.7222222 0.3055556 1.0000000</code></pre>
</div>
</div>
</section>
<section id="get-environmental-data" class="level3">
<h3 class="anchored" data-anchor-id="get-environmental-data">Get environmental data</h3>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Decision time:
</div>
</div>
<div class="callout-body-container callout-body">
<p>What are the variables you want to use? Considerations: Prior/expert knowledge, hypotheses, species distribution models (SDMs), most important variables, data availability, think about surrogates/proxies if specific data not available, raw variables, PCA, other transformations?</p>
<p>In this case we are going to keep it simple and just use a few WorldClim variables Bio01 (annual mean temperature), Bio05 (temperature in hottest month), Bio15 (rainfall seasonality) and Bio19 (rainfall in coldest quarter)</p>
</div>
</div>
<div class="cell" data-hash="session07_cache/html/unnamed-chunk-7_b983bbb421175499faec3c5f3dbd0cd1">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set the data dir and extend the waiting time (sometimes the database can be slow to respond)</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">sdmpredictors_datadir=</span><span class="st">"data/spatial_data"</span>)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">timeout =</span> <span class="fu">max</span>(<span class="dv">300</span>, <span class="fu">getOption</span>(<span class="st">"timeout"</span>)))</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Explore datasets in the package </span></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>sdmpredictors<span class="sc">::</span><span class="fu">list_datasets</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  dataset_code terrestrial marine                              url
1    WorldClim        TRUE  FALSE        http://www.worldclim.org/
2   Bio-ORACLE       FALSE   TRUE          https://bio-oracle.org/
3      MARSPEC       FALSE   TRUE              http://marspec.org/
4      ENVIREM        TRUE  FALSE       https://envirem.github.io/
5   Freshwater        TRUE  FALSE https://www.earthenv.org/streams
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               description
1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   WorldClim is a set of global climate layers (climate grids). Note that all data has been transformed back to real values, so there is no need to e.g. divide temperature layers by 10.
2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                Bio-ORACLE is a set of GIS rasters providing geophysical, biotic and environmental data for surface and benthic marine realms at a spatial resolution 5 arcmin (9.2 km) in the ESRI ascii and tif format.
3 MARSPEC is a set of high resolution climatic and geophysical GIS data layers for the world ocean. Seven geophysical variables were derived from the SRTM30_PLUS high resolution bathymetry dataset.  These layers characterize the horizontal orientation (aspect), slope, and curvature of the seafloor and the distance from shore.  Ten "bioclimatic" variables were derived from NOAA's World Ocean Atlas and NASA's MODIS satellite imagery and characterize the inter-annual means, extremes, and variances in sea surface temperature and salinity. These variables will be useful to those interested in the spatial ecology of marine shallow-water and surface-associated pelagic organisms across the globe. Note that, in contrary to the original MARSPEC, all layers have unscaled values.
4                                                                                                                                 The ENVIREM dataset is a set of 16 climatic and 2 topographic variables that can be used in modeling species' distributions. The strengths of this dataset include their close ties to ecological processes, and their availability at a global scale, at several spatial resolutions, and for several time periods. The underlying temperature and precipitation data that went into their construction comes from the WorldClim dataset (www.worldclim.org), and the solar radiation data comes from the Consortium for Spatial Information (www.cgiar-csi.org). The data are compatible with and expand the set of variables from WorldClim v1.4 (www.worldclim.org).
5                                                      The dataset consists of near-global, spatially continuous, and freshwater-specific environmental variables in a standardized 1km grid. We delineated the sub-catchment for each grid cell along the HydroSHEDS river network and summarized the upstream environment (climate, topography, land cover, surface geology and soil) to each grid cell using various metrics (average, minimum, maximum, range, sum, inverse distance-weighted average and sum). All variables were subsequently averaged across single lakes and reservoirs of the Global lakes and Wetlands Database that are connected to the river network. Monthly climate variables were summarized into 19 long-term climatic variables following the \xd2bioclim\xd3 framework.
                                                                                                                                                                                                                                  citation
1                                   Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
2 Tyberghein L., Verbruggen H., Pauly K., Troupin C., Mineur F. &amp; De Clerck O. Bio-ORACLE: a global environmental dataset for marine species distribution modeling. Global Ecology and Biogeography. doi: 10.1111/j.1466-8238.2011.00656.x
3                                                                                                      Sbrocco, EJ and Barber, PH (2013) MARSPEC: Ocean climate layers for marine spatial ecology. Ecology 94: 979. doi: 10.1890/12-1358.1
4                       Title, P.O., Bemmels, J.B. 2017. ENVIREM: An expanded set of bioclimatic and topographic variables increases flexibility and improves performance of ecological niche modeling. Ecography doi: 10.1111/ecog.02880.
5                                 Domisch, S., Amatulli, G., and Jetz, W. (2015) Near-global freshwater-specific environmental variables for biodiversity analyses in 1 km resolution. Scientific Data 2:150073 doi: 10.1038/sdata.2015.73</code></pre>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>sdmpredictors<span class="sc">::</span><span class="fu">list_layers</span>(<span class="st">"WorldClim"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   dataset_code layer_code                                name
1     WorldClim     WC_alt                            Altitude
2     WorldClim    WC_bio1             Annual mean temperature
3     WorldClim    WC_bio2      Mean diurnal temperature range
4     WorldClim    WC_bio3                       Isothermality
5     WorldClim    WC_bio4             Temperature seasonality
6     WorldClim    WC_bio5                 Maximum temperature
7     WorldClim    WC_bio6                 Minimum temperature
8     WorldClim    WC_bio7            Annual temperature range
9     WorldClim    WC_bio8 Mean temperature of wettest quarter
10    WorldClim    WC_bio9  Mean temperature of driest quarter
11    WorldClim   WC_bio10 Mean temperature of warmest quarter
12    WorldClim   WC_bio11 Mean temperature of coldest quarter
13    WorldClim   WC_bio12                Annual precipitation
14    WorldClim   WC_bio13      Precipitation of wettest month
15    WorldClim   WC_bio14       Precipitation of driest month
16    WorldClim   WC_bio15           Precipitation seasonality
17    WorldClim   WC_bio16    Precipitation of wettest quarter
18    WorldClim   WC_bio17     Precipitation of driest quarter
19    WorldClim   WC_bio18    Precipitation of warmest quarter
20    WorldClim   WC_bio19    Precipitation of coldest quarter
21    WorldClim   WC_prec1             Precipitation (January)
22    WorldClim   WC_prec2            Precipitation (February)
23    WorldClim   WC_prec3               Precipitation (March)
24    WorldClim   WC_prec4               Precipitation (April)
25    WorldClim   WC_prec5                 Precipitation (May)
26    WorldClim   WC_prec6                Precipitation (June)
27    WorldClim   WC_prec7                Precipitation (July)
28    WorldClim   WC_prec8              Precipitation (August)
29    WorldClim   WC_prec9           Precipitation (September)
30    WorldClim  WC_prec10             Precipitation (October)
31    WorldClim  WC_prec11            Precipitation (November)
32    WorldClim  WC_prec12            Precipitation (December)
33    WorldClim   WC_tmax1       Maximum temperature (January)
34    WorldClim   WC_tmax2      Maximum temperature (February)
35    WorldClim   WC_tmax3         Maximum temperature (March)
36    WorldClim   WC_tmax4         Maximum temperature (April)
37    WorldClim   WC_tmax5           Maximum temperature (May)
38    WorldClim   WC_tmax6          Maximum temperature (June)
39    WorldClim   WC_tmax7          Maximum temperature (July)
40    WorldClim   WC_tmax8        Maximum temperature (August)
41    WorldClim   WC_tmax9     Maximum temperature (September)
42    WorldClim  WC_tmax10       Maximum temperature (October)
43    WorldClim  WC_tmax11      Maximum temperature (November)
44    WorldClim  WC_tmax12      Maximum temperature (December)
45    WorldClim  WC_tmean1          Mean temperature (January)
46    WorldClim  WC_tmean2         Mean temperature (February)
47    WorldClim  WC_tmean3            Mean temperature (March)
48    WorldClim  WC_tmean4            Mean temperature (April)
49    WorldClim  WC_tmean5              Mean temperature (May)
50    WorldClim  WC_tmean6             Mean temperature (June)
51    WorldClim  WC_tmean7             Mean temperature (July)
52    WorldClim  WC_tmean8           Mean temperature (August)
53    WorldClim  WC_tmean9        Mean temperature (September)
54    WorldClim WC_tmean10          Mean temperature (October)
55    WorldClim WC_tmean11         Mean temperature (November)
56    WorldClim WC_tmean12         Mean temperature (December)
57    WorldClim   WC_tmin1       Minimum temperature (January)
58    WorldClim   WC_tmin2      Minimum temperature (February)
59    WorldClim   WC_tmin3         Minimum temperature (March)
60    WorldClim   WC_tmin4         Minimum temperature (April)
61    WorldClim   WC_tmin5           Minimum temperature (May)
62    WorldClim   WC_tmin6          Minimum temperature (June)
63    WorldClim   WC_tmin7          Minimum temperature (July)
64    WorldClim   WC_tmin8        Minimum temperature (August)
65    WorldClim   WC_tmin9     Minimum temperature (September)
66    WorldClim  WC_tmin10       Minimum temperature (October)
67    WorldClim  WC_tmin11      Minimum temperature (November)
68    WorldClim  WC_tmin12      Minimum temperature (December)
                                                               description
1                                                                 Altitude
2                                                  Annual mean temperature
3          Mean of the monthly (maximum temperature - minimum temperature)
4  Mean diurnal temperature range (bio2) / Annual temperature range (bio7)
5                        Standard deviation of the annual mean temperature
6                                 Maximum temperature of the warmest month
7                                 Minimum temperature of the coldest month
8                  Maximum temperature (bio5) - minimum temperature (bio6)
9                                      Mean temperature of wettest quarter
10                                      Mean temperature of driest quarter
11                                     Mean temperature of warmest quarter
12                                     Mean temperature of coldest quarter
13                                                    Annual precipitation
14                                          Precipitation of wettest month
15                                           Precipitation of driest month
16                   Coefficient of variation of the monthly precipitation
17                                        Precipitation of wettest quarter
18                                         Precipitation of driest quarter
19                                        Precipitation of warmest quarter
20                                        Precipitation of coldest quarter
21                                           Average January precipitation
22                                          Average February precipitation
23                                             Average March precipitation
24                                             Average April precipitation
25                                               Average May precipitation
26                                              Average June precipitation
27                                              Average July precipitation
28                                            Average August precipitation
29                                         Average September precipitation
30                                           Average October precipitation
31                                          Average November precipitation
32                                                        Average December
33                                             Maximum January temperature
34                                            Maximum February temperature
35                                               Maximum March temperature
36                                               Maximum April temperature
37                                                 Maximum May temperature
38                                                Maximum June temperature
39                                                Maximum July temperature
40                                              Maximum August temperature
41                                           Maximum September temperature
42                                             Maximum October temperature
43                                            Maximum November temperature
44                                            Maximum December temperature
45                                             Average January temperature
46                                            Average February temperature
47                                               Average March temperature
48                                               Average April temperature
49                                                 Average May temperature
50                                                Average June temperature
51                                                Average July temperature
52                                              Average August temperature
53                                           Average September temperature
54                                             Average October temperature
55                                            Average November temperature
56                                            Average December temperature
57                                             Minimum January temperature
58                                            Minimum February temperature
59                                               Minimum March temperature
60                                               Minimum April temperature
61                                                 Minimum May temperature
62                                                Minimum June temperature
63                                                Minimum July temperature
64                                              Minimum August temperature
65                                           Minimum September temperature
66                                             Minimum October temperature
67                                            Minimum November temperature
68                                            Minimum December temperature
   terrestrial marine freshwater cellsize_equalarea cellsize_lonlat   units
1         TRUE  FALSE      FALSE               7000      0.08333333  meters
2         TRUE  FALSE      FALSE               7000      0.08333333 Celsius
3         TRUE  FALSE      FALSE               7000      0.08333333 Celsius
4         TRUE  FALSE      FALSE               7000      0.08333333 Celsius
5         TRUE  FALSE      FALSE               7000      0.08333333 Celsius
6         TRUE  FALSE      FALSE               7000      0.08333333 Celsius
7         TRUE  FALSE      FALSE               7000      0.08333333 Celsius
8         TRUE  FALSE      FALSE               7000      0.08333333 Celsius
9         TRUE  FALSE      FALSE               7000      0.08333333 Celsius
10        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
11        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
12        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
13        TRUE  FALSE      FALSE               7000      0.08333333      mm
14        TRUE  FALSE      FALSE               7000      0.08333333      mm
15        TRUE  FALSE      FALSE               7000      0.08333333      mm
16        TRUE  FALSE      FALSE               7000      0.08333333      mm
17        TRUE  FALSE      FALSE               7000      0.08333333      mm
18        TRUE  FALSE      FALSE               7000      0.08333333      mm
19        TRUE  FALSE      FALSE               7000      0.08333333      mm
20        TRUE  FALSE      FALSE               7000      0.08333333      mm
21        TRUE  FALSE      FALSE               7000      0.08333333      mm
22        TRUE  FALSE      FALSE               7000      0.08333333      mm
23        TRUE  FALSE      FALSE               7000      0.08333333      mm
24        TRUE  FALSE      FALSE               7000      0.08333333      mm
25        TRUE  FALSE      FALSE               7000      0.08333333      mm
26        TRUE  FALSE      FALSE               7000      0.08333333      mm
27        TRUE  FALSE      FALSE               7000      0.08333333      mm
28        TRUE  FALSE      FALSE               7000      0.08333333      mm
29        TRUE  FALSE      FALSE               7000      0.08333333      mm
30        TRUE  FALSE      FALSE               7000      0.08333333      mm
31        TRUE  FALSE      FALSE               7000      0.08333333      mm
32        TRUE  FALSE      FALSE               7000      0.08333333      mm
33        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
34        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
35        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
36        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
37        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
38        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
39        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
40        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
41        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
42        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
43        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
44        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
45        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
46        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
47        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
48        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
49        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
50        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
51        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
52        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
53        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
54        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
55        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
56        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
57        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
58        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
59        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
60        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
61        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
62        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
63        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
64        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
65        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
66        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
67        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
68        TRUE  FALSE      FALSE               7000      0.08333333 Celsius
                primary_type primary_spatial_resolution
1  Satellite (SRTM), GTOPO30   3 arcsecond, 30 arsecond
2        in situ measurement                         ''
3        in situ measurement                         ''
4        in situ measurement                         ''
5        in situ measurement                         ''
6        in situ measurement                         ''
7        in situ measurement                         ''
8        in situ measurement                         ''
9        in situ measurement                         ''
10       in situ measurement                         ''
11       in situ measurement                         ''
12       in situ measurement                         ''
13       in situ measurement                         ''
14       in situ measurement                         ''
15       in situ measurement                         ''
16       in situ measurement                         ''
17       in situ measurement                         ''
18       in situ measurement                         ''
19       in situ measurement                         ''
20       in situ measurement                         ''
21       in situ measurement                         ''
22       in situ measurement                         ''
23       in situ measurement                         ''
24       in situ measurement                         ''
25       in situ measurement                         ''
26       in situ measurement                         ''
27       in situ measurement                         ''
28       in situ measurement                         ''
29       in situ measurement                         ''
30       in situ measurement                         ''
31       in situ measurement                         ''
32       in situ measurement                         ''
33       in situ measurement                         ''
34       in situ measurement                         ''
35       in situ measurement                         ''
36       in situ measurement                         ''
37       in situ measurement                         ''
38       in situ measurement                         ''
39       in situ measurement                         ''
40       in situ measurement                         ''
41       in situ measurement                         ''
42       in situ measurement                         ''
43       in situ measurement                         ''
44       in situ measurement                         ''
45       in situ measurement                         ''
46       in situ measurement                         ''
47       in situ measurement                         ''
48       in situ measurement                         ''
49       in situ measurement                         ''
50       in situ measurement                         ''
51       in situ measurement                         ''
52       in situ measurement                         ''
53       in situ measurement                         ''
54       in situ measurement                         ''
55       in situ measurement                         ''
56       in situ measurement                         ''
57       in situ measurement                         ''
58       in situ measurement                         ''
59       in situ measurement                         ''
60       in situ measurement                         ''
61       in situ measurement                         ''
62       in situ measurement                         ''
63       in situ measurement                         ''
64       in situ measurement                         ''
65       in situ measurement                         ''
66       in situ measurement                         ''
67       in situ measurement                         ''
68       in situ measurement                         ''
                                                                                                                                                                                                                                           primary_source
1   SRTM: CGIAR-CSI, reference: Jarvis, A., H.I. Reuter, A. Nelson, E. Guevara, 2008, Hole-filled SRTM for the globe Version 4, available from the CGIAR-CSI SRTM 90m Database (http://srtm.csi.cgiar.org). GTOPO30: url: https://lta.cr.usgs.gov/GTOPO30
2  Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
3  Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
4  Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
5  Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
6  Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
7  Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
8  Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
9  Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
10 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
11 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
12 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
13 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
14 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
15 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
16 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
17 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
18 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
19 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
20 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
21 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
22 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
23 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
24 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
25 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
26 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
27 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
28 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
29 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
30 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
31 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
32 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
33 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
34 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
35 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
36 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
37 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
38 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
39 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
40 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
41 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
42 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
43 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
44 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
45 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
46 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
47 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
48 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
49 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
50 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
51 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
52 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
53 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
54 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
55 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
56 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
57 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
58 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
59 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
60 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
61 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
62 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
63 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
64 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
65 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
66 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
67 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
68 Weather station data from multiple sources, see Hijmans, R.J., S.E. Cameron, J.L. Parra, P.G. Jones and A. Jarvis, 2005. Very high resolution interpolated climate surfaces for global land areas. International Journal of Climatology 25: 1965-1978.
   start_year start_month start_day end_year end_month end_day
1        1950           1         1     2000         1       1
2        1950           1         1     2000         1       1
3        1950           1         1     2000         1       1
4        1950           1         1     2000         1       1
5        1950           1         1     2000         1       1
6        1950           1         1     2000         1       1
7        1950           1         1     2000         1       1
8        1950           1         1     2000         1       1
9        1950           1         1     2000         1       1
10       1950           1         1     2000         1       1
11       1950           1         1     2000         1       1
12       1950           1         1     2000         1       1
13       1950           1         1     2000         1       1
14       1950           1         1     2000         1       1
15       1950           1         1     2000         1       1
16       1950           1         1     2000         1       1
17       1950           1         1     2000         1       1
18       1950           1         1     2000         1       1
19       1950           1         1     2000         1       1
20       1950           1         1     2000         1       1
21       1950           1         1     2000         1       1
22       1950           1         1     2000         1       1
23       1950           1         1     2000         1       1
24       1950           1         1     2000         1       1
25       1950           1         1     2000         1       1
26       1950           1         1     2000         1       1
27       1950           1         1     2000         1       1
28       1950           1         1     2000         1       1
29       1950           1         1     2000         1       1
30       1950           1         1     2000         1       1
31       1950           1         1     2000         1       1
32       1950           1         1     2000         1       1
33       1950           1         1     2000         1       1
34       1950           1         1     2000         1       1
35       1950           1         1     2000         1       1
36       1950           1         1     2000         1       1
37       1950           1         1     2000         1       1
38       1950           1         1     2000         1       1
39       1950           1         1     2000         1       1
40       1950           1         1     2000         1       1
41       1950           1         1     2000         1       1
42       1950           1         1     2000         1       1
43       1950           1         1     2000         1       1
44       1950           1         1     2000         1       1
45       1950           1         1     2000         1       1
46       1950           1         1     2000         1       1
47       1950           1         1     2000         1       1
48       1950           1         1     2000         1       1
49       1950           1         1     2000         1       1
50       1950           1         1     2000         1       1
51       1950           1         1     2000         1       1
52       1950           1         1     2000         1       1
53       1950           1         1     2000         1       1
54       1950           1         1     2000         1       1
55       1950           1         1     2000         1       1
56       1950           1         1     2000         1       1
57       1950           1         1     2000         1       1
58       1950           1         1     2000         1       1
59       1950           1         1     2000         1       1
60       1950           1         1     2000         1       1
61       1950           1         1     2000         1       1
62       1950           1         1     2000         1       1
63       1950           1         1     2000         1       1
64       1950           1         1     2000         1       1
65       1950           1         1     2000         1       1
66       1950           1         1     2000         1       1
67       1950           1         1     2000         1       1
68       1950           1         1     2000         1       1
                                         derivation month is_surface version
1                                                ''    NA       TRUE      10
2                mean, thin-plate smoothing splines    NA       TRUE      10
3          mean range, thin-plate smoothing splines    NA       TRUE      10
4                                                ''    NA       TRUE      10
5  standard deviation, thin-plate smoothing splines    NA       TRUE      10
6             maximum, thin-plate smoothing splines    NA       TRUE      10
7             minimum, thin-plate smoothing splines    NA       TRUE      10
8               range, thin-plate smoothing splines    NA       TRUE      10
9                mean, thin-plate smoothing splines    NA       TRUE      10
10               mean, thin-plate smoothing splines    NA       TRUE      10
11               mean, thin-plate smoothing splines    NA       TRUE      10
12               mean, thin-plate smoothing splines    NA       TRUE      10
13                                               ''    NA       TRUE      10
14            maximum, thin-plate smoothing splines    NA       TRUE      10
15            minimum, thin-plate smoothing splines    NA       TRUE      10
16                                               ''    NA       TRUE      10
17            maximum, thin-plate smoothing splines    NA       TRUE      10
18            minimum, thin-plate smoothing splines    NA       TRUE      10
19                                               ''    NA       TRUE      10
20                                               ''    NA       TRUE      10
21                                               ''     1       TRUE      10
22                                               ''     2       TRUE      10
23                                               ''     3       TRUE      10
24                                               ''     4       TRUE      10
25                                               ''     5       TRUE      10
26                                               ''     6       TRUE      10
27                                               ''     7       TRUE      10
28                                               ''     8       TRUE      10
29                                               ''     9       TRUE      10
30                                               ''    10       TRUE      10
31                                               ''    11       TRUE      10
32                                               ''    12       TRUE      10
33            maximum, thin-plate smoothing splines     1       TRUE      10
34            maximum, thin-plate smoothing splines     2       TRUE      10
35            maximum, thin-plate smoothing splines     3       TRUE      10
36            maximum, thin-plate smoothing splines     4       TRUE      10
37            maximum, thin-plate smoothing splines     5       TRUE      10
38            maximum, thin-plate smoothing splines     6       TRUE      10
39            maximum, thin-plate smoothing splines     7       TRUE      10
40            maximum, thin-plate smoothing splines     8       TRUE      10
41            maximum, thin-plate smoothing splines     9       TRUE      10
42            maximum, thin-plate smoothing splines    10       TRUE      10
43            maximum, thin-plate smoothing splines    11       TRUE      10
44            maximum, thin-plate smoothing splines    12       TRUE      10
45               mean, thin-plate smoothing splines     1       TRUE      10
46               mean, thin-plate smoothing splines     2       TRUE      10
47               mean, thin-plate smoothing splines     3       TRUE      10
48               mean, thin-plate smoothing splines     4       TRUE      10
49               mean, thin-plate smoothing splines     5       TRUE      10
50               mean, thin-plate smoothing splines     6       TRUE      10
51               mean, thin-plate smoothing splines     7       TRUE      10
52               mean, thin-plate smoothing splines     8       TRUE      10
53               mean, thin-plate smoothing splines     9       TRUE      10
54               mean, thin-plate smoothing splines    10       TRUE      10
55               mean, thin-plate smoothing splines    11       TRUE      10
56               mean, thin-plate smoothing splines    12       TRUE      10
57            minimum, thin-plate smoothing splines     1       TRUE      10
58            minimum, thin-plate smoothing splines     2       TRUE      10
59            minimum, thin-plate smoothing splines     3       TRUE      10
60            minimum, thin-plate smoothing splines     4       TRUE      10
61            minimum, thin-plate smoothing splines     5       TRUE      10
62            minimum, thin-plate smoothing splines     6       TRUE      10
63            minimum, thin-plate smoothing splines     7       TRUE      10
64            minimum, thin-plate smoothing splines     8       TRUE      10
65            minimum, thin-plate smoothing splines     9       TRUE      10
66            minimum, thin-plate smoothing splines    10       TRUE      10
67            minimum, thin-plate smoothing splines    11       TRUE      10
68            minimum, thin-plate smoothing splines    12       TRUE      10
                                                      layer_url
1      https://www.lifewatch.be/sdmpredictors/WC_alt_lonlat.tif
2     https://www.lifewatch.be/sdmpredictors/WC_bio1_lonlat.tif
3     https://www.lifewatch.be/sdmpredictors/WC_bio2_lonlat.tif
4     https://www.lifewatch.be/sdmpredictors/WC_bio3_lonlat.tif
5     https://www.lifewatch.be/sdmpredictors/WC_bio4_lonlat.tif
6     https://www.lifewatch.be/sdmpredictors/WC_bio5_lonlat.tif
7     https://www.lifewatch.be/sdmpredictors/WC_bio6_lonlat.tif
8     https://www.lifewatch.be/sdmpredictors/WC_bio7_lonlat.tif
9     https://www.lifewatch.be/sdmpredictors/WC_bio8_lonlat.tif
10    https://www.lifewatch.be/sdmpredictors/WC_bio9_lonlat.tif
11   https://www.lifewatch.be/sdmpredictors/WC_bio10_lonlat.tif
12   https://www.lifewatch.be/sdmpredictors/WC_bio11_lonlat.tif
13   https://www.lifewatch.be/sdmpredictors/WC_bio12_lonlat.tif
14   https://www.lifewatch.be/sdmpredictors/WC_bio13_lonlat.tif
15   https://www.lifewatch.be/sdmpredictors/WC_bio14_lonlat.tif
16   https://www.lifewatch.be/sdmpredictors/WC_bio15_lonlat.tif
17   https://www.lifewatch.be/sdmpredictors/WC_bio16_lonlat.tif
18   https://www.lifewatch.be/sdmpredictors/WC_bio17_lonlat.tif
19   https://www.lifewatch.be/sdmpredictors/WC_bio18_lonlat.tif
20   https://www.lifewatch.be/sdmpredictors/WC_bio19_lonlat.tif
21   https://www.lifewatch.be/sdmpredictors/WC_prec1_lonlat.tif
22   https://www.lifewatch.be/sdmpredictors/WC_prec2_lonlat.tif
23   https://www.lifewatch.be/sdmpredictors/WC_prec3_lonlat.tif
24   https://www.lifewatch.be/sdmpredictors/WC_prec4_lonlat.tif
25   https://www.lifewatch.be/sdmpredictors/WC_prec5_lonlat.tif
26   https://www.lifewatch.be/sdmpredictors/WC_prec6_lonlat.tif
27   https://www.lifewatch.be/sdmpredictors/WC_prec7_lonlat.tif
28   https://www.lifewatch.be/sdmpredictors/WC_prec8_lonlat.tif
29   https://www.lifewatch.be/sdmpredictors/WC_prec9_lonlat.tif
30  https://www.lifewatch.be/sdmpredictors/WC_prec10_lonlat.tif
31  https://www.lifewatch.be/sdmpredictors/WC_prec11_lonlat.tif
32  https://www.lifewatch.be/sdmpredictors/WC_prec12_lonlat.tif
33   https://www.lifewatch.be/sdmpredictors/WC_tmax1_lonlat.tif
34   https://www.lifewatch.be/sdmpredictors/WC_tmax2_lonlat.tif
35   https://www.lifewatch.be/sdmpredictors/WC_tmax3_lonlat.tif
36   https://www.lifewatch.be/sdmpredictors/WC_tmax4_lonlat.tif
37   https://www.lifewatch.be/sdmpredictors/WC_tmax5_lonlat.tif
38   https://www.lifewatch.be/sdmpredictors/WC_tmax6_lonlat.tif
39   https://www.lifewatch.be/sdmpredictors/WC_tmax7_lonlat.tif
40   https://www.lifewatch.be/sdmpredictors/WC_tmax8_lonlat.tif
41   https://www.lifewatch.be/sdmpredictors/WC_tmax9_lonlat.tif
42  https://www.lifewatch.be/sdmpredictors/WC_tmax10_lonlat.tif
43  https://www.lifewatch.be/sdmpredictors/WC_tmax11_lonlat.tif
44  https://www.lifewatch.be/sdmpredictors/WC_tmax12_lonlat.tif
45  https://www.lifewatch.be/sdmpredictors/WC_tmean1_lonlat.tif
46  https://www.lifewatch.be/sdmpredictors/WC_tmean2_lonlat.tif
47  https://www.lifewatch.be/sdmpredictors/WC_tmean3_lonlat.tif
48  https://www.lifewatch.be/sdmpredictors/WC_tmean4_lonlat.tif
49  https://www.lifewatch.be/sdmpredictors/WC_tmean5_lonlat.tif
50  https://www.lifewatch.be/sdmpredictors/WC_tmean6_lonlat.tif
51  https://www.lifewatch.be/sdmpredictors/WC_tmean7_lonlat.tif
52  https://www.lifewatch.be/sdmpredictors/WC_tmean8_lonlat.tif
53  https://www.lifewatch.be/sdmpredictors/WC_tmean9_lonlat.tif
54 https://www.lifewatch.be/sdmpredictors/WC_tmean10_lonlat.tif
55 https://www.lifewatch.be/sdmpredictors/WC_tmean11_lonlat.tif
56 https://www.lifewatch.be/sdmpredictors/WC_tmean12_lonlat.tif
57   https://www.lifewatch.be/sdmpredictors/WC_tmin1_lonlat.tif
58   https://www.lifewatch.be/sdmpredictors/WC_tmin2_lonlat.tif
59   https://www.lifewatch.be/sdmpredictors/WC_tmin3_lonlat.tif
60   https://www.lifewatch.be/sdmpredictors/WC_tmin4_lonlat.tif
61   https://www.lifewatch.be/sdmpredictors/WC_tmin5_lonlat.tif
62   https://www.lifewatch.be/sdmpredictors/WC_tmin6_lonlat.tif
63   https://www.lifewatch.be/sdmpredictors/WC_tmin7_lonlat.tif
64   https://www.lifewatch.be/sdmpredictors/WC_tmin8_lonlat.tif
65   https://www.lifewatch.be/sdmpredictors/WC_tmin9_lonlat.tif
66  https://www.lifewatch.be/sdmpredictors/WC_tmin10_lonlat.tif
67  https://www.lifewatch.be/sdmpredictors/WC_tmin11_lonlat.tif
68  https://www.lifewatch.be/sdmpredictors/WC_tmin12_lonlat.tif</code></pre>
</div>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Explore environmental layers</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>layers <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(sdmpredictors<span class="sc">::</span><span class="fu">list_layers</span>(<span class="st">"WorldClim"</span>))</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(layers, <span class="st">"./data/WorldClim.csv"</span>)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Download specific layers to the datadir</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Bio01 (annual mean temperature), Bio05 (temperature in hottest month), Bio15 (rainfall seasonality) and Bio19 (rainfall in coldest quarter)</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>WC <span class="ot">&lt;-</span> <span class="fu">load_layers</span>(<span class="fu">c</span>(<span class="st">"WC_bio1"</span>, <span class="st">"WC_bio5"</span>, <span class="st">"WC_bio15"</span>, <span class="st">"WC_bio19"</span>))</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(WC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session07_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Crop rasters to study area and align CRS</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="co"># get Murray-Darling Basin polygon</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>mdb <span class="ot">&lt;-</span> <span class="fu">st_read</span>(<span class="st">"data/spatial_data/MDB_polygon/MDB.shp"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Reading layer `MDB' from data source 
  `C:\repositories\dartR_workshop\kioloa\data\spatial_data\MDB_polygon\MDB.shp' 
  using driver `ESRI Shapefile'
Simple feature collection with 1 feature and 18 fields
Geometry type: MULTIPOLYGON
Dimension:     XY
Bounding box:  xmin: 138.57 ymin: -37.68 xmax: 152.49 ymax: -24.585
Geodetic CRS:  GDA94</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set CRS</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>mdb <span class="ot">&lt;-</span> <span class="fu">st_transform</span>(mdb, <span class="dv">4326</span>)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="co"># convert to sf format</span></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>mdb <span class="ot">&lt;-</span> <span class="fu">as</span>(mdb, <span class="st">"Spatial"</span>)</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a><span class="co"># crop WorldClim rasters to MDB extent, then mask pixels outside of the MDB</span></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>ENV <span class="ot">&lt;-</span> <span class="fu">crop</span>(WC, <span class="fu">extent</span>(mdb))</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>ENV <span class="ot">&lt;-</span> <span class="fu">mask</span>(ENV, mdb)</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a><span class="co"># get sampling sites, format as sf</span></span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>sites <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"data/Mf_xy.csv"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a>sites_sf <span class="ot">&lt;-</span> <span class="fu">st_as_sf</span>(sites, <span class="at">coords =</span> <span class="fu">c</span>(<span class="st">"X"</span>, <span class="st">"Y"</span>), <span class="at">crs =</span> <span class="dv">4326</span>)</span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-17"><a href="#cb35-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate a nice color ramp and plot the rasters </span></span>
<span id="cb35-18"><a href="#cb35-18" aria-hidden="true" tabindex="-1"></a>my.colors <span class="ot">=</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"#5E85B8"</span>,<span class="st">"#EDF0C0"</span>,<span class="st">"#C13127"</span>))</span>
<span id="cb35-19"><a href="#cb35-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-20"><a href="#cb35-20" aria-hidden="true" tabindex="-1"></a><span class="co"># include other spatial data that you might want to show</span></span>
<span id="cb35-21"><a href="#cb35-21" aria-hidden="true" tabindex="-1"></a>rivers <span class="ot">&lt;-</span> <span class="fu">st_read</span>(<span class="st">"data/spatial_data/Mf_streams/Mf_streams.shp"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Reading layer `Mf_streams' from data source 
  `C:\repositories\dartR_workshop\kioloa\data\spatial_data\Mf_streams\Mf_streams.shp' 
  using driver `ESRI Shapefile'
Simple feature collection with 2826 features and 31 fields
Geometry type: LINESTRING
Dimension:     XY
Bounding box:  xmin: 138.7888 ymin: -36.52625 xmax: 152.2712 ymax: -26.75375
Geodetic CRS:  GDA94</code></pre>
</div>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>rivers <span class="ot">&lt;-</span> <span class="fu">st_transform</span>(rivers, <span class="dv">4326</span>)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>rivers <span class="ot">&lt;-</span> <span class="fu">as</span>(rivers, <span class="st">"Spatial"</span>)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a><span class="co"># plot a layer</span></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ENV<span class="sc">$</span>WC_bio1, <span class="at">col=</span><span class="fu">my.colors</span>(<span class="dv">100</span>), <span class="at">main=</span><span class="fu">names</span>(ENV<span class="sc">$</span>WC_bio1))</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(rivers, <span class="at">col=</span><span class="st">"blue"</span>, <span class="at">lwd=</span><span class="fl">0.3</span>)</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="fu">st_coordinates</span>(sites_sf)[,<span class="dv">1</span>], <span class="fu">st_coordinates</span>(sites_sf)[,<span class="dv">2</span>], <span class="at">labels=</span>sites_sf<span class="sc">$</span>site, <span class="at">cex=</span><span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session07_files/figure-html/unnamed-chunk-7-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<section id="save-pdf" class="level5">
<h5 class="anchored" data-anchor-id="save-pdf">Save PDF</h5>
<div class="cell" data-hash="session07_cache/html/unnamed-chunk-8_ace0a88a20cfc478ad4c5ad4f46c79d3">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># format nicely and export as pdf</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>{<span class="fu">pdf</span>(<span class="st">"./images/env_rasters.pdf"</span>)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nlayers</span>(ENV)) {</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>  rasterLayer <span class="ot">&lt;-</span> ENV[[i]]</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># skew the colour ramp to improve visualisation (not necessary, but nice for data like this)</span></span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>  p95 <span class="ot">&lt;-</span> <span class="fu">quantile</span>(<span class="fu">values</span>(rasterLayer), <span class="at">probs =</span> <span class="fl">0.05</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>  breaks <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">minValue</span>(rasterLayer), <span class="fu">seq</span>(p95, <span class="fu">maxValue</span>(rasterLayer), <span class="at">length.out =</span> <span class="dv">50</span>))</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>  breaks <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">seq</span>(<span class="fu">minValue</span>(rasterLayer), p95, <span class="at">length.out =</span> <span class="dv">10</span>), breaks))</span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a>  color.palette <span class="ot">&lt;-</span> <span class="fu">my.colors</span>(<span class="fu">length</span>(breaks) <span class="sc">-</span> <span class="dv">1</span>)</span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a>  layerColors <span class="ot">&lt;-</span> <span class="cf">if</span>(i <span class="sc">&lt;</span> <span class="dv">4</span>) color.palette <span class="cf">else</span> <span class="fu">rev</span>(color.palette)</span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># tidy up the values displayed in the legend</span></span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a>  simplifiedBreaks <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">min</span>(breaks), <span class="fu">quantile</span>(breaks, <span class="at">probs =</span> <span class="fu">c</span>(<span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="fl">0.75</span>)), <span class="fu">max</span>(breaks))</span>
<span id="cb38-15"><a href="#cb38-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb38-16"><a href="#cb38-16" aria-hidden="true" tabindex="-1"></a>  <span class="co"># plot the rasters</span></span>
<span id="cb38-17"><a href="#cb38-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot</span>(rasterLayer, <span class="at">breaks=</span>breaks, <span class="at">col=</span>layerColors, <span class="at">main=</span><span class="fu">names</span>(rasterLayer),</span>
<span id="cb38-18"><a href="#cb38-18" aria-hidden="true" tabindex="-1"></a>       <span class="at">axis.args=</span><span class="fu">list</span>(<span class="at">at=</span>simplifiedBreaks, <span class="at">labels=</span><span class="fu">as.character</span>(<span class="fu">round</span>(simplifiedBreaks))),</span>
<span id="cb38-19"><a href="#cb38-19" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend.args=</span><span class="fu">list</span>(<span class="at">text=</span><span class="st">''</span>, <span class="at">side=</span><span class="dv">3</span>, <span class="at">line=</span><span class="dv">2</span>, <span class="at">at=</span>simplifiedBreaks))</span>
<span id="cb38-20"><a href="#cb38-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lines</span>(rivers, <span class="at">col=</span><span class="st">"blue"</span>, <span class="at">lwd=</span><span class="fl">0.3</span>, <span class="fu">labels</span>(rivers<span class="sc">$</span>Name))</span>
<span id="cb38-21"><a href="#cb38-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">text</span>(<span class="fu">st_coordinates</span>(sites_sf)[,<span class="dv">1</span>], <span class="fu">st_coordinates</span>(sites_sf)[,<span class="dv">2</span>], <span class="at">labels=</span>sites_sf<span class="sc">$</span>site, <span class="at">cex=</span><span class="fl">0.5</span>)</span>
<span id="cb38-22"><a href="#cb38-22" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb38-23"><a href="#cb38-23" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="extract-site-environmental-data" class="level3">
<h3 class="anchored" data-anchor-id="extract-site-environmental-data">Extract site environmental data</h3>
<div class="cell" data-hash="session07_cache/html/unnamed-chunk-9_35326369d6559b34c855a62bb5f47876">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># and finally extract the environmental data for your sample sites</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>env.site <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">site =</span> sites_sf<span class="sc">$</span>site,</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">X =</span> <span class="fu">st_coordinates</span>(sites_sf)[,<span class="dv">1</span>],</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">Y =</span> <span class="fu">st_coordinates</span>(sites_sf)[,<span class="dv">2</span>],</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>                       <span class="at">omegaX =</span> sites_sf<span class="sc">$</span>omegaX,</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>                       <span class="at">omegaY =</span> sites_sf<span class="sc">$</span>omegaY)</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>env.dat <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(raster<span class="sc">::</span><span class="fu">extract</span>(ENV,<span class="fu">st_coordinates</span>(sites_sf)))</span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>env.site <span class="ot">&lt;-</span> <span class="fu">cbind</span>(env.site, env.dat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="no-missing-data" class="level3">
<h3 class="anchored" data-anchor-id="no-missing-data">No missing data</h3>
<div class="cell" data-hash="session07_cache/html/unnamed-chunk-10_137d628b2a23cc835abd2dc314f091db">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># we now have our SNP data with no missing genotypes</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="co"># to get a feel for the data we can run quick PCA using the rda function</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>pc <span class="ot">&lt;-</span> <span class="fu">rda</span>(AF)</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pc)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session07_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set some plotting variables to make it easier to interpret</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="co"># set colours, markers and labels for plotting and add to env.site df</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>env.site<span class="sc">$</span>cols <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'darkturquoise'</span>, <span class="st">'darkturquoise'</span>, <span class="st">'darkturquoise'</span>, <span class="st">'limegreen'</span>, <span class="st">'limegreen'</span>, <span class="st">'darkorange1'</span>, <span class="st">'slategrey'</span>, <span class="st">'slategrey'</span>, <span class="st">'slategrey'</span>,</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'slategrey'</span>, <span class="st">'slategrey'</span>, <span class="st">'dodgerblue3'</span>, <span class="st">'gold'</span>, <span class="st">'gold'</span>)</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>env.site<span class="sc">$</span>pch <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">15</span>, <span class="dv">16</span>, <span class="dv">17</span>, <span class="dv">17</span>, <span class="dv">15</span>, <span class="dv">15</span>, <span class="dv">16</span>, <span class="dv">15</span>, <span class="dv">18</span>, <span class="dv">16</span>, <span class="dv">16</span>, <span class="dv">8</span>, <span class="dv">16</span>, <span class="dv">16</span>)</span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a><span class="co"># generate labels for each axis</span></span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>x.lab <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"PC1 ("</span>, <span class="fu">paste</span>(<span class="fu">round</span>((pc<span class="sc">$</span>CA<span class="sc">$</span>eig[<span class="dv">1</span>]<span class="sc">/</span>pc<span class="sc">$</span>tot.chi<span class="sc">*</span><span class="dv">100</span>),<span class="dv">2</span>)),<span class="st">"%)"</span>)</span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>y.lab <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"PC2 ("</span>, <span class="fu">paste</span>(<span class="fu">round</span>((pc<span class="sc">$</span>CA<span class="sc">$</span>eig[<span class="dv">2</span>]<span class="sc">/</span>pc<span class="sc">$</span>tot.chi<span class="sc">*</span><span class="dv">100</span>),<span class="dv">2</span>)),<span class="st">"%)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plot-pca" class="level3">
<h3 class="anchored" data-anchor-id="plot-pca">Plot PCA</h3>
<div class="cell" data-hash="session07_cache/html/unnamed-chunk-11_f0f2a118fd84dd2365321989cbf7ef27">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot PCA</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>{<span class="fu">pdf</span>(<span class="at">file =</span> <span class="st">"./images/Mf_PCA.pdf"</span>, <span class="at">height =</span> <span class="dv">6</span>, <span class="at">width =</span> <span class="dv">6</span>)</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>pcaplot <span class="ot">&lt;-</span> <span class="fu">plot</span>(pc, <span class="at">choices =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">type =</span> <span class="st">"n"</span>, <span class="at">xlab=</span>x.lab, <span class="at">ylab=</span>y.lab, <span class="at">cex.lab=</span><span class="dv">1</span>)</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a><span class="fu">with</span>(env.site, <span class="fu">points</span>(pc, <span class="at">display =</span> <span class="st">"sites"</span>, <span class="at">col =</span> env.site<span class="sc">$</span>cols, <span class="at">pch =</span> env.site<span class="sc">$</span>pch, <span class="at">cex=</span><span class="fl">1.5</span>, <span class="at">bg =</span> env.site<span class="sc">$</span>cols))</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topleft"</span>, <span class="at">legend =</span> env.site<span class="sc">$</span>site, <span class="at">col=</span>env.site<span class="sc">$</span>cols, <span class="at">pch=</span>env.site<span class="sc">$</span>pch, <span class="at">pt.cex=</span><span class="dv">1</span>, <span class="at">cex=</span><span class="fl">0.50</span>, <span class="at">xpd=</span><span class="dv">1</span>, <span class="at">box.lty =</span> <span class="dv">0</span>, <span class="at">bg=</span> <span class="st">"transparent"</span>)</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="images/Mf_PCA.png" class="img-fluid"> ## 2. Variable filtering</p>
<div class="cell" data-hash="session07_cache/html/unnamed-chunk-13_7444275287b5b611b3bb26f3ed022942">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># extract predictor variables to matrix</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>env_var <span class="ot">&lt;-</span> env.site[ , <span class="fu">c</span>(<span class="st">"WC_bio1"</span>, <span class="st">"WC_bio5"</span>, <span class="st">"WC_bio15"</span>, <span class="st">"WC_bio19"</span>)]</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a><span class="co"># check for obvious correlations between variables</span></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a><span class="fu">pairs.panels</span>(env_var, <span class="at">scale=</span>T, <span class="at">lm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session07_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="do">## reduce variance associated with correlated environmental PCs using VIF analyses</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="co"># run procedure to remove variables with the highest VIF one at a time until all remaining variables are below 10</span></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>keep.env <span class="ot">&lt;-</span><span class="fu">vif_func</span>(<span class="at">in_frame=</span>env_var,<span class="at">thresh=</span><span class="dv">2</span>,<span class="at">trace=</span>T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> var      vif             
 WC_bio1  21.5371434397055
 WC_bio5  10.4639698247348
 WC_bio15 7.69781293748316
 WC_bio19 3.97039440920372

removed:  WC_bio1 21.53714 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>keep.env  <span class="co"># the retained environmental variables</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "WC_bio5"  "WC_bio15" "WC_bio19"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>reduced.env <span class="ot">&lt;-</span> <span class="fu">subset</span>(<span class="fu">as.data.frame</span>(env_var), <span class="at">select=</span><span class="fu">c</span>(keep.env))</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>scaled.env <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">scale</span>(reduced.env))</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a><span class="co"># lets have another look</span></span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a><span class="fu">pairs.panels</span>(reduced.env, <span class="at">scale=</span>T, <span class="at">lm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session07_files/figure-html/unnamed-chunk-13-2.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="run-analysis" class="level2">
<h2 class="anchored" data-anchor-id="run-analysis">3. Run analysis</h2>
<p>So we have formatted genotypes, environmental data</p>
<p>What about controlling for spatial population structure in the model?</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Decision time:
</div>
</div>
<div class="callout-body-container callout-body">
<p>Decision time: 1. do I need to control for spatial structure?</p>
<pre><code> No? Great!

Yes? How do I do that?
 </code></pre>
<ul>
<li>simple xy coordinates (or some transformation, MDS, polynomial…)</li>
<li>Principal Coordinates of Neighbourhood Matrix (PCNM; R package vegan)</li>
<li>Moran’s Eigenvector Maps (R package memgene)</li>
<li>Fst, admixture Q values</li>
<li>population allele frequency covariance matrix (BayPass omega)</li>
</ul>
</div>
</div>
<p>We are going to use the scaled allelic covariance (Ω) estimated by BayPass to control for spatial structure in the data, see BayPass and Gates et al.&nbsp;2023</p>
<div class="cell" data-hash="session07_cache/html/unnamed-chunk-14_bb5bdf0ccdcecb21a43161d9a8346c7c">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># format spatial coordinates as matrix</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>xy <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">cbind</span>(env.site<span class="sc">$</span>omegaX, env.site<span class="sc">$</span>omegaY))</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a><span class="do">################################################################################</span></span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a><span class="co"># reduced RDA model using the retained environmental PCs </span></span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a><span class="co"># conditioned on the retained spatial variables</span></span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a>Mf.RDA <span class="ot">&lt;-</span> <span class="fu">rda</span>(AF <span class="sc">~</span> WC_bio5 <span class="sc">+</span> WC_bio15 <span class="sc">+</span> WC_bio19 <span class="sc">+</span> <span class="fu">Condition</span>(xy), <span class="at">data =</span> scaled.env)</span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a>Mf.RDA</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Call: rda(formula = AF ~ WC_bio5 + WC_bio15 + WC_bio19 + Condition(xy),
data = scaled.env)

                Inertia Proportion Rank
Total         202.83442    1.00000     
Conditional     9.32930    0.04599    2
Constrained   102.05924    0.50317    3
Unconstrained  91.44588    0.45084    8
Inertia is variance 

Eigenvalues for constrained axes:
 RDA1  RDA2  RDA3 
85.00 12.97  4.09 

Eigenvalues for unconstrained axes:
  PC1   PC2   PC3   PC4   PC5   PC6   PC7   PC8 
66.70  6.36  5.17  3.75  3.10  2.86  2.41  1.10 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co"># So how much genetic variation can be explained by our environmental model?</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="fu">RsquareAdj</span>(Mf.RDA)<span class="sc">$</span>r.squared</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.5031653</code></pre>
</div>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># how much inertia is associated with each axis</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="fu">screeplot</span>(Mf.RDA)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session07_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate significance of the reduced model, marginal effect of each term and significance of each axis</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="co">#   (this can take several hours, depending on your computer)</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>mod_perm <span class="ot">&lt;-</span> <span class="fu">anova.cca</span>(Mf.RDA, <span class="at">nperm=</span><span class="dv">1000</span>) <span class="co">#test significance of the model</span></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>mod_perm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Permutation test for rda under reduced model
Permutation: free
Number of permutations: 999

Model: rda(formula = AF ~ WC_bio5 + WC_bio15 + WC_bio19 + Condition(xy), data = scaled.env)
         Df Variance      F Pr(&gt;F)  
Model     3  102.059 2.9762   0.04 *
Residual  8   91.446                
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co">#generate x and y labels (% constrained variation)</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>x.lab <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"RDA1 ("</span>, <span class="fu">paste</span>(<span class="fu">round</span>((Mf.RDA<span class="sc">$</span>CCA<span class="sc">$</span>eig[<span class="dv">1</span>]<span class="sc">/</span>Mf.RDA<span class="sc">$</span>CCA<span class="sc">$</span>tot.chi<span class="sc">*</span><span class="dv">100</span>),<span class="dv">2</span>)),<span class="st">"%)"</span>)</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>y.lab <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"RDA2 ("</span>, <span class="fu">paste</span>(<span class="fu">round</span>((Mf.RDA<span class="sc">$</span>CCA<span class="sc">$</span>eig[<span class="dv">2</span>]<span class="sc">/</span>Mf.RDA<span class="sc">$</span>CCA<span class="sc">$</span>tot.chi<span class="sc">*</span><span class="dv">100</span>),<span class="dv">2</span>)),<span class="st">"%)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="plot-rda" class="level3">
<h3 class="anchored" data-anchor-id="plot-rda">Plot RDA</h3>
<div class="cell" data-hash="session07_cache/html/unnamed-chunk-15_97e631460a79f897f58b3edcc8cc31b5">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co">#plot RDA1, RDA2</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>{<span class="fu">pdf</span>(<span class="at">file =</span> <span class="st">"./images/Mf_RDA.pdf"</span>, <span class="at">height =</span> <span class="dv">6</span>, <span class="at">width =</span> <span class="dv">6</span>)</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>pRDAplot <span class="ot">&lt;-</span> <span class="fu">plot</span>(Mf.RDA, <span class="at">choices =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">type=</span><span class="st">"n"</span>, <span class="at">cex.lab=</span><span class="dv">1</span>, <span class="at">xlab=</span>x.lab, <span class="at">ylab=</span>y.lab)</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a><span class="fu">with</span>(env.site, <span class="fu">points</span>(Mf.RDA, <span class="at">display =</span> <span class="st">"sites"</span>, <span class="at">col =</span> env.site<span class="sc">$</span>cols, <span class="at">pch =</span> env.site<span class="sc">$</span>pch, <span class="at">cex=</span><span class="fl">1.5</span>, <span class="at">bg =</span> env.site<span class="sc">$</span>cols))</span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(Mf.RDA, <span class="st">"bp"</span>,<span class="at">choices =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"WC_bio5"</span>, <span class="st">"WC_bio15"</span>, <span class="st">"WC_bio19"</span>), <span class="at">col=</span><span class="st">"blue"</span>, <span class="at">cex=</span><span class="fl">0.6</span>)</span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topleft"</span>, <span class="at">legend =</span> env.site<span class="sc">$</span>site, <span class="at">col=</span>env.site<span class="sc">$</span>cols, <span class="at">pch=</span>env.site<span class="sc">$</span>pch, <span class="at">pt.cex=</span><span class="dv">1</span>, <span class="at">cex=</span><span class="fl">0.50</span>, <span class="at">xpd=</span><span class="dv">1</span>, <span class="at">box.lty =</span> <span class="dv">0</span>, <span class="at">bg=</span> <span class="st">"transparent"</span>)</span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="images/Mf_RDA.png" class="img-fluid"></p>
</section>
</section>
<section id="identify-candidate-loci" class="level2">
<h2 class="anchored" data-anchor-id="identify-candidate-loci">4. Identify candidate loci</h2>
<p>Estimate p values following:&nbsp;<a href="https://kjschmidlab.gitlab.io/b1k-gbs/GenomeEnvAssociation.html#RDA_approaches" class="uri">https://kjschmidlab.gitlab.io/b1k-gbs/GenomeEnvAssociation.html#RDA_approaches</a> and the R function rdadapt (see <a href="https://github.com/Capblancq/RDA-genome-scan" class="uri">https://github.com/Capblancq/RDA-genome-scan</a>)</p>
<div class="cell" data-hash="session07_cache/html/unnamed-chunk-17_7c537923cbe454339783f8a462dfe9b0">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>k <span class="ot">=</span> <span class="dv">2</span> <span class="co"># number of RDA axes to retain</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>rda.q <span class="ot">&lt;-</span> <span class="fu">rdadapt</span>(Mf.RDA, <span class="at">K =</span> k)</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(rda.q<span class="sc">$</span>q.values <span class="sc">&lt;=</span> <span class="fl">0.01</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 373</code></pre>
</div>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>candidate.loci <span class="ot">&lt;-</span> <span class="fu">colnames</span>(AF)[<span class="fu">which</span>(rda.q[,<span class="dv">2</span>] <span class="sc">&lt;</span> <span class="fl">0.01</span>)]</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a><span class="co"># check the distribution of candidates and q-values</span></span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x=</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(rda.q[,<span class="dv">2</span>])), </span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a>                 <span class="at">y=</span><span class="sc">-</span><span class="fu">log10</span>(rda.q[,<span class="dv">2</span>])), <span class="at">col=</span><span class="st">"gray83"</span>) <span class="sc">+</span></span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x=</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(rda.q[,<span class="dv">2</span>]))[<span class="fu">which</span>(rda.q[,<span class="dv">2</span>] <span class="sc">&lt;</span> <span class="fl">0.01</span>)], </span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a>                 <span class="at">y=</span><span class="sc">-</span><span class="fu">log10</span>(rda.q[<span class="fu">which</span>(rda.q[,<span class="dv">2</span>] <span class="sc">&lt;</span> <span class="fl">0.01</span>),<span class="dv">2</span>])), <span class="at">col=</span><span class="st">"orange"</span>) <span class="sc">+</span></span>
<span id="cb61-10"><a href="#cb61-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"SNPs"</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">"-log10(q.values"</span>) <span class="sc">+</span></span>
<span id="cb61-11"><a href="#cb61-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session07_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="co"># check correlations between candidate loci and environmental predictors</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>npred <span class="ot">&lt;-</span> <span class="dv">3</span></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>env_mat <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">nrow=</span>(<span class="fu">sum</span>(rda.q<span class="sc">$</span>q.values <span class="sc">&lt;</span> <span class="fl">0.01</span>)), </span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>                  <span class="at">ncol=</span>npred)  <span class="co"># n columns for n predictors</span></span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(env_mat) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(reduced.env)</span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-8"><a href="#cb62-8" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate correlation between candidate snps and environmental variables</span></span>
<span id="cb62-9"><a href="#cb62-9" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(candidate.loci)) {</span>
<span id="cb62-10"><a href="#cb62-10" aria-hidden="true" tabindex="-1"></a>  nam <span class="ot">&lt;-</span> candidate.loci[i]</span>
<span id="cb62-11"><a href="#cb62-11" aria-hidden="true" tabindex="-1"></a>  snp.gen <span class="ot">&lt;-</span> AF[,nam]</span>
<span id="cb62-12"><a href="#cb62-12" aria-hidden="true" tabindex="-1"></a>  env_mat[i,] <span class="ot">&lt;-</span> <span class="fu">apply</span>(reduced.env,<span class="dv">2</span>,<span class="cf">function</span>(x) <span class="fu">cor</span>(x,snp.gen))</span>
<span id="cb62-13"><a href="#cb62-13" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb62-14"><a href="#cb62-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-15"><a href="#cb62-15" aria-hidden="true" tabindex="-1"></a>cand <span class="ot">&lt;-</span> <span class="fu">cbind.data.frame</span>(<span class="fu">as.data.frame</span>(candidate.loci),env_mat)  </span>
<span id="cb62-16"><a href="#cb62-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-17"><a href="#cb62-17" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(cand<span class="sc">$</span>candidate.loci)) {</span>
<span id="cb62-18"><a href="#cb62-18" aria-hidden="true" tabindex="-1"></a>  bar <span class="ot">&lt;-</span> cand[i,]</span>
<span id="cb62-19"><a href="#cb62-19" aria-hidden="true" tabindex="-1"></a>  cand[i,npred<span class="sc">+</span><span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="fu">names</span>(<span class="fu">which.max</span>(<span class="fu">abs</span>(bar[,<span class="dv">2</span><span class="sc">:</span><span class="dv">4</span>]))) <span class="co"># gives the variable</span></span>
<span id="cb62-20"><a href="#cb62-20" aria-hidden="true" tabindex="-1"></a>  cand[i,npred<span class="sc">+</span><span class="dv">3</span>] <span class="ot">&lt;-</span> <span class="fu">max</span>(<span class="fu">abs</span>(bar[,<span class="dv">2</span><span class="sc">:</span><span class="dv">4</span>]))              <span class="co"># gives the correlation</span></span>
<span id="cb62-21"><a href="#cb62-21" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb62-22"><a href="#cb62-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-23"><a href="#cb62-23" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(cand)[npred<span class="sc">+</span><span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="st">"predictor"</span></span>
<span id="cb62-24"><a href="#cb62-24" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(cand)[npred<span class="sc">+</span><span class="dv">3</span>] <span class="ot">&lt;-</span> <span class="st">"correlation"</span></span>
<span id="cb62-25"><a href="#cb62-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-26"><a href="#cb62-26" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(cand<span class="sc">$</span>predictor) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
WC_bio15 WC_bio19  WC_bio5 
     216       47      110 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(cand, <span class="st">"./data/RDA_candidates.csv"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="individual-level" class="level3">
<h3 class="anchored" data-anchor-id="individual-level">Individual level</h3>
<p>Now let’s have a look at an individual level biplot first expand environmental data from population to individual dataframe.</p>
<div class="cell" data-hash="session07_cache/html/unnamed-chunk-18_d7847aa50f6b06ab6f99fc31348bfce2">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>env.ind <span class="ot">&lt;-</span> <span class="fu">expand_pop2ind</span>(Mf5000_gl, env.site)</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a><span class="co"># now subset our genotype objects to the candidate loci</span></span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>candidate.gl <span class="ot">&lt;-</span> <span class="fu">gl.keep.loc</span>(<span class="fu">gi2gl</span>(imputed.sorted.gi), <span class="at">loc.list=</span>candidate.loci)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.keep.loc 
Starting gi2gl 
Starting gl.compliance.check 
  Processing genlight object with SNP data
  The slot loc.all, which stores allele name for each locus, is empty. Creating a dummy variable (A/C) to insert in this slot. 
  Checking coding of SNPs
    SNP data scored NA, 0, 1 or 2 confirmed
  Checking for population assignments
    Population assignments confirmed
  Checking locus metrics and flags
  Recalculating locus metrics
  Checking for monomorphic loci
    No monomorphic loci detected
  Checking for loci with all missing data
    No loci with all missing data detected
  Checking whether individual names are unique.
  Checking for individual metrics
  Warning: Creating a slot for individual metrics
  Spelling of coordinates checked and changed if necessary to 
            lat/lon
Completed: gl.compliance.check 
Completed: gi2gl 
  Processing genlight object with SNP data
  List of loci to keep has been specified
  Deleting all but the specified loci
Completed: gl.keep.loc </code></pre>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>candidate.gi <span class="ot">&lt;-</span> <span class="fu">gl2gi</span>(candidate.gl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl2gi 
  Processing genlight object with SNP data
Matrix converted.. Prepare genind object...
Completed: gl2gi </code></pre>
</div>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get genotypes (counts of reference allele) and clean up locus names</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>candidate.alleles <span class="ot">&lt;-</span> candidate.gi<span class="sc">@</span>tab</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>candidate.snps <span class="ot">&lt;-</span> candidate.alleles[,<span class="fu">seq</span>(<span class="dv">1</span>,<span class="fu">ncol</span>(candidate.alleles),<span class="dv">2</span>)]</span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(candidate.snps) <span class="ot">&lt;-</span> <span class="fu">locNames</span>(candidate.gi)</span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a><span class="co"># finally run the individual based RDA</span></span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a><span class="co"># no need to control for pop structure this time</span></span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a>Mfcandidate.RDA <span class="ot">&lt;-</span> <span class="fu">rda</span>(candidate.snps <span class="sc">~</span> WC_bio5 <span class="sc">+</span> WC_bio15 <span class="sc">+</span> WC_bio19, <span class="at">data =</span> env.ind)</span>
<span id="cb69-9"><a href="#cb69-9" aria-hidden="true" tabindex="-1"></a>Mfcandidate.RDA</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Call: rda(formula = candidate.snps ~ WC_bio5 + WC_bio15 + WC_bio19,
data = env.ind)

               Inertia Proportion Rank
Total         207.1000     1.0000     
Constrained    59.6948     0.2882    3
Unconstrained 147.4052     0.7118  245
Inertia is variance 

Eigenvalues for constrained axes:
 RDA1  RDA2  RDA3 
44.71 12.02  2.97 

Eigenvalues for unconstrained axes:
   PC1    PC2    PC3    PC4    PC5    PC6    PC7    PC8 
16.227  7.584  5.579  4.597  3.536  3.373  3.075  2.695 
(Showing 8 of 245 unconstrained eigenvalues)</code></pre>
</div>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>ind.mod_perm <span class="ot">&lt;-</span> <span class="fu">anova.cca</span>(Mfcandidate.RDA, <span class="at">nperm=</span><span class="dv">1000</span>, </span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">parallel=</span><span class="dv">4</span>) <span class="co">#test significance of the model</span></span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>ind.mod_perm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Permutation test for rda under reduced model
Permutation: free
Number of permutations: 999

Model: rda(formula = candidate.snps ~ WC_bio5 + WC_bio15 + WC_bio19, data = env.ind)
          Df Variance      F Pr(&gt;F)    
Model      3   59.695 33.073  0.001 ***
Residual 245  147.405                  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>x.lab <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"RDA1 ("</span>,</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>                <span class="fu">paste</span>(<span class="fu">round</span>((Mfcandidate.RDA<span class="sc">$</span>CA<span class="sc">$</span>eig[<span class="dv">1</span>]<span class="sc">/</span>Mfcandidate.RDA<span class="sc">$</span>tot.chi<span class="sc">*</span><span class="dv">100</span>),<span class="dv">2</span>)),<span class="st">"%)"</span>)</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>y.lab <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"RDA2 ("</span>,</span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a>                <span class="fu">paste</span>(<span class="fu">round</span>((Mfcandidate.RDA<span class="sc">$</span>CA<span class="sc">$</span>eig[<span class="dv">2</span>]<span class="sc">/</span>Mfcandidate.RDA<span class="sc">$</span>tot.chi<span class="sc">*</span><span class="dv">100</span>),<span class="dv">2</span>)),<span class="st">"%)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plot-rda-1" class="level3">
<h3 class="anchored" data-anchor-id="plot-rda-1">Plot RDA</h3>
<div class="cell" data-hash="session07_cache/html/unnamed-chunk-19_914515ccef97dd74ca106f85efb2f748">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="co">#plot RDA1, RDA2</span></span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>{<span class="fu">pdf</span>(<span class="at">file =</span> <span class="st">"Mfcandidate_RDA.pdf"</span>, <span class="at">height =</span> <span class="dv">6</span>, <span class="at">width =</span> <span class="dv">6</span>)</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>pRDAplot <span class="ot">&lt;-</span> <span class="fu">plot</span>(Mfcandidate.RDA, <span class="at">choices =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), </span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">type=</span><span class="st">"n"</span>, <span class="at">cex.lab=</span><span class="dv">1</span>, <span class="at">xlab=</span>x.lab, <span class="at">ylab=</span>y.lab)</span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(Mfcandidate.RDA, <span class="at">display =</span> <span class="st">"sites"</span>, <span class="at">col =</span> env.ind<span class="sc">$</span>cols,</span>
<span id="cb74-6"><a href="#cb74-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">pch =</span> env.ind<span class="sc">$</span>pch, <span class="at">cex=</span><span class="fl">0.8</span>, <span class="at">bg =</span> env.ind<span class="sc">$</span>cols)</span>
<span id="cb74-7"><a href="#cb74-7" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(Mfcandidate.RDA, <span class="st">"bp"</span>,<span class="at">choices =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), </span>
<span id="cb74-8"><a href="#cb74-8" aria-hidden="true" tabindex="-1"></a>     <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"WC_bio5"</span>, <span class="st">"WC_bio15"</span>, <span class="st">"WC_bio19"</span>), <span class="at">col=</span><span class="st">"blue"</span>, <span class="at">cex=</span><span class="fl">0.6</span>)</span>
<span id="cb74-9"><a href="#cb74-9" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topleft"</span>, <span class="at">legend =</span> env.site<span class="sc">$</span>site, <span class="at">col=</span>env.site<span class="sc">$</span>cols, <span class="at">pch=</span>env.site<span class="sc">$</span>pch,</span>
<span id="cb74-10"><a href="#cb74-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">pt.cex=</span><span class="dv">1</span>, <span class="at">cex=</span><span class="fl">0.50</span>, <span class="at">xpd=</span><span class="dv">1</span>, <span class="at">box.lty =</span> <span class="dv">0</span>, <span class="at">bg=</span> <span class="st">"transparent"</span>)</span>
<span id="cb74-11"><a href="#cb74-11" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb74-12"><a href="#cb74-12" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="images/Mfcandidate_RDA.png" class="img-fluid"></p>
</section>
</section>
<section id="further-study" class="level2">
<h2 class="anchored" data-anchor-id="further-study"><em>Further Study</em></h2>
<p>still to come…</p>


</section>

</main> <!-- /main -->
<script>

/* update total correct if #webex-total_correct exists */
update_total_correct = function() {
  console.log("webex: update total_correct");

  var t = document.getElementsByClassName("webex-total_correct");
  for (var i = 0; i < t.length; i++) {
    p = t[i].parentElement;
    var correct = p.getElementsByClassName("webex-correct").length;
    var solvemes = p.getElementsByClassName("webex-solveme").length;
    var radiogroups = p.getElementsByClassName("webex-radiogroup").length;
    var selects = p.getElementsByClassName("webex-select").length;

    t[i].innerHTML = correct + " of " + (solvemes + radiogroups + selects) + " correct";
  }
}

/* webex-solution button toggling function */
b_func = function() {
  console.log("webex: toggle hide");

  var cl = this.parentElement.classList;
  if (cl.contains('open')) {
    cl.remove("open");
  } else {
    cl.add("open");
  }
}

/* check answers */
check_func = function() {
  console.log("webex: check answers");

  var cl = this.parentElement.classList;
  if (cl.contains('unchecked')) {
    cl.remove("unchecked");
    this.innerHTML = "Hide Answers";
  } else {
    cl.add("unchecked");
    this.innerHTML = "Show Answers";
  }
}

/* function for checking solveme answers */
solveme_func = function(e) {
  console.log("webex: check solveme");

  var real_answers = JSON.parse(this.dataset.answer);
  var my_answer = this.value;
  var cl = this.classList;
  if (cl.contains("ignorecase")) {
    my_answer = my_answer.toLowerCase();
  }
  if (cl.contains("nospaces")) {
    my_answer = my_answer.replace(/ /g, "")
  }

  if (my_answer == "") {
    cl.remove("webex-correct");
    cl.remove("webex-incorrect");
  } else if (real_answers.includes(my_answer)) {
    cl.add("webex-correct");
    cl.remove("webex-incorrect");
  } else {
    cl.add("webex-incorrect");
    cl.remove("webex-correct");
  }

  // match numeric answers within a specified tolerance
  if(this.dataset.tol > 0){
    var tol = JSON.parse(this.dataset.tol);
    var matches = real_answers.map(x => Math.abs(x - my_answer) < tol)
    if (matches.reduce((a, b) => a + b, 0) > 0) {
      cl.add("webex-correct");
    } else {
      cl.remove("webex-correct");
    }
  }

  // added regex bit
  if (cl.contains("regex")){
    answer_regex = RegExp(real_answers.join("|"))
    if (answer_regex.test(my_answer)) {
      cl.add("webex-correct");
    }
  }

  update_total_correct();
}

/* function for checking select answers */
select_func = function(e) {
  console.log("webex: check select");

  var cl = this.classList

  /* add style */
  cl.remove("webex-incorrect");
  cl.remove("webex-correct");
  if (this.value == "answer") {
    cl.add("webex-correct");
  } else if (this.value != "blank") {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

/* function for checking radiogroups answers */
radiogroups_func = function(e) {
  console.log("webex: check radiogroups");

  var checked_button = document.querySelector('input[name=' + this.id + ']:checked');
  var cl = checked_button.parentElement.classList;
  var labels = checked_button.parentElement.parentElement.children;

  /* get rid of styles */
  for (i = 0; i < labels.length; i++) {
    labels[i].classList.remove("webex-incorrect");
    labels[i].classList.remove("webex-correct");
  }

  /* add style */
  if (checked_button.value == "answer") {
    cl.add("webex-correct");
  } else {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

window.onload = function() {
  console.log("webex onload");
  /* set up solution buttons */
  var buttons = document.getElementsByTagName("button");

  for (var i = 0; i < buttons.length; i++) {
    if (buttons[i].parentElement.classList.contains('webex-solution')) {
      buttons[i].onclick = b_func;
    }
  }

  var check_sections = document.getElementsByClassName("webex-check");
  console.log("check:", check_sections.length);
  for (var i = 0; i < check_sections.length; i++) {
    check_sections[i].classList.add("unchecked");

    let btn = document.createElement("button");
    btn.innerHTML = "Show Answers";
    btn.classList.add("webex-check-button");
    btn.onclick = check_func;
    check_sections[i].appendChild(btn);

    let spn = document.createElement("span");
    spn.classList.add("webex-total_correct");
    check_sections[i].appendChild(spn);
  }

  /* set up webex-solveme inputs */
  var solveme = document.getElementsByClassName("webex-solveme");

  for (var i = 0; i < solveme.length; i++) {
    /* make sure input boxes don't auto-anything */
    solveme[i].setAttribute("autocomplete","off");
    solveme[i].setAttribute("autocorrect", "off");
    solveme[i].setAttribute("autocapitalize", "off");
    solveme[i].setAttribute("spellcheck", "false");
    solveme[i].value = "";

    /* adjust answer for ignorecase or nospaces */
    var cl = solveme[i].classList;
    var real_answer = solveme[i].dataset.answer;
    if (cl.contains("ignorecase")) {
      real_answer = real_answer.toLowerCase();
    }
    if (cl.contains("nospaces")) {
      real_answer = real_answer.replace(/ /g, "");
    }
    solveme[i].dataset.answer = real_answer;

    /* attach checking function */
    solveme[i].onkeyup = solveme_func;
    solveme[i].onchange = solveme_func;

    solveme[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  /* set up radiogroups */
  var radiogroups = document.getElementsByClassName("webex-radiogroup");
  for (var i = 0; i < radiogroups.length; i++) {
    radiogroups[i].onchange = radiogroups_func;
  }

  /* set up selects */
  var selects = document.getElementsByClassName("webex-select");
  for (var i = 0; i < selects.length; i++) {
    selects[i].onchange = select_func;
    selects[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  update_total_correct();
}

</script>
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
          // default icon
          link.classList.add("external");
      }
    }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./session06.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Management of Small Populations</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./session08.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Landscape Genetics</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>