<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Kioloa PopGen - 7&nbsp; Natural Selection</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./session08.html" rel="next">
<link href="./session06.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="css.css">
<link rel="stylesheet" href="include/webex.css">
</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./session07_CB_Yuma.html"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Natural Selection</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Kioloa PopGen</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome!</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./schedule.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Program</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./speakers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Speaker Bios</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./install.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Install dartRverse</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session01.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Intro to dartR</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session02.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Pop Gen In Conservation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session03.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Sequencing Technologies</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session04.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Data Management, Reproducibility &amp; Integrity</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session05.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Effective Population Size</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session06.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Management of Small Populations</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session07_CB_Yuma.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Natural Selection</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session08.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Landscape Genetics</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session09.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Lineage Divergence</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session10.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Sex Linked Markers</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session11.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">From Genes to Kin: Dissecting Relatedness &amp; Kinship</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session12.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Genetic Structure</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session13_JD.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Combining Genomic Resources</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./break.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">__________</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./useful.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">More info</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">dartR Tutorials</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#required-packages" id="toc-required-packages" class="nav-link active" data-scroll-target="#required-packages"><em>Required packages</em></a></li>
  <li><a href="#gea-analysis-of-snp-and-environmental-data-using-rda" id="toc-gea-analysis-of-snp-and-environmental-data-using-rda" class="nav-link" data-scroll-target="#gea-analysis-of-snp-and-environmental-data-using-rda"><strong>GEA analysis of SNP and environmental data using RDA</strong></a></li>
  <li><a href="#data-preparation" id="toc-data-preparation" class="nav-link" data-scroll-target="#data-preparation">1. Data preparation</a>
  <ul class="collapse">
  <li><a href="#format-snp-data-for-ordination-analysis-in-vegan" id="toc-format-snp-data-for-ordination-analysis-in-vegan" class="nav-link" data-scroll-target="#format-snp-data-for-ordination-analysis-in-vegan">Format SNP data for ordination analysis in vegan</a></li>
  <li><a href="#get-environmental-data" id="toc-get-environmental-data" class="nav-link" data-scroll-target="#get-environmental-data">Get environmental data</a></li>
  <li><a href="#extract-site-environmental-data" id="toc-extract-site-environmental-data" class="nav-link" data-scroll-target="#extract-site-environmental-data">Extract site environmental data</a></li>
  <li><a href="#pca" id="toc-pca" class="nav-link" data-scroll-target="#pca">PCA</a></li>
  <li><a href="#plot-pca" id="toc-plot-pca" class="nav-link" data-scroll-target="#plot-pca">Plot PCA</a></li>
  </ul></li>
  <li><a href="#run-analysis" id="toc-run-analysis" class="nav-link" data-scroll-target="#run-analysis">3. Run analysis</a>
  <ul class="collapse">
  <li><a href="#plot-rda" id="toc-plot-rda" class="nav-link" data-scroll-target="#plot-rda">Plot RDA</a></li>
  </ul></li>
  <li><a href="#identify-candidate-loci" id="toc-identify-candidate-loci" class="nav-link" data-scroll-target="#identify-candidate-loci">4. Identify candidate loci</a>
  <ul class="collapse">
  <li><a href="#individual-level" id="toc-individual-level" class="nav-link" data-scroll-target="#individual-level">Individual level</a></li>
  <li><a href="#plot-rda-1" id="toc-plot-rda-1" class="nav-link" data-scroll-target="#plot-rda-1">Plot RDA</a></li>
  </ul></li>
  <li><a href="#gwas-genotype-and-phenotype-association" id="toc-gwas-genotype-and-phenotype-association" class="nav-link" data-scroll-target="#gwas-genotype-and-phenotype-association"><strong>GWAS genotype and phenotype association</strong></a></li>
  <li><a href="#randomforest" id="toc-randomforest" class="nav-link" data-scroll-target="#randomforest">RandomForest</a></li>
  <li><a href="#input-file" id="toc-input-file" class="nav-link" data-scroll-target="#input-file">1 Input file</a></li>
  <li><a href="#model-tunning" id="toc-model-tunning" class="nav-link" data-scroll-target="#model-tunning">2 Model tunning</a>
  <ul class="collapse">
  <li><a href="#model-tuning-do-not-run-now" id="toc-model-tuning-do-not-run-now" class="nav-link" data-scroll-target="#model-tuning-do-not-run-now">MODEL TUNING: Do not RUN now</a></li>
  <li><a href="#back-to-the-exercise" id="toc-back-to-the-exercise" class="nav-link" data-scroll-target="#back-to-the-exercise">Back to the exercise</a></li>
  </ul></li>
  <li><a href="#final-model" id="toc-final-model" class="nav-link" data-scroll-target="#final-model">3 Final model</a></li>
  <li><a href="#candidates-snps" id="toc-candidates-snps" class="nav-link" data-scroll-target="#candidates-snps">4 Candidates SNPs</a></li>
  <li><a href="#rainbowr" id="toc-rainbowr" class="nav-link" data-scroll-target="#rainbowr">RAINBOWR</a></li>
  <li><a href="#inputs" id="toc-inputs" class="nav-link" data-scroll-target="#inputs">1 Inputs</a></li>
  <li><a href="#single-snp-base-gwas" id="toc-single-snp-base-gwas" class="nav-link" data-scroll-target="#single-snp-base-gwas">2 Single-SNP base GWAS</a></li>
  <li><a href="#genomic-prediction" id="toc-genomic-prediction" class="nav-link" data-scroll-target="#genomic-prediction">3 Genomic prediction</a>
  <ul class="collapse">
  <li><a href="#plot-genomic-prediction." id="toc-plot-genomic-prediction." class="nav-link" data-scroll-target="#plot-genomic-prediction.">Plot Genomic Prediction.</a></li>
  </ul></li>
  <li><a href="#further-steps" id="toc-further-steps" class="nav-link" data-scroll-target="#further-steps"><em>Further steps</em></a></li>
  <li><a href="#readings" id="toc-readings" class="nav-link" data-scroll-target="#readings"><em>Readings</em></a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Natural Selection</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p><em>Session Presenters</em></p>
<p><img src="images/Presenters7.png" class="img-fluid"></p>
<section id="required-packages" class="level2">
<h2 class="anchored" data-anchor-id="required-packages"><em>Required packages</em></h2>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-1_31309e00894832c880819ce3035102e3">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#devtools::install_github("pygmyperch/melfuR")</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co">#BiocManager::install('qvalue')</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(adegenet)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(LEA)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vegan)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(fmsb)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(psych)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dartRverse)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(melfuR)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sdmpredictors)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sf)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(raster)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(robust)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(qvalue)</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mlbench)</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(caret)</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RAINBOWR)</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(qqman)</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(randomForest)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>make sure you have the packages installed, see</em> <a href="./install.html">Install dartRverse</a></p>
</section>
<section id="gea-analysis-of-snp-and-environmental-data-using-rda" class="level2">
<h2 class="anchored" data-anchor-id="gea-analysis-of-snp-and-environmental-data-using-rda"><strong>GEA analysis of SNP and environmental data using RDA</strong></h2>
<p>First lets load in some functions we will be using.</p>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-2_f0a3303c69191a32dfc8f2b48db9ffd9">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"utils.R"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="data-preparation" class="level2">
<h2 class="anchored" data-anchor-id="data-preparation">1. Data preparation</h2>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-3_ead8a84bd26f5a87661b6cbee42a6e1a">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load genlight object</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"data/Mf5000_gl.RData"</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>Mf5000_gl</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> /// GENLIGHT OBJECT /////////

 // 249 genotypes,  5,000 binary SNPs, size: 1.1 Mb
 12917 (1.04 %) missing data

 // Basic content
   @gen: list of 249 SNPbin
   @ploidy: ploidy of each individual  (range: 2-2)

 // Optional content
   @ind.names:  249 individual labels
   @loc.names:  5000 locus labels
   @pop: population of each individual (group size range: 9-30)
   @other: a list containing: X </code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>Mf5000_gl<span class="sc">@</span>pop</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR
 [19] MBR MBR MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC
 [37] MDC MDC MDC MDC MDC MDC GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL
 [55] GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL
 [73] WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK BEN BEN
 [91] BEN BEN BEN BEN BEN BEN BEN BEN BEN BEN BEN BEN BOG BOG BOG BOG BOG BOG
[109] BOG BOG BOG BOG BOG BOG BOG BOG BOG BOG PEL PEL PEL PEL PEL PEL PEL PEL
[127] PEL GWY GWY GWY GWY GWY GWY GWY GWY GWY GWY GWY GWY DUM DUM DUM DUM DUM
[145] DUM DUM DUM DUM DUM DUM DUM DUM DUM MIB MIB MIB MIB MIB MIB MIB MIB MIB
[163] MIB MIB MIB MIB MIB MIB MIB MIB MIB MIB MIB STG STG STG STG STG STG STG
[181] STG STG STG STG STG STG STG STG STG STG STG STG STG OAK OAK OAK OAK OAK
[199] OAK OAK OAK OAK OAK OAK OAK OAK OAK OAK OAK OAK OAK WAR WAR WAR WAR WAR
[217] WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR KIL KIL KIL
[235] KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL
Levels: BEN BOG DUM GGL GWY KIL MBR MDC MIB OAK PEL STG WAK WAR</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="do">## What do you notice about the factor levels?</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Factors in R can make you question your life choices like nothing else...</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co"># re-order the pop levels to match the order of individuals in your data</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>Mf5000_gl<span class="sc">@</span>pop <span class="ot">&lt;-</span> <span class="fu">factor</span>(Mf5000_gl<span class="sc">@</span>pop, <span class="at">levels =</span> <span class="fu">as.character</span>(<span class="fu">unique</span>(Mf5000_gl<span class="sc">@</span>pop)))</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>Mf5000_gl<span class="sc">@</span>pop</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR MBR
 [19] MBR MBR MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC MDC
 [37] MDC MDC MDC MDC MDC MDC GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL
 [55] GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL GGL
 [73] WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK WAK BEN BEN
 [91] BEN BEN BEN BEN BEN BEN BEN BEN BEN BEN BEN BEN BOG BOG BOG BOG BOG BOG
[109] BOG BOG BOG BOG BOG BOG BOG BOG BOG BOG PEL PEL PEL PEL PEL PEL PEL PEL
[127] PEL GWY GWY GWY GWY GWY GWY GWY GWY GWY GWY GWY GWY DUM DUM DUM DUM DUM
[145] DUM DUM DUM DUM DUM DUM DUM DUM DUM MIB MIB MIB MIB MIB MIB MIB MIB MIB
[163] MIB MIB MIB MIB MIB MIB MIB MIB MIB MIB MIB STG STG STG STG STG STG STG
[181] STG STG STG STG STG STG STG STG STG STG STG STG STG OAK OAK OAK OAK OAK
[199] OAK OAK OAK OAK OAK OAK OAK OAK OAK OAK OAK OAK OAK WAR WAR WAR WAR WAR
[217] WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR WAR KIL KIL KIL
[235] KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL KIL
Levels: MBR MDC GGL WAK BEN BOG PEL GWY DUM MIB STG OAK WAR KIL</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># convert to genind</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>Mf5000.genind <span class="ot">&lt;-</span> <span class="fu">gl2gi</span>(Mf5000_gl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl2gi 
  Processing genlight object with SNP data
Matrix converted.. Prepare genind object...
Completed: gl2gi </code></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>Mf5000.genind</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>/// GENIND OBJECT /////////

 // 249 individuals; 5,000 loci; 10,000 alleles; size: 12.2 Mb

 // Basic content
   @tab:  249 x 10000 matrix of allele counts
   @loc.n.all: number of alleles per locus (range: 2-2)
   @loc.fac: locus factor for the 10000 columns of @tab
   @all.names: list of allele names for each locus
   @ploidy: ploidy of each individual  (range: 2-2)
   @type:  codom
   @call: df2genind(X = xx[, ], sep = "/", ncode = 1, ind.names = x@ind.names, 
    pop = x@pop, NA.char = "-", ploidy = 2)

 // Optional content
   @pop: population of each individual (group size range: 9-30)
   @other: a list containing: X </code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Ok we have our genotypes</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Decision time:
</div>
</div>
<div class="callout-body-container callout-body">
<p>Ordination analyses cannot handle missing data…</p>
<p>How can we impute missing data?</p>
<ol type="1">
<li><p>remove loci with missing data?</p></li>
<li><p>remove individuals with missing data?</p></li>
<li><p>impute missing data, ok how?</p>
<ul>
<li><p>most common genotype across all data?</p></li>
<li><p>most common genotype per site/population/other?</p></li>
<li><p>based on population allele frequencies characterised using snmf (admixture)?<br>
</p></li>
</ul></li>
</ol>
</div>
</div>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-4_1b7db2cd6892f9d8bd97f497e196e6ff">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ADMIXTURE results most likely 6 pops</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>imputed.gi <span class="ot">&lt;-</span> melfuR<span class="sc">::</span><span class="fu">impute.data</span>(Mf5000.genind, <span class="at">K =</span> <span class="dv">6</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-5_fbf3fc5a2c41d0d538971c7e09e1afbe">
<div class="cell-output cell-output-stdout">
<pre><code>Total missing data =  1.04 %
The project is saved into :
 dat.snmfProject 

To load the project, use:
 project = load.snmfProject("dat.snmfProject")

To remove the project, use:
 remove.snmfProject("dat.snmfProject")

[1] 42
[1] "*************************************"
[1] "*          create.dataset            *"
[1] "*************************************"
summary of the options:


[...]

Missing genotype imputation for K = 6 
Missing genotype imputation for run = 3 
Results are written in the file:  ./dat.lfmm_imputed.lfmm 
Total missing data was =  1.04 %
Total missing data now =  0 %</code></pre>
</div>
</div>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-6_a49707654a4eb0b64a71c15499d12a2e">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># re-order each genotype as major/minor allele</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="co"># (generally no real reason to do this... but it's neat and might be useful for something)</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>imputed.sorted.gi <span class="ot">&lt;-</span> <span class="fu">sort_alleles</span>(imputed.gi)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="co"># check results</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>imputed.gi<span class="sc">@</span>tab[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       SNP_1.C SNP_1.G SNP_2.A SNP_2.T SNP_3.T SNP_3.A
MBR_10       1       1       1       1       2       0
MBR_11       1       1       1       1       2       0
MBR_12       1       1       2       0       2       0
MBR_14       0       2       2       0       2       0
MBR_15       0       2       1       1       2       0
MBR_16       0       2       2       0       2       0
MBR_17       0       2       0       2       2       0
MBR_18       0       2       1       1       2       0
MBR_19       1       1       2       0       2       0
MBR_1        0       2       1       1       2       0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>imputed.sorted.gi<span class="sc">@</span>tab[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       SNP_1.G SNP_1.C SNP_2.A SNP_2.T SNP_3.T SNP_3.A
MBR_10       1       1       1       1       2       0
MBR_11       1       1       1       1       2       0
MBR_12       1       1       2       0       2       0
MBR_14       2       0       2       0       2       0
MBR_15       2       0       1       1       2       0
MBR_16       2       0       2       0       2       0
MBR_17       2       0       0       2       2       0
MBR_18       2       0       1       1       2       0
MBR_19       1       1       2       0       2       0
MBR_1        2       0       1       1       2       0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The allele counts for SNP_1 were ordered C/T before sorting, now they are T/C</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="co"># the sort_alleles function also has an optional 'pops' argument where you can specify</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="co"># a population to use as the reference if you want to analyse relative to some focal pop</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="fu">unlink</span>(<span class="fu">c</span>(<span class="st">'dat.geno'</span>, <span class="st">'dat.lfmm'</span>, <span class="st">'dat.snmfProject'</span>, <span class="st">'dat.snmf'</span>, <span class="st">'dat.lfmm_imputed.lfmm'</span>,</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>         <span class="st">'individual_env_data.csv'</span>), <span class="at">recursive =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The sort_alleles function also has an optional ‘pops’ argument where you can specify a population to use as the reference if you want to analyse relative to some focal pop.</p>
<section id="format-snp-data-for-ordination-analysis-in-vegan" class="level3">
<h3 class="anchored" data-anchor-id="format-snp-data-for-ordination-analysis-in-vegan">Format SNP data for ordination analysis in vegan</h3>
<p>So we have our genotypes and have filled in the missing data. We now need to format the data for use in the RDA. We can do the analysis at an individual level using a matrix of allele counts per locus, or at a population level using population allele frequencies. Both of those options are easy to format from a genind object.</p>
<section id="individual-based" class="level4">
<h4 class="anchored" data-anchor-id="individual-based">Individual-based</h4>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-7_0be10eaef6e75fda61dda1ae9992ce66">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># for individual based analyses</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="co"># get allele counts</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>alleles <span class="ot">&lt;-</span> imputed.sorted.gi<span class="sc">@</span>tab</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="co"># get genotypes (counts of reference allele) and clean up locus names</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>snps <span class="ot">&lt;-</span> alleles[,<span class="fu">seq</span>(<span class="dv">1</span>,<span class="fu">ncol</span>(alleles),<span class="dv">2</span>)]</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(snps) <span class="ot">&lt;-</span> <span class="fu">locNames</span>(imputed.sorted.gi)</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>snps[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       SNP_1 SNP_2 SNP_3 SNP_4 SNP_5 SNP_6 SNP_7 SNP_8 SNP_9 SNP_10
MBR_10     1     1     2     2     2     2     1     1     2      1
MBR_11     1     1     2     2     2     2     2     2     2      0
MBR_12     1     2     2     2     2     1     2     2     2      1
MBR_14     2     2     2     2     2     2     1     1     2      2
MBR_15     2     1     2     2     2     2     1     1     2      2
MBR_16     2     2     2     2     2     2     2     2     2      2
MBR_17     2     0     2     2     2     2     0     2     2      2
MBR_18     2     1     2     2     2     1     1     2     2      0
MBR_19     1     2     2     2     2     1     2     2     2      2
MBR_1      2     1     2     2     1     2     2     2     2      2</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Alternative: impute missing data with most common genotype across all data</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="co"># alleles &lt;- Mf5000.genind@tab</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="co"># snps &lt;- alleles[,seq(1,ncol(alleles),2)]</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="co"># colnames(snps) &lt;- locNames(Mf5000.genind)</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="co"># snps &lt;- apply(snps, 2, function(x) replace(x, is.na(x), as.numeric(names(which.max(table(x))))))</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a><span class="co"># check total % missing data</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="co"># (sum(is.na(snps)))/(dim(snps)[1]*dim(snps)[2])*100</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="population-based" class="level4">
<h4 class="anchored" data-anchor-id="population-based">Population-based</h4>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-8_a1ee4ac00821a2df24f18a4b06910d49">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># for population based analyses</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="co"># get pop allele frequencies</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>gp <span class="ot">&lt;-</span> <span class="fu">genind2genpop</span>(imputed.sorted.gi)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 Converting data from a genind to a genpop object... 

...done.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>AF <span class="ot">&lt;-</span> <span class="fu">makefreq</span>(gp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 Finding allelic frequencies from a genpop object... 

...done.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># drop one (redundant) allele per locus</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>AF <span class="ot">&lt;-</span> AF[,<span class="fu">seq</span>(<span class="dv">1</span>,<span class="fu">ncol</span>(AF),<span class="dv">2</span>)]</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(AF) <span class="ot">&lt;-</span> <span class="fu">locNames</span>(gp)</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(AF) <span class="ot">&lt;-</span> <span class="fu">levels</span>(imputed.sorted.gi<span class="sc">@</span>pop)</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>AF[<span class="dv">1</span><span class="sc">:</span><span class="dv">14</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         SNP_1     SNP_2     SNP_3     SNP_4     SNP_5     SNP_6
MBR 0.87500000 0.8250000 0.9750000 1.0000000 0.9750000 0.8250000
MDC 0.70454545 0.7045455 0.8409091 0.9545455 0.8863636 0.8181818
GGL 0.71666667 0.7833333 0.9000000 0.9166667 0.8500000 0.8000000
WAK 0.93750000 1.0000000 1.0000000 1.0000000 1.0000000 0.3125000
BEN 0.92857143 1.0000000 1.0000000 1.0000000 1.0000000 0.6071429
BOG 0.75000000 0.8125000 0.7812500 1.0000000 0.7812500 0.9062500
PEL 0.83333333 0.7222222 0.9444444 0.9444444 0.8888889 1.0000000
GWY 0.75000000 0.9166667 1.0000000 1.0000000 1.0000000 0.8333333
DUM 0.82142857 1.0000000 0.9285714 0.9285714 1.0000000 0.8571429
MIB 0.92500000 0.7750000 0.9250000 0.8250000 0.9750000 0.8500000
STG 0.97500000 0.8750000 0.6500000 0.8000000 1.0000000 0.8000000
OAK 0.02777778 0.6388889 0.6388889 0.6666667 0.3333333 1.0000000
WAR 0.20000000 0.9000000 0.7000000 0.5750000 0.2750000 1.0000000
KIL 0.25000000 0.7777778 0.5277778 0.7222222 0.3055556 1.0000000</code></pre>
</div>
</div>
</section>
</section>
<section id="get-environmental-data" class="level3">
<h3 class="anchored" data-anchor-id="get-environmental-data">Get environmental data</h3>
<p>Ok our SNP data are formatted and ready to go Now we need some environmental data…</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Decision time:
</div>
</div>
<div class="callout-body-container callout-body">
<p>What are the variables you want to use? Considerations: Prior/expert knowledge, existing hypotheses, species distribution models (SDMs), most important variables Data availability? Think about surrogates/proxies if specific data not available, raw variables, PCA, other transformations?</p>
<p>In this case we are going to keep it simple and just use a few WorldClim variables Bio01 (annual mean temperature), Bio05 (temperature in hottest month), Bio15 (rainfall seasonality) and Bio19 (rainfall in coldest quarter)</p>
</div>
</div>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-9_a403e7f7ca19790c78a66245ac3e0275">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set the data directory and extend the response timeout (sometimes the database can be slow to respond)</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">sdmpredictors_datadir=</span><span class="st">"data/spatial_data"</span>)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">timeout =</span> <span class="fu">max</span>(<span class="dv">300</span>, <span class="fu">getOption</span>(<span class="st">"timeout"</span>)))</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Explore environmental layers</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>sdmpredictors<span class="sc">::</span><span class="fu">list_layers</span>(<span class="st">"WorldClim"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-10_e0b3cb2e39b76a31d8a8839ff113a49c">
<div class="cell-output cell-output-stdout">
<pre><code>   dataset_code layer_code                                name
1     WorldClim     WC_alt                            Altitude
2     WorldClim    WC_bio1             Annual mean temperature
3     WorldClim    WC_bio2      Mean diurnal temperature range
4     WorldClim    WC_bio3                       Isothermality
5     WorldClim    WC_bio4             Temperature seasonality
6     WorldClim    WC_bio5                 Maximum temperature
7     WorldClim    WC_bio6                 Minimum temperature
8     WorldClim    WC_bio7            Annual temperature range
9     WorldClim    WC_bio8 Mean temperature of wettest quarter
10    WorldClim    WC_bio9  Mean temperature of driest quarter
11    WorldClim   WC_bio10 Mean temperature of warmest quarter
12    WorldClim   WC_bio11 Mean temperature of coldest quarter
13    WorldClim   WC_bio12                Annual precipitation
14    WorldClim   WC_bio13      Precipitation of wettest month
15    WorldClim   WC_bio14       Precipitation of driest month
16    WorldClim   WC_bio15           Precipitation seasonality
17    WorldClim   WC_bio16    Precipitation of wettest quarter
18    WorldClim   WC_bio17     Precipitation of driest quarter
19    WorldClim   WC_bio18    Precipitation of warmest quarter


[...]</code></pre>
</div>
</div>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-11_ee7a713ffe427fe95e40d5e6b329031b">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Explore environmental layers</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>layers <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(sdmpredictors<span class="sc">::</span><span class="fu">list_layers</span>(<span class="st">"WorldClim"</span>))</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(layers, <span class="st">"./data/WorldClim.csv"</span>)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Download specific layers to the datadir</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Bio01 (annual mean temperature), Bio05 (temperature in hottest month), Bio15 (rainfall seasonality) and Bio19 (rainfall in coldest quarter)</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>WC <span class="ot">&lt;-</span> <span class="fu">load_layers</span>(<span class="fu">c</span>(<span class="st">"WC_bio1"</span>, <span class="st">"WC_bio5"</span>, <span class="st">"WC_bio15"</span>, <span class="st">"WC_bio19"</span>))</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(WC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session07_CB_Yuma_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Crop rasters to study area and align CRS</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="co"># get Murray-Darling Basin polygon</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>mdb <span class="ot">&lt;-</span> <span class="fu">st_read</span>(<span class="st">"data/spatial_data/MDB_polygon/MDB.shp"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Reading layer `MDB' from data source 
  `C:\repositories\dartR_workshop\kioloa\data\spatial_data\MDB_polygon\MDB.shp' 
  using driver `ESRI Shapefile'
Simple feature collection with 1 feature and 18 fields
Geometry type: MULTIPOLYGON
Dimension:     XY
Bounding box:  xmin: 138.57 ymin: -37.68 xmax: 152.49 ymax: -24.585
Geodetic CRS:  GDA94</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set CRS</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>mdb <span class="ot">&lt;-</span> <span class="fu">st_transform</span>(mdb, <span class="dv">4326</span>)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="co"># convert to sf format</span></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>mdb <span class="ot">&lt;-</span> <span class="fu">as</span>(mdb, <span class="st">"Spatial"</span>)</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a><span class="co"># crop WorldClim rasters to MDB extent, then mask pixels outside of the MDB</span></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>ENV <span class="ot">&lt;-</span> <span class="fu">crop</span>(WC, <span class="fu">extent</span>(mdb))</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>ENV <span class="ot">&lt;-</span> <span class="fu">mask</span>(ENV, mdb)</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a><span class="co"># get sampling sites, format as sf</span></span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>sites <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"data/Mf_xy.csv"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a>sites_sf <span class="ot">&lt;-</span> <span class="fu">st_as_sf</span>(sites, <span class="at">coords =</span> <span class="fu">c</span>(<span class="st">"X"</span>, <span class="st">"Y"</span>), <span class="at">crs =</span> <span class="dv">4326</span>)</span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-17"><a href="#cb35-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate a nice color ramp and plot the rasters </span></span>
<span id="cb35-18"><a href="#cb35-18" aria-hidden="true" tabindex="-1"></a>my.colors <span class="ot">=</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"#5E85B8"</span>,<span class="st">"#EDF0C0"</span>,<span class="st">"#C13127"</span>))</span>
<span id="cb35-19"><a href="#cb35-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-20"><a href="#cb35-20" aria-hidden="true" tabindex="-1"></a><span class="co"># include other spatial data that you might want to show</span></span>
<span id="cb35-21"><a href="#cb35-21" aria-hidden="true" tabindex="-1"></a>rivers <span class="ot">&lt;-</span> <span class="fu">st_read</span>(<span class="st">"data/spatial_data/Mf_streams/Mf_streams.shp"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Reading layer `Mf_streams' from data source 
  `C:\repositories\dartR_workshop\kioloa\data\spatial_data\Mf_streams\Mf_streams.shp' 
  using driver `ESRI Shapefile'
Simple feature collection with 2826 features and 31 fields
Geometry type: LINESTRING
Dimension:     XY
Bounding box:  xmin: 138.7888 ymin: -36.52625 xmax: 152.2712 ymax: -26.75375
Geodetic CRS:  GDA94</code></pre>
</div>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>rivers <span class="ot">&lt;-</span> <span class="fu">st_transform</span>(rivers, <span class="dv">4326</span>)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>rivers <span class="ot">&lt;-</span> <span class="fu">as</span>(rivers, <span class="st">"Spatial"</span>)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a><span class="co"># plot a layer</span></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ENV<span class="sc">$</span>WC_bio1, <span class="at">col=</span><span class="fu">my.colors</span>(<span class="dv">100</span>), <span class="at">main=</span><span class="fu">names</span>(ENV<span class="sc">$</span>WC_bio1))</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(rivers, <span class="at">col=</span><span class="st">"blue"</span>, <span class="at">lwd=</span><span class="fl">0.3</span>)</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="fu">st_coordinates</span>(sites_sf)[,<span class="dv">1</span>], <span class="fu">st_coordinates</span>(sites_sf)[,<span class="dv">2</span>], <span class="at">labels=</span>sites_sf<span class="sc">$</span>site, <span class="at">cex=</span><span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session07_CB_Yuma_files/figure-html/unnamed-chunk-11-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<section id="save-pdf" class="level5">
<h5 class="anchored" data-anchor-id="save-pdf">Save PDF</h5>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-12_572e0c523e7c6b9a5d83dba6b4077693">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># format nicely and export as pdf</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>{<span class="fu">pdf</span>(<span class="st">"./images/env_rasters.pdf"</span>)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nlayers</span>(ENV)) {</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>  rasterLayer <span class="ot">&lt;-</span> ENV[[i]]</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># skew the colour ramp to improve visualisation (not necessary, but nice for data like this)</span></span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>  p95 <span class="ot">&lt;-</span> <span class="fu">quantile</span>(<span class="fu">values</span>(rasterLayer), <span class="at">probs =</span> <span class="fl">0.05</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>  breaks <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">minValue</span>(rasterLayer), <span class="fu">seq</span>(p95, <span class="fu">maxValue</span>(rasterLayer), <span class="at">length.out =</span> <span class="dv">50</span>))</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>  breaks <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(<span class="fu">seq</span>(<span class="fu">minValue</span>(rasterLayer), p95, <span class="at">length.out =</span> <span class="dv">10</span>), breaks))</span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a>  color.palette <span class="ot">&lt;-</span> <span class="fu">my.colors</span>(<span class="fu">length</span>(breaks) <span class="sc">-</span> <span class="dv">1</span>)</span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a>  layerColors <span class="ot">&lt;-</span> <span class="cf">if</span>(i <span class="sc">&lt;</span> <span class="dv">4</span>) color.palette <span class="cf">else</span> <span class="fu">rev</span>(color.palette)</span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># tidy up the values displayed in the legend</span></span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a>  simplifiedBreaks <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">min</span>(breaks), <span class="fu">quantile</span>(breaks, <span class="at">probs =</span> <span class="fu">c</span>(<span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="fl">0.75</span>)), <span class="fu">max</span>(breaks))</span>
<span id="cb38-15"><a href="#cb38-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb38-16"><a href="#cb38-16" aria-hidden="true" tabindex="-1"></a>  <span class="co"># plot the rasters</span></span>
<span id="cb38-17"><a href="#cb38-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot</span>(rasterLayer, <span class="at">breaks=</span>breaks, <span class="at">col=</span>layerColors, <span class="at">main=</span><span class="fu">names</span>(rasterLayer),</span>
<span id="cb38-18"><a href="#cb38-18" aria-hidden="true" tabindex="-1"></a>       <span class="at">axis.args=</span><span class="fu">list</span>(<span class="at">at=</span>simplifiedBreaks, <span class="at">labels=</span><span class="fu">as.character</span>(<span class="fu">round</span>(simplifiedBreaks))),</span>
<span id="cb38-19"><a href="#cb38-19" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend.args=</span><span class="fu">list</span>(<span class="at">text=</span><span class="st">''</span>, <span class="at">side=</span><span class="dv">3</span>, <span class="at">line=</span><span class="dv">2</span>, <span class="at">at=</span>simplifiedBreaks))</span>
<span id="cb38-20"><a href="#cb38-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lines</span>(rivers, <span class="at">col=</span><span class="st">"blue"</span>, <span class="at">lwd=</span><span class="fl">0.3</span>, <span class="fu">labels</span>(rivers<span class="sc">$</span>Name))</span>
<span id="cb38-21"><a href="#cb38-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">text</span>(<span class="fu">st_coordinates</span>(sites_sf)[,<span class="dv">1</span>], <span class="fu">st_coordinates</span>(sites_sf)[,<span class="dv">2</span>], <span class="at">labels=</span>sites_sf<span class="sc">$</span>site, <span class="at">cex=</span><span class="fl">0.5</span>)</span>
<span id="cb38-22"><a href="#cb38-22" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb38-23"><a href="#cb38-23" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="extract-site-environmental-data" class="level3">
<h3 class="anchored" data-anchor-id="extract-site-environmental-data">Extract site environmental data</h3>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-13_be3d686afc707c571eafa68af2451a4b">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># and finally extract the environmental data for your sample sites</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>env.site <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">site =</span> sites_sf<span class="sc">$</span>site,</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">X =</span> <span class="fu">st_coordinates</span>(sites_sf)[,<span class="dv">1</span>],</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">Y =</span> <span class="fu">st_coordinates</span>(sites_sf)[,<span class="dv">2</span>],</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>                       <span class="at">omegaX =</span> sites_sf<span class="sc">$</span>omegaX,</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>                       <span class="at">omegaY =</span> sites_sf<span class="sc">$</span>omegaY)</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>env.dat <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(raster<span class="sc">::</span><span class="fu">extract</span>(ENV,<span class="fu">st_coordinates</span>(sites_sf)))</span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>env.site <span class="ot">&lt;-</span> <span class="fu">cbind</span>(env.site, env.dat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Have a look at the env.site data frame This is where we will keep all of the metadata for use in the analyses Keep an eye out, we are going to add a few more columns to it and eventually munge it into individual-level metadata</p>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-14_51d1f2bdf75fccbe0de56259bf8865d8">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># let's also just do a sanity check to make sure the genomic and environmental data are in the same order</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">all</span>(env.site<span class="sc">$</span>site <span class="sc">==</span> <span class="fu">rownames</span>(AF)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="pca" class="level3">
<h3 class="anchored" data-anchor-id="pca">PCA</h3>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-15_3f69415f3fae7cb73d7cf5fb609a3baf">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># to get a feel for the data we can run quick PCA using the rda function in vegan</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>pc <span class="ot">&lt;-</span> <span class="fu">rda</span>(AF)</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pc)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session07_CB_Yuma_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Ok, not really that impressive… let’s set some plotting variables to make it easier to interpret</p>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-16_aea9f1504cd46aa022a8987aa371a2a5">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set colours, markers and labels for plotting and add to env.site df</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>env.site<span class="sc">$</span>cols <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'darkturquoise'</span>, <span class="st">'darkturquoise'</span>, <span class="st">'darkturquoise'</span>, <span class="st">'limegreen'</span>, <span class="st">'limegreen'</span>, <span class="st">'darkorange1'</span>, <span class="st">'slategrey'</span>, <span class="st">'slategrey'</span>, <span class="st">'slategrey'</span>,</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>                   <span class="st">'slategrey'</span>, <span class="st">'slategrey'</span>, <span class="st">'dodgerblue3'</span>, <span class="st">'gold'</span>, <span class="st">'gold'</span>)</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>env.site<span class="sc">$</span>pch <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">15</span>, <span class="dv">16</span>, <span class="dv">17</span>, <span class="dv">17</span>, <span class="dv">15</span>, <span class="dv">15</span>, <span class="dv">16</span>, <span class="dv">15</span>, <span class="dv">18</span>, <span class="dv">16</span>, <span class="dv">16</span>, <span class="dv">8</span>, <span class="dv">16</span>, <span class="dv">16</span>)</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a><span class="co"># generate labels for each axis</span></span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a>x.lab <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"PC1 ("</span>, <span class="fu">paste</span>(<span class="fu">round</span>((pc<span class="sc">$</span>CA<span class="sc">$</span>eig[<span class="dv">1</span>]<span class="sc">/</span>pc<span class="sc">$</span>tot.chi<span class="sc">*</span><span class="dv">100</span>),<span class="dv">2</span>)),<span class="st">"%)"</span>)</span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a>y.lab <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"PC2 ("</span>, <span class="fu">paste</span>(<span class="fu">round</span>((pc<span class="sc">$</span>CA<span class="sc">$</span>eig[<span class="dv">2</span>]<span class="sc">/</span>pc<span class="sc">$</span>tot.chi<span class="sc">*</span><span class="dv">100</span>),<span class="dv">2</span>)),<span class="st">"%)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plot-pca" class="level3">
<h3 class="anchored" data-anchor-id="plot-pca">Plot PCA</h3>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-17_ec5e4930eb113d7b5940dccead88098a">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot PCA</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>{<span class="fu">pdf</span>(<span class="at">file =</span> <span class="st">"./images/Mf_PCA.pdf"</span>, <span class="at">height =</span> <span class="dv">6</span>, <span class="at">width =</span> <span class="dv">6</span>)</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>pcaplot <span class="ot">&lt;-</span> <span class="fu">plot</span>(pc, <span class="at">choices =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">type =</span> <span class="st">"n"</span>, <span class="at">xlab=</span>x.lab, <span class="at">ylab=</span>y.lab, <span class="at">cex.lab=</span><span class="dv">1</span>)</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a><span class="fu">with</span>(env.site, <span class="fu">points</span>(pc, <span class="at">display =</span> <span class="st">"sites"</span>, <span class="at">col =</span> env.site<span class="sc">$</span>cols, <span class="at">pch =</span> env.site<span class="sc">$</span>pch, <span class="at">cex=</span><span class="fl">1.5</span>, <span class="at">bg =</span> env.site<span class="sc">$</span>cols))</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topleft"</span>, <span class="at">legend =</span> env.site<span class="sc">$</span>site, <span class="at">col=</span>env.site<span class="sc">$</span>cols, <span class="at">pch=</span>env.site<span class="sc">$</span>pch, <span class="at">pt.cex=</span><span class="dv">1</span>, <span class="at">cex=</span><span class="fl">0.50</span>, <span class="at">xpd=</span><span class="dv">1</span>, <span class="at">box.lty =</span> <span class="dv">0</span>, <span class="at">bg=</span> <span class="st">"transparent"</span>)</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="Session 07 Natural Selection/Mf_PCA.png" class="img-fluid"> ## 2. Variable filtering</p>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-19_c92be64dacb4dc9bdb0c455ba448e882">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># extract predictor variables to matrix</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>env_var <span class="ot">&lt;-</span> env.site[ , <span class="fu">c</span>(<span class="st">"WC_bio1"</span>, <span class="st">"WC_bio5"</span>, <span class="st">"WC_bio15"</span>, <span class="st">"WC_bio19"</span>)]</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a><span class="co"># check for obvious correlations between variables</span></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a><span class="fu">pairs.panels</span>(env_var, <span class="at">scale=</span>T, <span class="at">lm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session07_CB_Yuma_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="do">## reduce variance associated with covarying variables using variance inflation factor analyses</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="co"># run procedure to remove variables with the highest VIF one at a time until all remaining variables are below your threshold</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>keep.env <span class="ot">&lt;-</span><span class="fu">vif_func</span>(<span class="at">in_frame=</span>env_var,<span class="at">thresh=</span><span class="dv">2</span>,<span class="at">trace=</span>T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> var      vif             
 WC_bio1  21.5371434397055
 WC_bio5  10.4639698247348
 WC_bio15 7.69781293748316
 WC_bio19 3.97039440920372

removed:  WC_bio1 21.53714 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>keep.env  <span class="co"># the retained environmental variables</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "WC_bio5"  "WC_bio15" "WC_bio19"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>reduced.env <span class="ot">&lt;-</span> <span class="fu">subset</span>(<span class="fu">as.data.frame</span>(env_var), <span class="at">select=</span><span class="fu">c</span>(keep.env))</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>scaled.env <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">scale</span>(reduced.env))</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a><span class="co"># lets have another look</span></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a><span class="fu">pairs.panels</span>(reduced.env, <span class="at">scale=</span>T, <span class="at">lm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session07_CB_Yuma_files/figure-html/unnamed-chunk-19-2.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="run-analysis" class="level2">
<h2 class="anchored" data-anchor-id="run-analysis">3. Run analysis</h2>
<p>So we have formatted genotypes, environmental data</p>
<p>What about controlling for spatial population structure in the model?</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Decision time:
</div>
</div>
<div class="callout-body-container callout-body">
<p>Do I need to control for spatial structure?</p>
<pre><code> No? Great!

 Yes? How do I do that?
 </code></pre>
<ul>
<li>simple xy coordinates (or some transformation, MDS, polynomial…)</li>
<li>Principal Coordinates of Neighbourhood Matrix (PCNM; R package vegan)</li>
<li>Moran’s Eigenvector Maps (R package memgene)</li>
<li>Fst, admixture Q values</li>
<li>population allele frequency covariance matrix (BayPass omega)</li>
</ul>
</div>
</div>
<p>We are going to use the scaled allelic covariance (Ω) estimated by BayPass to control for spatial structure in the data, see BayPass and Gates et al.&nbsp;2023</p>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-20_c446f536ec291c0ea9e3633b289e331c">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># format spatial coordinates as matrix</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>xy <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">cbind</span>(env.site<span class="sc">$</span>omegaX, env.site<span class="sc">$</span>omegaY))</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a><span class="co"># reduced RDA model using the retained environmental PCs </span></span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a><span class="co"># conditioned on the retained spatial variables</span></span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a>Mf.RDA <span class="ot">&lt;-</span> <span class="fu">rda</span>(AF <span class="sc">~</span> WC_bio5 <span class="sc">+</span> WC_bio15 <span class="sc">+</span> WC_bio19 <span class="sc">+</span> <span class="fu">Condition</span>(xy), <span class="at">data =</span> scaled.env)</span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a>Mf.RDA</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Call: rda(formula = AF ~ WC_bio5 + WC_bio15 + WC_bio19 + Condition(xy),
data = scaled.env)

                Inertia Proportion Rank
Total         202.83442    1.00000     
Conditional     9.32930    0.04599    2
Constrained   102.05924    0.50317    3
Unconstrained  91.44588    0.45084    8
Inertia is variance 

Eigenvalues for constrained axes:
 RDA1  RDA2  RDA3 
85.00 12.97  4.09 

Eigenvalues for unconstrained axes:
  PC1   PC2   PC3   PC4   PC5   PC6   PC7   PC8 
66.70  6.36  5.17  3.75  3.10  2.86  2.41  1.10 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co"># So how much genetic variation can be explained by our environmental model?</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a><span class="fu">RsquareAdj</span>(Mf.RDA)<span class="sc">$</span>r.squared</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.5031653</code></pre>
</div>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># how much inertia is associated with each axis</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a><span class="fu">screeplot</span>(Mf.RDA)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session07_CB_Yuma_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate significance of the reduced model</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>mod_perm <span class="ot">&lt;-</span> <span class="fu">anova.cca</span>(Mf.RDA, <span class="at">nperm=</span><span class="dv">1000</span>) <span class="co">#test significance of the model</span></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>mod_perm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Permutation test for rda under reduced model
Permutation: free
Number of permutations: 999

Model: rda(formula = AF ~ WC_bio5 + WC_bio15 + WC_bio19 + Condition(xy), data = scaled.env)
         Df Variance      F Pr(&gt;F)  
Model     3  102.059 2.9762   0.04 *
Residual  8   91.446                
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co">#generate x and y labels (% constrained variation)</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>x.lab <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"RDA1 ("</span>, <span class="fu">paste</span>(<span class="fu">round</span>((Mf.RDA<span class="sc">$</span>CCA<span class="sc">$</span>eig[<span class="dv">1</span>]<span class="sc">/</span>Mf.RDA<span class="sc">$</span>CCA<span class="sc">$</span>tot.chi<span class="sc">*</span><span class="dv">100</span>),<span class="dv">2</span>)),<span class="st">"%)"</span>)</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>y.lab <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"RDA2 ("</span>, <span class="fu">paste</span>(<span class="fu">round</span>((Mf.RDA<span class="sc">$</span>CCA<span class="sc">$</span>eig[<span class="dv">2</span>]<span class="sc">/</span>Mf.RDA<span class="sc">$</span>CCA<span class="sc">$</span>tot.chi<span class="sc">*</span><span class="dv">100</span>),<span class="dv">2</span>)),<span class="st">"%)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="plot-rda" class="level3">
<h3 class="anchored" data-anchor-id="plot-rda">Plot RDA</h3>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-21_e25d64b9e2b0f5a85b323c28412183b6">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co">#plot RDA1, RDA2</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>{<span class="fu">pdf</span>(<span class="at">file =</span> <span class="st">"./images/Mf_RDA.pdf"</span>, <span class="at">height =</span> <span class="dv">6</span>, <span class="at">width =</span> <span class="dv">6</span>)</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>pRDAplot <span class="ot">&lt;-</span> <span class="fu">plot</span>(Mf.RDA, <span class="at">choices =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">type=</span><span class="st">"n"</span>, <span class="at">cex.lab=</span><span class="dv">1</span>, <span class="at">xlab=</span>x.lab, <span class="at">ylab=</span>y.lab)</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a><span class="fu">with</span>(env.site, <span class="fu">points</span>(Mf.RDA, <span class="at">display =</span> <span class="st">"sites"</span>, <span class="at">col =</span> env.site<span class="sc">$</span>cols, <span class="at">pch =</span> env.site<span class="sc">$</span>pch, <span class="at">cex=</span><span class="fl">1.5</span>, <span class="at">bg =</span> env.site<span class="sc">$</span>cols))</span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(Mf.RDA, <span class="st">"bp"</span>,<span class="at">choices =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"WC_bio5"</span>, <span class="st">"WC_bio15"</span>, <span class="st">"WC_bio19"</span>), <span class="at">col=</span><span class="st">"blue"</span>, <span class="at">cex=</span><span class="fl">0.6</span>)</span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topleft"</span>, <span class="at">legend =</span> env.site<span class="sc">$</span>site, <span class="at">col=</span>env.site<span class="sc">$</span>cols, <span class="at">pch=</span>env.site<span class="sc">$</span>pch, <span class="at">pt.cex=</span><span class="dv">1</span>, <span class="at">cex=</span><span class="fl">0.50</span>, <span class="at">xpd=</span><span class="dv">1</span>, <span class="at">box.lty =</span> <span class="dv">0</span>, <span class="at">bg=</span> <span class="st">"transparent"</span>)</span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="images/Mf_RDA.png" class="img-fluid"></p>
</section>
</section>
<section id="identify-candidate-loci" class="level2">
<h2 class="anchored" data-anchor-id="identify-candidate-loci">4. Identify candidate loci</h2>
<p>Estimate q-values following:&nbsp;<a href="https://kjschmidlab.gitlab.io/b1k-gbs/GenomeEnvAssociation.html#RDA_approaches" class="uri">https://kjschmidlab.gitlab.io/b1k-gbs/GenomeEnvAssociation.html#RDA_approaches</a> and the R function rdadapt (see <a href="https://github.com/Capblancq/RDA-genome-scan" class="uri">https://github.com/Capblancq/RDA-genome-scan</a>)</p>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-23_b4935a8488696651587be76f1c152b03">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>k <span class="ot">=</span> <span class="dv">2</span> <span class="co"># number of RDA axes to retain</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>rda.q <span class="ot">&lt;-</span> <span class="fu">rdadapt</span>(Mf.RDA, <span class="at">K =</span> k)</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(rda.q<span class="sc">$</span>q.values <span class="sc">&lt;=</span> <span class="fl">0.01</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 373</code></pre>
</div>
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>candidate.loci <span class="ot">&lt;-</span> <span class="fu">colnames</span>(AF)[<span class="fu">which</span>(rda.q[,<span class="dv">2</span>] <span class="sc">&lt;</span> <span class="fl">0.01</span>)]</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a><span class="co"># check the distribution of candidates and q-values</span></span>
<span id="cb62-5"><a href="#cb62-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb62-6"><a href="#cb62-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x=</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(rda.q[,<span class="dv">2</span>])), </span>
<span id="cb62-7"><a href="#cb62-7" aria-hidden="true" tabindex="-1"></a>                 <span class="at">y=</span><span class="sc">-</span><span class="fu">log10</span>(rda.q[,<span class="dv">2</span>])), <span class="at">col=</span><span class="st">"gray83"</span>) <span class="sc">+</span></span>
<span id="cb62-8"><a href="#cb62-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x=</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(rda.q[,<span class="dv">2</span>]))[<span class="fu">which</span>(rda.q[,<span class="dv">2</span>] <span class="sc">&lt;</span> <span class="fl">0.01</span>)], </span>
<span id="cb62-9"><a href="#cb62-9" aria-hidden="true" tabindex="-1"></a>                 <span class="at">y=</span><span class="sc">-</span><span class="fu">log10</span>(rda.q[<span class="fu">which</span>(rda.q[,<span class="dv">2</span>] <span class="sc">&lt;</span> <span class="fl">0.01</span>),<span class="dv">2</span>])), <span class="at">col=</span><span class="st">"orange"</span>) <span class="sc">+</span></span>
<span id="cb62-10"><a href="#cb62-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"SNPs"</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">"-log10(q.values"</span>) <span class="sc">+</span></span>
<span id="cb62-11"><a href="#cb62-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session07_CB_Yuma_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="co"># check correlations between candidate loci and environmental predictors</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a><span class="co"># this section relies on code by Brenna Forester, found at</span></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a><span class="co"># https://popgen.nescent.org/2018-03-27_RDA_GEA.html</span></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>npred <span class="ot">&lt;-</span> <span class="dv">3</span></span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a>env_mat <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">nrow=</span>(<span class="fu">sum</span>(rda.q<span class="sc">$</span>q.values <span class="sc">&lt;</span> <span class="fl">0.01</span>)), </span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a>                  <span class="at">ncol=</span>npred)  <span class="co"># n columns for n predictors</span></span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(env_mat) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(reduced.env)</span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-10"><a href="#cb63-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-11"><a href="#cb63-11" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate correlation between candidate snps and environmental variables</span></span>
<span id="cb63-12"><a href="#cb63-12" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(candidate.loci)) {</span>
<span id="cb63-13"><a href="#cb63-13" aria-hidden="true" tabindex="-1"></a>  nam <span class="ot">&lt;-</span> candidate.loci[i]</span>
<span id="cb63-14"><a href="#cb63-14" aria-hidden="true" tabindex="-1"></a>  snp.gen <span class="ot">&lt;-</span> AF[,nam]</span>
<span id="cb63-15"><a href="#cb63-15" aria-hidden="true" tabindex="-1"></a>  env_mat[i,] <span class="ot">&lt;-</span> <span class="fu">apply</span>(reduced.env,<span class="dv">2</span>,<span class="cf">function</span>(x) <span class="fu">cor</span>(x,snp.gen))</span>
<span id="cb63-16"><a href="#cb63-16" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb63-17"><a href="#cb63-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-18"><a href="#cb63-18" aria-hidden="true" tabindex="-1"></a>cand <span class="ot">&lt;-</span> <span class="fu">cbind.data.frame</span>(<span class="fu">as.data.frame</span>(candidate.loci),env_mat)  </span>
<span id="cb63-19"><a href="#cb63-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-20"><a href="#cb63-20" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(cand<span class="sc">$</span>candidate.loci)) {</span>
<span id="cb63-21"><a href="#cb63-21" aria-hidden="true" tabindex="-1"></a>  bar <span class="ot">&lt;-</span> cand[i,]</span>
<span id="cb63-22"><a href="#cb63-22" aria-hidden="true" tabindex="-1"></a>  cand[i,npred<span class="sc">+</span><span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="fu">names</span>(<span class="fu">which.max</span>(<span class="fu">abs</span>(bar[,<span class="dv">2</span><span class="sc">:</span><span class="dv">4</span>]))) <span class="co"># gives the variable</span></span>
<span id="cb63-23"><a href="#cb63-23" aria-hidden="true" tabindex="-1"></a>  cand[i,npred<span class="sc">+</span><span class="dv">3</span>] <span class="ot">&lt;-</span> <span class="fu">max</span>(<span class="fu">abs</span>(bar[,<span class="dv">2</span><span class="sc">:</span><span class="dv">4</span>]))              <span class="co"># gives the correlation</span></span>
<span id="cb63-24"><a href="#cb63-24" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb63-25"><a href="#cb63-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-26"><a href="#cb63-26" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(cand)[npred<span class="sc">+</span><span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="st">"predictor"</span></span>
<span id="cb63-27"><a href="#cb63-27" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(cand)[npred<span class="sc">+</span><span class="dv">3</span>] <span class="ot">&lt;-</span> <span class="st">"correlation"</span></span>
<span id="cb63-28"><a href="#cb63-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-29"><a href="#cb63-29" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(cand<span class="sc">$</span>predictor) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
WC_bio15 WC_bio19  WC_bio5 
     216       47      110 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(cand, <span class="st">"./data/RDA_candidates.csv"</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="individual-level" class="level3">
<h3 class="anchored" data-anchor-id="individual-level">Individual level</h3>
<p>Now let’s have a look at an individual level biplot first expand environmental data from population to individual dataframe.</p>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-24_9031607b67d9c1a0057bdfe0b1241749">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>env.ind <span class="ot">&lt;-</span> <span class="fu">expand_pop2ind</span>(Mf5000_gl, env.site)</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a><span class="co"># now subset our genotype objects to the candidate loci</span></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>candidate.gl <span class="ot">&lt;-</span> <span class="fu">gl.keep.loc</span>(<span class="fu">gi2gl</span>(imputed.sorted.gi), <span class="at">loc.list=</span>candidate.loci)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.keep.loc 
Starting gi2gl 
Starting gl.compliance.check 
  Processing genlight object with SNP data
  The slot loc.all, which stores allele name for each locus, is empty. Creating a dummy variable (A/C) to insert in this slot. 
  Checking coding of SNPs
    SNP data scored NA, 0, 1 or 2 confirmed
  Checking for population assignments
    Population assignments confirmed
  Checking locus metrics and flags
  Recalculating locus metrics
  Checking for monomorphic loci
    No monomorphic loci detected
  Checking for loci with all missing data
    No loci with all missing data detected
  Checking whether individual names are unique.
  Checking for individual metrics
  Warning: Creating a slot for individual metrics
  Spelling of coordinates checked and changed if necessary to 
            lat/lon
Completed: gl.compliance.check 
Completed: gi2gl 
  Processing genlight object with SNP data
  List of loci to keep has been specified
  Deleting all but the specified loci
Completed: gl.keep.loc </code></pre>
</div>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>candidate.gi <span class="ot">&lt;-</span> <span class="fu">gl2gi</span>(candidate.gl)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl2gi 
  Processing genlight object with SNP data
Matrix converted.. Prepare genind object...
Completed: gl2gi </code></pre>
</div>
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get genotypes (counts of reference allele) and clean up locus names</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>candidate.alleles <span class="ot">&lt;-</span> candidate.gi<span class="sc">@</span>tab</span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>candidate.snps <span class="ot">&lt;-</span> candidate.alleles[,<span class="fu">seq</span>(<span class="dv">1</span>,<span class="fu">ncol</span>(candidate.alleles),<span class="dv">2</span>)]</span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(candidate.snps) <span class="ot">&lt;-</span> <span class="fu">locNames</span>(candidate.gi)</span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a><span class="co"># another quick sanity check</span></span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">all</span>(env.ind<span class="sc">$</span>ind.names <span class="sc">==</span> <span class="fu">indNames</span>(candidate.gi)))</span>
<span id="cb70-8"><a href="#cb70-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-9"><a href="#cb70-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-10"><a href="#cb70-10" aria-hidden="true" tabindex="-1"></a><span class="co"># finally run the individual based RDA</span></span>
<span id="cb70-11"><a href="#cb70-11" aria-hidden="true" tabindex="-1"></a><span class="co"># no need to control for pop structure this time</span></span>
<span id="cb70-12"><a href="#cb70-12" aria-hidden="true" tabindex="-1"></a>Mfcandidate.RDA <span class="ot">&lt;-</span> <span class="fu">rda</span>(candidate.snps <span class="sc">~</span> WC_bio5 <span class="sc">+</span> WC_bio15 <span class="sc">+</span> WC_bio19, <span class="at">data =</span> env.ind)</span>
<span id="cb70-13"><a href="#cb70-13" aria-hidden="true" tabindex="-1"></a>Mfcandidate.RDA</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Call: rda(formula = candidate.snps ~ WC_bio5 + WC_bio15 + WC_bio19,
data = env.ind)

               Inertia Proportion Rank
Total         207.1000     1.0000     
Constrained    59.6948     0.2882    3
Unconstrained 147.4052     0.7118  245
Inertia is variance 

Eigenvalues for constrained axes:
 RDA1  RDA2  RDA3 
44.71 12.02  2.97 

Eigenvalues for unconstrained axes:
   PC1    PC2    PC3    PC4    PC5    PC6    PC7    PC8 
16.227  7.584  5.579  4.597  3.536  3.373  3.075  2.695 
(Showing 8 of 245 unconstrained eigenvalues)</code></pre>
</div>
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>ind.mod_perm <span class="ot">&lt;-</span> <span class="fu">anova.cca</span>(Mfcandidate.RDA, <span class="at">nperm=</span><span class="dv">1000</span>, </span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">parallel=</span><span class="dv">4</span>) <span class="co">#test significance of the model</span></span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>ind.mod_perm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Permutation test for rda under reduced model
Permutation: free
Number of permutations: 999

Model: rda(formula = candidate.snps ~ WC_bio5 + WC_bio15 + WC_bio19, data = env.ind)
          Df Variance      F Pr(&gt;F)    
Model      3   59.695 33.073  0.001 ***
Residual 245  147.405                  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>x.lab <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"RDA1 ("</span>,</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>                <span class="fu">paste</span>(<span class="fu">round</span>((Mfcandidate.RDA<span class="sc">$</span>CA<span class="sc">$</span>eig[<span class="dv">1</span>]<span class="sc">/</span>Mfcandidate.RDA<span class="sc">$</span>tot.chi<span class="sc">*</span><span class="dv">100</span>),<span class="dv">2</span>)),<span class="st">"%)"</span>)</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>y.lab <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"RDA2 ("</span>,</span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>                <span class="fu">paste</span>(<span class="fu">round</span>((Mfcandidate.RDA<span class="sc">$</span>CA<span class="sc">$</span>eig[<span class="dv">2</span>]<span class="sc">/</span>Mfcandidate.RDA<span class="sc">$</span>tot.chi<span class="sc">*</span><span class="dv">100</span>),<span class="dv">2</span>)),<span class="st">"%)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="plot-rda-1" class="level3">
<h3 class="anchored" data-anchor-id="plot-rda-1">Plot RDA</h3>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-25_3a31da729181dee9cf5f9b62d95c6e59">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="co">#plot RDA1, RDA2</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>{<span class="fu">pdf</span>(<span class="at">file =</span> <span class="st">"./images/Mfcandidate_RDA.pdf"</span>, <span class="at">height =</span> <span class="dv">6</span>, <span class="at">width =</span> <span class="dv">6</span>)</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>pRDAplot <span class="ot">&lt;-</span> <span class="fu">plot</span>(Mfcandidate.RDA, <span class="at">choices =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), </span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">type=</span><span class="st">"n"</span>, <span class="at">cex.lab=</span><span class="dv">1</span>, <span class="at">xlab=</span>x.lab, <span class="at">ylab=</span>y.lab)</span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(Mfcandidate.RDA, <span class="at">display =</span> <span class="st">"sites"</span>, <span class="at">col =</span> env.ind<span class="sc">$</span>cols,</span>
<span id="cb75-6"><a href="#cb75-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">pch =</span> env.ind<span class="sc">$</span>pch, <span class="at">cex=</span><span class="fl">0.8</span>, <span class="at">bg =</span> env.ind<span class="sc">$</span>cols)</span>
<span id="cb75-7"><a href="#cb75-7" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(Mfcandidate.RDA, <span class="st">"bp"</span>,<span class="at">choices =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), </span>
<span id="cb75-8"><a href="#cb75-8" aria-hidden="true" tabindex="-1"></a>     <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"WC_bio5"</span>, <span class="st">"WC_bio15"</span>, <span class="st">"WC_bio19"</span>), <span class="at">col=</span><span class="st">"blue"</span>, <span class="at">cex=</span><span class="fl">0.6</span>)</span>
<span id="cb75-9"><a href="#cb75-9" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topleft"</span>, <span class="at">legend =</span> env.site<span class="sc">$</span>site, <span class="at">col=</span>env.site<span class="sc">$</span>cols, <span class="at">pch=</span>env.site<span class="sc">$</span>pch,</span>
<span id="cb75-10"><a href="#cb75-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">pt.cex=</span><span class="dv">1</span>, <span class="at">cex=</span><span class="fl">0.50</span>, <span class="at">xpd=</span><span class="dv">1</span>, <span class="at">box.lty =</span> <span class="dv">0</span>, <span class="at">bg=</span> <span class="st">"transparent"</span>)</span>
<span id="cb75-11"><a href="#cb75-11" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb75-12"><a href="#cb75-12" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="images/Mfcandidate_RDA.png" class="img-fluid"></p>
</section>
</section>
<section id="gwas-genotype-and-phenotype-association" class="level2">
<h2 class="anchored" data-anchor-id="gwas-genotype-and-phenotype-association"><strong>GWAS genotype and phenotype association</strong></h2>
</section>
<section id="randomforest" class="level2">
<h2 class="anchored" data-anchor-id="randomforest">RandomForest</h2>
</section>
<section id="input-file" class="level2">
<h2 class="anchored" data-anchor-id="input-file">1 Input file</h2>
<p>Let’s start with randomForest.</p>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-27_3744197d33e47cff6869df8b930db5c6">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Get input files </span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="at">file =</span> <span class="st">"./data/GWAS.RData"</span>)</span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a><span class="do">##randomForest</span></span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a>CaRF<span class="sc">$</span>Size <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(CaRF<span class="sc">$</span>Size)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The input is a structure file with genotypes codified as counts of the major allele per individual per locus (0 hom minor allele (mA), 1 het, 2 hom major allele (MA)), with sample’s phenotype or trait in the first column instead of the population.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>If the trait or phenotype is categorical instead of continuous, it should be transformed into a factor.</p>
</div>
</div>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-28_e1ebf1025a19590113e5fdc3c0ec17bf">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">See</span>(CaRF)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              Size LG1.22769 LG1.197846 LG1.341740 LG1.428005 LG1.606435
class     &lt;factor&gt; &lt;integer&gt;  &lt;integer&gt;  &lt;integer&gt;  &lt;integer&gt;  &lt;integer&gt;
popL_L026        2        -1         -1         -1          0         -1
popL_L004        2        -1         -1         -1         -1          0
popL_L024        2        -1         -1         -1         -1         -1
popL_L010        2        -1         -1         -1         -1         -1
popL_L025        2        -1         -1         -1         -1          0
popL_L016        2        -1         -1         -1         -1         -1
[1] "class: data.frame"
[1] "dimension: 363 x 2063"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(CaRF<span class="sc">$</span>Size)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
  1   2 
180 183 </code></pre>
</div>
</div>
<p>So here we have the 363 samples, (180 small and 183 large), genotyped for a subset of 2,062 markers from the 17,490 original dataset.</p>
</section>
<section id="model-tunning" class="level2">
<h2 class="anchored" data-anchor-id="model-tunning">2 Model tunning</h2>
<p>The first step is to crate our training and validation data in a 70:25 proportion.</p>
<p>Then we run a simple forest to see performance without tuning.</p>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-29_f3a785d69e173162a640236d299482ac">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>ind <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">2</span>, <span class="fu">nrow</span>(CaRF), <span class="at">replace =</span> <span class="cn">TRUE</span>, <span class="at">prob =</span> <span class="fu">c</span>(<span class="fl">0.75</span>, <span class="fl">0.25</span>))</span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>train <span class="ot">&lt;-</span> CaRF[ind<span class="sc">==</span><span class="dv">1</span>,]</span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> CaRF[ind<span class="sc">==</span><span class="dv">2</span>,]</span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-6"><a href="#cb81-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-7"><a href="#cb81-7" aria-hidden="true" tabindex="-1"></a>rf <span class="ot">&lt;-</span> <span class="fu">randomForest</span>(Size<span class="sc">~</span>., <span class="at">data=</span>train,</span>
<span id="cb81-8"><a href="#cb81-8" aria-hidden="true" tabindex="-1"></a>                   <span class="at">ntree =</span> <span class="dv">50</span>,       <span class="co"># number of trees</span></span>
<span id="cb81-9"><a href="#cb81-9" aria-hidden="true" tabindex="-1"></a>                   <span class="at">mtry =</span> <span class="dv">10</span>,          <span class="co"># number of variables tried at each split</span></span>
<span id="cb81-10"><a href="#cb81-10" aria-hidden="true" tabindex="-1"></a>                   <span class="at">importance =</span> <span class="cn">TRUE</span>,</span>
<span id="cb81-11"><a href="#cb81-11" aria-hidden="true" tabindex="-1"></a>                   <span class="at">proximity =</span> <span class="cn">TRUE</span>)</span>
<span id="cb81-12"><a href="#cb81-12" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(rf)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
 randomForest(formula = Size ~ ., data = train, ntree = 50, mtry = 10,      importance = TRUE, proximity = TRUE) 
               Type of random forest: classification
                     Number of trees: 50
No. of variables tried at each split: 10

        OOB estimate of  error rate: 27.44%
Confusion matrix:
   1   2 class.error
1 98  39   0.2846715
2 37 103   0.2642857</code></pre>
</div>
</div>
<p>But if you run this again you get slightly different results, so we need to do some replications using different training data.</p>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-30_118c3fa5589d67f1588bdef56c177c84">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>control <span class="ot">&lt;-</span> <span class="fu">trainControl</span>(<span class="at">method=</span><span class="st">"oob"</span>, <span class="at">number=</span><span class="dv">5</span>)</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>) </span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a>metric <span class="ot">&lt;-</span> <span class="st">"Accuracy"</span></span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a>mtry <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb83-5"><a href="#cb83-5" aria-hidden="true" tabindex="-1"></a>tunegrid <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="at">.mtry=</span>mtry )</span>
<span id="cb83-6"><a href="#cb83-6" aria-hidden="true" tabindex="-1"></a>rf_small <span class="ot">&lt;-</span> <span class="fu">train</span>(Size<span class="sc">~</span>., <span class="at">data=</span>CaRF, <span class="at">method=</span><span class="st">"rf"</span>, <span class="at">metric=</span>metric, <span class="at">tuneGrid=</span>tunegrid, <span class="at">trControl=</span>control, <span class="at">ntree=</span><span class="dv">50</span>)</span>
<span id="cb83-7"><a href="#cb83-7" aria-hidden="true" tabindex="-1"></a>rf_small</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Random Forest 

 363 samples
2062 predictors
   2 classes: '1', '2' 

No pre-processing
Resampling results:

  Accuracy   Kappa    
  0.7410468  0.4818695

Tuning parameter 'mtry' was held constant at a value of 10</code></pre>
</div>
</div>
<p>Here we have also Kappa, a measure of how much better the model classification is than a random classification. This is important when there is a significant skew in sample sizes.</p>
<p>Now let’s check what the default parameter can do.</p>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-31_86295c77f7b385748ed1249edd13f6b9">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>control <span class="ot">&lt;-</span> <span class="fu">trainControl</span>(<span class="at">method=</span><span class="st">"oob"</span>, <span class="at">number=</span><span class="dv">5</span>)</span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>) </span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a>metric <span class="ot">&lt;-</span> <span class="st">"Accuracy"</span></span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a>mtry <span class="ot">&lt;-</span> <span class="fu">sqrt</span>(<span class="fu">ncol</span>(CaRF))</span>
<span id="cb85-5"><a href="#cb85-5" aria-hidden="true" tabindex="-1"></a>tunegrid <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="at">.mtry=</span>mtry )</span>
<span id="cb85-6"><a href="#cb85-6" aria-hidden="true" tabindex="-1"></a>rf_default <span class="ot">&lt;-</span> <span class="fu">train</span>(Size<span class="sc">~</span>., <span class="at">data=</span>CaRF, <span class="at">method=</span><span class="st">"rf"</span>, <span class="at">metric=</span>metric, <span class="at">tuneGrid=</span>tunegrid, <span class="at">trControl=</span>control)</span>
<span id="cb85-7"><a href="#cb85-7" aria-hidden="true" tabindex="-1"></a>rf_default</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Random Forest 

 363 samples
2062 predictors
   2 classes: '1', '2' 

No pre-processing
Resampling results:

  Accuracy   Kappa    
  0.7465565  0.4928012

Tuning parameter 'mtry' was held constant at a value of 45.42026</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p>70% accuracy is no too bad, but can we improve it?</p>
</blockquote>
<section id="model-tuning-do-not-run-now" class="level3">
<h3 class="anchored" data-anchor-id="model-tuning-do-not-run-now">MODEL TUNING: Do not RUN now</h3>
<p>The next couple of steps could take <strong>too long</strong>, so here are the commands for you to try in your own time.</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
run in <strong>your own time</strong>
</div>
</div>
<div class="callout-body-container callout-body">
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-32_2ef5e2fbb132c99e94f314298ed7a6a1">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>control <span class="ot">&lt;-</span> <span class="fu">trainControl</span>(<span class="at">method=</span><span class="st">"oob"</span>, <span class="at">number=</span><span class="dv">10</span>, <span class="at">search=</span><span class="st">"random"</span>)</span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a>mtry <span class="ot">&lt;-</span> <span class="fu">sqrt</span>(<span class="fu">ncol</span>(CaRF))</span>
<span id="cb87-4"><a href="#cb87-4" aria-hidden="true" tabindex="-1"></a>rf_random <span class="ot">&lt;-</span> <span class="fu">train</span>(Size<span class="sc">~</span>., <span class="at">data=</span>CaRF, <span class="at">method=</span><span class="st">"rf"</span>, <span class="at">metric=</span>metric, <span class="at">tuneLength=</span><span class="dv">15</span>, <span class="at">trControl=</span>control)</span>
<span id="cb87-5"><a href="#cb87-5" aria-hidden="true" tabindex="-1"></a>(rf_random)</span>
<span id="cb87-6"><a href="#cb87-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(rf_random)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-33_185e0c5f30d64e11b53456cc34581c11">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="do">####Expected output </span></span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a><span class="co"># mtry Accuracy Kappa</span></span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a><span class="co"># 195 0.7703453 0.5403670 </span></span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a><span class="co"># 348 0.7685185 0.5367943 </span></span>
<span id="cb88-5"><a href="#cb88-5" aria-hidden="true" tabindex="-1"></a><span class="co"># 526 0.7684685 0.5367281 </span></span>
<span id="cb88-6"><a href="#cb88-6" aria-hidden="true" tabindex="-1"></a><span class="co"># 665 0.7721972 0.5441283 </span></span>
<span id="cb88-7"><a href="#cb88-7" aria-hidden="true" tabindex="-1"></a><span class="co"># 953 0.7730731 0.5458833</span></span>
<span id="cb88-8"><a href="#cb88-8" aria-hidden="true" tabindex="-1"></a><span class="co"># 1011 0.7694444 0.5386467</span></span>
<span id="cb88-9"><a href="#cb88-9" aria-hidden="true" tabindex="-1"></a><span class="co"># 1017 0.7702202 0.5401206 </span></span>
<span id="cb88-10"><a href="#cb88-10" aria-hidden="true" tabindex="-1"></a><span class="co"># 1038 0.7685435 0.5367490 </span></span>
<span id="cb88-11"><a href="#cb88-11" aria-hidden="true" tabindex="-1"></a><span class="co"># 1115 0.7740240 0.5477867 </span></span>
<span id="cb88-12"><a href="#cb88-12" aria-hidden="true" tabindex="-1"></a><span class="co"># 1142 0.7750000 0.5497308 </span></span>
<span id="cb88-13"><a href="#cb88-13" aria-hidden="true" tabindex="-1"></a><span class="co"># 1253 0.7740490 0.5478857 </span></span>
<span id="cb88-14"><a href="#cb88-14" aria-hidden="true" tabindex="-1"></a><span class="co"># 1268 0.7740240 0.5477339 </span></span>
<span id="cb88-15"><a href="#cb88-15" aria-hidden="true" tabindex="-1"></a><span class="co"># 1627 0.7768018 0.5533787 </span></span>
<span id="cb88-16"><a href="#cb88-16" aria-hidden="true" tabindex="-1"></a><span class="co"># 1842 0.7768268 0.5533826 </span></span>
<span id="cb88-17"><a href="#cb88-17" aria-hidden="true" tabindex="-1"></a><span class="co"># 2013 0.7749750 0.5496527</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can test a combination of ntree and mtry values. Unfortunately, this is the most time-consuming part.</p>
<p>But here is the code for you to try.</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
run in <strong>your own time</strong>
</div>
</div>
<div class="callout-body-container callout-body">
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-34_076cb79e3e4cce0681699d629a72ee06">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>control <span class="ot">&lt;-</span> <span class="fu">trainControl</span>(<span class="at">method=</span><span class="st">"repeatedcv"</span>, <span class="at">number=</span><span class="dv">10</span>, <span class="at">repeats=</span><span class="dv">3</span>, <span class="at">search=</span><span class="st">"grid"</span>)</span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>tunegrid <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="at">.mtry=</span> <span class="fu">c</span>(<span class="dv">565</span>, <span class="dv">665</span>, <span class="dv">765</span>))</span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a>modellist <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a>seed<span class="ot">&lt;-</span><span class="dv">123</span></span>
<span id="cb89-5"><a href="#cb89-5" aria-hidden="true" tabindex="-1"></a>metric <span class="ot">&lt;-</span> <span class="st">"Accuracy"</span></span>
<span id="cb89-6"><a href="#cb89-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (ntree <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">200</span>, <span class="dv">300</span>, <span class="dv">500</span>, <span class="dv">1000</span>)) {</span>
<span id="cb89-7"><a href="#cb89-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(seed)</span>
<span id="cb89-8"><a href="#cb89-8" aria-hidden="true" tabindex="-1"></a>  fit <span class="ot">&lt;-</span> <span class="fu">train</span>(Size<span class="sc">~</span>., <span class="at">data=</span>CaRF, <span class="at">method=</span><span class="st">"rf"</span>, <span class="at">metric=</span>metric, <span class="at">tuneGrid=</span>tunegrid, <span class="at">trControl=</span>control, <span class="at">ntree=</span>ntree)</span>
<span id="cb89-9"><a href="#cb89-9" aria-hidden="true" tabindex="-1"></a>  key <span class="ot">&lt;-</span> <span class="fu">toString</span>(ntree)</span>
<span id="cb89-10"><a href="#cb89-10" aria-hidden="true" tabindex="-1"></a>  modellist[[key]] <span class="ot">&lt;-</span> fit</span>
<span id="cb89-11"><a href="#cb89-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb89-12"><a href="#cb89-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Then you can compare tuning results</span></span>
<span id="cb89-13"><a href="#cb89-13" aria-hidden="true" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">resamples</span>(modellist)</span>
<span id="cb89-14"><a href="#cb89-14" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(results)</span>
<span id="cb89-15"><a href="#cb89-15" aria-hidden="true" tabindex="-1"></a><span class="fu">dotplot</span>(results)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</section>
<section id="back-to-the-exercise" class="level3">
<h3 class="anchored" data-anchor-id="back-to-the-exercise">Back to the exercise</h3>
<p><img src="images/AccuracyKappa.jpg" class="img-fluid"></p>
</section>
</section>
<section id="final-model" class="level2">
<h2 class="anchored" data-anchor-id="final-model">3 Final model</h2>
<p>But the best combination was ntree = 200 and mtry= 665.</p>
<p>Now we run our forest with the optimal settings based on our tuning.</p>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-35_c689c42390a62b143e14b11783c4b0db">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>control <span class="ot">&lt;-</span> <span class="fu">trainControl</span>(<span class="at">method=</span><span class="st">"oob"</span>, <span class="at">number=</span><span class="dv">5</span>)</span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>) </span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a>metric <span class="ot">&lt;-</span> <span class="st">"Accuracy"</span></span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a>mtry <span class="ot">&lt;-</span> <span class="dv">665</span></span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a>tunegrid <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="at">.mtry=</span>mtry)</span>
<span id="cb90-6"><a href="#cb90-6" aria-hidden="true" tabindex="-1"></a>rf_best <span class="ot">&lt;-</span> <span class="fu">train</span>(Size<span class="sc">~</span>., <span class="at">data=</span>CaRF, <span class="at">method=</span><span class="st">"rf"</span>, <span class="at">metric=</span>metric, <span class="at">tuneGrid=</span>tunegrid, <span class="at">trControl=</span>control, <span class="at">ntree=</span><span class="dv">200</span>)</span>
<span id="cb90-7"><a href="#cb90-7" aria-hidden="true" tabindex="-1"></a>rf_best</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Random Forest 

 363 samples
2062 predictors
   2 classes: '1', '2' 

No pre-processing
Resampling results:

  Accuracy   Kappa    
  0.7823691  0.5645301

Tuning parameter 'mtry' was held constant at a value of 665</code></pre>
</div>
</div>
<p>~ 3% accuracy gain looks like not much, but a disclaimer, these are just SNPs from the 4 chromosomes we already identified as important for growth.</p>
</section>
<section id="candidates-snps" class="level2">
<h2 class="anchored" data-anchor-id="candidates-snps">4 Candidates SNPs</h2>
<p>Now just check what are this markers.</p>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-36_237049fe3264cd1049c8ed28501632a1">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="fu">varImpPlot</span>(rf_best<span class="sc">$</span>finalModel)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session07_CB_Yuma_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="fu">varImpPlot</span>(rf)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session07_CB_Yuma_files/figure-html/unnamed-chunk-36-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We can determine their importance using their effect on two metrics:</p>
<p>Accuracy - Ratio of correct classifications to total number of samples Gini - Measurement of the cleanliness of feature split in the tree</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Questions!?
</div>
</div>
<div class="callout-body-container callout-body">

</div>
</div>
<p>Ok let’s change approach and check concordance:</p>
</section>
<section id="rainbowr" class="level2">
<h2 class="anchored" data-anchor-id="rainbowr">RAINBOWR</h2>
</section>
<section id="inputs" class="level2">
<h2 class="anchored" data-anchor-id="inputs">1 Inputs</h2>
<p>First, we integrate our inputs in to a GWAS format for RAINBOWR.</p>
<p>And then we check each dataset format:</p>
<p>Phenotype is just two columns, ID and phenotype. Genotype is same format as randomforest, but -1 hom mA 0 het 1 hom MA. Map file with the chromosome and position of each SNP.</p>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-37_0da64ba1176e1b335639c3c8b50ec1cd">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>modify.CaGW.res <span class="ot">&lt;-</span> <span class="fu">modify.data</span>(<span class="at">pheno.mat =</span> CaGW_phe, <span class="at">geno.mat =</span> CaGW_geno, <span class="at">map =</span> CaGW_map, <span class="at">return.ZETA =</span> <span class="cn">TRUE</span>, <span class="at">return.GWAS.format =</span> <span class="cn">TRUE</span>)</span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a>pheno.GWAS <span class="ot">&lt;-</span> modify.CaGW.res<span class="sc">$</span>pheno.GWAS</span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a>geno.GWAS <span class="ot">&lt;-</span> modify.CaGW.res<span class="sc">$</span>geno.GWAS</span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-5"><a href="#cb94-5" aria-hidden="true" tabindex="-1"></a><span class="do">### View each dataset</span></span>
<span id="cb94-6"><a href="#cb94-6" aria-hidden="true" tabindex="-1"></a><span class="fu">See</span>(pheno.GWAS)  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          Sample_names      Size
class      &lt;character&gt; &lt;integer&gt;
popL_L026    popL_L026         2
popL_L004    popL_L004         2
popL_L024    popL_L024         2
popL_L010    popL_L010         2
popL_L025    popL_L025         2
popL_L016    popL_L016         2
[1] "class: data.frame"
[1] "dimension: 363 x 2"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="fu">See</span>(geno.GWAS)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            marker     chrom       pos popL_L026 popL_L004 popL_L024
class  &lt;character&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
1     LG1:10070023         1  10070023        -1        -1        -1
2     LG1:10196551         1  10196551        -1        -1        -1
3     LG1:10271823         1  10271823        -1        -1        -1
4     LG1:10361941         1  10361941         0        -1        -1
5     LG1:10398562         1  10398562        -1         0        -1
6     LG1:10529180         1  10529180        -1        -1        -1
[1] "class: data.frame"
[1] "dimension: 17490 x 366"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="fu">See</span>(CaGW_map)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                   marker       chr       pos
class         &lt;character&gt; &lt;integer&gt; &lt;integer&gt;
LG1:10070023 LG1:10070023         1  10070023
LG1:10196551 LG1:10196551         1  10196551
LG1:10271823 LG1:10271823         1  10271823
LG1:10361941 LG1:10361941         1  10361941
LG1:10398562 LG1:10398562         1  10398562
LG1:10529180 LG1:10529180         1  10529180
[1] "class: data.frame"
[1] "dimension: 17490 x 3"</code></pre>
</div>
</div>
</section>
<section id="single-snp-base-gwas" class="level2">
<h2 class="anchored" data-anchor-id="single-snp-base-gwas">2 Single-SNP base GWAS</h2>
<p>Now we calculate ZETA, a list of genomic relationship matrix (GRM). ZETA is basically a random effect and accounts for possible confounding/covariant factors like population structure and relatedness.</p>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-38_7a7903b96afdfba386e789fd5b2388fc">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>ZETA <span class="ot">&lt;-</span> modify.CaGW.res<span class="sc">$</span>ZETA</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
There are many ways to control for cofactors, here some examples.
</div>
</div>
<div class="callout-body-container callout-body">
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-39_f0c037c7f7da2513ee6aa112b0d9647a">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="do">### Population structure, using a structure Q values matrix, n.PC should be &gt; 0</span></span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a>normal.res <span class="ot">&lt;-</span> <span class="fu">RGWAS.normal</span>(<span class="at">pheno =</span> pheno.GWAS, <span class="at">geno =</span> geno.GWAS, <span class="at">ZETA =</span> ZETA, <span class="at">structure.matrix=</span> QMATRIX, <span class="at">n.PC =</span> <span class="dv">4</span>, <span class="at">P3D =</span> <span class="cn">TRUE</span>)</span>
<span id="cb101-3"><a href="#cb101-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb101-4"><a href="#cb101-4" aria-hidden="true" tabindex="-1"></a><span class="do">#### other cofactors, like sex or age (a two-column table similar to phenotype table)</span></span>
<span id="cb101-5"><a href="#cb101-5" aria-hidden="true" tabindex="-1"></a>normal.res <span class="ot">&lt;-</span> <span class="fu">RGWAS.normal</span>(<span class="at">pheno =</span> pheno.GWAS, <span class="at">geno =</span> geno.GWAS, <span class="at">ZETA =</span> ZETA, <span class="at">structure.matrix=</span> QMATRIX, <span class="at">covariate =</span> SEX_AGE,  <span class="at">n.PC =</span> <span class="dv">4</span>, <span class="at">P3D =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<p>Now, to perform single-SNP GWAS.</p>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-40_5b99ad6e8403de89b128d7496e639be6">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>normal.res <span class="ot">&lt;-</span> <span class="fu">RGWAS.normal</span>(<span class="at">pheno =</span> pheno.GWAS, <span class="at">geno =</span> geno.GWAS, <span class="at">ZETA =</span> ZETA, <span class="at">n.PC =</span> <span class="dv">4</span>, <span class="at">P3D =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can see, first the QQ plot, it shows that our data has a little deviation from the expected distribution, but this is just at the end of the tail.</p>
<p>Second, the Manhattan plot shows few SNPs that are significantly associated to size, but more importantly, it shows a peak of relatively high-importance SNPs at the end of chromosome 16.</p>
<p>So, what are the most important important SNPs?</p>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-41_03487d40c76a6406038701ffdc280fae">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>normal.res<span class="sc">$</span>D[normal.res<span class="sc">$</span>D<span class="sc">$</span>Size  <span class="sc">&gt;=</span> <span class="dv">4</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                             marker chrom      pos     Size
543                    LG1:34671481     1 34671481 6.266137
8335                  LG11:34898993    11 34898993 4.304650
11893                  LG16:4658138    16  4658138 4.758395
11942                  LG16:6898540    16  6898540 5.501243
11992                  LG16:8984283    16  8984283 4.677513
11995                  LG16:8986211    16  8986211 5.995467
12004                  LG16:9625373    16  9625373 4.453957
12005                  LG16:9719769    16  9719769 4.451432
12007                  LG16:9838071    16  9838071 4.261391
11793                 LG16:28880624    16 28880624 5.004830
15566                  LG22:2463072    22  2463072 4.568311
17287 CauratusV1_scaffold_3461:3888 10141     3888 4.214943</code></pre>
</div>
</div>
</section>
<section id="genomic-prediction" class="level2">
<h2 class="anchored" data-anchor-id="genomic-prediction">3 Genomic prediction</h2>
<p>How much of the phenotype can we really predict?</p>
<p>To determine that, we use a mixed model approach to calculate heritability, in other words the genetic contribution to our trait variation.</p>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-42_35bb911e6562db6f2f9d0acbe7cafd26">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>K.A2<span class="ot">&lt;-</span><span class="fu">calcGRM</span>(<span class="at">genoMat =</span> CaGW_geno) <span class="do">### aditive effects</span></span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a>k.AA2 <span class="ot">&lt;-</span> K.A2 <span class="sc">*</span> K.A2 <span class="do">### epistatic effects</span></span>
<span id="cb105-3"><a href="#cb105-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb105-4"><a href="#cb105-4" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(pheno.GWAS[,<span class="st">'Size'</span>, <span class="at">drop =</span> <span class="cn">FALSE</span>])</span>
<span id="cb105-5"><a href="#cb105-5" aria-hidden="true" tabindex="-1"></a>Z <span class="ot">&lt;-</span> <span class="fu">design.Z</span>(<span class="at">pheno.labels =</span> <span class="fu">rownames</span>(y), <span class="at">geno.names =</span> <span class="fu">rownames</span>(K.A2))</span>
<span id="cb105-6"><a href="#cb105-6" aria-hidden="true" tabindex="-1"></a>pheno.mat2 <span class="ot">&lt;-</span> y[<span class="fu">rownames</span>(Z), , drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb105-7"><a href="#cb105-7" aria-hidden="true" tabindex="-1"></a>ZETA2 <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">A =</span> <span class="fu">list</span>(<span class="at">Z =</span> Z, <span class="at">K =</span> K.A2),<span class="at">AA =</span> <span class="fu">list</span>(<span class="at">Z =</span> Z, <span class="at">K =</span> k.AA2))</span>
<span id="cb105-8"><a href="#cb105-8" aria-hidden="true" tabindex="-1"></a>EM3.res <span class="ot">&lt;-</span> <span class="fu">EM3.cpp</span>(<span class="at">y =</span> pheno.mat2, <span class="at">X =</span> <span class="cn">NULL</span>, <span class="at">ZETA =</span> ZETA2) <span class="do">### multi-kernel linear mixed effects model</span></span>
<span id="cb105-9"><a href="#cb105-9" aria-hidden="true" tabindex="-1"></a>(Vu <span class="ot">&lt;-</span> EM3.res<span class="sc">$</span>Vu)   <span class="do">### estimated genetic variance</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1655835</code></pre>
</div>
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>(Ve <span class="ot">&lt;-</span> EM3.res<span class="sc">$</span>Ve)   <span class="do">### estimated residual variance</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.07379584</code></pre>
</div>
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a>(weights <span class="ot">&lt;-</span> EM3.res<span class="sc">$</span>weights) <span class="do">### estimated proportion of two genetic variances</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.91355965 0.08644035</code></pre>
</div>
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a>(herit <span class="ot">&lt;-</span> Vu <span class="sc">*</span> weights <span class="sc">/</span> (Vu <span class="sc">+</span> Ve)) <span class="do">### genomic heritability</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.63192756 0.05979253</code></pre>
</div>
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a>herit</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.63192756 0.05979253</code></pre>
</div>
</div>
<p>Now we can test the performance of our genomic prediction with 10-fold cross-validation.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>If you have NAs in your phenotype dataset you have to remove them.</p>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-43_6fb443eaf9c41f61925f47b884b45cce">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a>phenoNoNA <span class="ot">&lt;-</span> pheno.mat2[noNA, , drop <span class="ot">=</span> <span class="cn">FALSE</span>]   <span class="do">### remove NA</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<section id="but-our-data-doesnt-have-any-nas." class="level4">
<h4 class="anchored" data-anchor-id="but-our-data-doesnt-have-any-nas."><em>But our data doesn’t have any NAs.</em></h4>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-44_b6ec1cf8cda35fc9646fd6909023bb5c">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a>nFold <span class="ot">&lt;-</span> <span class="dv">10</span> </span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a>nLine <span class="ot">&lt;-</span> <span class="fu">nrow</span>(pheno.mat2)</span>
<span id="cb116-3"><a href="#cb116-3" aria-hidden="true" tabindex="-1"></a>idCV <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span>nLine <span class="sc">%%</span> nFold) <span class="do">### assign random ids for cross-validation</span></span>
<span id="cb116-4"><a href="#cb116-4" aria-hidden="true" tabindex="-1"></a>idCV[idCV <span class="sc">==</span> <span class="dv">0</span>] <span class="ot">&lt;-</span> nFold</span>
<span id="cb116-5"><a href="#cb116-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb116-6"><a href="#cb116-6" aria-hidden="true" tabindex="-1"></a>yPred <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA</span>, nLine)</span>
<span id="cb116-7"><a href="#cb116-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (noCV <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>nFold) {</span>
<span id="cb116-8"><a href="#cb116-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"Fold: "</span>, noCV))</span>
<span id="cb116-9"><a href="#cb116-9" aria-hidden="true" tabindex="-1"></a>  yTrain <span class="ot">&lt;-</span> pheno.mat2</span>
<span id="cb116-10"><a href="#cb116-10" aria-hidden="true" tabindex="-1"></a>  yTrain[idCV <span class="sc">==</span> noCV, ] <span class="ot">&lt;-</span> <span class="cn">NA</span> <span class="do">### prepare test data</span></span>
<span id="cb116-11"><a href="#cb116-11" aria-hidden="true" tabindex="-1"></a>  EM3.gaston.resCV <span class="ot">&lt;-</span> <span class="fu">EM3.general</span>(<span class="at">y =</span> yTrain, <span class="at">X0 =</span> <span class="cn">NULL</span>, <span class="at">ZETA =</span> ZETA,</span>
<span id="cb116-12"><a href="#cb116-12" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">package =</span> <span class="st">"gaston"</span>, <span class="at">return.u.always =</span> <span class="cn">TRUE</span>,</span>
<span id="cb116-13"><a href="#cb116-13" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">pred =</span> <span class="cn">TRUE</span>, <span class="at">return.u.each =</span> <span class="cn">TRUE</span>,</span>
<span id="cb116-14"><a href="#cb116-14" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">return.Hinv =</span> <span class="cn">TRUE</span>) </span>
<span id="cb116-15"><a href="#cb116-15" aria-hidden="true" tabindex="-1"></a>  yTest <span class="ot">&lt;-</span> EM3.gaston.resCV<span class="sc">$</span>y.pred <span class="do">### predicted values</span></span>
<span id="cb116-16"><a href="#cb116-16" aria-hidden="true" tabindex="-1"></a>  yPred[idCV <span class="sc">==</span> noCV] <span class="ot">&lt;-</span> yTest[idCV <span class="sc">==</span> noCV]</span>
<span id="cb116-17"><a href="#cb116-17" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Fold: 1"
[1] "Fold: 2"
[1] "Fold: 3"
[1] "Fold: 4"
[1] "Fold: 5"
[1] "Fold: 6"
[1] "Fold: 7"
[1] "Fold: 8"
[1] "Fold: 9"
[1] "Fold: 10"</code></pre>
</div>
</div>
</section>
<section id="plot-genomic-prediction." class="level3">
<h3 class="anchored" data-anchor-id="plot-genomic-prediction.">Plot Genomic Prediction.</h3>
<div class="cell" data-hash="session07_CB_Yuma_cache/html/unnamed-chunk-45_36a2f5195ed91705ab5ce80b8c69496a">
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a>plotRange <span class="ot">&lt;-</span> <span class="fu">range</span>(pheno.mat2, yPred)</span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> pheno.mat2, <span class="at">y =</span> yPred,<span class="at">xlim =</span> plotRange, <span class="at">ylim =</span> plotRange,</span>
<span id="cb118-3"><a href="#cb118-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Observed values"</span>, <span class="at">ylab =</span> <span class="st">"Predicted values"</span>,</span>
<span id="cb118-4"><a href="#cb118-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Results of Genomic Prediction (multi-kernel)"</span>,</span>
<span id="cb118-5"><a href="#cb118-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">cex.lab =</span> <span class="fl">1.5</span>, <span class="at">cex.main =</span> <span class="fl">1.5</span>, <span class="at">cex.axis =</span> <span class="fl">1.3</span>)</span>
<span id="cb118-6"><a href="#cb118-6" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">a =</span> <span class="dv">0</span>, <span class="at">b =</span> <span class="dv">1</span>, <span class="at">col =</span> <span class="dv">2</span>, <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">lty =</span> <span class="dv">2</span>)</span>
<span id="cb118-7"><a href="#cb118-7" aria-hidden="true" tabindex="-1"></a>R2 <span class="ot">&lt;-</span> <span class="fu">cor</span>(<span class="at">x =</span> pheno.mat2[, <span class="dv">1</span>], <span class="at">y =</span> yPred) <span class="sc">^</span> <span class="dv">2</span></span>
<span id="cb118-8"><a href="#cb118-8" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="at">x =</span> plotRange[<span class="dv">2</span>] <span class="sc">-</span> <span class="dv">10</span>,</span>
<span id="cb118-9"><a href="#cb118-9" aria-hidden="true" tabindex="-1"></a>     <span class="at">y =</span> plotRange[<span class="dv">1</span>] <span class="sc">+</span> <span class="dv">10</span>,</span>
<span id="cb118-10"><a href="#cb118-10" aria-hidden="true" tabindex="-1"></a>     <span class="fu">paste0</span>(<span class="st">"R2 = "</span>, <span class="fu">round</span>(R2, <span class="dv">3</span>)),</span>
<span id="cb118-11"><a href="#cb118-11" aria-hidden="true" tabindex="-1"></a>     <span class="at">cex =</span> <span class="fl">1.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session07_CB_Yuma_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Not really the best prediction power, and even worse than randomForest just using ~2K SNPs.</p>
<p>It is important to note that combinations of methods can improve our power to predict complex traits that are polygenic.</p>
<p>Using haplotype-based approaches and Bayesian methods helps a lot.</p>
<p>Annotation of the regions of chromosome 16 shows presence of several genes related to growth and body development.</p>
</section>
</section>
<section id="further-steps" class="level2">
<h2 class="anchored" data-anchor-id="further-steps"><em>Further steps</em></h2>
<p>It is time to test how structural variances can improve the power of GWAS.</p>
</section>
<section id="readings" class="level2">
<h2 class="anchored" data-anchor-id="readings"><em>Readings</em></h2>
<p><span class="citation" data-cites="attard_fish_2022">Attard et al. (<a href="references.html#ref-attard_fish_2022" role="doc-biblioref">2022</a>)</span></p>
<p><span class="citation" data-cites="brauer_roles_2018">Brauer et al. (<a href="references.html#ref-brauer_roles_2018" role="doc-biblioref">2018</a>)</span></p>
<p><span class="citation" data-cites="batley_whole_2021">Batley et al. (<a href="references.html#ref-batley_whole_2021" role="doc-biblioref">2021</a>)</span></p>
<p><span class="citation" data-cites="smith_latitudinal_2020">Smith et al. (<a href="references.html#ref-smith_latitudinal_2020" role="doc-biblioref">2020</a>)</span></p>
<p><span class="citation" data-cites="brauer_natural_2023">Brauer et al. (<a href="references.html#ref-brauer_natural_2023" role="doc-biblioref">2023</a>)</span></p>
<p><span class="citation" data-cites="grummer_aquatic_2019">Grummer et al. (<a href="references.html#ref-grummer_aquatic_2019" role="doc-biblioref">2019</a>)</span></p>
<p><span class="citation" data-cites="pratt_seascape_2022">Pratt et al. (<a href="references.html#ref-pratt_seascape_2022" role="doc-biblioref">2022</a>)</span></p>
<p><span class="citation" data-cites="sandoval-castillo_seascape_2018">Sandoval-Castillo et al. (<a href="references.html#ref-sandoval-castillo_seascape_2018" role="doc-biblioref">2018</a>)</span></p>


<div id="refs" class="references csl-bib-body hanging-indent" role="list" style="display: none">
<div id="ref-attard_fish_2022" class="csl-entry" role="listitem">
Attard, Catherine R. M., Jonathan Sandoval-Castillo, Chris J. Brauer, Peter J. Unmack, David Schmarr, Louis Bernatchez, and Luciano B. Beheregaray. 2022. <span>“Fish Out of Water: <span>Genomic</span> Insights into Persistence of Rainbowfish Populations in the Desert.”</span> <em>Evolution; International Journal of Organic Evolution</em> 76 (1): 171–83. <a href="https://doi.org/10.1111/evo.14399">https://doi.org/10.1111/evo.14399</a>.
</div>
<div id="ref-batley_whole_2021" class="csl-entry" role="listitem">
Batley, Kimberley C., Jonathan Sandoval-Castillo, Catherine M. Kemper, Nikki Zanardo, Ikuko Tomo, Luciano B. Beheregaray, and Luciana M. Möller. 2021. <span>“Whole Genomes Reveal Multiple Candidate Genes and Pathways Involved in the Immune Response of Dolphins to a Highly Infectious Virus.”</span> <em>Molecular Ecology</em> 30 (23): 6434–48. <a href="https://doi.org/10.1111/mec.15873">https://doi.org/10.1111/mec.15873</a>.
</div>
<div id="ref-brauer_natural_2023" class="csl-entry" role="listitem">
Brauer, Chris J., Jonathan Sandoval-Castillo, Katie Gates, Michael P. Hammer, Peter J. Unmack, Louis Bernatchez, and Luciano B. Beheregaray. 2023. <span>“Natural Hybridization Reduces Vulnerability to Climate Change.”</span> <em>Nature Climate Change</em> 13 (3): 282–89. <a href="https://doi.org/10.1038/s41558-022-01585-1">https://doi.org/10.1038/s41558-022-01585-1</a>.
</div>
<div id="ref-brauer_roles_2018" class="csl-entry" role="listitem">
Brauer, Chris J., Peter J. Unmack, Steve Smith, Louis Bernatchez, and Luciano B. Beheregaray. 2018. <span>“On the Roles of Landscape Heterogeneity and Environmental Variation in Determining Population Genomic Structure in a Dendritic System.”</span> <em>Molecular Ecology</em> 27 (17): 3484–97. <a href="https://doi.org/10.1111/mec.14808">https://doi.org/10.1111/mec.14808</a>.
</div>
<div id="ref-grummer_aquatic_2019" class="csl-entry" role="listitem">
Grummer, Jared A., Luciano B. Beheregaray, Louis Bernatchez, Brian K. Hand, Gordon Luikart, Shawn R. Narum, and Eric B. Taylor. 2019. <span>“Aquatic <span>Landscape</span> <span>Genomics</span> and <span>Environmental</span> <span>Effects</span> on <span>Genetic</span> <span>Variation</span>.”</span> <em>Trends in Ecology &amp; Evolution</em> 34 (7): 641–54. <a href="https://doi.org/10.1016/j.tree.2019.02.013">https://doi.org/10.1016/j.tree.2019.02.013</a>.
</div>
<div id="ref-pratt_seascape_2022" class="csl-entry" role="listitem">
Pratt, Eleanor A. L., Luciano B. Beheregaray, Kerstin Bilgmann, Nikki Zanardo, Fernando Diaz-Aguirre, Chris Brauer, Jonathan Sandoval-Castillo, and Luciana M. Möller. 2022. <span>“Seascape Genomics of Coastal Bottlenose Dolphins Along Strong Gradients of Temperature and Salinity.”</span> <em>Molecular Ecology</em> 31 (8): 2223–41. <a href="https://doi.org/10.1111/mec.16389">https://doi.org/10.1111/mec.16389</a>.
</div>
<div id="ref-sandoval-castillo_seascape_2018" class="csl-entry" role="listitem">
Sandoval-Castillo, Jonathan, Nick A. Robinson, Anthony M. Hart, Lachlan W. S. Strain, and Luciano B. Beheregaray. 2018. <span>“Seascape Genomics Reveals Adaptive Divergence in a Connected and Commercially Important Mollusc, the Greenlip Abalone (<span>Haliotis</span> Laevigata), Along a Longitudinal Environmental Gradient.”</span> <em>Molecular Ecology</em> 27 (7): 1603–20. <a href="https://doi.org/10.1111/mec.14526">https://doi.org/10.1111/mec.14526</a>.
</div>
<div id="ref-smith_latitudinal_2020" class="csl-entry" role="listitem">
Smith, Steve, Chris J. Brauer, Minami Sasaki, Peter J. Unmack, Gilles Guillot, Martin Laporte, Louis Bernatchez, and Luciano B. Beheregaray. 2020. <span>“Latitudinal Variation in Climate-Associated Genes Imperils Range Edge Populations.”</span> <em>Molecular Ecology</em> 29 (22): 4337–49. <a href="https://doi.org/10.1111/mec.15637">https://doi.org/10.1111/mec.15637</a>.
</div>
</div>
</section>

</main> <!-- /main -->
<script>

/* update total correct if #webex-total_correct exists */
update_total_correct = function() {
  console.log("webex: update total_correct");

  var t = document.getElementsByClassName("webex-total_correct");
  for (var i = 0; i < t.length; i++) {
    p = t[i].parentElement;
    var correct = p.getElementsByClassName("webex-correct").length;
    var solvemes = p.getElementsByClassName("webex-solveme").length;
    var radiogroups = p.getElementsByClassName("webex-radiogroup").length;
    var selects = p.getElementsByClassName("webex-select").length;

    t[i].innerHTML = correct + " of " + (solvemes + radiogroups + selects) + " correct";
  }
}

/* webex-solution button toggling function */
b_func = function() {
  console.log("webex: toggle hide");

  var cl = this.parentElement.classList;
  if (cl.contains('open')) {
    cl.remove("open");
  } else {
    cl.add("open");
  }
}

/* check answers */
check_func = function() {
  console.log("webex: check answers");

  var cl = this.parentElement.classList;
  if (cl.contains('unchecked')) {
    cl.remove("unchecked");
    this.innerHTML = "Hide Answers";
  } else {
    cl.add("unchecked");
    this.innerHTML = "Show Answers";
  }
}

/* function for checking solveme answers */
solveme_func = function(e) {
  console.log("webex: check solveme");

  var real_answers = JSON.parse(this.dataset.answer);
  var my_answer = this.value;
  var cl = this.classList;
  if (cl.contains("ignorecase")) {
    my_answer = my_answer.toLowerCase();
  }
  if (cl.contains("nospaces")) {
    my_answer = my_answer.replace(/ /g, "")
  }

  if (my_answer == "") {
    cl.remove("webex-correct");
    cl.remove("webex-incorrect");
  } else if (real_answers.includes(my_answer)) {
    cl.add("webex-correct");
    cl.remove("webex-incorrect");
  } else {
    cl.add("webex-incorrect");
    cl.remove("webex-correct");
  }

  // match numeric answers within a specified tolerance
  if(this.dataset.tol > 0){
    var tol = JSON.parse(this.dataset.tol);
    var matches = real_answers.map(x => Math.abs(x - my_answer) < tol)
    if (matches.reduce((a, b) => a + b, 0) > 0) {
      cl.add("webex-correct");
    } else {
      cl.remove("webex-correct");
    }
  }

  // added regex bit
  if (cl.contains("regex")){
    answer_regex = RegExp(real_answers.join("|"))
    if (answer_regex.test(my_answer)) {
      cl.add("webex-correct");
    }
  }

  update_total_correct();
}

/* function for checking select answers */
select_func = function(e) {
  console.log("webex: check select");

  var cl = this.classList

  /* add style */
  cl.remove("webex-incorrect");
  cl.remove("webex-correct");
  if (this.value == "answer") {
    cl.add("webex-correct");
  } else if (this.value != "blank") {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

/* function for checking radiogroups answers */
radiogroups_func = function(e) {
  console.log("webex: check radiogroups");

  var checked_button = document.querySelector('input[name=' + this.id + ']:checked');
  var cl = checked_button.parentElement.classList;
  var labels = checked_button.parentElement.parentElement.children;

  /* get rid of styles */
  for (i = 0; i < labels.length; i++) {
    labels[i].classList.remove("webex-incorrect");
    labels[i].classList.remove("webex-correct");
  }

  /* add style */
  if (checked_button.value == "answer") {
    cl.add("webex-correct");
  } else {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

window.onload = function() {
  console.log("webex onload");
  /* set up solution buttons */
  var buttons = document.getElementsByTagName("button");

  for (var i = 0; i < buttons.length; i++) {
    if (buttons[i].parentElement.classList.contains('webex-solution')) {
      buttons[i].onclick = b_func;
    }
  }

  var check_sections = document.getElementsByClassName("webex-check");
  console.log("check:", check_sections.length);
  for (var i = 0; i < check_sections.length; i++) {
    check_sections[i].classList.add("unchecked");

    let btn = document.createElement("button");
    btn.innerHTML = "Show Answers";
    btn.classList.add("webex-check-button");
    btn.onclick = check_func;
    check_sections[i].appendChild(btn);

    let spn = document.createElement("span");
    spn.classList.add("webex-total_correct");
    check_sections[i].appendChild(spn);
  }

  /* set up webex-solveme inputs */
  var solveme = document.getElementsByClassName("webex-solveme");

  for (var i = 0; i < solveme.length; i++) {
    /* make sure input boxes don't auto-anything */
    solveme[i].setAttribute("autocomplete","off");
    solveme[i].setAttribute("autocorrect", "off");
    solveme[i].setAttribute("autocapitalize", "off");
    solveme[i].setAttribute("spellcheck", "false");
    solveme[i].value = "";

    /* adjust answer for ignorecase or nospaces */
    var cl = solveme[i].classList;
    var real_answer = solveme[i].dataset.answer;
    if (cl.contains("ignorecase")) {
      real_answer = real_answer.toLowerCase();
    }
    if (cl.contains("nospaces")) {
      real_answer = real_answer.replace(/ /g, "");
    }
    solveme[i].dataset.answer = real_answer;

    /* attach checking function */
    solveme[i].onkeyup = solveme_func;
    solveme[i].onchange = solveme_func;

    solveme[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  /* set up radiogroups */
  var radiogroups = document.getElementsByClassName("webex-radiogroup");
  for (var i = 0; i < radiogroups.length; i++) {
    radiogroups[i].onchange = radiogroups_func;
  }

  /* set up selects */
  var selects = document.getElementsByClassName("webex-select");
  for (var i = 0; i < selects.length; i++) {
    selects[i].onchange = select_func;
    selects[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  update_total_correct();
}

</script>
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
          // default icon
          link.classList.add("external");
      }
    }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./session06.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Management of Small Populations</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./session08.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Landscape Genetics</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>