<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Kioloa PopGen - 3&nbsp; Sequencing Technologies</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./session04.html" rel="next">
<link href="./session02.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="css.css">
<link rel="stylesheet" href="include/webex.css">
</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./session03.html"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Sequencing Technologies</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Kioloa PopGen</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome!</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./schedule.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Program</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./speakers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Speaker Bios</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./install.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Install dartRverse</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session01.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Intro to dartR</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session02.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Pop Gen In Conservation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session03.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Sequencing Technologies</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session04.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Data Management, Reproducibility &amp; Integrity</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session05.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Effective Population Size</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session06.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Management of Small Populations</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session07.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Natural Selection</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session08.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Landscape Genetics</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session09.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Lineage Divergence</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session10.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Sex Linked Markers</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session11.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">From Genes to Kin: Dissecting Relatedness &amp; Kinship</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session12.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Genetic Structure</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./session13.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Combining Genomic Resources</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./break.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">__________</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./useful.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">More info</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">dartR Tutorials</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#required-packages" id="toc-required-packages" class="nav-link active" data-scroll-target="#required-packages"><em>Required packages</em></a></li>
  <li><a href="#dart-sequencing" id="toc-dart-sequencing" class="nav-link" data-scroll-target="#dart-sequencing"><strong>DArT sequencing</strong></a>
  <ul class="collapse">
  <li><a href="#characteristics" id="toc-characteristics" class="nav-link" data-scroll-target="#characteristics">Characteristics</a></li>
  <li><a href="#standardised" id="toc-standardised" class="nav-link" data-scroll-target="#standardised">Standardised</a></li>
  </ul></li>
  <li><a href="#snp-based-heterozygosity-estimates" id="toc-snp-based-heterozygosity-estimates" class="nav-link" data-scroll-target="#snp-based-heterozygosity-estimates"><strong>SNP-based heterozygosity estimates</strong></a></li>
  <li><a href="#estimating-heterozygosity-70-callrate" id="toc-estimating-heterozygosity-70-callrate" class="nav-link" data-scroll-target="#estimating-heterozygosity-70-callrate">1. Estimating heterozygosity: 70% callrate</a>
  <ul class="collapse">
  <li><a href="#load-data" id="toc-load-data" class="nav-link" data-scroll-target="#load-data">Load data</a></li>
  <li><a href="#calculate-heterozygosity" id="toc-calculate-heterozygosity" class="nav-link" data-scroll-target="#calculate-heterozygosity">Calculate heterozygosity</a></li>
  <li><a href="#plotting-results" id="toc-plotting-results" class="nav-link" data-scroll-target="#plotting-results">Plotting results</a></li>
  </ul></li>
  <li><a href="#estimating-heterozygosity-95-callrate" id="toc-estimating-heterozygosity-95-callrate" class="nav-link" data-scroll-target="#estimating-heterozygosity-95-callrate">2. Estimating heterozygosity: 95% callrate</a>
  <ul class="collapse">
  <li><a href="#reload-data" id="toc-reload-data" class="nav-link" data-scroll-target="#reload-data">Reload data</a></li>
  <li><a href="#calculate-heterozygosity-1" id="toc-calculate-heterozygosity-1" class="nav-link" data-scroll-target="#calculate-heterozygosity-1">Calculate heterozygosity</a></li>
  <li><a href="#plotting-results-1" id="toc-plotting-results-1" class="nav-link" data-scroll-target="#plotting-results-1">Plotting results</a></li>
  </ul></li>
  <li><a href="#estimating-heterozygosity-even-sampling" id="toc-estimating-heterozygosity-even-sampling" class="nav-link" data-scroll-target="#estimating-heterozygosity-even-sampling">3. Estimating heterozygosity: Even sampling</a>
  <ul class="collapse">
  <li><a href="#subsampling" id="toc-subsampling" class="nav-link" data-scroll-target="#subsampling">Subsampling</a></li>
  <li><a href="#plotting-results-2" id="toc-plotting-results-2" class="nav-link" data-scroll-target="#plotting-results-2">Plotting results</a></li>
  <li><a href="#subsampling-more-populations" id="toc-subsampling-more-populations" class="nav-link" data-scroll-target="#subsampling-more-populations">Subsampling more populations</a></li>
  <li><a href="#plotting-results-3" id="toc-plotting-results-3" class="nav-link" data-scroll-target="#plotting-results-3">Plotting Results</a></li>
  </ul></li>
  <li><a href="#calling-populations-together" id="toc-calling-populations-together" class="nav-link" data-scroll-target="#calling-populations-together">Calling populations together</a>
  <ul class="collapse">
  <li><a href="#calling-snps-together-but-with-different-sample-sizes" id="toc-calling-snps-together-but-with-different-sample-sizes" class="nav-link" data-scroll-target="#calling-snps-together-but-with-different-sample-sizes">Calling SNPs together but with different sample sizes</a></li>
  </ul></li>
  <li><a href="#further-study" id="toc-further-study" class="nav-link" data-scroll-target="#further-study"><em>Further Study</em></a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Sequencing Technologies</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p><em>Session Presenters</em></p>
<p><img src="images/Presenters3.png" class="img-fluid"></p>
<section id="required-packages" class="level2">
<h2 class="anchored" data-anchor-id="required-packages"><em>Required packages</em></h2>
<div class="cell" data-hash="session03_cache/html/unnamed-chunk-1_daf36640fce10a477df63a670538399d">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dartRverse)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>make sure you have the packages installed, see</em> <a href="./install.html">Install dartRverse</a></p>
</section>
<section id="dart-sequencing" class="level2">
<h2 class="anchored" data-anchor-id="dart-sequencing"><strong>DArT sequencing</strong></h2>
<p><img src="images/DArT_logo.png" class="img-fluid"></p>
<p><img src="images/dart_sequencing.png" class="img-fluid"></p>
<p>DArTseq is our proprietary genome complexity reduction-based sequencing technology.</p>
<p>It differs from other methods through its ability to intelligently select the predominantly active – low copy sequence – areas in a genome, which are the ones containing the most useful information. At the same time, DArTseq masks the lesser value, repetitive sequences. It does this through the application of a combination of restriction enzymes to fragment DNA samples in a highly reproducible manner.</p>
<p>A key component of DArTseq is its complexity reduction step. This involves reducing complexity of the genome by cutting it down to a consistent and reproducible fraction from which markers are discovered and called. This process does not introduce ascertainment bias because we are discovering markers within the (restriction) fragment set without the bias.</p>
<section id="characteristics" class="level3">
<h3 class="anchored" data-anchor-id="characteristics">Characteristics</h3>
<p>DArTseq is designed to provide a consistent genomic representation across samples within and between experiments, even over a decade of operations. DArTseq has been used to process tens of thousands of samples from hundreds of organisms, and is capable of co-analysing thousands of samples together, with the great majority of markers called across all of the samples. These advantages come from the library processing methods of DArTseq which distinguish it from all other complexity reduced genotyping technologies using restriction enzymes.</p>
</section>
<section id="standardised" class="level3">
<h3 class="anchored" data-anchor-id="standardised">Standardised</h3>
<p>Laboratory sample processing is performed in a very high throughput streamlined operation utilising end-to-end pipetting robotics and informatics integration, with a Laboratory Information Management System (LIMS). All the physical sample processing steps undertaken in the lab are linked with the data storage and analysis components via tight integration with the LIMS. Sample processing robots have automated data exchange with the LIMS.</p>
<p>Manual operations have been minimised and plate based sample handling prevents the introduction of sample-tracking mistakes after arrival. The analytical components for processing DArTseq data and generating marker calls have been tailored specifically to the unique DArTseq complexity reduction and library construction methods. This results in a streamlined and highly effective data analysis pipeline, enabling rapid generation of high-quality marker data from thousands or even tens of thousands of assays in a single analysis. Ability to co-analysis with past services of 14 years despite 5 generations of the illumina sequences. The population of fragments in the genomic representation does not vary across multiple sample submissions despite the many years.</p>
<p>See more at <a href="https://www.diversityarrays.com/" class="uri">https://www.diversityarrays.com/</a></p>
</section>
</section>
<section id="snp-based-heterozygosity-estimates" class="level2">
<h2 class="anchored" data-anchor-id="snp-based-heterozygosity-estimates"><strong>SNP-based heterozygosity estimates</strong></h2>
</section>
<section id="estimating-heterozygosity-70-callrate" class="level2">
<h2 class="anchored" data-anchor-id="estimating-heterozygosity-70-callrate">1. Estimating heterozygosity: 70% callrate</h2>
<p><em>How do estimates of heterozygosity change as the number of individuals changes in your SNP calling?</em></p>
<p>This uses data from Litoria rubella, a very abundant and widespread frog species. We are using data following <a href="https://onlinelibrary.wiley.com/doi/10.1111/1755-0998.13947" class="uri">https://onlinelibrary.wiley.com/doi/10.1111/1755-0998.13947</a>. Let’s focus on the Kimberley, where we have lots of samples, and we expect the diversity to be really high because there are millions of them everywhere. SNPs were called on different numbers of individuals, in increments of 5 using the Stacks pipeline. We are only going to do a simple filter for call rates because we already filtered for allele depth in Stacks. SNPs were called independent of other populations, which we will get to later.</p>
<section id="load-data" class="level3">
<h3 class="anchored" data-anchor-id="load-data">Load data</h3>
<div class="cell" data-hash="session03_cache/html/unnamed-chunk-2_52228a6fc292bc1f97f79aded52f0377">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load data</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">'./data/Session3_data.RData'</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co">#create a list of the kimberley genlights</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>kimberley_names <span class="ot">&lt;-</span> <span class="fu">ls</span>(<span class="at">pattern =</span> <span class="st">"^Kimberley"</span>)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co">#put all the genlights into a mega list</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>kimberley <span class="ot">&lt;-</span> <span class="fu">mget</span>(kimberley_names)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="co">#we're going to do this in a loop for speed, applying the same filters</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Iterate over the names of the kimberley list</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(name <span class="cf">in</span> <span class="fu">names</span>(kimberley)){</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Extract the genlight object from the kimberley list using its name</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>  genlight_object <span class="ot">&lt;-</span> kimberley[[name]]</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Apply the filter call rate function</span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Assuming gl.filter.callrate is a function that operates on a genlight object</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>  filtered_object <span class="ot">&lt;-</span> <span class="fu">gl.filter.callrate</span>(genlight_object, <span class="at">threshold =</span> <span class="fl">0.7</span>, <span class="at">mono.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Assign the filtered object back to the environment with a new name</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">assign</span>(<span class="fu">paste0</span>(name, <span class="st">"_0.7"</span>), filtered_object)</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: data include loci that are scored NA across all individuals.
  Consider filtering using gl &lt;- gl.filter.allna(gl)
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-2-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: data include loci that are scored NA across all individuals.
  Consider filtering using gl &lt;- gl.filter.allna(gl)
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-2-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-2-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-2-4.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-2-5.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-2-6.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-2-7.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-2-8.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate </code></pre>
</div>
</div>
</section>
<section id="calculate-heterozygosity" class="level3">
<h3 class="anchored" data-anchor-id="calculate-heterozygosity">Calculate heterozygosity</h3>
<div class="cell" data-hash="session03_cache/html/unnamed-chunk-3_3d86f01cb5b46dd96cd6078e4985ba75">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># List all object names in the environment</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>all_names <span class="ot">&lt;-</span> <span class="fu">ls</span>()</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Use grep() to match names that start with "Kimberley" </span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="co"># and end with "0.7", .+ indicates any characters in between</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>kimberley_filtered <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">"^Kimberley.+0</span><span class="sc">\\</span><span class="st">.7$"</span>, all_names, <span class="at">value =</span> <span class="cn">TRUE</span>)<span class="co">#put all the genlights into a mega list</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="co">#create another list with the ones we want</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>kimberley <span class="ot">&lt;-</span> <span class="fu">mget</span>(kimberley_filtered)</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a><span class="co">#Initialize an empty data frame</span></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>heterozygosity_reports_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Iterate over the kimberley list to apply gl.report.heterozygosity</span></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a><span class="co"># and bind the results</span></span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(name <span class="cf">in</span> <span class="fu">names</span>(kimberley)) {</span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Apply the function</span></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>  report <span class="ot">&lt;-</span> <span class="fu">gl.report.heterozygosity</span>(kimberley[[name]])</span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add 'ObjectName' as the first column of the report</span></span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>  report <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="at">ObjectName =</span> name, report)</span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Bind this report to the main data frame</span></span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>  heterozygosity_reports_df <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(heterozygosity_reports_df, report)</span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 4.327947 28465         1   28465       0         0 0.233986 0.190371
         HoSE       He     HeSD     HeSE      uHe    uHeSD    uHeSE      FIS
pop1 0.001128 0.329375 0.112778 0.000668 0.372398 0.127509 0.000756 0.302454
        FISSD    FISSE
pop1 0.469498 0.002783
Completed: gl.report.heterozygosity 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-3-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 8.113689 47375         1   47375       0         0 0.174677 0.144488
         HoSE       He     HeSD    HeSE      uHe    uHeSD    uHeSE      FIS
pop1 0.000664 0.256162 0.135015 0.00062 0.272985 0.143882 0.000661 0.279239
        FISSD    FISSE
pop1 0.419149 0.001926
Completed: gl.report.heterozygosity 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-3-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc      Ho     HoSD
pop1 pop1 12.36041 35945         1   35945       0         0 0.15375 0.130734
        HoSE       He    HeSD     HeSE      uHe    uHeSD   uHeSE      FIS
pop1 0.00069 0.218107 0.14199 0.000749 0.227301 0.147975 0.00078 0.244173
        FISSD    FISSE
pop1 0.379656 0.002002
Completed: gl.report.heterozygosity 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-3-4.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 16.03257 46276         1   46276       0         0 0.136788 0.122952
         HoSE       He     HeSD     HeSE      uHe    uHeSD    uHeSE     FIS
pop1 0.000572 0.196149 0.143623 0.000668 0.202464 0.148246 0.000689 0.24159
        FISSD    FISSE
pop1 0.366035 0.001702
Completed: gl.report.heterozygosity 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-3-5.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 20.24648 39573         1   39573       0         0 0.128296 0.118505
         HoSE       He     HeSD     HeSE      uHe    uHeSD    uHeSE      FIS
pop1 0.000596 0.180369 0.144139 0.000725 0.184936 0.147788 0.000743 0.224197
        FISSD    FISSE
pop1 0.345555 0.001737
Completed: gl.report.heterozygosity 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-3-6.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 23.86971 44532         1   44532       0         0 0.122023 0.115572
         HoSE       He     HeSD     HeSE      uHe    uHeSD    uHeSE      FIS
pop1 0.000548 0.172406 0.144272 0.000684 0.176095 0.147359 0.000698 0.225027
        FISSD    FISSE
pop1 0.336198 0.001593
Completed: gl.report.heterozygosity 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-3-7.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 28.28105 46273         1   46273       0         0 0.114091 0.112716
         HoSE       He     HeSD     HeSE      uHe    uHeSD    uHeSE      FIS
pop1 0.000524 0.160215 0.143001 0.000665 0.163099 0.145574 0.000677 0.217536
        FISSD    FISSE
pop1 0.325623 0.001514
Completed: gl.report.heterozygosity 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-3-8.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 31.86356 49487         1   49487       0         0 0.110686 0.111441
         HoSE       He     HeSD     HeSE      uHe    uHeSD    uHeSE      FIS
pop1 0.000501 0.156579 0.143352 0.000644 0.159075 0.145637 0.000655 0.221427
       FISSD    FISSE
pop1 0.32204 0.001448
Completed: gl.report.heterozygosity </code></pre>
</div>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># heterozygosity_reports_df now contains all the reports with an </span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="co"># additional column for object names</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(heterozygosity_reports_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 4%">
<col style="width: 11%">
<col style="width: 2%">
<col style="width: 5%">
<col style="width: 3%">
<col style="width: 5%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 5%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">ObjectName</th>
<th style="text-align: left;">pop</th>
<th style="text-align: right;">n.Ind</th>
<th style="text-align: right;">n.Loc</th>
<th style="text-align: right;">n.Loc.adj</th>
<th style="text-align: right;">polyLoc</th>
<th style="text-align: right;">monoLoc</th>
<th style="text-align: right;">all_NALoc</th>
<th style="text-align: right;">Ho</th>
<th style="text-align: right;">HoSD</th>
<th style="text-align: right;">HoSE</th>
<th style="text-align: right;">He</th>
<th style="text-align: right;">HeSD</th>
<th style="text-align: right;">HeSE</th>
<th style="text-align: right;">uHe</th>
<th style="text-align: right;">uHeSD</th>
<th style="text-align: right;">uHeSE</th>
<th style="text-align: right;">FIS</th>
<th style="text-align: right;">FISSD</th>
<th style="text-align: right;">FISSE</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">pop1…1</td>
<td style="text-align: left;">Kimberley_n_05.vcf_0.7</td>
<td style="text-align: left;">pop1</td>
<td style="text-align: right;">4.327947</td>
<td style="text-align: right;">28465</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">28465</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.233986</td>
<td style="text-align: right;">0.190371</td>
<td style="text-align: right;">0.001128</td>
<td style="text-align: right;">0.329375</td>
<td style="text-align: right;">0.112778</td>
<td style="text-align: right;">0.000668</td>
<td style="text-align: right;">0.372398</td>
<td style="text-align: right;">0.127509</td>
<td style="text-align: right;">0.000756</td>
<td style="text-align: right;">0.302454</td>
<td style="text-align: right;">0.469498</td>
<td style="text-align: right;">0.002783</td>
</tr>
<tr class="even">
<td style="text-align: left;">pop1…2</td>
<td style="text-align: left;">Kimberley_n_10.vcf_0.7</td>
<td style="text-align: left;">pop1</td>
<td style="text-align: right;">8.113689</td>
<td style="text-align: right;">47375</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">47375</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.174677</td>
<td style="text-align: right;">0.144488</td>
<td style="text-align: right;">0.000664</td>
<td style="text-align: right;">0.256162</td>
<td style="text-align: right;">0.135015</td>
<td style="text-align: right;">0.000620</td>
<td style="text-align: right;">0.272985</td>
<td style="text-align: right;">0.143882</td>
<td style="text-align: right;">0.000661</td>
<td style="text-align: right;">0.279239</td>
<td style="text-align: right;">0.419149</td>
<td style="text-align: right;">0.001926</td>
</tr>
<tr class="odd">
<td style="text-align: left;">pop1…3</td>
<td style="text-align: left;">Kimberley_n_15.vcf_0.7</td>
<td style="text-align: left;">pop1</td>
<td style="text-align: right;">12.360412</td>
<td style="text-align: right;">35945</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">35945</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.153750</td>
<td style="text-align: right;">0.130734</td>
<td style="text-align: right;">0.000690</td>
<td style="text-align: right;">0.218107</td>
<td style="text-align: right;">0.141990</td>
<td style="text-align: right;">0.000749</td>
<td style="text-align: right;">0.227301</td>
<td style="text-align: right;">0.147975</td>
<td style="text-align: right;">0.000780</td>
<td style="text-align: right;">0.244173</td>
<td style="text-align: right;">0.379656</td>
<td style="text-align: right;">0.002002</td>
</tr>
<tr class="even">
<td style="text-align: left;">pop1…4</td>
<td style="text-align: left;">Kimberley_n_20.vcf_0.7</td>
<td style="text-align: left;">pop1</td>
<td style="text-align: right;">16.032565</td>
<td style="text-align: right;">46276</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">46276</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.136788</td>
<td style="text-align: right;">0.122952</td>
<td style="text-align: right;">0.000572</td>
<td style="text-align: right;">0.196149</td>
<td style="text-align: right;">0.143623</td>
<td style="text-align: right;">0.000668</td>
<td style="text-align: right;">0.202464</td>
<td style="text-align: right;">0.148246</td>
<td style="text-align: right;">0.000689</td>
<td style="text-align: right;">0.241590</td>
<td style="text-align: right;">0.366035</td>
<td style="text-align: right;">0.001702</td>
</tr>
<tr class="odd">
<td style="text-align: left;">pop1…5</td>
<td style="text-align: left;">Kimberley_n_25.vcf_0.7</td>
<td style="text-align: left;">pop1</td>
<td style="text-align: right;">20.246481</td>
<td style="text-align: right;">39573</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">39573</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.128296</td>
<td style="text-align: right;">0.118505</td>
<td style="text-align: right;">0.000596</td>
<td style="text-align: right;">0.180369</td>
<td style="text-align: right;">0.144139</td>
<td style="text-align: right;">0.000725</td>
<td style="text-align: right;">0.184936</td>
<td style="text-align: right;">0.147788</td>
<td style="text-align: right;">0.000743</td>
<td style="text-align: right;">0.224197</td>
<td style="text-align: right;">0.345555</td>
<td style="text-align: right;">0.001737</td>
</tr>
<tr class="even">
<td style="text-align: left;">pop1…6</td>
<td style="text-align: left;">Kimberley_n_30.vcf_0.7</td>
<td style="text-align: left;">pop1</td>
<td style="text-align: right;">23.869712</td>
<td style="text-align: right;">44532</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">44532</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.122023</td>
<td style="text-align: right;">0.115572</td>
<td style="text-align: right;">0.000548</td>
<td style="text-align: right;">0.172406</td>
<td style="text-align: right;">0.144272</td>
<td style="text-align: right;">0.000684</td>
<td style="text-align: right;">0.176095</td>
<td style="text-align: right;">0.147359</td>
<td style="text-align: right;">0.000698</td>
<td style="text-align: right;">0.225027</td>
<td style="text-align: right;">0.336198</td>
<td style="text-align: right;">0.001593</td>
</tr>
<tr class="odd">
<td style="text-align: left;">pop1…7</td>
<td style="text-align: left;">Kimberley_n_35.vcf_0.7</td>
<td style="text-align: left;">pop1</td>
<td style="text-align: right;">28.281049</td>
<td style="text-align: right;">46273</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">46273</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.114091</td>
<td style="text-align: right;">0.112716</td>
<td style="text-align: right;">0.000524</td>
<td style="text-align: right;">0.160215</td>
<td style="text-align: right;">0.143001</td>
<td style="text-align: right;">0.000665</td>
<td style="text-align: right;">0.163099</td>
<td style="text-align: right;">0.145574</td>
<td style="text-align: right;">0.000677</td>
<td style="text-align: right;">0.217536</td>
<td style="text-align: right;">0.325623</td>
<td style="text-align: right;">0.001514</td>
</tr>
<tr class="even">
<td style="text-align: left;">pop1…8</td>
<td style="text-align: left;">Kimberley_n_40.vcf_0.7</td>
<td style="text-align: left;">pop1</td>
<td style="text-align: right;">31.863560</td>
<td style="text-align: right;">49487</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">49487</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.110686</td>
<td style="text-align: right;">0.111441</td>
<td style="text-align: right;">0.000501</td>
<td style="text-align: right;">0.156579</td>
<td style="text-align: right;">0.143352</td>
<td style="text-align: right;">0.000644</td>
<td style="text-align: right;">0.159075</td>
<td style="text-align: right;">0.145637</td>
<td style="text-align: right;">0.000655</td>
<td style="text-align: right;">0.221427</td>
<td style="text-align: right;">0.322040</td>
<td style="text-align: right;">0.001448</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="plotting-results" class="level3">
<h3 class="anchored" data-anchor-id="plotting-results">Plotting results</h3>
<div class="cell" data-hash="session03_cache/html/unnamed-chunk-4_36a006897766bfcf045b3e1d564318a5">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>kimberley_Ho_0<span class="fl">.7</span>callrate <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(heterozygosity_reports_df, <span class="fu">aes</span>(<span class="at">x =</span> ObjectName, <span class="at">y =</span> Ho)) <span class="sc">+</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="cn">NA</span>)) <span class="sc">+</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">65</span>, <span class="at">hjust =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Observed Heterozygosity by Sample number"</span>, <span class="at">x =</span> <span class="st">"Sample Number"</span>, <span class="at">y =</span> <span class="st">"Observed Heterozygosity (Ho) at 0.7 Call Rate"</span>)</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>kimberley_Ho_0<span class="fl">.7</span>callrate</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Sampling
</div>
</div>
<div class="callout-body-container callout-body">
<p>As you can see, different numbers of samples can substantially change your heterozygosity estimate.</p>
</div>
</div>
</section>
</section>
<section id="estimating-heterozygosity-95-callrate" class="level2">
<h2 class="anchored" data-anchor-id="estimating-heterozygosity-95-callrate">2. Estimating heterozygosity: 95% callrate</h2>
<p><em>Lets redo our heterozygosity estimates, but see how changing your call rate filter impacts estimates</em></p>
<section id="reload-data" class="level3">
<h3 class="anchored" data-anchor-id="reload-data">Reload data</h3>
<div class="cell" data-hash="session03_cache/html/unnamed-chunk-5_e8da821f9ebac8d99cafde431c1d4b03">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># reload data</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">'./data/Session3_data.RData'</span>)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="co"># List all object names in the environment</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>all_names <span class="ot">&lt;-</span> <span class="fu">ls</span>()</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Use grep() to match names that start with "Kimberley" </span></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>kimberley_names <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">"^Kimberley.*</span><span class="sc">\\</span><span class="st">.vcf$"</span>, all_names, <span class="at">value =</span> <span class="cn">TRUE</span>) <span class="co">#put all the genlights into a mega list</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a><span class="co">#create another list with the ones we want</span></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>kimberley <span class="ot">&lt;-</span> <span class="fu">mget</span>(kimberley_names)</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a><span class="co">#we're going to do this in a loop for speed, applying the same filters</span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Iterate over the names of the kimberley list</span></span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(name <span class="cf">in</span> <span class="fu">names</span>(kimberley)){</span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Extract the genlight object from the kimberley list using its name</span></span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>  genlight_object <span class="ot">&lt;-</span> kimberley[[name]]</span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Apply the filter call rate function</span></span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Assuming gl.filter.callrate is a function that operates on a genlight object</span></span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a>   filtered_object <span class="ot">&lt;-</span> <span class="fu">gl.filter.callrate</span>(genlight_object, <span class="at">threshold =</span> .<span class="dv">95</span>, <span class="at">mono.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Assign the filtered object back to the environment with a new name</span></span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">assign</span>(<span class="fu">paste0</span>(name, <span class="st">"_0.95"</span>), filtered_object)</span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: data include loci that are scored NA across all individuals.
  Consider filtering using gl &lt;- gl.filter.allna(gl)
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.95 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: data include loci that are scored NA across all individuals.
  Consider filtering using gl &lt;- gl.filter.allna(gl)
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.95 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-5-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.95 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-5-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.95 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-5-4.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.95 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-5-5.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.95 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-5-6.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.95 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-5-7.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.95 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-5-8.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate </code></pre>
</div>
</div>
</section>
<section id="calculate-heterozygosity-1" class="level3">
<h3 class="anchored" data-anchor-id="calculate-heterozygosity-1">Calculate heterozygosity</h3>
<div class="cell" data-hash="session03_cache/html/unnamed-chunk-6_a9babc2e76974d332e06686e8b11cac8">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># List all object names in the environment</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>all_names <span class="ot">&lt;-</span> <span class="fu">ls</span>()</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Use grep() to match names that start with "Kimberley" </span></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a><span class="co"># and end with "0.95", .+ indicates any characters in between</span></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>kimberley_filtered <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">"^Kimberley.+0</span><span class="sc">\\</span><span class="st">.95$"</span>, all_names, <span class="at">value =</span> <span class="cn">TRUE</span>)<span class="co">#put all the genlights into a mega list</span></span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a><span class="co">#create another list with the ones we want</span></span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>kimberley <span class="ot">&lt;-</span> <span class="fu">mget</span>(kimberley_filtered)</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a><span class="co">#Initialize an empty data frame</span></span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>heterozygosity_reports_df_0<span class="fl">.95</span> <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Iterate over the kimberley list to apply gl.report.heterozygosity</span></span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a><span class="co"># and bind the results</span></span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(name <span class="cf">in</span> <span class="fu">names</span>(kimberley)) {</span>
<span id="cb34-17"><a href="#cb34-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Apply the function</span></span>
<span id="cb34-18"><a href="#cb34-18" aria-hidden="true" tabindex="-1"></a>  report <span class="ot">&lt;-</span> <span class="fu">gl.report.heterozygosity</span>(kimberley[[name]])</span>
<span id="cb34-19"><a href="#cb34-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb34-20"><a href="#cb34-20" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add 'ObjectName' as the first column of the report</span></span>
<span id="cb34-21"><a href="#cb34-21" aria-hidden="true" tabindex="-1"></a>  report <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="at">ObjectName =</span> name, report)</span>
<span id="cb34-22"><a href="#cb34-22" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb34-23"><a href="#cb34-23" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Bind this report to the main data frame</span></span>
<span id="cb34-24"><a href="#cb34-24" aria-hidden="true" tabindex="-1"></a>  heterozygosity_reports_df_0<span class="fl">.95</span> <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(heterozygosity_reports_df_0<span class="fl">.95</span>, report)</span>
<span id="cb34-25"><a href="#cb34-25" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1     5  9335         1    9335       0         0 0.229502 0.173222
         HoSE       He     HeSD    HeSE      uHe    uHeSD    uHeSE      FIS
pop1 0.001793 0.307878 0.119766 0.00124 0.342087 0.133073 0.001377 0.261799
        FISSD    FISSE
pop1 0.439324 0.004547
Completed: gl.report.heterozygosity 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-6-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1    10  5814         1    5814       0         0 0.176402 0.137187
         HoSE       He     HeSD    HeSE      uHe    uHeSD    uHeSE      FIS
pop1 0.001799 0.228789 0.139511 0.00183 0.240831 0.146854 0.001926 0.192591
        FISSD    FISSE
pop1 0.359841 0.004719
Completed: gl.report.heterozygosity 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-6-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1    15  2003         1    2003       0         0 0.152804 0.122813
         HoSE       He     HeSD     HeSE      uHe   uHeSD    uHeSE      FIS
pop1 0.002744 0.196298 0.145174 0.003244 0.203067 0.15018 0.003356 0.161721
        FISSD    FISSE
pop1 0.315393 0.007047
Completed: gl.report.heterozygosity 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-6-4.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 19.27864  4199         1    4199       0         0 0.140333 0.124561
         HoSE       He     HeSD     HeSE      uHe    uHeSD    uHeSE      FIS
pop1 0.001922 0.177053 0.144114 0.002224 0.181767 0.147951 0.002283 0.153933
        FISSD    FISSE
pop1 0.306136 0.004724
Completed: gl.report.heterozygosity 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-6-5.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop   n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 24.2354  2243         1    2243       0         0 0.129226 0.119719
         HoSE       He     HeSD     HeSE      uHe    uHeSD    uHeSE      FIS
pop1 0.002528 0.162692 0.145793 0.003078 0.166119 0.148864 0.003143 0.135561
        FISSD    FISSE
pop1 0.276653 0.005841
Completed: gl.report.heterozygosity 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-6-6.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 29.20428  1356         1    1356       0         0 0.117792 0.114525
        HoSE       He     HeSD     HeSE      uHe    uHeSD  uHeSE      FIS
pop1 0.00311 0.147605 0.141139 0.003833 0.150176 0.143597 0.0039 0.129329
        FISSD    FISSE
pop1 0.267582 0.007267
Completed: gl.report.heterozygosity 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-6-7.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 34.21357  1105         1    1105       0         0 0.111003 0.113222
         HoSE       He     HeSD     HeSE      uHe    uHeSD    uHeSE      FIS
pop1 0.003406 0.136806 0.139159 0.004186 0.138835 0.141223 0.004248 0.112563
        FISSD    FISSE
pop1 0.246423 0.007413
Completed: gl.report.heterozygosity 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-6-8.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 38.42546  1838         1    1838       0         0 0.110599 0.115467
         HoSE       He     HeSD     HeSE      uHe    uHeSD    uHeSE      FIS
pop1 0.002693 0.137218 0.141699 0.003305 0.139027 0.143567 0.003349 0.114931
        FISSD    FISSE
pop1 0.246381 0.005747
Completed: gl.report.heterozygosity </code></pre>
</div>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># heterozygosity_reports_df now contains all the reports with an </span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="co"># additional column for object names</span></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(heterozygosity_reports_df_0<span class="fl">.95</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 4%">
<col style="width: 12%">
<col style="width: 2%">
<col style="width: 4%">
<col style="width: 3%">
<col style="width: 5%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 5%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
<col style="width: 4%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">ObjectName</th>
<th style="text-align: left;">pop</th>
<th style="text-align: right;">n.Ind</th>
<th style="text-align: right;">n.Loc</th>
<th style="text-align: right;">n.Loc.adj</th>
<th style="text-align: right;">polyLoc</th>
<th style="text-align: right;">monoLoc</th>
<th style="text-align: right;">all_NALoc</th>
<th style="text-align: right;">Ho</th>
<th style="text-align: right;">HoSD</th>
<th style="text-align: right;">HoSE</th>
<th style="text-align: right;">He</th>
<th style="text-align: right;">HeSD</th>
<th style="text-align: right;">HeSE</th>
<th style="text-align: right;">uHe</th>
<th style="text-align: right;">uHeSD</th>
<th style="text-align: right;">uHeSE</th>
<th style="text-align: right;">FIS</th>
<th style="text-align: right;">FISSD</th>
<th style="text-align: right;">FISSE</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">pop1…1</td>
<td style="text-align: left;">Kimberley_n_05.vcf_0.95</td>
<td style="text-align: left;">pop1</td>
<td style="text-align: right;">5.00000</td>
<td style="text-align: right;">9335</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">9335</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.229502</td>
<td style="text-align: right;">0.173222</td>
<td style="text-align: right;">0.001793</td>
<td style="text-align: right;">0.307878</td>
<td style="text-align: right;">0.119766</td>
<td style="text-align: right;">0.001240</td>
<td style="text-align: right;">0.342087</td>
<td style="text-align: right;">0.133073</td>
<td style="text-align: right;">0.001377</td>
<td style="text-align: right;">0.261799</td>
<td style="text-align: right;">0.439324</td>
<td style="text-align: right;">0.004547</td>
</tr>
<tr class="even">
<td style="text-align: left;">pop1…2</td>
<td style="text-align: left;">Kimberley_n_10.vcf_0.95</td>
<td style="text-align: left;">pop1</td>
<td style="text-align: right;">10.00000</td>
<td style="text-align: right;">5814</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">5814</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.176402</td>
<td style="text-align: right;">0.137187</td>
<td style="text-align: right;">0.001799</td>
<td style="text-align: right;">0.228789</td>
<td style="text-align: right;">0.139511</td>
<td style="text-align: right;">0.001830</td>
<td style="text-align: right;">0.240831</td>
<td style="text-align: right;">0.146854</td>
<td style="text-align: right;">0.001926</td>
<td style="text-align: right;">0.192591</td>
<td style="text-align: right;">0.359841</td>
<td style="text-align: right;">0.004719</td>
</tr>
<tr class="odd">
<td style="text-align: left;">pop1…3</td>
<td style="text-align: left;">Kimberley_n_15.vcf_0.95</td>
<td style="text-align: left;">pop1</td>
<td style="text-align: right;">15.00000</td>
<td style="text-align: right;">2003</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2003</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.152804</td>
<td style="text-align: right;">0.122813</td>
<td style="text-align: right;">0.002744</td>
<td style="text-align: right;">0.196298</td>
<td style="text-align: right;">0.145174</td>
<td style="text-align: right;">0.003244</td>
<td style="text-align: right;">0.203067</td>
<td style="text-align: right;">0.150180</td>
<td style="text-align: right;">0.003356</td>
<td style="text-align: right;">0.161721</td>
<td style="text-align: right;">0.315393</td>
<td style="text-align: right;">0.007047</td>
</tr>
<tr class="even">
<td style="text-align: left;">pop1…4</td>
<td style="text-align: left;">Kimberley_n_20.vcf_0.95</td>
<td style="text-align: left;">pop1</td>
<td style="text-align: right;">19.27864</td>
<td style="text-align: right;">4199</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">4199</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.140333</td>
<td style="text-align: right;">0.124561</td>
<td style="text-align: right;">0.001922</td>
<td style="text-align: right;">0.177053</td>
<td style="text-align: right;">0.144114</td>
<td style="text-align: right;">0.002224</td>
<td style="text-align: right;">0.181767</td>
<td style="text-align: right;">0.147951</td>
<td style="text-align: right;">0.002283</td>
<td style="text-align: right;">0.153933</td>
<td style="text-align: right;">0.306136</td>
<td style="text-align: right;">0.004724</td>
</tr>
<tr class="odd">
<td style="text-align: left;">pop1…5</td>
<td style="text-align: left;">Kimberley_n_25.vcf_0.95</td>
<td style="text-align: left;">pop1</td>
<td style="text-align: right;">24.23540</td>
<td style="text-align: right;">2243</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">2243</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.129226</td>
<td style="text-align: right;">0.119719</td>
<td style="text-align: right;">0.002528</td>
<td style="text-align: right;">0.162692</td>
<td style="text-align: right;">0.145793</td>
<td style="text-align: right;">0.003078</td>
<td style="text-align: right;">0.166119</td>
<td style="text-align: right;">0.148864</td>
<td style="text-align: right;">0.003143</td>
<td style="text-align: right;">0.135561</td>
<td style="text-align: right;">0.276653</td>
<td style="text-align: right;">0.005841</td>
</tr>
<tr class="even">
<td style="text-align: left;">pop1…6</td>
<td style="text-align: left;">Kimberley_n_30.vcf_0.95</td>
<td style="text-align: left;">pop1</td>
<td style="text-align: right;">29.20428</td>
<td style="text-align: right;">1356</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1356</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.117792</td>
<td style="text-align: right;">0.114525</td>
<td style="text-align: right;">0.003110</td>
<td style="text-align: right;">0.147605</td>
<td style="text-align: right;">0.141139</td>
<td style="text-align: right;">0.003833</td>
<td style="text-align: right;">0.150176</td>
<td style="text-align: right;">0.143597</td>
<td style="text-align: right;">0.003900</td>
<td style="text-align: right;">0.129329</td>
<td style="text-align: right;">0.267582</td>
<td style="text-align: right;">0.007267</td>
</tr>
<tr class="odd">
<td style="text-align: left;">pop1…7</td>
<td style="text-align: left;">Kimberley_n_35.vcf_0.95</td>
<td style="text-align: left;">pop1</td>
<td style="text-align: right;">34.21357</td>
<td style="text-align: right;">1105</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1105</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.111003</td>
<td style="text-align: right;">0.113222</td>
<td style="text-align: right;">0.003406</td>
<td style="text-align: right;">0.136806</td>
<td style="text-align: right;">0.139159</td>
<td style="text-align: right;">0.004186</td>
<td style="text-align: right;">0.138835</td>
<td style="text-align: right;">0.141223</td>
<td style="text-align: right;">0.004248</td>
<td style="text-align: right;">0.112563</td>
<td style="text-align: right;">0.246423</td>
<td style="text-align: right;">0.007413</td>
</tr>
<tr class="even">
<td style="text-align: left;">pop1…8</td>
<td style="text-align: left;">Kimberley_n_40.vcf_0.95</td>
<td style="text-align: left;">pop1</td>
<td style="text-align: right;">38.42546</td>
<td style="text-align: right;">1838</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1838</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.110599</td>
<td style="text-align: right;">0.115467</td>
<td style="text-align: right;">0.002693</td>
<td style="text-align: right;">0.137218</td>
<td style="text-align: right;">0.141699</td>
<td style="text-align: right;">0.003305</td>
<td style="text-align: right;">0.139027</td>
<td style="text-align: right;">0.143567</td>
<td style="text-align: right;">0.003349</td>
<td style="text-align: right;">0.114931</td>
<td style="text-align: right;">0.246381</td>
<td style="text-align: right;">0.005747</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="plotting-results-1" class="level3">
<h3 class="anchored" data-anchor-id="plotting-results-1">Plotting results</h3>
<div class="cell" data-hash="session03_cache/html/unnamed-chunk-7_586c37063cd34e3345fd8b7ec4de144a">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Example using ggplot2 to plot the data</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>kimberley_Ho_0<span class="fl">.95</span>callrate <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(heterozygosity_reports_df_0<span class="fl">.95</span>, <span class="fu">aes</span>(<span class="at">x =</span> ObjectName, <span class="at">y =</span> Ho)) <span class="sc">+</span></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="cn">NA</span>)) <span class="sc">+</span></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">65</span>, <span class="at">hjust =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Observed Heterozygosity by Sample number 0.95 Call Rate"</span>,</span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Sample Number"</span>, <span class="at">y =</span> <span class="st">"Observed Heterozygosity (Ho) at 0.95 Call Rate"</span>)</span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>kimberley_Ho_0<span class="fl">.95</span>callrate</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">1</span>))</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>kimberley_Ho_0<span class="fl">.7</span>callrate <span class="sc">+</span>kimberley_Ho_0<span class="fl">.95</span>callrate</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-7-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Call rate filters
</div>
</div>
<div class="callout-body-container callout-body">
<p>Higher call rate filters can reduce your slightly reduce your heterozygosity estimate, but the major issue is sample number.</p>
</div>
</div>
</section>
</section>
<section id="estimating-heterozygosity-even-sampling" class="level2">
<h2 class="anchored" data-anchor-id="estimating-heterozygosity-even-sampling">3. Estimating heterozygosity: Even sampling</h2>
<p><em>does subsetting to have the same sample numbers fix the issue?</em></p>
<p>Let’s compare heterozygosity when SNPs are called on 5 individuals, versus called on 40 individuals and then subset to 5 individuals.</p>
<p>If you called SNPs on more individuals than you wanted to make them equal, can you just remove individuals without re-calling SNPS? Let’s test.</p>
<section id="subsampling" class="level3">
<h3 class="anchored" data-anchor-id="subsampling">Subsampling</h3>
<div class="cell" data-hash="session03_cache/html/unnamed-chunk-8_b31bb93eb6396f86e79470ddaadb10fc">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Use grep() to match names that start with "Kimberley" </span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>kimberley_names <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">"^Kimberley.*</span><span class="sc">\\</span><span class="st">.vcf$"</span>, all_names, <span class="at">value =</span> <span class="cn">TRUE</span>) </span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a><span class="co"># put all the genlights into a mega list</span></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a><span class="co">#create another list with the ones we want</span></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>kimberley <span class="ot">&lt;-</span> <span class="fu">mget</span>(kimberley_names)</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Now lets subsample the datasets down to the same five individuals</span></span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a><span class="co"># so that the only difference is our SNP calling</span></span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a><span class="co"># who are the individuals</span></span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>inds <span class="ot">&lt;-</span> <span class="fu">indNames</span>(Kimberley_n_05.vcf_0<span class="fl">.7</span>)</span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a><span class="co">#Initialize an empty data frame</span></span>
<span id="cb47-14"><a href="#cb47-14" aria-hidden="true" tabindex="-1"></a>heterozygosity_when_subsampling <span class="ot">&lt;-</span> <span class="fu">data.frame</span>()</span>
<span id="cb47-15"><a href="#cb47-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-16"><a href="#cb47-16" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(name <span class="cf">in</span> <span class="fu">names</span>(kimberley)) {</span>
<span id="cb47-17"><a href="#cb47-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Access the genlight object from your list</span></span>
<span id="cb47-18"><a href="#cb47-18" aria-hidden="true" tabindex="-1"></a>  genlight_object <span class="ot">&lt;-</span> kimberley[[name]]</span>
<span id="cb47-19"><a href="#cb47-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb47-20"><a href="#cb47-20" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Subset the individuals</span></span>
<span id="cb47-21"><a href="#cb47-21" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">gl.keep.ind</span>(genlight_object, <span class="at">ind.list =</span> inds, <span class="at">mono.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb47-22"><a href="#cb47-22" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb47-23"><a href="#cb47-23" aria-hidden="true" tabindex="-1"></a>  <span class="co">#filter on call rate</span></span>
<span id="cb47-24"><a href="#cb47-24" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb47-25"><a href="#cb47-25" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">gl.filter.callrate</span>(x, <span class="at">threshold =</span> .<span class="dv">7</span>)</span>
<span id="cb47-26"><a href="#cb47-26" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb47-27"><a href="#cb47-27" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Apply the function</span></span>
<span id="cb47-28"><a href="#cb47-28" aria-hidden="true" tabindex="-1"></a>  report <span class="ot">&lt;-</span> <span class="fu">gl.report.heterozygosity</span>(x)</span>
<span id="cb47-29"><a href="#cb47-29" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb47-30"><a href="#cb47-30" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add 'ObjectName' as the first column of the report</span></span>
<span id="cb47-31"><a href="#cb47-31" aria-hidden="true" tabindex="-1"></a>  report<span class="sc">$</span>ObjectName <span class="ot">&lt;-</span> name </span>
<span id="cb47-32"><a href="#cb47-32" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Adjusting to add column without cbind to maintain data frame classes</span></span>
<span id="cb47-33"><a href="#cb47-33" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb47-34"><a href="#cb47-34" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Bind this report to the main data frame</span></span>
<span id="cb47-35"><a href="#cb47-35" aria-hidden="true" tabindex="-1"></a>  heterozygosity_when_subsampling <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(heterozygosity_when_subsampling,</span>
<span id="cb47-36"><a href="#cb47-36" aria-hidden="true" tabindex="-1"></a>                                               report)</span>
<span id="cb47-37"><a href="#cb47-37" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.keep.ind 
  Processing genlight object with SNP data
  Warning: data include loci that are scored NA across all individuals.
  Consider filtering using gl &lt;- gl.filter.allna(gl)
  Deleting all but the listed individuals Kimberley_ABTC102537_DLit20-5353_CE4Y2ANXX_1, Kimberley_ABTC11911_DLit20-5353_CE7VEANXX_1, Kimberley_ABTC86488_DLit20-5353_CECHMANXX_4, Kimberley_CCM7524_DLit20-5353_CECHMANXX_2, Kimberley_ABTC135731_DLit20-5353_CE7VEANXX_1 
  Deleting monomorphic loc
  Locus metrics not recalculated
Completed: gl.keep.ind 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-8-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 4.327947 28465         1   28465       0         0 0.233986 0.190371
         HoSE       He     HeSD     HeSE      uHe    uHeSD    uHeSE      FIS
pop1 0.001128 0.329375 0.112778 0.000668 0.372398 0.127509 0.000756 0.302454
        FISSD    FISSE
pop1 0.469498 0.002783
Completed: gl.report.heterozygosity 
Starting gl.keep.ind 
  Processing genlight object with SNP data
  Warning: data include loci that are scored NA across all individuals.
  Consider filtering using gl &lt;- gl.filter.allna(gl)
  Deleting all but the listed individuals Kimberley_ABTC102537_DLit20-5353_CE4Y2ANXX_1, Kimberley_ABTC11911_DLit20-5353_CE7VEANXX_1, Kimberley_ABTC86488_DLit20-5353_CECHMANXX_4, Kimberley_CCM7524_DLit20-5353_CECHMANXX_2, Kimberley_ABTC135731_DLit20-5353_CE7VEANXX_1 
  Deleting monomorphic loc
  Locus metrics not recalculated
Completed: gl.keep.ind 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-8-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-8-4.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 4.360384 25656         1   25656       0         0 0.236861 0.185631
         HoSE       He     HeSD     HeSE      uHe    uHeSD    uHeSE     FIS
pop1 0.001159 0.324224 0.113105 0.000706 0.366218 0.127754 0.000798 0.28369
        FISSD    FISSE
pop1 0.460341 0.002874
Completed: gl.report.heterozygosity 
Starting gl.keep.ind 
  Processing genlight object with SNP data
  Deleting all but the listed individuals Kimberley_ABTC102537_DLit20-5353_CE4Y2ANXX_1, Kimberley_ABTC11911_DLit20-5353_CE7VEANXX_1, Kimberley_ABTC86488_DLit20-5353_CECHMANXX_4, Kimberley_CCM7524_DLit20-5353_CECHMANXX_2, Kimberley_ABTC135731_DLit20-5353_CE7VEANXX_1 
  Deleting monomorphic loc
  Locus metrics not recalculated
Completed: gl.keep.ind 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-8-5.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-8-6.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 4.404846 21171         1   21171       0         0 0.241755 0.181273
         HoSE       He     HeSD     HeSE      uHe    uHeSD    uHeSE      FIS
pop1 0.001246 0.318786 0.113593 0.000781 0.359605 0.128138 0.000881 0.259282
       FISSD    FISSE
pop1 0.44866 0.003084
Completed: gl.report.heterozygosity 
Starting gl.keep.ind 
  Processing genlight object with SNP data
  Deleting all but the listed individuals Kimberley_ABTC102537_DLit20-5353_CE4Y2ANXX_1, Kimberley_ABTC11911_DLit20-5353_CE7VEANXX_1, Kimberley_ABTC86488_DLit20-5353_CECHMANXX_4, Kimberley_CCM7524_DLit20-5353_CECHMANXX_2, Kimberley_ABTC135731_DLit20-5353_CE7VEANXX_1 
  Deleting monomorphic loc
  Locus metrics not recalculated
Completed: gl.keep.ind 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-8-7.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-8-8.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 4.390371 22184         1   22184       0         0 0.241731 0.181534
         HoSE       He     HeSD     HeSE      uHe    uHeSD    uHeSE      FIS
pop1 0.001219 0.319324 0.113423 0.000762 0.360365 0.128001 0.000859 0.260639
        FISSD    FISSE
pop1 0.449285 0.003016
Completed: gl.report.heterozygosity 
Starting gl.keep.ind 
  Processing genlight object with SNP data
  Deleting all but the listed individuals Kimberley_ABTC102537_DLit20-5353_CE4Y2ANXX_1, Kimberley_ABTC11911_DLit20-5353_CE7VEANXX_1, Kimberley_ABTC86488_DLit20-5353_CECHMANXX_4, Kimberley_CCM7524_DLit20-5353_CECHMANXX_2, Kimberley_ABTC135731_DLit20-5353_CE7VEANXX_1 
  Deleting monomorphic loc
  Locus metrics not recalculated
Completed: gl.keep.ind 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-8-9.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-8-10.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 4.414982 19731         1   19731       0         0 0.243949 0.180217
         HoSE       He     HeSD     HeSE     uHe    uHeSD    uHeSE      FIS
pop1 0.001283 0.317031 0.113598 0.000809 0.35752 0.128106 0.000912 0.250234
        FISSD    FISSE
pop1 0.444447 0.003164
Completed: gl.report.heterozygosity 
Starting gl.keep.ind 
  Processing genlight object with SNP data
  Deleting all but the listed individuals Kimberley_ABTC102537_DLit20-5353_CE4Y2ANXX_1, Kimberley_ABTC11911_DLit20-5353_CE7VEANXX_1, Kimberley_ABTC86488_DLit20-5353_CECHMANXX_4, Kimberley_CCM7524_DLit20-5353_CECHMANXX_2, Kimberley_ABTC135731_DLit20-5353_CE7VEANXX_1 
  Deleting monomorphic loc
  Locus metrics not recalculated
Completed: gl.keep.ind 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-8-11.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-8-12.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 4.404714 20239         1   20239       0         0 0.242991 0.179167
         HoSE       He     HeSD     HeSE     uHe    uHeSD    uHeSE      FIS
pop1 0.001259 0.316909 0.113413 0.000797 0.35749 0.127936 0.000899 0.251896
        FISSD    FISSE
pop1 0.444278 0.003123
Completed: gl.report.heterozygosity 
Starting gl.keep.ind 
  Processing genlight object with SNP data
  Deleting all but the listed individuals Kimberley_ABTC102537_DLit20-5353_CE4Y2ANXX_1, Kimberley_ABTC11911_DLit20-5353_CE7VEANXX_1, Kimberley_ABTC86488_DLit20-5353_CECHMANXX_4, Kimberley_CCM7524_DLit20-5353_CECHMANXX_2, Kimberley_ABTC135731_DLit20-5353_CE7VEANXX_1 
  Deleting monomorphic loc
  Locus metrics not recalculated
Completed: gl.keep.ind 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-8-13.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-8-14.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 4.407437 19873         1   19873       0         0 0.244377 0.179303
         HoSE       He     HeSD     HeSE      uHe    uHeSD    uHeSE      FIS
pop1 0.001272 0.316308 0.113527 0.000805 0.356783 0.128054 0.000908 0.247269
       FISSD   FISSE
pop1 0.44271 0.00314
Completed: gl.report.heterozygosity 
Starting gl.keep.ind 
  Processing genlight object with SNP data
  Deleting all but the listed individuals Kimberley_ABTC102537_DLit20-5353_CE4Y2ANXX_1, Kimberley_ABTC11911_DLit20-5353_CE7VEANXX_1, Kimberley_ABTC86488_DLit20-5353_CECHMANXX_4, Kimberley_CCM7524_DLit20-5353_CECHMANXX_2, Kimberley_ABTC135731_DLit20-5353_CE7VEANXX_1 
  Deleting monomorphic loc
  Locus metrics not recalculated
Completed: gl.keep.ind 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-8-15.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-8-16.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 4.404922 20073         1   20073       0         0 0.244488 0.179823
         HoSE       He     HeSD     HeSE      uHe    uHeSD    uHeSE      FIS
pop1 0.001269 0.316304 0.113509 0.000801 0.356805 0.128043 0.000904 0.247144
        FISSD    FISSE
pop1 0.443154 0.003128
Completed: gl.report.heterozygosity </code></pre>
</div>
</div>
</section>
<section id="plotting-results-2" class="level3">
<h3 class="anchored" data-anchor-id="plotting-results-2">Plotting results</h3>
<div class="cell" data-hash="session03_cache/html/unnamed-chunk-9_83231383465074347e683241519e7372">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>kimberley_Ho_subsampling <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(heterozygosity_when_subsampling, <span class="fu">aes</span>(<span class="at">x =</span> ObjectName, <span class="at">y =</span> Ho)) <span class="sc">+</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="cn">NA</span>)) <span class="sc">+</span></span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">65</span>, <span class="at">hjust =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Observed Heterozygosity when subsampling"</span>, </span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"SNP Calling"</span>, <span class="at">y =</span> <span class="st">"Observed Heterozygosity (Ho)"</span>)</span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-8"><a href="#cb65-8" aria-hidden="true" tabindex="-1"></a>kimberley_Ho_subsampling</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
subsampling
</div>
</div>
<div class="callout-body-container callout-body">
<p>There are some minor differences but it’s not too bad so this may be okay in some circumstances.</p>
</div>
</div>
</section>
<section id="subsampling-more-populations" class="level3">
<h3 class="anchored" data-anchor-id="subsampling-more-populations">Subsampling more populations</h3>
<p><em>There appears to be minor differences between sample sizes for the Kimberley population, but is this true for all populations?</em></p>
<section id="southeast" class="level4">
<h4 class="anchored" data-anchor-id="southeast">Southeast</h4>
<div class="cell" data-hash="session03_cache/html/unnamed-chunk-10_dddc15a7bc13f6a0ffe1c17705841d7a">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="co">#create a list of the southeast genlights</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>Southeast_names <span class="ot">&lt;-</span> <span class="fu">ls</span>(<span class="at">pattern =</span> <span class="st">"^SouthEast"</span>)</span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a><span class="co">#put all the genlights into a mega list</span></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>southeast <span class="ot">&lt;-</span> <span class="fu">mget</span>(Southeast_names)</span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a><span class="co">#Now lets subsample the datasets down to the same five individuals</span></span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a><span class="co">#so that the only difference is our SNP calling</span></span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a><span class="co">#who are the individuals</span></span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a>inds <span class="ot">&lt;-</span> <span class="fu">indNames</span>(SouthEast_n_05.vcf)</span>
<span id="cb66-10"><a href="#cb66-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-11"><a href="#cb66-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-12"><a href="#cb66-12" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(name <span class="cf">in</span> <span class="fu">names</span>(southeast)) {</span>
<span id="cb66-13"><a href="#cb66-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Access the genlight object from your list</span></span>
<span id="cb66-14"><a href="#cb66-14" aria-hidden="true" tabindex="-1"></a>  genlight_object <span class="ot">&lt;-</span> southeast[[name]]</span>
<span id="cb66-15"><a href="#cb66-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb66-16"><a href="#cb66-16" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Subset the individuals</span></span>
<span id="cb66-17"><a href="#cb66-17" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">gl.keep.ind</span>(genlight_object, <span class="at">ind.list =</span> inds, <span class="at">mono.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb66-18"><a href="#cb66-18" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb66-19"><a href="#cb66-19" aria-hidden="true" tabindex="-1"></a>  <span class="co">#filter on call rate</span></span>
<span id="cb66-20"><a href="#cb66-20" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb66-21"><a href="#cb66-21" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">gl.filter.callrate</span>(x, <span class="at">threshold =</span> .<span class="dv">7</span>)</span>
<span id="cb66-22"><a href="#cb66-22" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb66-23"><a href="#cb66-23" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Apply the function</span></span>
<span id="cb66-24"><a href="#cb66-24" aria-hidden="true" tabindex="-1"></a>  report <span class="ot">&lt;-</span> <span class="fu">gl.report.heterozygosity</span>(x)</span>
<span id="cb66-25"><a href="#cb66-25" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb66-26"><a href="#cb66-26" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add 'ObjectName' as the first column of the report</span></span>
<span id="cb66-27"><a href="#cb66-27" aria-hidden="true" tabindex="-1"></a>  report<span class="sc">$</span>ObjectName <span class="ot">&lt;-</span> name <span class="co"># Adjusting to add column without cbind to maintain data frame classes</span></span>
<span id="cb66-28"><a href="#cb66-28" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb66-29"><a href="#cb66-29" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Bind this report to the main data frame</span></span>
<span id="cb66-30"><a href="#cb66-30" aria-hidden="true" tabindex="-1"></a>  heterozygosity_when_subsampling <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(heterozygosity_when_subsampling, report)</span>
<span id="cb66-31"><a href="#cb66-31" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.keep.ind 
  Processing genlight object with SNP data
  Warning: data include loci that are scored NA across all individuals.
  Consider filtering using gl &lt;- gl.filter.allna(gl)
  Deleting all but the listed individuals SouthEast_ABTC92145_DLit20-5353_CECHMANXX_3, SouthEast_ABTC112638_DLit20-5353_CE7VEANXX_1, SouthEast_AMR164735_DLit20-5353_CE7VEANXX_1, SouthEast_ABTC38934_DLit20-5353_CECHMANXX_3, SouthEast_ABTC111142_DLit20-5353_CECHMANXX_3 
  Deleting monomorphic loc
  Locus metrics not recalculated
Completed: gl.keep.ind 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-10-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 4.543063 23907         1   23907       0         0 0.306134 0.227794
         HoSE       He     HeSD     HeSE     uHe    uHeSD    uHeSE     FIS
pop1 0.001473 0.345947 0.118783 0.000768 0.38873 0.133473 0.000863 0.18011
        FISSD    FISSE
pop1 0.458069 0.002963
Completed: gl.report.heterozygosity 
Starting gl.keep.ind 
  Processing genlight object with SNP data
  Warning: data include loci that are scored NA across all individuals.
  Consider filtering using gl &lt;- gl.filter.allna(gl)
  Deleting all but the listed individuals SouthEast_ABTC92145_DLit20-5353_CECHMANXX_3, SouthEast_ABTC112638_DLit20-5353_CE7VEANXX_1, SouthEast_AMR164735_DLit20-5353_CE7VEANXX_1, SouthEast_ABTC38934_DLit20-5353_CECHMANXX_3, SouthEast_ABTC111142_DLit20-5353_CECHMANXX_3 
  Deleting monomorphic loc
  Locus metrics not recalculated
Completed: gl.keep.ind 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-10-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-10-4.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc      Ho    HoSD
pop1 pop1 4.560711 23167         1   23167       0         0 0.30276 0.21977
         HoSE       He     HeSD     HeSE      uHe    uHeSD    uHeSE      FIS
pop1 0.001444 0.340234 0.119545 0.000785 0.382127 0.134265 0.000882 0.172587
        FISSD    FISSE
pop1 0.445217 0.002925
Completed: gl.report.heterozygosity 
Starting gl.keep.ind 
  Processing genlight object with SNP data
  Warning: data include loci that are scored NA across all individuals.
  Consider filtering using gl &lt;- gl.filter.allna(gl)
  Deleting all but the listed individuals SouthEast_ABTC92145_DLit20-5353_CECHMANXX_3, SouthEast_ABTC112638_DLit20-5353_CE7VEANXX_1, SouthEast_AMR164735_DLit20-5353_CE7VEANXX_1, SouthEast_ABTC38934_DLit20-5353_CECHMANXX_3, SouthEast_ABTC111142_DLit20-5353_CECHMANXX_3 
  Deleting monomorphic loc
  Locus metrics not recalculated
Completed: gl.keep.ind 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-10-5.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-10-6.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 4.584442 21956         1   21956       0         0 0.300644 0.212385
         HoSE       He     HeSD    HeSE      uHe    uHeSD    uHeSE      FIS
pop1 0.001433 0.337159 0.119975 0.00081 0.378432 0.134662 0.000909 0.168427
        FISSD    FISSE
pop1 0.433861 0.002928
Completed: gl.report.heterozygosity 
Starting gl.keep.ind 
  Processing genlight object with SNP data
  Deleting all but the listed individuals SouthEast_ABTC92145_DLit20-5353_CECHMANXX_3, SouthEast_ABTC112638_DLit20-5353_CE7VEANXX_1, SouthEast_AMR164735_DLit20-5353_CE7VEANXX_1, SouthEast_ABTC38934_DLit20-5353_CECHMANXX_3, SouthEast_ABTC111142_DLit20-5353_CECHMANXX_3 
  Deleting monomorphic loc
  Locus metrics not recalculated
Completed: gl.keep.ind 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-10-7.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-10-8.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 4.575179 22513         1   22513       0         0 0.300346 0.212014
         HoSE       He     HeSD  HeSE     uHe    uHeSD    uHeSE      FIS
pop1 0.001413 0.336459 0.120052 8e-04 0.37774 0.134782 0.000898 0.167435
        FISSD    FISSE
pop1 0.432974 0.002886
Completed: gl.report.heterozygosity </code></pre>
</div>
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>southeast_Ho_subsampling <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(heterozygosity_when_subsampling, <span class="fu">aes</span>(<span class="at">x =</span> ObjectName, <span class="at">y =</span> Ho)) <span class="sc">+</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="cn">NA</span>)) <span class="sc">+</span></span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">65</span>, <span class="at">hjust =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Observed Heterozygosity when subsampling"</span>,</span>
<span id="cb76-6"><a href="#cb76-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"SNP Calling"</span>, <span class="at">y =</span> <span class="st">"Observed Heterozygosity (Ho)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="central" class="level4">
<h4 class="anchored" data-anchor-id="central">Central</h4>
<div class="cell" data-hash="session03_cache/html/unnamed-chunk-11_cf3f0fd1a5a15af8dfa34567b3a9c9f2">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co">#create a list of the central australian genlights</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>central_names <span class="ot">&lt;-</span> <span class="fu">ls</span>(<span class="at">pattern =</span> <span class="st">"^Central"</span>)</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a><span class="co">#put all the genlights into a mega list</span></span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>central <span class="ot">&lt;-</span> <span class="fu">mget</span>(central_names)</span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-6"><a href="#cb77-6" aria-hidden="true" tabindex="-1"></a><span class="co">#Now lets subsample the datasets down to the same five individuals</span></span>
<span id="cb77-7"><a href="#cb77-7" aria-hidden="true" tabindex="-1"></a><span class="co">#so that the only difference is our SNP calling</span></span>
<span id="cb77-8"><a href="#cb77-8" aria-hidden="true" tabindex="-1"></a><span class="co">#who are the individuals</span></span>
<span id="cb77-9"><a href="#cb77-9" aria-hidden="true" tabindex="-1"></a>inds <span class="ot">&lt;-</span> <span class="fu">indNames</span>(Central_n_05.vcf)</span>
<span id="cb77-10"><a href="#cb77-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-11"><a href="#cb77-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(name <span class="cf">in</span> <span class="fu">names</span>(central)) {</span>
<span id="cb77-12"><a href="#cb77-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Access the genlight object from your list</span></span>
<span id="cb77-13"><a href="#cb77-13" aria-hidden="true" tabindex="-1"></a>  genlight_object <span class="ot">&lt;-</span> central[[name]]</span>
<span id="cb77-14"><a href="#cb77-14" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb77-15"><a href="#cb77-15" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Subset the individuals</span></span>
<span id="cb77-16"><a href="#cb77-16" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">gl.keep.ind</span>(genlight_object, <span class="at">ind.list =</span> inds, <span class="at">mono.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb77-17"><a href="#cb77-17" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb77-18"><a href="#cb77-18" aria-hidden="true" tabindex="-1"></a>  <span class="co">#filter on call rate</span></span>
<span id="cb77-19"><a href="#cb77-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb77-20"><a href="#cb77-20" aria-hidden="true" tabindex="-1"></a> x <span class="ot">&lt;-</span> <span class="fu">gl.filter.callrate</span>(x, <span class="at">threshold =</span> .<span class="dv">7</span>)</span>
<span id="cb77-21"><a href="#cb77-21" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb77-22"><a href="#cb77-22" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Apply the function</span></span>
<span id="cb77-23"><a href="#cb77-23" aria-hidden="true" tabindex="-1"></a>  report <span class="ot">&lt;-</span> <span class="fu">gl.report.heterozygosity</span>(x)</span>
<span id="cb77-24"><a href="#cb77-24" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb77-25"><a href="#cb77-25" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add 'ObjectName' as the first column of the report</span></span>
<span id="cb77-26"><a href="#cb77-26" aria-hidden="true" tabindex="-1"></a>  report<span class="sc">$</span>ObjectName <span class="ot">&lt;-</span> name <span class="co"># Adjusting to add column without cbind to maintain data frame classes</span></span>
<span id="cb77-27"><a href="#cb77-27" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb77-28"><a href="#cb77-28" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Bind this report to the main data frame</span></span>
<span id="cb77-29"><a href="#cb77-29" aria-hidden="true" tabindex="-1"></a>  heterozygosity_when_subsampling <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(heterozygosity_when_subsampling, report)</span>
<span id="cb77-30"><a href="#cb77-30" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.keep.ind 
  Processing genlight object with SNP data
  Warning: data include loci that are scored NA across all individuals.
  Consider filtering using gl &lt;- gl.filter.allna(gl)
  Deleting all but the listed individuals Central_ABTC12035_DLit20-5353_CECHMANXX_3, Central_ABTC36027_DLit20-5353_CECHMANXX_4, Central_ABTC9964_DLit20-5353_CE4Y2ANXX_5, Central_ABTC30246_DLit20-5353_CECHMANXX_3, Central_ABTC58795_DLit20-5353_CE4Y2ANXX_5 
  Deleting monomorphic loc
  Locus metrics not recalculated
Completed: gl.keep.ind 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-11-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 4.591516 20745         1   20745       0         0 0.293408 0.265273
         HoSE      He     HeSD     HeSE      uHe    uHeSD    uHeSE      FIS
pop1 0.001842 0.35118 0.115651 0.000803 0.394095 0.129784 0.000901 0.246186
        FISSD    FISSE
pop1 0.526134 0.003653
Completed: gl.report.heterozygosity 
Starting gl.keep.ind 
  Processing genlight object with SNP data
  Warning: data include loci that are scored NA across all individuals.
  Consider filtering using gl &lt;- gl.filter.allna(gl)
  Deleting all but the listed individuals Central_ABTC12035_DLit20-5353_CECHMANXX_3, Central_ABTC36027_DLit20-5353_CECHMANXX_4, Central_ABTC9964_DLit20-5353_CE4Y2ANXX_5, Central_ABTC30246_DLit20-5353_CECHMANXX_3, Central_ABTC58795_DLit20-5353_CE4Y2ANXX_5 
  Deleting monomorphic loc
  Locus metrics not recalculated
Completed: gl.keep.ind 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-11-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-11-4.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 4.612141 20064         1   20064       0         0 0.281935 0.253619
        HoSE       He     HeSD    HeSE     uHe    uHeSD    uHeSE      FIS
pop1 0.00179 0.346597 0.116089 0.00082 0.38874 0.130204 0.000919 0.258105
        FISSD    FISSE
pop1 0.514939 0.003635
Completed: gl.report.heterozygosity 
Starting gl.keep.ind 
  Processing genlight object with SNP data
  Warning: data include loci that are scored NA across all individuals.
  Consider filtering using gl &lt;- gl.filter.allna(gl)
  Deleting all but the listed individuals Central_ABTC12035_DLit20-5353_CECHMANXX_3, Central_ABTC36027_DLit20-5353_CECHMANXX_4, Central_ABTC9964_DLit20-5353_CE4Y2ANXX_5, Central_ABTC30246_DLit20-5353_CECHMANXX_3, Central_ABTC58795_DLit20-5353_CE4Y2ANXX_5 
  Deleting monomorphic loc
  Locus metrics not recalculated
Completed: gl.keep.ind 
Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.7 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-11-5.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate 
Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-11-6.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      pop   n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho     HoSD
pop1 pop1 4.63602 19067         1   19067       0         0 0.276449 0.249131
         HoSE       He     HeSD     HeSE      uHe   uHeSD    uHeSE      FIS
pop1 0.001804 0.345069 0.116176 0.000841 0.386784 0.13022 0.000943 0.266283
        FISSD    FISSE
pop1 0.511497 0.003704
Completed: gl.report.heterozygosity </code></pre>
</div>
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>central_Ho_subsampling <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(heterozygosity_when_subsampling, <span class="fu">aes</span>(<span class="at">x =</span> ObjectName, <span class="at">y =</span> Ho)) <span class="sc">+</span></span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="cn">NA</span>)) <span class="sc">+</span></span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">65</span>, <span class="at">hjust =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb85-5"><a href="#cb85-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Observed Heterozygosity when subsampling 5 individuals"</span>,</span>
<span id="cb85-6"><a href="#cb85-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"SNP Calling"</span>, <span class="at">y =</span> <span class="st">"Observed Heterozygosity (Ho)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="plotting-results-3" class="level3">
<h3 class="anchored" data-anchor-id="plotting-results-3">Plotting Results</h3>
<div class="cell" data-hash="session03_cache/html/unnamed-chunk-12_c69c4e3dda6f6f8d033eb179c97a332f">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>southeast_Ho_subsampling</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>central_Ho_subsampling</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-12-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="callout callout-style-default callout-caution callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Subsampling
</div>
</div>
<div class="callout-body-container callout-body">
<p>So we’re not seeing too much impact of subsampling if SNPs are called separately on populations.</p>
<p><strong>BUT</strong>, BUT, BUT, BUT, BUT, BUT, BUT…</p>
<p>The Kimberley population has the highest actual heterozygosity (see the paper). The difference here is that this population lost a lot of variable sites because more than 5% (or &gt;35k) of the variable sites have more than two alleles. So a comparison between Central, Southeast and Kimberley Litoria rubella would find the highest heterozygosity in the wrong population.</p>
<p><strong>This is very problematic and is an ongoing bioinformatic issue.</strong></p>
<p>Therefore, calling SNPs separately on populations and keeping sample sizes the same does not resolve the issues with heterozygosity based on SNPs.</p>
</div>
</div>
</section>
</section>
<section id="calling-populations-together" class="level2">
<h2 class="anchored" data-anchor-id="calling-populations-together">Calling populations together</h2>
<p><em>Does this resolve the issue?</em></p>
<p>Lets look at a file where SNPs were called on all three populations at once, and the samples were equal numbers.</p>
<div class="cell" data-hash="session03_cache/html/unnamed-chunk-13_85337f8c2883b80794534391377cb0d8">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="co">#first we assign populations to the individuals in the genlight</span></span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>individual_names <span class="ot">&lt;-</span> <span class="fu">as.vector</span>(<span class="fu">indNames</span>(combined_kim_15_cen_15_se_15.vcf))</span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Split each name at the underscore and keep the first part</span></span>
<span id="cb88-5"><a href="#cb88-5" aria-hidden="true" tabindex="-1"></a>modified_names <span class="ot">&lt;-</span> <span class="fu">sapply</span>(individual_names, <span class="cf">function</span>(name) {</span>
<span id="cb88-6"><a href="#cb88-6" aria-hidden="true" tabindex="-1"></a>  parts <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(name, <span class="st">"_"</span>)[[<span class="dv">1</span>]]</span>
<span id="cb88-7"><a href="#cb88-7" aria-hidden="true" tabindex="-1"></a>  parts[<span class="dv">1</span>]</span>
<span id="cb88-8"><a href="#cb88-8" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb88-9"><a href="#cb88-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-10"><a href="#cb88-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert the output to a factor</span></span>
<span id="cb88-11"><a href="#cb88-11" aria-hidden="true" tabindex="-1"></a>modified_names <span class="ot">&lt;-</span> <span class="fu">factor</span>(modified_names)</span>
<span id="cb88-12"><a href="#cb88-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-13"><a href="#cb88-13" aria-hidden="true" tabindex="-1"></a><span class="co">#then we assign the populations</span></span>
<span id="cb88-14"><a href="#cb88-14" aria-hidden="true" tabindex="-1"></a>combined_kim_15_cen_15_se_15.vcf<span class="sc">@</span>pop <span class="ot">&lt;-</span> modified_names</span>
<span id="cb88-15"><a href="#cb88-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-16"><a href="#cb88-16" aria-hidden="true" tabindex="-1"></a><span class="co">#then we apply a call rate filter</span></span>
<span id="cb88-17"><a href="#cb88-17" aria-hidden="true" tabindex="-1"></a>combined_kim_15_cen_15_se_15.vcf_0<span class="fl">.7</span> <span class="ot">&lt;-</span> <span class="fu">gl.filter.callrate</span>(combined_kim_15_cen_15_se_15.vcf,</span>
<span id="cb88-18"><a href="#cb88-18" aria-hidden="true" tabindex="-1"></a>                                                           <span class="at">threshold =</span> <span class="fl">0.95</span>, <span class="at">mono.rm =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.95 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate </code></pre>
</div>
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="co">#then we look at the heterozygosity estimates by population</span></span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>diversity_equal_sample_15_ind <span class="ot">&lt;-</span> <span class="fu">gl.report.heterozygosity</span>(combined_kim_15_cen_15_se_15.vcf_0<span class="fl">.7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-13-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>                pop    n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc       Ho
Central     Central 14.74289   914         1     205     709         0 0.037671
Kimberley Kimberley 14.07878   914         1     656     258         0 0.117284
SouthEast SouthEast 14.62144   914         1     372     542         0 0.079583
              HoSD     HoSE       He     HeSD     HeSE      uHe    uHeSD
Central   0.112102 0.003708 0.071363 0.153583 0.005080 0.073868 0.158975
Kimberley 0.134002 0.004432 0.143542 0.152636 0.005049 0.148827 0.158257
SouthEast 0.134695 0.004455 0.089745 0.145624 0.004817 0.092922 0.150781
             uHeSE      FIS    FISSD FISSE
Central   0.005258 0.400900 0.452247    NA
Kimberley 0.005235 0.128504 0.310302    NA
SouthEast 0.004987 0.091166 0.254158    NA
Completed: gl.report.heterozygosity </code></pre>
</div>
</div>
<p>We now see that our heterozygosity is found to be highest in the Kimberley and lowest in Central Australia, which is true.</p>
<section id="calling-snps-together-but-with-different-sample-sizes" class="level3">
<h3 class="anchored" data-anchor-id="calling-snps-together-but-with-different-sample-sizes">Calling SNPs together but with different sample sizes</h3>
<p>Lots of times we have different numbers of samples… so what does this do to our estimates? Lets look at what happens when we have different sample numbers across all populations</p>
<div class="cell" data-hash="session03_cache/html/unnamed-chunk-14_0e6a546597d10d354dd12771bc8a899f">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>individual_names <span class="ot">&lt;-</span> <span class="fu">as.vector</span>(<span class="fu">indNames</span>(combined_kim_15_cen_5_se_10.vcf))</span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Split each name at the underscore and keep the first part</span></span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a>modified_names <span class="ot">&lt;-</span> <span class="fu">sapply</span>(individual_names, <span class="cf">function</span>(name) {</span>
<span id="cb94-5"><a href="#cb94-5" aria-hidden="true" tabindex="-1"></a>  parts <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(name, <span class="st">"_"</span>)[[<span class="dv">1</span>]]</span>
<span id="cb94-6"><a href="#cb94-6" aria-hidden="true" tabindex="-1"></a>  parts[<span class="dv">1</span>]</span>
<span id="cb94-7"><a href="#cb94-7" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb94-8"><a href="#cb94-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-9"><a href="#cb94-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert the output to a factor</span></span>
<span id="cb94-10"><a href="#cb94-10" aria-hidden="true" tabindex="-1"></a>modified_names <span class="ot">&lt;-</span> <span class="fu">factor</span>(modified_names)</span>
<span id="cb94-11"><a href="#cb94-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-12"><a href="#cb94-12" aria-hidden="true" tabindex="-1"></a><span class="co">#then we assign the populations</span></span>
<span id="cb94-13"><a href="#cb94-13" aria-hidden="true" tabindex="-1"></a>combined_kim_15_cen_5_se_10.vcf<span class="sc">@</span>pop <span class="ot">&lt;-</span> modified_names</span>
<span id="cb94-14"><a href="#cb94-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-15"><a href="#cb94-15" aria-hidden="true" tabindex="-1"></a><span class="co">#then we apply a call rate filter</span></span>
<span id="cb94-16"><a href="#cb94-16" aria-hidden="true" tabindex="-1"></a>combined_kim_15_cen_5_se_10.vcf_0<span class="fl">.95</span> <span class="ot">&lt;-</span> <span class="fu">gl.filter.callrate</span>(combined_kim_15_cen_5_se_10.vcf, <span class="at">threshold =</span> <span class="fl">0.95</span>, <span class="at">mono.rm =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.95 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate </code></pre>
</div>
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="co">#then we look at the heterozygosity estimates by population</span></span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>diversity_unequal_sample <span class="ot">&lt;-</span> <span class="fu">gl.report.heterozygosity</span>(combined_kim_15_cen_5_se_10.vcf_0<span class="fl">.95</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="session03_files/figure-html/unnamed-chunk-14-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>                pop     n.Ind n.Loc n.Loc.adj polyLoc monoLoc all_NALoc
Central     Central  4.980290   964         1     206     758         0
Kimberley Kimberley 14.429461   964         1     709     255         0
SouthEast SouthEast  9.854772   964         1     373     591         0
                Ho     HoSD     HoSE       He     HeSD     HeSE      uHe
Central   0.047355 0.138961 0.004476 0.067501 0.141209 0.004548 0.075034
Kimberley 0.113975 0.131324 0.004230 0.140553 0.147435 0.004749 0.145598
SouthEast 0.089523 0.153169 0.004933 0.094679 0.151447 0.004878 0.099740
             uHeSD    uHeSE      FIS    FISSD FISSE
Central   0.156968 0.005056 0.325189 0.510923    NA
Kimberley 0.152727 0.004919 0.130096 0.317600    NA
SouthEast 0.159541 0.005138 0.065369 0.267902    NA
Completed: gl.report.heterozygosity </code></pre>
</div>
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="co">#how much more heterozygosity does kimberley have than central in different analyses?</span></span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a><span class="co">#where our samples are equal?</span></span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a>diversity_equal_sample_15_ind<span class="sc">$</span>Ho[<span class="dv">1</span>]<span class="sc">/</span>diversity_equal_sample_15_ind<span class="sc">$</span>Ho[<span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.4733549</code></pre>
</div>
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="co">#where our samples are equal?</span></span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a>diversity_unequal_sample<span class="sc">$</span>Ho[<span class="dv">1</span>]<span class="sc">/</span>diversity_unequal_sample<span class="sc">$</span>Ho[<span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.5289702</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Take away
</div>
</div>
<div class="callout-body-container callout-body">
<p>What you need to take away is that as the sample size of our most diverse population grows, the less diverse populations become more similar. In this case, if you use really unequal sample sizes, your results get so biased that your less diverse population can be calculated as more diverse than your most diverse population.</p>
<section id="think-of-the-conservation-implications" class="level4">
<h4 class="anchored" data-anchor-id="think-of-the-conservation-implications"><strong>Think of the conservation implications!</strong></h4>
<p>In a full analysis of all the data, with unequal sample numbers and SNPs called on all populations, we got a result that the Kimberley population was the least diverse. This would then result in a direction that the most genetically important population is actually one of the least.</p>
</section>
</div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
SNP-based heterozygosity
</div>
</div>
<div class="callout-body-container callout-body">
<p>SNP-based heterozygosity is extremely biased and cannot be used to compare different populations, and there are no clear workarounds that will give you reliable answers.</p>
<p>You need to report autosomal heterozygosity at a minimum, which means doing your own bioinformatics from raw data at this time.</p>
</div>
</div>
</section>
</section>
<section id="further-study" class="level2">
<h2 class="anchored" data-anchor-id="further-study"><em>Further Study</em></h2>
<p>Galaxy Australia:</p>
<ul>
<li><p>Bioinformatics resources available to through the AAF login (your university)</p></li>
<li><p>Lots of compute power</p></li>
<li><p>Has the Stacks2 pipeline available</p></li>
</ul>
<p>Links:</p>
<p><a href="https://australianbiocommons.github.io/how-to-guides/stacks_workflows/stacks" class="uri">https://australianbiocommons.github.io/how-to-guides/stacks_workflows/stacks</a></p>
<p><a href="https://catchenlab.life.illinois.edu/stacks/manual/" class="uri">https://catchenlab.life.illinois.edu/stacks/manual/</a></p>


</section>

</main> <!-- /main -->
<script>

/* update total correct if #webex-total_correct exists */
update_total_correct = function() {
  console.log("webex: update total_correct");

  var t = document.getElementsByClassName("webex-total_correct");
  for (var i = 0; i < t.length; i++) {
    p = t[i].parentElement;
    var correct = p.getElementsByClassName("webex-correct").length;
    var solvemes = p.getElementsByClassName("webex-solveme").length;
    var radiogroups = p.getElementsByClassName("webex-radiogroup").length;
    var selects = p.getElementsByClassName("webex-select").length;

    t[i].innerHTML = correct + " of " + (solvemes + radiogroups + selects) + " correct";
  }
}

/* webex-solution button toggling function */
b_func = function() {
  console.log("webex: toggle hide");

  var cl = this.parentElement.classList;
  if (cl.contains('open')) {
    cl.remove("open");
  } else {
    cl.add("open");
  }
}

/* check answers */
check_func = function() {
  console.log("webex: check answers");

  var cl = this.parentElement.classList;
  if (cl.contains('unchecked')) {
    cl.remove("unchecked");
    this.innerHTML = "Hide Answers";
  } else {
    cl.add("unchecked");
    this.innerHTML = "Show Answers";
  }
}

/* function for checking solveme answers */
solveme_func = function(e) {
  console.log("webex: check solveme");

  var real_answers = JSON.parse(this.dataset.answer);
  var my_answer = this.value;
  var cl = this.classList;
  if (cl.contains("ignorecase")) {
    my_answer = my_answer.toLowerCase();
  }
  if (cl.contains("nospaces")) {
    my_answer = my_answer.replace(/ /g, "")
  }

  if (my_answer == "") {
    cl.remove("webex-correct");
    cl.remove("webex-incorrect");
  } else if (real_answers.includes(my_answer)) {
    cl.add("webex-correct");
    cl.remove("webex-incorrect");
  } else {
    cl.add("webex-incorrect");
    cl.remove("webex-correct");
  }

  // match numeric answers within a specified tolerance
  if(this.dataset.tol > 0){
    var tol = JSON.parse(this.dataset.tol);
    var matches = real_answers.map(x => Math.abs(x - my_answer) < tol)
    if (matches.reduce((a, b) => a + b, 0) > 0) {
      cl.add("webex-correct");
    } else {
      cl.remove("webex-correct");
    }
  }

  // added regex bit
  if (cl.contains("regex")){
    answer_regex = RegExp(real_answers.join("|"))
    if (answer_regex.test(my_answer)) {
      cl.add("webex-correct");
    }
  }

  update_total_correct();
}

/* function for checking select answers */
select_func = function(e) {
  console.log("webex: check select");

  var cl = this.classList

  /* add style */
  cl.remove("webex-incorrect");
  cl.remove("webex-correct");
  if (this.value == "answer") {
    cl.add("webex-correct");
  } else if (this.value != "blank") {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

/* function for checking radiogroups answers */
radiogroups_func = function(e) {
  console.log("webex: check radiogroups");

  var checked_button = document.querySelector('input[name=' + this.id + ']:checked');
  var cl = checked_button.parentElement.classList;
  var labels = checked_button.parentElement.parentElement.children;

  /* get rid of styles */
  for (i = 0; i < labels.length; i++) {
    labels[i].classList.remove("webex-incorrect");
    labels[i].classList.remove("webex-correct");
  }

  /* add style */
  if (checked_button.value == "answer") {
    cl.add("webex-correct");
  } else {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

window.onload = function() {
  console.log("webex onload");
  /* set up solution buttons */
  var buttons = document.getElementsByTagName("button");

  for (var i = 0; i < buttons.length; i++) {
    if (buttons[i].parentElement.classList.contains('webex-solution')) {
      buttons[i].onclick = b_func;
    }
  }

  var check_sections = document.getElementsByClassName("webex-check");
  console.log("check:", check_sections.length);
  for (var i = 0; i < check_sections.length; i++) {
    check_sections[i].classList.add("unchecked");

    let btn = document.createElement("button");
    btn.innerHTML = "Show Answers";
    btn.classList.add("webex-check-button");
    btn.onclick = check_func;
    check_sections[i].appendChild(btn);

    let spn = document.createElement("span");
    spn.classList.add("webex-total_correct");
    check_sections[i].appendChild(spn);
  }

  /* set up webex-solveme inputs */
  var solveme = document.getElementsByClassName("webex-solveme");

  for (var i = 0; i < solveme.length; i++) {
    /* make sure input boxes don't auto-anything */
    solveme[i].setAttribute("autocomplete","off");
    solveme[i].setAttribute("autocorrect", "off");
    solveme[i].setAttribute("autocapitalize", "off");
    solveme[i].setAttribute("spellcheck", "false");
    solveme[i].value = "";

    /* adjust answer for ignorecase or nospaces */
    var cl = solveme[i].classList;
    var real_answer = solveme[i].dataset.answer;
    if (cl.contains("ignorecase")) {
      real_answer = real_answer.toLowerCase();
    }
    if (cl.contains("nospaces")) {
      real_answer = real_answer.replace(/ /g, "");
    }
    solveme[i].dataset.answer = real_answer;

    /* attach checking function */
    solveme[i].onkeyup = solveme_func;
    solveme[i].onchange = solveme_func;

    solveme[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  /* set up radiogroups */
  var radiogroups = document.getElementsByClassName("webex-radiogroup");
  for (var i = 0; i < radiogroups.length; i++) {
    radiogroups[i].onchange = radiogroups_func;
  }

  /* set up selects */
  var selects = document.getElementsByClassName("webex-select");
  for (var i = 0; i < selects.length; i++) {
    selects[i].onchange = select_func;
    selects[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  update_total_correct();
}

</script>
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
          // default icon
          link.classList.add("external");
      }
    }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./session02.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Pop Gen In Conservation</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./session04.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Data Management, Reproducibility &amp; Integrity</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>